<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Karissa Barthelson" />

<meta name="date" content="2022-12-12" />

<title>_BCAC</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/master/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>








<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">2022_MPSIII_3mBrainRNAseq</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="QC.html">QC</a>
</li>
<li>
  <a href="analysis.html">Analysis</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">_BCAC</h1>
<h4 class="author">Karissa Barthelson</h4>
<h4 class="date">2022-12-12</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2022-12-20
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 6 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 1
</p>
<p>
<strong>Knit directory:</strong> <code>2022_MPSIII_3mBrainRNAseq/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.6.2). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguncommittedchanges"> <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> <strong>R Markdown file:</strong> uncommitted changes </a>
</p>
</div>
<div id="strongRMarkdownfilestronguncommittedchanges" class="panel-collapse collapse">
<div class="panel-body">
<p>The R Markdown is untracked by Git. To know which version of the R Markdown file created these results, you’ll want to first commit it to the Git repo. If you’re still working on the analysis, you can ignore this warning. When you’re finished, you can run <code>wflow_publish</code> to commit the R Markdown file and build the HTML.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20220423code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20220423)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20220423code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20220423)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomkarissab2022MPSIII3mBrainRNAseqtreeb4f1f28c43d8b85636b53c7e8f83692dbe7e757atargetblankb4f1f28a"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/karissa-b/2022_MPSIII_3mBrainRNAseq/tree/b4f1f28c43d8b85636b53c7e8f83692dbe7e757a" target="_blank">b4f1f28</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomkarissab2022MPSIII3mBrainRNAseqtreeb4f1f28c43d8b85636b53c7e8f83692dbe7e757atargetblankb4f1f28a" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility.
</p>
<p>
The results in this page were generated with repository version <a href="https://github.com/karissa-b/2022_MPSIII_3mBrainRNAseq/tree/b4f1f28c43d8b85636b53c7e8f83692dbe7e757a" target="_blank">b4f1f28</a>. See the <em>Past versions</em> tab to see a history of the changes made to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Rapp.history
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    data/.DS_Store
    Ignored:    data/R_objetcs/
    Ignored:    output/.DS_Store

Untracked files:
    Untracked:  analysis/QC.Rmd
    Untracked:  analysis/QCofBC_comb.Rmd
    Untracked:  analysis/_BCAC.rmd
    Untracked:  analysis/analysis.Rmd
    Untracked:  analysis/analysis_AB.Rmd
    Untracked:  analysis/bashProcessing.Rmd
    Untracked:  analysis/explorationCQN.Rmd
    Untracked:  code/intervals.R
    Untracked:  code/mergeFiles.sh
    Untracked:  code/prelimPlots.R
    Untracked:  data/AB/
    Untracked:  data/BC_AC_run1/
    Untracked:  data/BC_AC_run2/
    Untracked:  data/RNASeqMetaData.xlsx
    Untracked:  data/R_objects/
    Untracked:  data/all_samples.vcf.gz
    Untracked:  data/gene_sets/
    Untracked:  data/mergedBCAC/
    Untracked:  data/meta_final.xlsx
    Untracked:  data/v4.2.1_geneinformation.tab
    Untracked:  data/~$RNASeqMetaData.xlsx
    Untracked:  output/SCFplots/
    Untracked:  output/plots/

Unstaged changes:
    Modified:   analysis/_site.yml
    Modified:   analysis/index.Rmd

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
There are no past versions. Publish this analysis with <code>wflow_publish()</code> to start tracking its development.
</p>
<hr>
</div>
</div>
</div>
<div id="introduction" class="section level2">
<h2>Introduction</h2>
<pre class="r"><code>library(tidyverse)
library(magrittr)
library(readxl)
library(ngsReports)
library(AnnotationHub)
library(biomaRt)
library(msigdbr)
library(pander)
library(scales)
library(pheatmap)
library(ggpubr)
library(ggrepel)
library(ggfortify)
library(RColorBrewer)
library(UpSetR)
library(edgeR)
library(cqn)
library(goseq)
library(fgsea)
library(harmonicmeanp)
theme_set(theme_bw())</code></pre>
<pre class="r"><code>ah &lt;- AnnotationHub() %&gt;%
    subset(species == &quot;Danio rerio&quot;) %&gt;%
    subset(rdataclass == &quot;EnsDb&quot;)

ensDb &lt;- ah[[&quot;AH83189&quot;]] # for release 101, latest version in Ah
grTrans &lt;- transcripts(ensDb)
trLengths &lt;- exonsBy(ensDb, &quot;tx&quot;) %&gt;%
    width() %&gt;%
    vapply(sum, integer(1))
mcols(grTrans)$length &lt;- trLengths[names(grTrans)]
gcGene &lt;- grTrans %&gt;%
  mcols() %&gt;%
  as.data.frame() %&gt;%
  dplyr::select(gene_id, tx_id, gc_content, length) %&gt;%
  as_tibble() %&gt;%
  group_by(gene_id) %&gt;%
  summarise(
    gc_content = sum(gc_content*length) / sum(length),
    length = ceiling(median(length))
  )
grGenes &lt;- genes(ensDb)
mcols(grGenes) %&lt;&gt;%
  as.data.frame() %&gt;%
  left_join(gcGene) %&gt;%
  as.data.frame() %&gt;%
  DataFrame()</code></pre>
<pre class="r"><code>meta &lt;- 
  read_excel(&quot;data/meta_final.xlsx&quot;) %&gt;% 
  mutate(genotype = case_when(
    genotype == &quot;wt&quot; ~ &quot;WT&quot;, 
    genotype == &quot;het&quot; ~ &quot;sgsh/+&quot;,
    genotype == &quot;naglu&quot; ~ &quot;MPS-IIIB&quot;, 
    genotype == &quot;sgsh&quot; ~ &quot;MPS-IIIA&quot;,
    genotype == &quot;hgsnat&quot; ~ &quot;MPS-IIIC&quot;
  ) %&gt;% 
    factor(levels = c(&quot;WT&quot;, &quot;sgsh/+&quot;, &quot;MPS-IIIA&quot;, &quot;MPS-IIIB&quot;, &quot;MPS-IIIC&quot;)),
  Tank = as.factor(`Home tank`)
  ) #%&gt;% 
  # mutate(Tank = as.factor(Tank),
  #        batchKilled = as.factor(batchKilled), 
  #        Sex = as.factor(Sex)) %&gt;% 
  # dplyr::filter(!is.na(ULN))

featureCounts &lt;- 
  read_delim(file = &quot;data/mergedBCAC/fc/counts.out&quot;, delim = &quot;\t&quot;, skip = 1) %&gt;%
  set_names(basename(names(.))) %&gt;% 
  set_names(names(.) %&gt;% str_remove(pattern = &quot;_S[0-9]+_merged.Aligned.sortedByCoord.dedup.out.bam&quot;)) %&gt;% 
  as_tibble() %&gt;% 
  dplyr::select(-c(Chr, Start, End, Length, Strand)) %&gt;% 
  gather(key = &quot;ULN&quot;, value = &quot;counts&quot;, starts_with(&quot;22&quot;)) %&gt;% 
  left_join(meta) %&gt;% 
  dplyr::select(Geneid, counts, sample) %&gt;% 
  spread(key = &quot;sample&quot;, value = &quot;counts&quot;) %&gt;% 
  column_to_rownames(&quot;Geneid&quot;) %&gt;% 
  .[,-(dim(.)[2])]# omit the SAGC neg control lane, the final col of the df. </code></pre>
</div>
<div id="filter-lowly-expressed-genes" class="section level2">
<h2>filter lowly expressed genes</h2>
<pre class="r"><code>min.lib.size = featureCounts %&gt;% 
    as.matrix() %&gt;% 
    DGEList() %&gt;%
    calcNormFactors() %&gt;% 
  .$samples %&gt;% 
  .$lib.size %&gt;% 
  min()</code></pre>
<p>Genes which are lowly expressed are uninformative for differential gene expression analysis. A general rule of thumb for considering a gene to be lowly expressed if it contains a count per million (CPM) of more than 10/(min_lib_size/1000000) in at least one of the experimental groups. This will allow us . The min lib size in this analysis is 4,106,543. So the minimum CPM in this analysis will be 2.4351383.</p>
<pre class="r"><code>a &lt;- featureCounts %&gt;% 
    cpm(log = TRUE) %&gt;%
    as.data.frame() %&gt;% 
  pivot_longer(
    cols = everything(),
    names_to = &quot;sample&quot;,
    values_to = &quot;logCPM&quot;
  ) %&gt;%
  split(f = .$sample) %&gt;%
  lapply(function(x){
    d &lt;- density(x$logCPM)
    tibble(
      sample = unique(x$sample),
      x = d$x,
      y = d$y
    )
  }) %&gt;%
  bind_rows() %&gt;%
  left_join(meta) %&gt;% 
  ggplot(aes(x, y, colour = genotype, group = sample)) +
  geom_line() +
  scale_y_continuous(limits = c(0,0.35)) +
  labs(
    x = &quot;logCPM&quot;,
    y = &quot;Density&quot;,
    colour = &quot;Genotype&quot;
  )+
  ggtitle(&quot;Before filtering&quot;) +
  theme_bw()

b &lt;- featureCounts %&gt;% 
  .[rowSums(cpm(.) &gt;= 2.435138) &gt;= 6,] %&gt;% 
    cpm(log = TRUE) %&gt;%
    as.data.frame() %&gt;% 
  pivot_longer(
    cols = everything(),
    names_to = &quot;sample&quot;,
    values_to = &quot;logCPM&quot;
  ) %&gt;%
  split(f = .$sample) %&gt;%
  lapply(function(x){
    d &lt;- density(x$logCPM)
    tibble(
      sample = unique(x$sample),
      x = d$x,
      y = d$y
    )
  }) %&gt;%
  bind_rows() %&gt;%
  left_join(meta) %&gt;% 
  ggplot(aes(x, y, colour = genotype, group = sample)) +
  geom_line() +
  scale_y_continuous(limits = c(0,0.35)) +
  labs(
    x = &quot;logCPM&quot;,
    y = &quot;Density&quot;,
    colour = &quot;Genotype&quot;
  )+
  ggtitle(&quot;After filtering&quot;)

ggarrange(a, b, common.legend = TRUE)</code></pre>
<p><img src="figure/_BCAC.rmd/filt-1.png" width="75%" height="75%" style="display: block; margin: auto;" /></p>
<pre class="r"><code>x &lt;- featureCounts %&gt;% 
  as.matrix() %&gt;% 
  .[rowSums(cpm(.) &gt;= 3) &gt;= 6,] %&gt;% 
  DGEList(
    samples = tibble(sample = colnames(.)) %&gt;%
      left_join(meta),
    genes = grGenes[rownames(.)] %&gt;%
      as.data.frame() %&gt;%
      dplyr::select(
        chromosome = seqnames, start, end, 
        gene_id, gene_name, gene_biotype, description, entrezid
      ) %&gt;% 
      left_join(gcGene) %&gt;% 
      as_tibble()
  ) %&gt;%
  calcNormFactors()</code></pre>
</div>
<div id="pca" class="section level2">
<h2>PCA</h2>
<div id="both-ac-and-bc" class="section level3">
<h3>Both AC and BC</h3>
<p>2 distinct groups are observed across PC2. And this is explained by the 2 families of fish which comprise this dataset. Therefore, I should analyse them seperately</p>
<pre class="r"><code>x$counts %&gt;% 
  cpm(log=TRUE) %&gt;% 
  t() %&gt;% 
  prcomp() %&gt;% 
  autoplot(data = tibble(sample = rownames(.$x)) %&gt;%
      left_join(x$samples),
    colour = &quot;MPS&quot;, 
    size = 6
  ) +
  theme(aspect.ratio = 1, 
        legend.position = &quot;bottom&quot;) +
  labs(colour = &quot;MPS comparison family&quot;) +
  scale_color_brewer(palette = &quot;Set2&quot;)</code></pre>
<p><img src="figure/_BCAC.rmd/unnamed-chunk-2-1.png" width="75%" height="75%" style="display: block; margin: auto;" /></p>
</div>
<div id="ac-only" class="section level3">
<h3>AC only</h3>
<p>The AC group came from 1 family, raised across 5 tanks side by side in the same system. The MPS-IIIA samples appear to be clustering away from the other 2.</p>
<pre class="r"><code>BCsamps &lt;- meta %&gt;% 
  dplyr::filter(MPS == &quot;BC&quot;) %&gt;% 
  .$sample

ACsamps &lt;- meta %&gt;% 
  dplyr::filter(MPS == &quot;AC&quot;) %&gt;% 
  .$sample

ggarrange(
x$counts %&gt;% 
  .[,ACsamps] %&gt;% 
  cpm(log=TRUE) %&gt;% 
  t() %&gt;% 
  prcomp() %&gt;% 
  autoplot(data = tibble(sample = rownames(.$x)) %&gt;%
      left_join(x$samples),
    colour = &quot;genotype&quot;, 
    size = 6
  ) +
  theme(aspect.ratio = 1) +
  ggtitle(&quot;by genotype&quot;) +
  labs(colour = &quot;&quot;),

x$counts %&gt;% 
  .[,ACsamps] %&gt;% 
  cpm(log=TRUE) %&gt;% 
  t() %&gt;% 
  prcomp() %&gt;% 
  autoplot(data = tibble(sample = rownames(.$x)) %&gt;%
      left_join(x$samples),
    colour = &quot;Tank&quot;, 
    size = 6
  ) +
  theme(aspect.ratio = 1) +
  labs(colour = &quot;&quot;) +
  ggtitle(&quot;by tank&quot;)
)</code></pre>
<p><img src="figure/_BCAC.rmd/unnamed-chunk-3-1.png" width="75%" height="75%" style="display: block; margin: auto;" /></p>
</div>
<div id="bc-only" class="section level3">
<h3>BC only</h3>
<p>The BC group of fish came from 2 sets of parents, across multiple lays/ This is evident in the plot below. The effect of “Home Tank” cannot be ignored in a DE analysis.</p>
<pre class="r"><code>ggarrange(
x$counts %&gt;% 
  .[,BCsamps] %&gt;% 
  cpm(log=TRUE) %&gt;% 
  t() %&gt;% 
  prcomp() %&gt;% 
  autoplot(data = tibble(sample = rownames(.$x)) %&gt;%
      left_join(x$samples),
    colour = &quot;genotype&quot;, 
    size = 6
  ) +
  theme(aspect.ratio = 1) +
  ggtitle(&quot;by genotype&quot;) +
  labs(colour = &quot;&quot;),

x$counts %&gt;% 
  .[,BCsamps] %&gt;% 
  cpm(log=TRUE) %&gt;% 
  t() %&gt;% 
  prcomp() %&gt;% 
  autoplot(data = tibble(sample = rownames(.$x)) %&gt;%
      left_join(x$samples),
    colour = &quot;Tank&quot;, 
    size = 6
  ) +
  theme(aspect.ratio = 1) +
  labs(colour = &quot;&quot;) +
  ggtitle(&quot;by tank&quot;)
)</code></pre>
<p><img src="figure/_BCAC.rmd/unnamed-chunk-4-1.png" width="75%" height="75%" style="display: block; margin: auto;" /> From hereout, I will process the 2 comparisons separately.</p>
</div>
</div>
<div id="ac" class="section level1">
<h1>AC</h1>
<div id="initial-de-analysis" class="section level2">
<h2>initial DE analysis</h2>
<pre class="r"><code>x.ac &lt;- x[,ACsamps]

design.ac &lt;- model.matrix(~genotype + Tank, data = x.ac$samples %&gt;% 
                            droplevels()) %&gt;% # get rid of the unused factor levels
  set_colnames(gsub(colnames(.), pattern = &quot;genotype&quot;, replacement = &quot;&quot;))

fit_1_ac &lt;- x.ac %&gt;% 
  estimateDisp(design.ac) %&gt;% 
  glmFit(design.ac)

toptable_1.ac &lt;- design.ac %&gt;% colnames() %&gt;% .[2:3] %&gt;% 
  sapply(function(x){
    glmLRT(fit_1_ac, coef = x) %&gt;%
      topTags(n = Inf) %&gt;%
      .[[&quot;table&quot;]] %&gt;%
      as_tibble() %&gt;%
      arrange(PValue) %&gt;%
       mutate(
        coef = x,
        DE = FDR &lt; 0.05
      ) %&gt;% 
      dplyr::select(
        gene_name, logFC, logCPM, PValue, FDR, DE, everything()  
      )
      
  }, simplify = FALSE)</code></pre>
<div id="vis" class="section level3">
<h3>vis</h3>
</div>
<div id="number-of-de-genes" class="section level3">
<h3>Number of DE genes</h3>
<pre class="r"><code>toptable_1.ac %&gt;% 
  lapply(function(y) {
    y %&gt;% 
      dplyr::filter(DE == T) %&gt;% 
      nrow()
  }) %&gt;% 
  bind_rows() %&gt;% 
  as_tibble() %&gt;% 
  gather(key = &quot;coef&quot;, value = &quot;numDE&quot;) %&gt;% 
  ggplot(aes(x = numDE, y = coef)) +
  geom_col(aes(fill = coef)) +
  geom_text(aes(label = numDE)) +
  ggtitle(&quot;Number of DE genes in each comparison&quot;,
          subtitle = &quot;The MPS-IIIA zebrafish samples have the most DE genes&quot;)</code></pre>
<p><img src="figure/_BCAC.rmd/unnamed-chunk-6-1.png" width="75%" height="75%" style="display: block; margin: auto;" /></p>
</div>
<div id="volcano-plots" class="section level3">
<h3>Volcano plots</h3>
<pre class="r"><code>toptable_1.ac %&gt;%
  bind_rows() %&gt;% 
  ggplot(aes(y = -log10(PValue), x = logFC, colour = DE)) +
  geom_point(
    alpha = 0.5, size = 1.25
  ) +
  facet_wrap(~coef) +
  # geom_label_repel(
  #   aes(label = gene_name),
  #   data = .  %&gt;% dplyr::filter(FDR &lt; 0.05),
  #   show.legend = FALSE
  #   ) +
  # coord_cartesian(xlim = c(-2.5,2.5), 
  #                 ylim = c(0, 20) )+
  theme(legend.position = &quot;none&quot;) +
  scale_color_manual(values = c(&quot;grey50&quot;, &quot;red&quot;)) </code></pre>
<p><img src="figure/_BCAC.rmd/unnamed-chunk-7-1.png" width="75%" height="75%" style="display: block; margin: auto;" /></p>
</div>
<div id="md-plots" class="section level3">
<h3>MD plots</h3>
<pre class="r"><code>toptable_1.ac %&gt;%
  bind_rows() %&gt;% 
  ggplot(aes(x = logCPM, y = logFC)) +
  geom_point(
    aes(colour = DE),
    alpha = 0.5
  ) +
  facet_wrap(~coef, ncol = 1) +
  theme(legend.position = &quot;none&quot;) +
  scale_color_manual(values = c(&quot;grey50&quot;, &quot;red&quot;))</code></pre>
<p><img src="figure/_BCAC.rmd/unnamed-chunk-8-1.png" width="75%" height="75%" style="display: block; margin: auto;" /></p>
</div>
<div id="gc-and-length-check" class="section level3">
<h3>%GC AND LENGTH CHECK</h3>
<pre class="r"><code>ggarrange(
toptable_1.ac %&gt;% 
  bind_rows() %&gt;% 
  mutate(rankstat = sign(logFC)*-log10(PValue)) %&gt;% 
  ggplot(aes(x = length, y = rankstat)) +
  geom_point(
    aes(colour = DE),
    alpha = 0.5
  ) +
  geom_smooth(se = FALSE, method = &quot;gam&quot;) +
  facet_grid(rows = vars(coef)) +
  theme_bw() +
  theme(legend.position = &quot;none&quot;) +
  scale_color_manual(values = c(&quot;grey50&quot;, &quot;red&quot;)) +
  scale_x_log10()+
  labs(x = &quot;Average transcript length per gene&quot;,
       colour = &quot;Differentially expressed?&quot;,
       y = &quot;sign(logFC)*-log10(PValue)&quot;) +
  coord_cartesian(ylim = c(-10, 10)),

toptable_1.ac %&gt;% 
  bind_rows() %&gt;% 
  mutate(rankstat = sign(logFC)*-log10(PValue)) %&gt;% 
  ggplot(aes(x = gc_content, y = rankstat)) +
  geom_point(
    aes(colour = DE),
    alpha = 0.5
  ) +
  geom_smooth(se = FALSE, method = &quot;gam&quot;) +
  facet_grid(rows = vars(coef)) +
  theme_bw() +
  theme(legend.position = &quot;none&quot;) +
  scale_color_manual(values = c(&quot;grey50&quot;, &quot;red&quot;)) +
  coord_cartesian(ylim = c(-10,10)) + 
  labs(x = &quot;Weighted average GC content (%) per gene&quot;, 
       colour = &quot;Differentially expressed?&quot;, 
       y = &quot;sign(logFC)*-log10(PValue)&quot;),
common.legend = TRUE, 
labels = &quot;AUTO&quot;
) </code></pre>
<p><img src="figure/_BCAC.rmd/unnamed-chunk-9-1.png" width="75%" height="75%" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="cqn" class="section level2">
<h2>CQN</h2>
<p>A little bit of bias is obsevred for %GC content and gene length, so will apply the <code>cqn</code> correction to adjust for this.</p>
<pre class="r"><code>cqn.ac &lt;-   
x.ac %&gt;% 
  with(
    cqn(
      counts = counts,
      x = genes$gc_content,
      lengths = genes$length,
      sizeFactors = samples$lib.size
    )
  )
logCPM.ac &lt;- cqn.ac$y + cqn.ac$offset</code></pre>
</div>
<div id="plot-out-the-model-fits" class="section level2">
<h2>plot out the model fits</h2>
<pre class="r"><code>par(mfrow = c(1, 2))
cqnplot(cqn.ac, n = 1, xlab = &quot;GC Content&quot;)
cqnplot(cqn.ac, n = 2, xlab = &quot;Length&quot;)</code></pre>
<div class="figure" style="text-align: center">
<img src="figure/_BCAC.rmd/plotCQN_AC-1.png" alt="*Model fits for GC content and gene length under the CQN model. Variability is clearly visible at either end*" width="75%" height="75%" />
<p class="caption">
<em>Model fits for GC content and gene length under the CQN model. Variability is clearly visible at either end</em>
</p>
</div>
<pre class="r"><code>par(mfrow = c(1, 1))</code></pre>
</div>
<div id="pca-1" class="section level2">
<h2>PCA</h2>
<p>PCA analysis was repeated. Similar clustering of samples was observed across PC1. However, samples appeared closer together.</p>
<pre class="r"><code>ggarrange(
  cpm(x.ac, log = T) %&gt;%
    t() %&gt;%
    prcomp() %&gt;% 
    autoplot(data = tibble(sample = rownames(.$x)) %&gt;%
               left_join(x$samples),
             colour = &quot;genotype&quot;, 
             size = 4
    ) +
    scale_color_viridis_d(option = &quot;turbo&quot;, end = 0.8) +
    ggtitle(&quot;Before CQN&quot;),
  
  logCPM.ac %&gt;% 
    t() %&gt;% 
    prcomp() %&gt;%
    autoplot(data = tibble(sample = rownames(.$x)) %&gt;%
               left_join(x$samples),
             colour = &quot;genotype&quot;, 
             size = 4
    ) +
    scale_color_viridis_d(option = &quot;turbo&quot;, end = 0.8) +
    ggtitle(&quot;After CQN&quot;),
  common.legend = T, 
  labels = &quot;AUTO&quot;
) </code></pre>
<p><img src="figure/_BCAC.rmd/unnamed-chunk-11-1.png" width="75%" height="75%" style="display: block; margin: auto;" /></p>
<pre class="r"><code>#ggsave(&quot;plotsforpub/PCA_cqn.png&quot;, width = 18, height = 8, units = &quot;cm&quot;, scale = 1.4)</code></pre>
</div>
</div>
<div id="de-with-cqn" class="section level1">
<h1>DE with cqn</h1>
<pre class="r"><code># Add the offset to the dge object 
x.ac$offset &lt;- cqn.ac$glm.offset

# fit the glm
fit_cqn.ac &lt;- x.ac %&gt;% 
  estimateDisp(design.ac) %&gt;% 
  glmFit(design.ac)

toptables_cqn.ac &lt;- design.ac %&gt;% colnames() %&gt;% .[2:3] %&gt;% 
  sapply(function(x){
    glmLRT(fit_cqn.ac, coef = x) %&gt;%
      topTags(n = Inf) %&gt;%
      .[[&quot;table&quot;]] %&gt;%
      as_tibble() %&gt;%
      arrange(PValue) %&gt;%
       mutate(
        coef = x,
        DE = FDR &lt; 0.05
      ) %&gt;% 
      dplyr::select(
        gene_name, logFC, logCPM, PValue, FDR, DE, everything()  
      )
      
  }, simplify = FALSE)</code></pre>
<div id="vis-with-cqn" class="section level2">
<h2>Vis with CQN</h2>
<div id="num-de-genes" class="section level3">
<h3>Num DE genes</h3>
<pre class="r"><code>toptables_cqn.ac %&gt;% 
  lapply(function(x){
    x %&gt;%
      mutate(Direction = case_when(
        sign(logFC) == &quot;1&quot; ~ &quot;up&quot;,
        sign(logFC) == &quot;-1&quot; ~ &quot;down&quot;
      ))
  }) %&gt;%
  bind_rows() %&gt;%
  dplyr::filter(DE == TRUE) %&gt;%
  ggplot(aes(y = `coef`)) +
  geom_bar(stat = &quot;count&quot;,
           width = 0.5,
           aes(fill = Direction),
           position = &quot;dodge&quot;) +
  theme(legend.position = &quot;bottom&quot;) +
  ggtitle(&quot;Number of DE genes in each comparison&quot;) +
  labs(x = &quot;Number of DE genes&quot;,
       y = &quot;Contrast&quot;,
       colour = &quot;Direction of change&quot;) </code></pre>
<p><img src="figure/_BCAC.rmd/unnamed-chunk-13-1.png" width="75%" height="75%" style="display: block; margin: auto;" /></p>
</div>
<div id="volcano-plots-1" class="section level3">
<h3>Volcano plots</h3>
<pre class="r"><code>toptables_cqn.ac %&gt;%
  bind_rows() %&gt;% 
  ggplot(aes(y = -log10(PValue), x = logFC, colour = DE)) +
  geom_point(
    alpha = 0.5, size = 1.25
  ) +
  facet_wrap(~coef) +
  # geom_label_repel(
  #   aes(label = gene_name),
  #   data = .  %&gt;% dplyr::filter(FDR &lt; 0.05),
  #   show.legend = FALSE
  #   ) +
  # coord_cartesian(xlim = c(-2.5,2.5), 
  #                 ylim = c(0, 20) )+
  theme(legend.position = &quot;none&quot;) +
  scale_color_manual(values = c(&quot;grey50&quot;, &quot;red&quot;)) </code></pre>
<p><img src="figure/_BCAC.rmd/unnamed-chunk-14-1.png" width="75%" height="75%" style="display: block; margin: auto;" /></p>
</div>
<div id="md-plots-1" class="section level3">
<h3>MD plots</h3>
<pre class="r"><code>toptables_cqn.ac %&gt;% 
  bind_rows() %&gt;% 
  ggplot(aes(x = logCPM, y = logFC)) +
  geom_point(
    aes(colour = DE),
    alpha = 0.5
  ) +
  facet_wrap(~coef, ncol = 1) +
  theme(legend.position = &quot;none&quot;) +
  scale_color_manual(values = c(&quot;grey50&quot;, &quot;red&quot;))</code></pre>
<p><img src="figure/_BCAC.rmd/unnamed-chunk-15-1.png" width="75%" height="75%" style="display: block; margin: auto;" /></p>
</div>
<div id="recheck-the-gc-and-length-bias" class="section level3">
<h3>Recheck the %GC and length bias</h3>
<pre class="r"><code>ggarrange(
toptables_cqn.ac %&gt;% 
  bind_rows() %&gt;% 
  mutate(rankstat = sign(logFC)*-log10(PValue)) %&gt;% 
  ggplot(aes(x = length, y = rankstat)) +
  geom_point(
    aes(colour = DE),
    alpha = 0.5
  ) +
  geom_smooth(se = FALSE, method = &quot;gam&quot;) +
  facet_grid(rows = vars(coef)) +
  theme_bw() +
  theme(legend.position = &quot;none&quot;) +
  scale_color_manual(values = c(&quot;grey50&quot;, &quot;red&quot;)) +
  scale_x_log10()+
  labs(x = &quot;Average transcript length per gene&quot;,
       colour = &quot;Differentially expressed?&quot;,
       y = &quot;sign(logFC)*-log10(PValue)&quot;) +
  coord_cartesian(ylim = c(-10, 10)),

toptables_cqn.ac %&gt;% 
  bind_rows() %&gt;% 
  mutate(rankstat = sign(logFC)*-log10(PValue)) %&gt;% 
  ggplot(aes(x = gc_content, y = rankstat)) +
  geom_point(
    aes(colour = DE),
    alpha = 0.5
  ) +
  geom_smooth(se = FALSE, method = &quot;gam&quot;) +
  facet_grid(rows = vars(coef)) +
  theme_bw() +
  theme(legend.position = &quot;none&quot;) +
  scale_color_manual(values = c(&quot;grey50&quot;, &quot;red&quot;)) +
  coord_cartesian(ylim = c(-10,10)) + 
  labs(x = &quot;Weighted average GC content (%) per gene&quot;, 
       colour = &quot;Differentially expressed?&quot;, 
       y = &quot;sign(logFC)*-log10(PValue)&quot;),
common.legend = TRUE, 
labels = &quot;AUTO&quot;
) </code></pre>
<p><img src="figure/_BCAC.rmd/unnamed-chunk-16-1.png" width="75%" height="75%" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="gene-set-enrichment-analysis" class="section level2">
<h2>Gene set enrichment analysis</h2>
<div id="gene-sets" class="section level3">
<h3>Gene sets</h3>
<pre class="r"><code>KEGG &lt;- msigdbr(&quot;Danio rerio&quot;, category = &quot;C2&quot;, subcategory = &quot;CP:KEGG&quot;) %&gt;% 
  left_join(x$genes, by = c(&quot;gene_symbol&quot; = &quot;gene_name&quot;)) %&gt;% 
  dplyr::filter(gene_id %in% rownames(x)) %&gt;% 
  distinct(gs_name, gene_id, .keep_all = TRUE) %&gt;% 
  split(f = .$gs_name) %&gt;%
  lapply(extract2, &quot;gene_id&quot;)

sizes &lt;- KEGG %&gt;% 
  lapply(length) %&gt;% 
  unlist %&gt;% 
  as.data.frame() %&gt;% 
  set_colnames( &quot;n_genes&quot;) %&gt;% 
  rownames_to_column(&quot;gs&quot;)


# retain gene sets with at least 10 genes in it
KEGG &lt;- KEGG[sizes %&gt;% dplyr::filter(n_genes &gt; 10) %&gt;% .$gs]

# GO Terms
goSummaries &lt;- url(&quot;https://uofabioinformaticshub.github.io/summaries2GO/data/goSummaries.RDS&quot;) %&gt;%
  readRDS() %&gt;%
  mutate(
    Term = Term(id),
    gs_name = Term %&gt;% str_to_upper() %&gt;% str_replace_all(&quot;[ -]&quot;, &quot;_&quot;),
    gs_name = paste0(&quot;GO_&quot;, gs_name)
    )

minPath &lt;- 3

GO &lt;- 
  msigdbr(&quot;Danio rerio&quot;, category = &quot;C5&quot;) %&gt;% 
  dplyr::filter(grepl(gs_name, pattern = &quot;^GO&quot;)) %&gt;% 
  left_join(x$genes, by = c(&quot;gene_symbol&quot; = &quot;gene_name&quot;)) %&gt;% 
  dplyr::filter(gene_id %in% rownames(x)) %&gt;% 
  mutate(gs_name = str_replace(gs_name, pattern = &quot;GOBP_&quot;, replacement = &quot;GO_&quot;)) %&gt;% 
  mutate(gs_name = str_replace(gs_name, pattern = &quot;GOMF_&quot;, replacement = &quot;GO_&quot;)) %&gt;%     
  mutate(gs_name = str_replace(gs_name, pattern = &quot;GOCC_&quot;, replacement = &quot;GO_&quot;)) %&gt;%     
  left_join(goSummaries) %&gt;% 
  dplyr::filter(shortest_path &gt;= minPath) %&gt;%
  distinct(gs_name, gene_id, .keep_all = TRUE) %&gt;% 
  split(f = .$gs_name) %&gt;%
  lapply(extract2, &quot;gene_id&quot;)

GOdf &lt;-  msigdbr(&quot;Danio rerio&quot;, category = &quot;C5&quot;) %&gt;% 
  dplyr::filter(grepl(gs_name, pattern = &quot;^GO&quot;)) %&gt;% 
  left_join(x$genes, by = c(&quot;gene_symbol&quot; = &quot;gene_name&quot;)) %&gt;% 
  dplyr::filter(gene_id %in% rownames(x)) %&gt;% 
  mutate(gs_name = str_replace(gs_name, pattern = &quot;GOBP_&quot;, replacement = &quot;GO_&quot;)) %&gt;% 
  mutate(gs_name = str_replace(gs_name, pattern = &quot;GOMF_&quot;, replacement = &quot;GO_&quot;)) %&gt;%     
  mutate(gs_name = str_replace(gs_name, pattern = &quot;GOCC_&quot;, replacement = &quot;GO_&quot;)) %&gt;%     
  left_join(goSummaries) %&gt;% 
  dplyr::filter(shortest_path &gt;= minPath) %&gt;%
  distinct(gs_name, gene_id, .keep_all = TRUE) %&gt;% 
  dplyr::select(gs_name, gene_id)

sizes_GO &lt;- GO %&gt;%
  lapply(length) %&gt;%
  unlist %&gt;%
  as.data.frame() %&gt;%
  set_colnames( &quot;n_genes&quot;) %&gt;%
  rownames_to_column(&quot;gs&quot;)


# retain gene sets with at least 5 genes in it
GO &lt;- GO[sizes_GO %&gt;% dplyr::filter(n_genes &gt; 5) %&gt;% .$gs]

ireGenes &lt;- 
  read_xlsx(&quot;data/gene_sets/ireGenes.xlsx&quot;, sheet = &quot;Zebrafish IREs&quot;) %&gt;% 
  gather %&gt;% 
  set_colnames(c(&quot;IRE&quot;, &quot;gene_id&quot;)) %&gt;% 
  mutate(IRE = case_when(
    IRE == &quot;All 3&#39; IREs&quot; ~ &quot;ire3_all&quot;, 
    IRE == &quot;All 5&#39; IREs&quot; ~ &quot;ire5_all&quot;, 
    IRE == &quot;HQ 3&#39; IREs&quot; ~ &quot;ire3_HQ&quot;, 
    IRE == &quot;HQ 5&#39; IREs&quot; ~ &quot;ire5_HQ&quot;
  )) %&gt;% 
  dplyr::filter(gene_id %in% rownames(x)) %&gt;% 
  split(f = .$IRE) %&gt;% 
  lapply(magrittr::extract2,&quot;gene_id&quot;) 

chr &lt;- x$genes %&gt;% 
  dplyr::select(gene_id, chromosome) %&gt;% 
  split(f = .$chromosome) %&gt;%
  lapply(extract2, &quot;gene_id&quot;) %&gt;% 
  .[c(&quot;MT&quot;, seq(1:25))] # omit the ALT chromosmes and scaffolds

#biomart to get the conversion between mouse and zebrafish IDs
mart &lt;- useMart(&quot;ensembl&quot;, &quot;drerio_gene_ensembl&quot;)
getFromBiomart &lt;- c(&quot;ensembl_gene_id&quot;, &quot;mmusculus_homolog_associated_gene_name&quot;)
zf_id_2_mm_genename &lt;- getBM(attributes = getFromBiomart,
                             mart = mart) %&gt;% 
  as_tibble() %&gt;% 
  set_colnames(c(&quot;gene_id&quot;, &quot;mouse_gene_name&quot;))

cell_type_markers &lt;- 
  read_xls(&quot;data/gene_sets/cell_type_genes41586_2007_BFnature05453_MOESM11_ESM.xls&quot;) %&gt;%
  as_tibble() %&gt;%
  dplyr::select(c(&quot;Neuron-enriched&quot;, &quot;Astrocyte-enriched&quot;, &quot;Oligodendrocyte-enriched&quot;)) %&gt;%
  gather(key = &quot;cell_type&quot;, value = &quot;mouse_gene_name&quot;) %&gt;% 
  left_join(zf_id_2_mm_genename) %&gt;% 
  na.omit %&gt;% 
  split(f = .$cell_type) %&gt;% 
  lapply(extract2, &quot;gene_id&quot;)

cell_type_markers$`Microglia-enriched` &lt;- 
  read_excel(&quot;data/gene_sets/microglia_oosterofetal.xlsx&quot;, sheet = 1) %&gt;%
  rbind(read_excel(&quot;data/gene_sets/microglia_oosterofetal.xlsx&quot;, sheet = 2)) %&gt;%
  rbind(read_excel(&quot;data/gene_sets/microglia_oosterofetal.xlsx&quot;, sheet = 3)) %&gt;%
  .$Zebrafish.Ensembl.Gene.ID %&gt;% 
  unique

cell_type_markers %&lt;&gt;% 
  lapply(function(y) {
    y[y %in% rownames(x)]
  })</code></pre>
</div>
</div>
<div id="goseq" class="section level2">
<h2>GOSEQ</h2>
<pre class="r"><code>pwf &lt;- toptables_cqn.ac$`MPS-IIIA` %&gt;%
      with(
        nullp(
          DEgenes = structure(DE, names = gene_id), 
          bias.data = length
        )
      )</code></pre>
<p><img src="figure/_BCAC.rmd/unnamed-chunk-18-1.png" width="75%" height="75%" style="display: block; margin: auto;" /></p>
<pre class="r"><code>goseq(pwf, gene2cat = GO) %&gt;% 
  as_tibble() %&gt;% 
  dplyr::filter(numDEInCat &gt; 0) %&gt;% 
  mutate(FDR = p.adjust(over_represented_pvalue, method = &quot;fdr&quot;)) %&gt;% 
  dplyr::select(-under_represented_pvalue) %&gt;% 
  dplyr::filter(FDR &lt; 0.05) </code></pre>
<pre><code># A tibble: 28 x 5
   category                     over_represented_p… numDEInCat numInCat      FDR
   &lt;chr&gt;                                      &lt;dbl&gt;      &lt;int&gt;    &lt;int&gt;    &lt;dbl&gt;
 1 GO_GLIAL_CELL_DEVELOPMENT          0.00000000361         10       86  6.93e-6
 2 GO_AXON_ENSHEATHMENT_IN_CEN…       0.0000000105           6       19  1.00e-5
 3 GO_GLIAL_CELL_DIFFERENTIATI…       0.0000000193          12      163  1.24e-5
 4 GO_OLIGODENDROCYTE_DIFFEREN…       0.0000000346           9       80  1.32e-5
 5 GO_OLIGODENDROCYTE_DEVELOPM…       0.0000000382           7       38  1.32e-5
 6 GO_GLIOGENESIS                     0.0000000412          13      212  1.32e-5
 7 GO_PERIPHERAL_NERVOUS_SYSTE…       0.00000862             6       56  2.28e-3
 8 GO_LONG_CHAIN_FATTY_ACYL_CO…       0.0000105              4       16  2.28e-3
 9 GO_REGULATION_OF_MYELINATION       0.0000107              5       34  2.28e-3
10 GO_ALCOHOL_BINDING                 0.0000178              6       63  3.42e-3
# … with 18 more rows</code></pre>
<pre class="r"><code>pwf &lt;- toptables_cqn.ac$`MPS-IIIC` %&gt;%
      with(
        nullp(
          DEgenes = structure(DE, names = gene_id), 
          bias.data = length
        )
      )</code></pre>
<p><img src="figure/_BCAC.rmd/unnamed-chunk-18-2.png" width="75%" height="75%" style="display: block; margin: auto;" /></p>
<pre class="r"><code>goseq(pwf, gene2cat = GO) %&gt;% 
  as_tibble() %&gt;% 
  dplyr::filter(numDEInCat &gt; 0) %&gt;% 
  mutate(FDR = p.adjust(over_represented_pvalue, method = &quot;fdr&quot;)) %&gt;% 
  dplyr::select(-under_represented_pvalue) %&gt;% 
  dplyr::filter(FDR &lt; 0.05) %&gt;% 
  head(10)</code></pre>
<pre><code># A tibble: 0 x 5
# … with 5 variables: category &lt;chr&gt;, over_represented_pvalue &lt;dbl&gt;,
#   numDEInCat &lt;int&gt;, numInCat &lt;int&gt;, FDR &lt;dbl&gt;</code></pre>
</div>
<div id="kegg" class="section level2">
<h2>KEGG</h2>
<pre class="r"><code>fryKEGG &lt;- 
  design.ac %&gt;% colnames() %&gt;% .[2:3] %&gt;% 
  sapply(function(y) {
    logCPM.ac %&gt;% 
      fry(
        index = KEGG,
        design = design.ac, 
        contrast = y, 
        sort = &quot;mixed&quot;
      ) %&gt;% 
      rownames_to_column(&quot;pathway&quot;) %&gt;% 
      as_tibble() %&gt;% 
      mutate(coef = y)
  }, simplify = FALSE)

fryKEGG$`MPS-IIIA` %&gt;% 
  dplyr::filter(FDR.Mixed &lt; 0.05) %&gt;%  
  mutate(pathway = str_remove(pathway, pattern = &quot;KEGG_&quot;)) %&gt;% 
  ggplot(aes(x = -log10(PValue.Mixed), y = reorder(pathway, -FDR.Mixed))) +
  geom_col(aes(fill = FDR.Mixed &lt; 0.05)) +
  scale_fill_manual(values = c(&quot;grey50&quot;, &#39;red&#39;)) +
  theme_pubr() +
  labs(fill = &quot;FDR &lt; 0.05&quot;, 
       x = &quot;-log10(PValue)&quot;, 
       y = &quot;&quot;)</code></pre>
<p><img src="figure/_BCAC.rmd/unnamed-chunk-19-1.png" width="75%" height="75%" style="display: block; margin: auto;" /></p>
<pre class="r"><code>fryKEGG$`MPS-IIIC` %&gt;% 
  dplyr::filter(FDR.Mixed &lt; 0.05) %&gt;%  
  mutate(pathway = str_remove(pathway, pattern = &quot;KEGG_&quot;)) %&gt;% 
  ggplot(aes(x = -log10(PValue.Mixed), y = reorder(pathway, -FDR.Mixed))) +
  geom_col(aes(fill = FDR.Mixed &lt; 0.05)) +
  scale_fill_manual(values = c(&quot;grey50&quot;, &#39;red&#39;)) +
  theme_pubr() +
  labs(fill = &quot;FDR &lt; 0.05&quot;, 
       x = &quot;-log10(PValue)&quot;, 
       y = &quot;&quot;)  </code></pre>
<p><img src="figure/_BCAC.rmd/unnamed-chunk-19-2.png" width="75%" height="75%" style="display: block; margin: auto;" /></p>
<pre class="r"><code>sigpaths &lt;- fryKEGG %&gt;% 
  bind_rows() %&gt;% 
  dplyr::filter(FDR.Mixed &lt; 0.05) %&gt;% 
    .$pathway %&gt;% 
    unique()
  
fryKEGG %&gt;% 
  bind_rows() %&gt;% 
  dplyr::filter(pathway %in% sigpaths) %&gt;% 
  mutate(pathway = str_remove(pathway, pattern = &quot;KEGG_&quot;)) %&gt;% 
  ggplot(aes(x = -log10(PValue.Mixed), y = reorder(pathway, -FDR.Mixed))) +
  geom_col(aes(fill = coef, alpha = FDR.Mixed &lt; 0.05), position = &quot;dodge&quot;) +
  scale_fill_viridis_d(end = 0.75) +
  scale_alpha_manual(values = c(0.5,1)) +
  theme_pubr() +
  geom_vline(xintercept = 2.3) +
  labs(fill = &quot;FDR &lt; 0.05&quot;, 
       x = &quot;-log10(PValue)&quot;, 
       y = &quot;&quot;)</code></pre>
<p><img src="figure/_BCAC.rmd/unnamed-chunk-19-3.png" width="75%" height="75%" style="display: block; margin: auto;" /></p>
<pre class="r"><code>  #ggsave(&quot;output/plots/AC_KEGG.png&quot;, width = 20, height = 20, units = &quot;cm&quot;, dpi = 400, scale = 2)</code></pre>
<div id="go" class="section level3">
<h3>GO</h3>
<pre class="r"><code>fry_GO &lt;-  design.ac %&gt;% colnames() %&gt;% .[2:3] %&gt;% 
  sapply(function(y) {
    logCPM.ac %&gt;% 
      fry(
        index = GO,
        design = design.ac, 
        contrast = y, 
        sort = &quot;mixed&quot;
      ) %&gt;% 
      rownames_to_column(&quot;pathway&quot;) %&gt;% 
      as_tibble() %&gt;% 
      mutate(coef = y, 
             sig = FDR.Mixed &lt; 0.05)
  }, simplify = FALSE)</code></pre>
<pre class="r"><code>toptables_cqn.ac %&gt;% 
  bind_rows() %&gt;% 
  dplyr::filter(gene_id %in% KEGG$KEGG_LYSOSOME) %&gt;% 
  dplyr::select(gene_name, logFC, coef) %&gt;% 
  dplyr::distinct(gene_name, coef, .keep_all = T) %&gt;% 
  spread(key = &quot;coef&quot;, value = &quot;logFC&quot;) %&gt;% 
  column_to_rownames(&quot;gene_name&quot;) %&gt;% 
  t() %&gt;% 
  pheatmap(color = colorRampPalette(rev(brewer.pal(n = 5,
                                               name = &quot;RdBu&quot;)))(100), 
           main = &quot;KEGG_LYSOSOME: FDR = 6.44e-5&quot;, 
           breaks = c(seq(min(.), 0, length.out=ceiling(100/2) + 1), 
              seq(max(.)/100, max(.), length.out=50)), 
           cellheight = 30, cellwidth = 12,angle_col = 45,
           treeheight_row = 0, treeheight_col = 0
           )</code></pre>
<p><img src="figure/_BCAC.rmd/unnamed-chunk-21-1.png" width="75%" height="75%" style="display: block; margin: auto;" /> ### IRE</p>
<pre class="r"><code>fryIRE.ac &lt;- design.ac %&gt;% colnames() %&gt;% .[2:3] %&gt;% 
  sapply(function(y) {
    logCPM.ac %&gt;% 
      fry(
        index = ireGenes,
        design = design.ac, 
        contrast = y, 
        sort = &quot;directional&quot;
      ) %&gt;% 
      rownames_to_column(&quot;pathway&quot;) %&gt;% 
      as_tibble() %&gt;% 
      mutate(coef = y)
  }, simplify = FALSE)

fryIRE.ac %&gt;% 
  bind_rows() %&gt;% 
  dplyr::select(pathway, coef, pvalue = PValue.Mixed, FDR = FDR.Mixed) %&gt;% 
  ggplot(aes(y = pathway, x = -log10(pvalue))) + 
  geom_col(aes(fill = coef, alpha = FDR &lt; 0.05), 
           position = &quot;dodge&quot;) +
  scale_fill_viridis_d(end = 0.75) +
  geom_vline(xintercept = 1.2) +
  scale_alpha_manual(values = c(0.5,1)) +
  annotate(geom = &quot;label&quot;, x = 1.1, y = &quot;ire5_HQ&quot;, label = &quot;FDR = 0.05&quot;) </code></pre>
<p><img src="figure/_BCAC.rmd/unnamed-chunk-22-1.png" width="75%" height="75%" style="display: block; margin: auto;" /></p>
<pre class="r"><code>  #ggsave(&quot;output/plots/AC_ire.png&quot;, width = 20, height = 10, units = &quot;cm&quot;, dpi = 400, scale = 2)


toptables_cqn.ac %&gt;% 
  bind_rows() %&gt;% 
    dplyr::filter(gene_id %in% ireGenes$ire3_all) %&gt;% 
  dplyr::select(gene_name, logFC, coef) %&gt;% 
  dplyr::distinct(gene_name, coef, .keep_all = T) %&gt;% 
  spread(key = &quot;coef&quot;, value = &quot;logFC&quot;) %&gt;% 
  column_to_rownames(&quot;gene_name&quot;) %&gt;% 
  t() %&gt;% 
  pheatmap(color = colorRampPalette(rev(brewer.pal(n = 5,
                                               name = &quot;RdBu&quot;)))(100), 
           main = &quot;3&#39; IRE genes: FDR = 0.066&quot;, 
           breaks = c(seq(min(.), 0, length.out=ceiling(100/2) + 1), 
              seq(max(.)/100, max(.), length.out=50)), 
           # cellheight = 30, 
           # cellwidth = 1,
           angle_col = 45,
           treeheight_row = 0, 
           treeheight_col = 0, 
           show_colnames = F
           )</code></pre>
<p><img src="figure/_BCAC.rmd/unnamed-chunk-22-2.png" width="75%" height="75%" style="display: block; margin: auto;" /></p>
</div>
<div id="cell-type-prop-check" class="section level3">
<h3>Cell type prop check</h3>
<p>Since this experiment was performed on whole brains, we need to have a look at whether changes to cell type proportions could explain cause some artefactual changes to gene expression. We have done this previously by assessing the expression of marker genes of specific cell types. So this is what I will do here. I have taken the lists of genes frmo day 5 only from the zebrafish single cell atlas (Farnsworth et al. 2020) which contain at least 10 genes. This leaves 4 gene sets (cell types) to test.</p>
<pre class="r"><code>celltype.ac &lt;- design.ac %&gt;% colnames() %&gt;% .[2:3] %&gt;% 
  sapply(function(y) {
    logCPM.ac %&gt;% 
      fry(
        index = cell_type_markers,
        design = design.ac, 
        contrast = y, 
        sort = &quot;directional&quot;
      ) %&gt;% 
      rownames_to_column(&quot;pathway&quot;) %&gt;% 
      as_tibble() %&gt;% 
      mutate(coef = y)
  }, simplify = FALSE)

celltype.ac %&gt;% 
  bind_rows() %&gt;% 
  ggplot(aes(y = pathway, x = -log10(PValue))) + 
  geom_col(aes(fill = coef, alpha = FDR &lt; 0.05), 
           position = &quot;dodge&quot;) +
  scale_fill_viridis_d(end = 0.75) +
  geom_vline(xintercept = 1.8) +
  annotate(geom = &quot;label&quot;, x = 2, y = 4.5, label = &quot;FDR = 0.05&quot;) </code></pre>
<p><img src="figure/_BCAC.rmd/unnamed-chunk-23-1.png" width="75%" height="75%" style="display: block; margin: auto;" /></p>
<p>Statistical significance of the oligodendrocyte gene set is observed in both models.</p>
<pre class="r"><code>toptables_cqn.ac %&gt;% 
  bind_rows() %&gt;% 
  dplyr::filter(gene_id %in% cell_type_markers$`Oligodendrocyte-enriched`) %&gt;% 
  dplyr::select(gene_name, logFC, coef) %&gt;% 
  dplyr::filter(coef == &quot;MPS-IIIA&quot;) %&gt;% 
  dplyr::distinct(gene_name, coef, .keep_all = T) %&gt;% 
  spread(key = &quot;coef&quot;, value = &quot;logFC&quot;) %&gt;% 
  column_to_rownames(&quot;gene_name&quot;) %&gt;% 
  arrange(desc(`MPS-IIIA`)) %&gt;% 
  t() %&gt;% 
  pheatmap(color = colorRampPalette(rev(brewer.pal(n = 5,
                                               name = &quot;RdBu&quot;)))(100), 
           main = &quot;Oligodendrocytes in MPS-IIIA&quot;, 
           breaks = c(seq(min(.), 0, length.out=ceiling(100/2) + 1), 
              seq(max(.)/100, max(.), length.out=50)), 
           cellheight = 30, cellwidth = 12,
           angle_col = 45,
           cluster_rows = F,
           cluster_cols = F,
           treeheight_row = 0, treeheight_col = 0
           )</code></pre>
<p><img src="figure/_BCAC.rmd/unnamed-chunk-24-1.png" width="75%" height="75%" style="display: block; margin: auto;" /></p>
<pre class="r"><code>toptables_cqn.ac %&gt;% 
  bind_rows() %&gt;% 
  dplyr::filter(gene_id %in% cell_type_markers$`Oligodendrocyte-enriched`) %&gt;% 
  dplyr::select(gene_name, logFC, coef) %&gt;% 
  dplyr::filter(coef == &quot;MPS-IIIC&quot;) %&gt;% 
  dplyr::distinct(gene_name, coef, .keep_all = T) %&gt;% 
  spread(key = &quot;coef&quot;, value = &quot;logFC&quot;) %&gt;% 
  column_to_rownames(&quot;gene_name&quot;) %&gt;% 
  arrange(desc(`MPS-IIIC`)) %&gt;% 
  t() %&gt;% 
  pheatmap(color = colorRampPalette(rev(brewer.pal(n = 5,
                                               name = &quot;RdBu&quot;)))(100), 
           main = &quot;Oligodendrocytes in MPS-IIIC&quot;, 
           breaks = c(seq(min(.), 0, length.out=ceiling(100/2) + 1), 
              seq(max(.)/100, max(.), length.out=50)), 
           cellheight = 30, cellwidth = 12,
           angle_col = 45,
           cluster_rows = F,
           cluster_cols = F,
           treeheight_row = 0, treeheight_col = 0
           )</code></pre>
<p><img src="figure/_BCAC.rmd/unnamed-chunk-24-2.png" width="75%" height="75%" style="display: block; margin: auto;" /></p>
</div>
<div id="chrosomosomes-gene-sets" class="section level3">
<h3>Chrosomosomes gene sets</h3>
<pre class="r"><code>fry_chr.ac &lt;- design.ac %&gt;% colnames() %&gt;% .[2:3] %&gt;% 
  sapply(function(y) {
    logCPM.ac %&gt;% 
      fry(
        index = c(chr),
        design = design.ac, 
        contrast = y, 
        sort = &quot;mixed&quot;
      ) %&gt;% 
      rownames_to_column(&quot;pathway&quot;) %&gt;% 
      as_tibble() %&gt;% 
      mutate(coef = y)
  }, simplify = FALSE)


fry_chr.ac %&gt;% 
  bind_rows() %&gt;% 
  mutate(pathway = factor(pathway, levels = c(&quot;MT&quot;, seq(1:25)))) %&gt;% 
  ggplot(aes(y = pathway, x = -log10(PValue.Mixed), fill = coef)) +
  geom_col(position = &quot;dodge&quot;)</code></pre>
<p><img src="figure/_BCAC.rmd/unnamed-chunk-25-1.png" width="75%" height="75%" style="display: block; margin: auto;" /></p>
<p>An enrichment of DE genes is observed on the same chromosome as the mutation. This is obsevred in the manhat plots below.</p>
<pre class="r"><code>manhat_input &lt;- 
    toptables_cqn.ac %&gt;% 
  bind_rows() %&gt;% 
    left_join(x$genes %&gt;% 
                as_tibble() %&gt;% 
                dplyr::select(start, end, chromosome, gene_id) %&gt;% 
                group_by(gene_id) %&gt;% 
                mutate(mid = mean(c(start, end)))
    ) %&gt;% 
  dplyr::filter(chromosome %in% seq(1:25)) %&gt;% 
  group_by(gene_id) %&gt;% 
  mutate(mid = mean(c(start, end))) %&gt;% 
  ungroup %&gt;% 
  mutate(chromosome = factor(chromosome, levels = seq(1:25)))  %&gt;% 
  group_by(chromosome) %&gt;% 
  summarise(chrLen = max(mid)) %&gt;% 
  mutate(chrSt = cumsum(chrLen)-chrLen) %&gt;% 
    dplyr::select(-chrLen) %&gt;% 
  left_join(toptables_cqn.ac %&gt;% 
              bind_rows() %&gt;% 
              left_join(x$genes %&gt;% 
                          as_tibble() %&gt;% 
                          dplyr::select(start, end, chromosome, gene_id) %&gt;% 
                          group_by(gene_id) %&gt;% 
                          mutate(mid = mean(c(start, end)))
              )
  ) %&gt;% 
  group_by(gene_id) %&gt;% 
  mutate(mid = mean(c(start, end))) %&gt;% 
  ungroup %&gt;% 
  dplyr::arrange(mid, chromosome) %&gt;% 
  mutate(midCum = chrSt + mid)

axis &lt;- manhat_input %&gt;%
  group_by(chromosome) %&gt;%
  dplyr::arrange(chromosome) %&gt;%
  summarize(center = (max(midCum) + min(midCum)) / 2) %&gt;%
  mutate(
    seqnames = as.numeric(chromosome),
    colour = rep(c(&quot;grey40&quot;, &quot;black&quot;), length.out = 25)
  )

ggplot(manhat_input, aes(x = midCum, y = -log10(PValue))) +
  geom_point(aes(color=chromosome), alpha = 0.5, size = 1) +
  geom_point(alpha = 0.5, size = 1, colour = &quot;red&quot;, 
             data = . %&gt;% 
               dplyr::filter(coef == &quot;MPS-IIIC&quot; &amp; chromosome == 1)) +
  geom_point(alpha = 0.5, size = 1, colour = &quot;red&quot;, 
             data = . %&gt;% 
               dplyr::filter(coef ==  &quot;MPS-IIIA&quot; &amp; chromosome == 22)) +
  scale_color_manual(values = axis$colour) +
  scale_x_continuous(label = axis$chromosome, breaks = axis$center) +
  scale_y_continuous(limits =c(0, 10)) + # zoom in
  facet_wrap(~coef, ncol = 1) +
  labs(x = &quot;Chromosome&quot;, y = expression(paste(-log[10], &quot;(p)&quot;))) +
  theme( 
    legend.position=&quot;none&quot;,
    panel.border = element_blank(),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.x = element_blank()
  ) </code></pre>
<p><img src="figure/_BCAC.rmd/unnamed-chunk-26-1.png" width="75%" height="75%" style="display: block; margin: auto;" /></p>
</div>
</div>
</div>
<div id="bc" class="section level1">
<h1>BC</h1>
<pre class="r"><code>x.bc &lt;- x[,BCsamps]

design.bc &lt;- model.matrix(~genotype + Tank, data = x.bc$samples %&gt;% 
                            droplevels()) %&gt;% # get rid of the unused factor levels
  set_colnames(gsub(colnames(.), pattern = &quot;genotype&quot;, replacement = &quot;&quot;))

fit_1_bc &lt;- x.bc %&gt;% 
  estimateDisp(design.bc) %&gt;% 
  glmFit(design.bc)

toptable_1.bc &lt;- design.bc %&gt;% colnames() %&gt;% .[2:3] %&gt;% 
  sapply(function(x){
    glmLRT(fit_1_bc, coef = x) %&gt;%
      topTags(n = Inf) %&gt;%
      .[[&quot;table&quot;]] %&gt;%
      as_tibble() %&gt;%
      arrange(PValue) %&gt;%
       mutate(
        coef = x,
        DE = FDR &lt; 0.05
      ) %&gt;% 
      dplyr::select(
        gene_name, logFC, logCPM, PValue, FDR, DE, everything()  
      )
      
  }, simplify = FALSE)</code></pre>
<div id="vis-1" class="section level3">
<h3>vis</h3>
</div>
<div id="number-of-de-genes-1" class="section level3">
<h3>Number of DE genes</h3>
<pre class="r"><code>toptable_1.bc %&gt;% 
  lapply(function(y) {
    y %&gt;% 
      dplyr::filter(DE == T) %&gt;% 
      nrow()
  }) %&gt;% 
  bind_rows() %&gt;% 
  as_tibble() %&gt;% 
  gather(key = &quot;coef&quot;, value = &quot;numDE&quot;) %&gt;% 
  ggplot(aes(x = numDE, y = coef)) +
  geom_col(aes(fill = coef)) +
  geom_text(aes(label = numDE)) +
  ggtitle(&quot;Number of DE genes in each comparison&quot;,
          subtitle = &quot;The MPS-IIIA zebrafish samples have the most DE genes&quot;)</code></pre>
<p><img src="figure/_BCAC.rmd/unnamed-chunk-28-1.png" width="75%" height="75%" style="display: block; margin: auto;" /></p>
</div>
<div id="volcano-plots-2" class="section level3">
<h3>Volcano plots</h3>
<p>The volcano plot below is zoomed in for vis purposes.</p>
<pre class="r"><code>toptable_1.bc %&gt;%
  bind_rows() %&gt;% 
  ggplot(aes(y = -log10(PValue), x = logFC, colour = DE)) +
  geom_point(
    alpha = 0.5, size = 1.25
  ) +
  facet_wrap(~coef) +
  # geom_label_repel(
  #   aes(label = gene_name),
  #   data = .  %&gt;% dplyr::filter(FDR &lt; 0.05),
  #   show.legend = FALSE
  #   ) +
  coord_cartesian(xlim = c(-2.5,2.5),
                  ylim = c(0, 20) )+
  theme(legend.position = &quot;none&quot;) +
  scale_color_manual(values = c(&quot;grey50&quot;, &quot;red&quot;)) </code></pre>
<p><img src="figure/_BCAC.rmd/unnamed-chunk-29-1.png" width="75%" height="75%" style="display: block; margin: auto;" /></p>
</div>
<div id="md-plots-2" class="section level3">
<h3>MD plots</h3>
<pre class="r"><code>toptable_1.bc %&gt;%
  bind_rows() %&gt;% 
  ggplot(aes(x = logCPM, y = logFC)) +
  geom_point(
    aes(colour = DE),
    alpha = 0.5
  ) +
  facet_wrap(~coef, ncol = 1) +
  theme(legend.position = &quot;none&quot;) +
  scale_color_manual(values = c(&quot;grey50&quot;, &quot;red&quot;))</code></pre>
<p><img src="figure/_BCAC.rmd/unnamed-chunk-30-1.png" width="75%" height="75%" style="display: block; margin: auto;" /></p>
</div>
<div id="gc-and-length-check-1" class="section level3">
<h3>%GC AND LENGTH CHECK</h3>
<pre class="r"><code>ggarrange(
toptable_1.bc %&gt;% 
  bind_rows() %&gt;% 
  mutate(rankstat = sign(logFC)*-log10(PValue)) %&gt;% 
  ggplot(aes(x = length, y = rankstat)) +
  geom_point(
    aes(colour = DE),
    alpha = 0.5
  ) +
  geom_smooth(se = FALSE, method = &quot;gam&quot;) +
  facet_grid(rows = vars(coef)) +
  theme_bw() +
  theme(legend.position = &quot;none&quot;) +
  scale_color_manual(values = c(&quot;grey50&quot;, &quot;red&quot;)) +
  scale_x_log10()+
  labs(x = &quot;Average transcript length per gene&quot;,
       colour = &quot;Differentially expressed?&quot;,
       y = &quot;sign(logFC)*-log10(PValue)&quot;) +
  coord_cartesian(ylim = c(-10, 10)),

toptable_1.bc %&gt;% 
  bind_rows() %&gt;% 
  mutate(rankstat = sign(logFC)*-log10(PValue)) %&gt;% 
  ggplot(aes(x = gc_content, y = rankstat)) +
  geom_point(
    aes(colour = DE),
    alpha = 0.5
  ) +
  geom_smooth(se = FALSE, method = &quot;gam&quot;) +
  facet_grid(rows = vars(coef)) +
  theme_bw() +
  theme(legend.position = &quot;none&quot;) +
  scale_color_manual(values = c(&quot;grey50&quot;, &quot;red&quot;)) +
  coord_cartesian(ylim = c(-10,10)) + 
  labs(x = &quot;Weighted average GC content (%) per gene&quot;, 
       colour = &quot;Differentially expressed?&quot;, 
       y = &quot;sign(logFC)*-log10(PValue)&quot;),
common.legend = TRUE, 
labels = &quot;AUTO&quot;
) </code></pre>
<p><img src="figure/_BCAC.rmd/unnamed-chunk-31-1.png" width="75%" height="75%" style="display: block; margin: auto;" /></p>
</div>
<div id="cqn-1" class="section level2">
<h2>CQN</h2>
<p>A bit of bias is obsevred for %GC content and gene length, so will apply the <code>cqn</code> correction to adjust for this.</p>
<pre class="r"><code>cqn.bc &lt;-   
x.bc %&gt;% 
  with(
    cqn(
      counts = counts,
      x = genes$gc_content,
      lengths = genes$length,
      sizeFactors = samples$lib.size
    )
  )
logCPM.bc &lt;- cqn.bc$y + cqn.bc$offset</code></pre>
</div>
<div id="plot-out-the-model-fits-1" class="section level2">
<h2>plot out the model fits</h2>
<pre class="r"><code>par(mfrow = c(1, 2))
cqnplot(cqn.bc, n = 1, xlab = &quot;GC Content&quot;)
cqnplot(cqn.bc, n = 2, xlab = &quot;Length&quot;)</code></pre>
<div class="figure" style="text-align: center">
<img src="figure/_BCAC.rmd/plotCQN-1.png" alt="*Model fits for GC content and gene length under the CQN model. Variability is clearly visible at either end*" width="75%" height="75%" />
<p class="caption">
<em>Model fits for GC content and gene length under the CQN model. Variability is clearly visible at either end</em>
</p>
</div>
<pre class="r"><code>par(mfrow = c(1, 1))</code></pre>
</div>
<div id="pca-2" class="section level2">
<h2>PCA</h2>
<p>PCA analysis was repeated. Similar clustering of samples was observed across PC1. However, samples appeared closer together.</p>
<pre class="r"><code>ggarrange(
  cpm(x.bc, log = T) %&gt;%
    t() %&gt;%
    prcomp() %&gt;% 
    autoplot(data = tibble(sample = rownames(.$x)) %&gt;%
               left_join(x$samples),
             colour = &quot;genotype&quot;, 
             size = 4
    ) +
    scale_color_viridis_d(option = &quot;turbo&quot;, end = 0.8) +
    ggtitle(&quot;Before CQN&quot;),
  
  logCPM.bc %&gt;% 
    t() %&gt;% 
    prcomp() %&gt;%
    autoplot(data = tibble(sample = rownames(.$x)) %&gt;%
               left_join(x$samples),
             colour = &quot;genotype&quot;, 
             size = 4
    ) +
    scale_color_viridis_d(option = &quot;turbo&quot;, end = 0.8) +
    ggtitle(&quot;After CQN&quot;),
  common.legend = T, 
  labels = &quot;AUTO&quot;
) </code></pre>
<p><img src="figure/_BCAC.rmd/unnamed-chunk-33-1.png" width="75%" height="75%" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="de-with-cqn-1" class="section level1">
<h1>DE with cqn</h1>
<pre class="r"><code># Add the offset to the dge object 
x.bc$offset &lt;- cqn.bc$glm.offset

# fit the glm
fit_cqn.bc &lt;- x.bc %&gt;% 
  estimateDisp(design.bc) %&gt;% 
  glmFit(design.bc)

toptables_cqn.bc &lt;- design.bc %&gt;% colnames() %&gt;% .[2:3] %&gt;% 
  sapply(function(x){
    glmLRT(fit_cqn.bc, coef = x) %&gt;%
      topTags(n = Inf) %&gt;%
      .[[&quot;table&quot;]] %&gt;%
      as_tibble() %&gt;%
      arrange(PValue) %&gt;%
       mutate(
        coef = x,
        DE = FDR &lt; 0.05
      ) %&gt;% 
      dplyr::select(
        gene_name, logFC, logCPM, PValue, FDR, DE, everything()  
      )
      
  }, simplify = FALSE)</code></pre>
<div id="vis-with-cqn-1" class="section level2">
<h2>Vis with CQN</h2>
<div id="num-de-genes-1" class="section level3">
<h3>Num DE genes</h3>
<pre class="r"><code>toptables_cqn.bc %&gt;% 
  lapply(function(x){
    x %&gt;%
      mutate(Direction = case_when(
        sign(logFC) == &quot;1&quot; ~ &quot;up&quot;,
        sign(logFC) == &quot;-1&quot; ~ &quot;down&quot;
      ))
  }) %&gt;%
  bind_rows() %&gt;%
  dplyr::filter(DE == TRUE) %&gt;%
  ggplot(aes(y = `coef`)) +
  geom_bar(stat = &quot;count&quot;,
           width = 0.5,
           aes(fill = Direction),
           position = &quot;dodge&quot;) +
  theme(legend.position = &quot;bottom&quot;) +
  ggtitle(&quot;Number of DE genes in each comparison&quot;) +
  labs(x = &quot;Number of DE genes&quot;,
       y = &quot;Contrast&quot;,
       colour = &quot;Direction of change&quot;) </code></pre>
<p><img src="figure/_BCAC.rmd/unnamed-chunk-35-1.png" width="75%" height="75%" style="display: block; margin: auto;" /></p>
</div>
<div id="volcano-plots-3" class="section level3">
<h3>Volcano plots</h3>
<p>These are zoomed in again for vis purposes</p>
<pre class="r"><code>toptables_cqn.bc %&gt;%
  bind_rows() %&gt;% 
  ggplot(aes(y = -log10(PValue), x = logFC, colour = DE)) +
  geom_point(
    alpha = 0.5, size = 1.25
  ) +
  facet_wrap(~coef) +
  # geom_label_repel(
  #   aes(label = gene_name),
  #   data = .  %&gt;% dplyr::filter(FDR &lt; 0.05),
  #   show.legend = FALSE
  #   ) +
  coord_cartesian(xlim = c(-2.5,2.5),
                  ylim = c(0, 20) )+
  theme(legend.position = &quot;none&quot;) +
  scale_color_manual(values = c(&quot;grey50&quot;, &quot;red&quot;)) </code></pre>
<p><img src="figure/_BCAC.rmd/unnamed-chunk-36-1.png" width="75%" height="75%" style="display: block; margin: auto;" /></p>
</div>
<div id="md-plots-3" class="section level3">
<h3>MD plots</h3>
<pre class="r"><code>toptables_cqn.bc %&gt;% 
  bind_rows() %&gt;% 
  ggplot(aes(x = logCPM, y = logFC)) +
  geom_point(
    aes(colour = DE),
    alpha = 0.5
  ) +
  facet_wrap(~coef, ncol = 1) +
  theme(legend.position = &quot;none&quot;) +
  scale_color_manual(values = c(&quot;grey50&quot;, &quot;red&quot;))</code></pre>
<p><img src="figure/_BCAC.rmd/unnamed-chunk-37-1.png" width="75%" height="75%" style="display: block; margin: auto;" /></p>
</div>
<div id="recheck-the-gc-and-length-bias-1" class="section level3">
<h3>Recheck the %GC and length bias</h3>
<pre class="r"><code>ggarrange(
toptables_cqn.bc %&gt;% 
  bind_rows() %&gt;% 
  mutate(rankstat = sign(logFC)*-log10(PValue)) %&gt;% 
  ggplot(aes(x = length, y = rankstat)) +
  geom_point(
    aes(colour = DE),
    alpha = 0.5
  ) +
  geom_smooth(se = FALSE, method = &quot;gam&quot;) +
  facet_grid(rows = vars(coef)) +
  theme_bw() +
  theme(legend.position = &quot;none&quot;) +
  scale_color_manual(values = c(&quot;grey50&quot;, &quot;red&quot;)) +
  scale_x_log10()+
  labs(x = &quot;Average transcript length per gene&quot;,
       colour = &quot;Differentially expressed?&quot;,
       y = &quot;sign(logFC)*-log10(PValue)&quot;) +
  coord_cartesian(ylim = c(-10, 10)),

toptables_cqn.bc %&gt;% 
  bind_rows() %&gt;% 
  mutate(rankstat = sign(logFC)*-log10(PValue)) %&gt;% 
  ggplot(aes(x = gc_content, y = rankstat)) +
  geom_point(
    aes(colour = DE),
    alpha = 0.5
  ) +
  geom_smooth(se = FALSE, method = &quot;gam&quot;) +
  facet_grid(rows = vars(coef)) +
  theme_bw() +
  theme(legend.position = &quot;none&quot;) +
  scale_color_manual(values = c(&quot;grey50&quot;, &quot;red&quot;)) +
  coord_cartesian(ylim = c(-10,10)) + 
  labs(x = &quot;Weighted average GC content (%) per gene&quot;, 
       colour = &quot;Differentially expressed?&quot;, 
       y = &quot;sign(logFC)*-log10(PValue)&quot;),
common.legend = TRUE, 
labels = &quot;AUTO&quot;
) </code></pre>
<p><img src="figure/_BCAC.rmd/unnamed-chunk-38-1.png" width="75%" height="75%" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="gene-set-enrichment-analysis-1" class="section level2">
<h2>Gene set enrichment analysis</h2>
<div id="gene-sets-1" class="section level3">
<h3>Gene sets</h3>
<pre class="r"><code>KEGG &lt;- msigdbr(&quot;Danio rerio&quot;, category = &quot;C2&quot;, subcategory = &quot;CP:KEGG&quot;) %&gt;% 
  left_join(x$genes, by = c(&quot;gene_symbol&quot; = &quot;gene_name&quot;)) %&gt;% 
  dplyr::filter(gene_id %in% rownames(x)) %&gt;% 
  distinct(gs_name, gene_id, .keep_all = TRUE) %&gt;% 
  split(f = .$gs_name) %&gt;%
  lapply(extract2, &quot;gene_id&quot;)

sizes &lt;- KEGG %&gt;% 
  lapply(length) %&gt;% 
  unlist %&gt;% 
  as.data.frame() %&gt;% 
  set_colnames( &quot;n_genes&quot;) %&gt;% 
  rownames_to_column(&quot;gs&quot;)


# retain gene sets with at least 10 genes in it
KEGG &lt;- KEGG[sizes %&gt;% dplyr::filter(n_genes &gt; 10) %&gt;% .$gs]

# GO Terms
goSummaries &lt;- url(&quot;https://uofabioinformaticshub.github.io/summaries2GO/data/goSummaries.RDS&quot;) %&gt;%
  readRDS() %&gt;%
  mutate(
    Term = Term(id),
    gs_name = Term %&gt;% str_to_upper() %&gt;% str_replace_all(&quot;[ -]&quot;, &quot;_&quot;),
    gs_name = paste0(&quot;GO_&quot;, gs_name)
    )

minPath &lt;- 3

GO &lt;- 
  msigdbr(&quot;Danio rerio&quot;, category = &quot;C5&quot;) %&gt;% 
  dplyr::filter(grepl(gs_name, pattern = &quot;^GO&quot;)) %&gt;% 
  left_join(x$genes, by = c(&quot;gene_symbol&quot; = &quot;gene_name&quot;)) %&gt;% 
  dplyr::filter(gene_id %in% rownames(x)) %&gt;% 
  mutate(gs_name = str_replace(gs_name, pattern = &quot;GOBP_&quot;, replacement = &quot;GO_&quot;)) %&gt;% 
  mutate(gs_name = str_replace(gs_name, pattern = &quot;GOMF_&quot;, replacement = &quot;GO_&quot;)) %&gt;%     
  mutate(gs_name = str_replace(gs_name, pattern = &quot;GOCC_&quot;, replacement = &quot;GO_&quot;)) %&gt;%     
  left_join(goSummaries) %&gt;% 
  dplyr::filter(shortest_path &gt;= minPath) %&gt;%
  distinct(gs_name, gene_id, .keep_all = TRUE) %&gt;% 
  split(f = .$gs_name) %&gt;%
  lapply(extract2, &quot;gene_id&quot;)

GOdf &lt;-  msigdbr(&quot;Danio rerio&quot;, category = &quot;C5&quot;) %&gt;% 
  dplyr::filter(grepl(gs_name, pattern = &quot;^GO&quot;)) %&gt;% 
  left_join(x$genes, by = c(&quot;gene_symbol&quot; = &quot;gene_name&quot;)) %&gt;% 
  dplyr::filter(gene_id %in% rownames(x)) %&gt;% 
  mutate(gs_name = str_replace(gs_name, pattern = &quot;GOBP_&quot;, replacement = &quot;GO_&quot;)) %&gt;% 
  mutate(gs_name = str_replace(gs_name, pattern = &quot;GOMF_&quot;, replacement = &quot;GO_&quot;)) %&gt;%     
  mutate(gs_name = str_replace(gs_name, pattern = &quot;GOCC_&quot;, replacement = &quot;GO_&quot;)) %&gt;%     
  left_join(goSummaries) %&gt;% 
  dplyr::filter(shortest_path &gt;= minPath) %&gt;%
  distinct(gs_name, gene_id, .keep_all = TRUE) %&gt;% 
  dplyr::select(gs_name, gene_id)

sizes_GO &lt;- GO %&gt;%
  lapply(length) %&gt;%
  unlist %&gt;%
  as.data.frame() %&gt;%
  set_colnames( &quot;n_genes&quot;) %&gt;%
  rownames_to_column(&quot;gs&quot;)


# retain gene sets with at least 5 genes in it
GO &lt;- GO[sizes_GO %&gt;% dplyr::filter(n_genes &gt; 5) %&gt;% .$gs]

ireGenes &lt;- 
  read_xlsx(&quot;data/gene_sets/ireGenes.xlsx&quot;, sheet = &quot;Zebrafish IREs&quot;) %&gt;% 
  gather %&gt;% 
  set_colnames(c(&quot;IRE&quot;, &quot;gene_id&quot;)) %&gt;% 
  mutate(IRE = case_when(
    IRE == &quot;All 3&#39; IREs&quot; ~ &quot;ire3_all&quot;, 
    IRE == &quot;All 5&#39; IREs&quot; ~ &quot;ire5_all&quot;, 
    IRE == &quot;HQ 3&#39; IREs&quot; ~ &quot;ire3_HQ&quot;, 
    IRE == &quot;HQ 5&#39; IREs&quot; ~ &quot;ire5_HQ&quot;
  )) %&gt;% 
  dplyr::filter(gene_id %in% rownames(x)) %&gt;% 
  split(f = .$IRE) %&gt;% 
  lapply(magrittr::extract2,&quot;gene_id&quot;) 

chr &lt;- x$genes %&gt;% 
  dplyr::select(gene_id, chromosome) %&gt;% 
  split(f = .$chromosome) %&gt;%
  lapply(extract2, &quot;gene_id&quot;) %&gt;% 
  .[c(&quot;MT&quot;, seq(1:25))] # omit the ALT chromosmes and scaffolds

#biomart to get the conversion between mouse and zebrafish IDs
mart &lt;- useMart(&quot;ensembl&quot;, &quot;drerio_gene_ensembl&quot;)
getFromBiomart &lt;- c(&quot;ensembl_gene_id&quot;, &quot;mmusculus_homolog_associated_gene_name&quot;)
zf_id_2_mm_genename &lt;- getBM(attributes = getFromBiomart,
                             mart = mart) %&gt;% 
  as_tibble() %&gt;% 
  set_colnames(c(&quot;gene_id&quot;, &quot;mouse_gene_name&quot;))

cell_type_markers &lt;- 
  read_xls(&quot;data/gene_sets/cell_type_genes41586_2007_BFnature05453_MOESM11_ESM.xls&quot;) %&gt;%
  as_tibble() %&gt;%
  dplyr::select(c(&quot;Neuron-enriched&quot;, &quot;Astrocyte-enriched&quot;, &quot;Oligodendrocyte-enriched&quot;)) %&gt;%
  gather(key = &quot;cell_type&quot;, value = &quot;mouse_gene_name&quot;) %&gt;% 
  left_join(zf_id_2_mm_genename) %&gt;% 
  na.omit %&gt;% 
  split(f = .$cell_type) %&gt;% 
  lapply(extract2, &quot;gene_id&quot;)

cell_type_markers$`Microglia-enriched` &lt;- 
  read_excel(&quot;data/gene_sets/microglia_oosterofetal.xlsx&quot;, sheet = 1) %&gt;%
  rbind(read_excel(&quot;data/gene_sets/microglia_oosterofetal.xlsx&quot;, sheet = 2)) %&gt;%
  rbind(read_excel(&quot;data/gene_sets/microglia_oosterofetal.xlsx&quot;, sheet = 3)) %&gt;%
  .$Zebrafish.Ensembl.Gene.ID %&gt;% 
  unique

cell_type_markers %&lt;&gt;% 
  lapply(function(y) {
    y[y %in% rownames(x)]
  })</code></pre>
</div>
</div>
<div id="goseq-1" class="section level2">
<h2>GOSEQ</h2>
<pre class="r"><code>pwf &lt;- toptables_cqn.bc$`MPS-IIIB` %&gt;%
      with(
        nullp(
          DEgenes = structure(DE, names = gene_id), 
          bias.data = length
        )
      )</code></pre>
<p><img src="figure/_BCAC.rmd/unnamed-chunk-40-1.png" width="75%" height="75%" style="display: block; margin: auto;" /></p>
<pre class="r"><code>goseq(pwf, gene2cat = GO) %&gt;% 
  as_tibble() %&gt;% 
  dplyr::filter(numDEInCat &gt; 0) %&gt;% 
  mutate(FDR = p.adjust(over_represented_pvalue, method = &quot;fdr&quot;)) %&gt;% 
  dplyr::select(-under_represented_pvalue) %&gt;% 
  dplyr::filter(FDR &lt; 0.05) </code></pre>
<pre><code># A tibble: 9 x 5
  category                        over_represented_… numDEInCat numInCat     FDR
  &lt;chr&gt;                                        &lt;dbl&gt;      &lt;int&gt;    &lt;int&gt;   &lt;dbl&gt;
1 GO_CALCIUM_INDEPENDENT_CELL_CE…          0.0000167          2        7 0.00485
2 GO_LONG_CHAIN_FATTY_ACID_BIOSY…          0.0000525          2       12 0.00763
3 GO_FATTY_ACID_BIOSYNTHETIC_PRO…          0.0000867          3       93 0.00841
4 GO_AXON_ENSHEATHMENT_IN_CENTRA…          0.000124           2       19 0.00901
5 GO_MONOCARBOXYLIC_ACID_BIOSYNT…          0.000207           3      128 0.0121 
6 GO_OLIGODENDROCYTE_DEVELOPMENT           0.000488           2       38 0.0237 
7 GO_ORGANIC_ACID_BIOSYNTHETIC_P…          0.000737           3      198 0.0306 
8 GO_FATTY_ACID_METABOLIC_PROCESS          0.00121            3      240 0.0415 
9 GO_LONG_CHAIN_FATTY_ACID_METAB…          0.00128            2       62 0.0415 </code></pre>
<pre class="r"><code>pwf &lt;- toptables_cqn.bc$`MPS-IIIC` %&gt;%
      with(
        nullp(
          DEgenes = structure(DE, names = gene_id), 
          bias.data = length
        )
      )</code></pre>
<p><img src="figure/_BCAC.rmd/unnamed-chunk-40-2.png" width="75%" height="75%" style="display: block; margin: auto;" /></p>
<pre class="r"><code>goseq(pwf, gene2cat = GO) %&gt;% 
  as_tibble() %&gt;% 
  dplyr::filter(numDEInCat &gt; 0) %&gt;% 
  mutate(FDR = p.adjust(over_represented_pvalue, method = &quot;fdr&quot;)) %&gt;% 
  dplyr::select(-under_represented_pvalue) %&gt;% 
  dplyr::filter(FDR &lt; 0.05) %&gt;% 
  head(10)</code></pre>
<pre><code># A tibble: 10 x 5
   category                        over_represented_… numDEInCat numInCat    FDR
   &lt;chr&gt;                                        &lt;dbl&gt;      &lt;int&gt;    &lt;int&gt;  &lt;dbl&gt;
 1 GO_HYPERSENSITIVITY                        0.00168          1        6 0.0299
 2 GO_ANTIGEN_PROCESSING_AND_PRES…            0.00169          1        6 0.0299
 3 GO_POSITIVE_REGULATION_OF_NATU…            0.00169          1        6 0.0299
 4 GO_MHC_PROTEIN_COMPLEX                     0.00169          1        6 0.0299
 5 GO_POSITIVE_REGULATION_OF_INFL…            0.00196          1        7 0.0299
 6 GO_POSITIVE_REGULATION_OF_INTE…            0.00197          1        7 0.0299
 7 GO_PROTECTION_FROM_NATURAL_KIL…            0.00197          1        7 0.0299
 8 GO_NATURAL_KILLER_CELL_PROLIFE…            0.00197          1        7 0.0299
 9 GO_PRIMARY_MIRNA_BINDING                   0.00224          1        8 0.0299
10 GO_REGULATION_OF_ACUTE_INFLAMM…            0.00224          1        8 0.0299</code></pre>
</div>
<div id="kegg-1" class="section level2">
<h2>KEGG</h2>
<pre class="r"><code>fryKEGG &lt;- 
  design.bc %&gt;% colnames() %&gt;% .[2:3] %&gt;% 
  sapply(function(y) {
    logCPM.bc %&gt;% 
      fry(
        index = KEGG,
        design = design.bc, 
        contrast = y, 
        sort = &quot;mixed&quot;
      ) %&gt;% 
      rownames_to_column(&quot;pathway&quot;) %&gt;% 
      as_tibble() %&gt;% 
      mutate(coef = y)
  }, simplify = FALSE)

fryKEGG$`MPS-IIIB` %&gt;% 
  dplyr::filter(FDR.Mixed &lt; 0.05) %&gt;%  
  mutate(pathway = str_remove(pathway, pattern = &quot;KEGG_&quot;)) %&gt;% 
  ggplot(aes(x = -log10(PValue.Mixed), y = reorder(pathway, -FDR.Mixed))) +
  geom_col(aes(fill = FDR.Mixed &lt; 0.05)) +
  scale_fill_manual(values = c(&quot;grey50&quot;, &#39;red&#39;)) +
  theme_pubr() +
  labs(fill = &quot;FDR &lt; 0.05&quot;, 
       x = &quot;-log10(PValue)&quot;, 
       y = &quot;&quot;)</code></pre>
<p><img src="figure/_BCAC.rmd/unnamed-chunk-41-1.png" width="75%" height="75%" style="display: block; margin: auto;" /></p>
<pre class="r"><code>fryKEGG$`MPS-IIIC` %&gt;% 
  dplyr::filter(FDR.Mixed &lt; 0.05) %&gt;%  
  mutate(pathway = str_remove(pathway, pattern = &quot;KEGG_&quot;)) %&gt;% 
  ggplot(aes(x = -log10(PValue.Mixed), y = reorder(pathway, -FDR.Mixed))) +
  geom_col(aes(fill = FDR.Mixed &lt; 0.05)) +
  scale_fill_manual(values = c(&quot;grey50&quot;, &#39;red&#39;)) +
  theme_pubr() +
  labs(fill = &quot;FDR &lt; 0.05&quot;, 
       x = &quot;-log10(PValue)&quot;, 
       y = &quot;&quot;)  </code></pre>
<p><img src="figure/_BCAC.rmd/unnamed-chunk-41-2.png" width="75%" height="75%" style="display: block; margin: auto;" /></p>
<pre class="r"><code>sigpaths &lt;- fryKEGG %&gt;% 
  bind_rows() %&gt;% 
  dplyr::filter(FDR.Mixed &lt; 0.05) %&gt;% 
    .$pathway %&gt;% 
    unique()
  
fryKEGG %&gt;% 
  bind_rows() %&gt;% 
  dplyr::filter(pathway %in% sigpaths) %&gt;% 
  mutate(pathway = str_remove(pathway, pattern = &quot;KEGG_&quot;)) %&gt;% 
  ggplot(aes(x = -log10(PValue.Mixed), y = reorder(pathway, -FDR.Mixed))) +
  geom_col(aes(fill = coef, alpha = FDR.Mixed &lt; 0.05), position = &quot;dodge&quot;) +
  scale_fill_viridis_d(end = 0.75) +
  scale_alpha_manual(values = c(0.5,1)) +
  theme_pubr() +
  geom_vline(xintercept = 2.3) +
  labs(fill = &quot;FDR &lt; 0.05&quot;, 
       x = &quot;-log10(PValue)&quot;, 
       y = &quot;&quot;)</code></pre>
<p><img src="figure/_BCAC.rmd/unnamed-chunk-41-3.png" width="75%" height="75%" style="display: block; margin: auto;" /></p>
<pre class="r"><code>  #ggsave(&quot;output/plots/AC_KEGG.png&quot;, width = 20, height = 20, units = &quot;cm&quot;, dpi = 400, scale = 2)</code></pre>
<div id="go-1" class="section level3">
<h3>GO</h3>
<pre class="r"><code>fry_GO &lt;-  design.ac %&gt;% colnames() %&gt;% .[2:3] %&gt;% 
  sapply(function(y) {
    logCPM.ac %&gt;% 
      fry(
        index = GO,
        design = design.ac, 
        contrast = y, 
        sort = &quot;mixed&quot;
      ) %&gt;% 
      rownames_to_column(&quot;pathway&quot;) %&gt;% 
      as_tibble() %&gt;% 
      mutate(coef = y, 
             sig = FDR.Mixed &lt; 0.05)
  }, simplify = FALSE)</code></pre>
<pre class="r"><code>toptables_cqn.ac %&gt;% 
  bind_rows() %&gt;% 
  dplyr::filter(gene_id %in% KEGG$KEGG_LYSOSOME) %&gt;% 
  dplyr::select(gene_name, logFC, coef) %&gt;% 
  dplyr::distinct(gene_name, coef, .keep_all = T) %&gt;% 
  spread(key = &quot;coef&quot;, value = &quot;logFC&quot;) %&gt;% 
  column_to_rownames(&quot;gene_name&quot;) %&gt;% 
  t() %&gt;% 
  pheatmap(color = colorRampPalette(rev(brewer.pal(n = 5,
                                               name = &quot;RdBu&quot;)))(100), 
           main = &quot;KEGG_LYSOSOME: FDR = 6.44e-5&quot;, 
           breaks = c(seq(min(.), 0, length.out=ceiling(100/2) + 1), 
              seq(max(.)/100, max(.), length.out=50)), 
           cellheight = 30, cellwidth = 12,angle_col = 45,
           treeheight_row = 0, treeheight_col = 0
           )</code></pre>
<p><img src="figure/_BCAC.rmd/unnamed-chunk-43-1.png" width="75%" height="75%" style="display: block; margin: auto;" /> ### IRE</p>
<pre class="r"><code>fryIRE.bc &lt;- design.bc %&gt;% colnames() %&gt;% .[2:3] %&gt;% 
  sapply(function(y) {
    logCPM.bc %&gt;% 
      fry(
        index = ireGenes,
        design = design.bc, 
        contrast = y, 
        sort = &quot;directional&quot;
      ) %&gt;% 
      rownames_to_column(&quot;pathway&quot;) %&gt;% 
      as_tibble() %&gt;% 
      mutate(coef = y)
  }, simplify = FALSE)

fryIRE.bc %&gt;% 
  bind_rows() %&gt;% 
  dplyr::select(pathway, coef, pvalue = PValue.Mixed, FDR = FDR.Mixed) %&gt;% 
  ggplot(aes(y = pathway, x = -log10(pvalue))) + 
  geom_col(aes(fill = coef, alpha = FDR &lt; 0.05), 
           position = &quot;dodge&quot;) +
  scale_fill_viridis_d(end = 0.75) +
  scale_alpha_manual(values = c(0.5, 1)) +
  geom_vline(xintercept = 1.5) +
  annotate(geom = &quot;label&quot;, x = 1.8, y = &quot;ire5_HQ&quot;, label = &quot;FDR = 0.05&quot;) </code></pre>
<p><img src="figure/_BCAC.rmd/unnamed-chunk-44-1.png" width="75%" height="75%" style="display: block; margin: auto;" /></p>
<pre class="r"><code>  #ggsave(&quot;output/plots/AC_ire.png&quot;, width = 20, height = 10, units = &quot;cm&quot;, dpi = 400, scale = 2)


toptables_cqn.bc %&gt;% 
  bind_rows() %&gt;% 
    dplyr::filter(gene_id %in% ireGenes$ire3_all) %&gt;% 
  dplyr::select(gene_name, logFC, coef) %&gt;% 
  dplyr::distinct(gene_name, coef, .keep_all = T) %&gt;% 
  spread(key = &quot;coef&quot;, value = &quot;logFC&quot;) %&gt;% 
  column_to_rownames(&quot;gene_name&quot;) %&gt;% 
  t() %&gt;% 
  pheatmap(color = colorRampPalette(rev(brewer.pal(n = 5,
                                               name = &quot;RdBu&quot;)))(100), 
           main = &quot;3&#39; IRE genes: FDR = 0.066&quot;, 
           breaks = c(seq(min(.), 0, length.out=ceiling(100/2) + 1), 
              seq(max(.)/100, max(.), length.out=50)), 
           # cellheight = 30, 
           # cellwidth = 1,
           angle_col = 45,
           treeheight_row = 0, 
           treeheight_col = 0, 
           show_colnames = F
           )</code></pre>
<p><img src="figure/_BCAC.rmd/unnamed-chunk-44-2.png" width="75%" height="75%" style="display: block; margin: auto;" /></p>
</div>
<div id="cell-type-prop-check-1" class="section level3">
<h3>Cell type prop check</h3>
<p>Since this experiment was performed on whole brains, we need to have a look at whether changes to cell type proportions could explain cause some artefactual changes to gene expression. We have done this previously by assessing the expression of marker genes of specific cell types. So this is what I will do here. I have taken the lists of genes frmo day 5 only from the zebrafish single cell atlas (Farnsworth et al. 2020) which contain at least 10 genes. This leaves 4 gene sets (cell types) to test.</p>
<pre class="r"><code>celltype.bc &lt;- design.bc %&gt;% colnames() %&gt;% .[2:3] %&gt;% 
  sapply(function(y) {
    logCPM.bc %&gt;% 
      fry(
        index = cell_type_markers,
        design = design.bc, 
        contrast = y, 
        sort = &quot;directional&quot;
      ) %&gt;% 
      rownames_to_column(&quot;pathway&quot;) %&gt;% 
      as_tibble() %&gt;% 
      mutate(coef = y)
  }, simplify = FALSE)

celltype.bc %&gt;% 
  bind_rows() %&gt;% 
  ggplot(aes(y = pathway, x = -log10(PValue))) + 
  geom_col(aes(fill = coef), 
           position = &quot;dodge&quot;) +
  scale_fill_viridis_d(end = 0.75) +
  geom_vline(xintercept = 1.7) +
  annotate(geom = &quot;label&quot;, x = 1.7, y = 4.5, label = &quot;FDR = 0.05&quot;) </code></pre>
<p><img src="figure/_BCAC.rmd/unnamed-chunk-45-1.png" width="75%" height="75%" style="display: block; margin: auto;" /></p>
</div>
<div id="chrosomosomes-gene-sets-1" class="section level3">
<h3>Chrosomosomes gene sets</h3>
<pre class="r"><code>fry_chr.bc &lt;- design.bc %&gt;% colnames() %&gt;% .[2:3] %&gt;% 
  sapply(function(y) {
    logCPM.bc %&gt;% 
      fry(
        index = c(chr),
        design = design.bc, 
        contrast = y, 
        sort = &quot;mixed&quot;
      ) %&gt;% 
      rownames_to_column(&quot;pathway&quot;) %&gt;% 
      as_tibble() %&gt;% 
      mutate(coef = y)
  }, simplify = FALSE)


fry_chr.bc %&gt;% 
  bind_rows() %&gt;% 
  mutate(pathway = factor(pathway, levels = c(&quot;MT&quot;, seq(1:25)))) %&gt;% 
  ggplot(aes(y = pathway, x = -log10(PValue.Mixed), fill = coef)) +
  geom_col(position = &quot;dodge&quot;)</code></pre>
<p><img src="figure/_BCAC.rmd/unnamed-chunk-46-1.png" width="75%" height="75%" style="display: block; margin: auto;" /></p>
<p>An enrichment of DE genes is observed on the same chromosome as the mutation. This is obsevred in the manhat plots below.</p>
<pre class="r"><code>manhat_input.bc &lt;- 
    toptables_cqn.bc %&gt;% 
  bind_rows() %&gt;% 
    left_join(x$genes %&gt;% 
                as_tibble() %&gt;% 
                dplyr::select(start, end, chromosome, gene_id) %&gt;% 
                group_by(gene_id) %&gt;% 
                mutate(mid = mean(c(start, end)))
    ) %&gt;% 
  dplyr::filter(chromosome %in% seq(1:25)) %&gt;% 
  group_by(gene_id) %&gt;% 
  mutate(mid = mean(c(start, end))) %&gt;% 
  ungroup %&gt;% 
  mutate(chromosome = factor(chromosome, levels = seq(1:25)))  %&gt;% 
  group_by(chromosome) %&gt;% 
  summarise(chrLen = max(mid)) %&gt;% 
  mutate(chrSt = cumsum(chrLen)-chrLen) %&gt;% 
    dplyr::select(-chrLen) %&gt;% 
  left_join(toptables_cqn.bc %&gt;% 
              bind_rows() %&gt;% 
              left_join(x$genes %&gt;% 
                          as_tibble() %&gt;% 
                          dplyr::select(start, end, chromosome, gene_id) %&gt;% 
                          group_by(gene_id) %&gt;% 
                          mutate(mid = mean(c(start, end)))
              )
  ) %&gt;% 
  group_by(gene_id) %&gt;% 
  mutate(mid = mean(c(start, end))) %&gt;% 
  ungroup %&gt;% 
  dplyr::arrange(mid, chromosome) %&gt;% 
  mutate(midCum = chrSt + mid)

axis.bc &lt;- manhat_input.bc %&gt;%
  group_by(chromosome) %&gt;%
  dplyr::arrange(chromosome) %&gt;%
  summarize(center = (max(midCum) + min(midCum)) / 2) %&gt;%
  mutate(
    seqnames = as.numeric(chromosome),
    colour = rep(c(&quot;grey40&quot;, &quot;black&quot;), length.out = 25)
  )

ggplot(manhat_input.bc, aes(x = midCum, y = -log10(PValue))) +
  geom_point(aes(color=chromosome), alpha = 0.5, size = 1) +
  geom_point(alpha = 0.5, size = 1, colour = &quot;red&quot;, 
             data = . %&gt;% 
               dplyr::filter(coef == &quot;MPS-IIIC&quot; &amp; chromosome == 1)) +
  geom_point(alpha = 0.5, size = 1, colour = &quot;red&quot;, 
             data = . %&gt;% 
               dplyr::filter(coef ==  &quot;MPS-IIIB&quot; &amp; chromosome == 24)) +
  scale_color_manual(values = axis.bc$colour) +
  scale_x_continuous(label = axis.bc$chromosome, breaks = axis.bc$center) +
  scale_y_continuous(limits =c(0, 10)) + # zoom in
  facet_wrap(~coef, ncol = 1) +
  labs(x = &quot;Chromosome&quot;, y = expression(paste(-log[10], &quot;(p)&quot;))) +
  theme( 
    legend.position=&quot;none&quot;,
    panel.border = element_blank(),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.x = element_blank()
  ) </code></pre>
<p><img src="figure/_BCAC.rmd/unnamed-chunk-47-1.png" width="75%" height="75%" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="data-export" class="section level2">
<h2>Data export</h2>
<pre class="r"><code>x.ac %&gt;% saveRDS(&quot;data/R_objects/dge_ac.rds&quot;)
toptables_cqn.ac %&gt;% saveRDS(&quot;data/R_objects/toptab_ac_cqn.rds&quot;)
fryKEGG %&gt;% saveRDS(&quot;data/R_objects/AC_fryKEGG.rds&quot;)
fryIRE.ac %&gt;% saveRDS(&quot;data/R_objects/AC_fryIRE.rds&quot;)
celltype.ac %&gt;% saveRDS(&quot;data/R_objects/AC_frycellTypes.rds&quot;)



x.bc %&gt;% saveRDS(&quot;data/R_objects/dge_bc.rds&quot;)
toptables_cqn.bc %&gt;% saveRDS(&quot;data/R_objects/toptab_bc_cqn.rds&quot;)
fryKEGG %&gt;% saveRDS(&quot;data/R_objects/BC_fryKEGG.rds&quot;)
fryIRE.bc %&gt;% saveRDS(&quot;data/R_objects/BC_fryIRE.rds&quot;)
celltype.bc %&gt;% saveRDS(&quot;data/R_objects/BC_frycellTypes.rds&quot;)</code></pre>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.0.2 (2020-06-22)
Platform: x86_64-apple-darwin17.0 (64-bit)
Running under: macOS Mojave 10.14.3

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRblas.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRlapack.dylib

locale:
[1] en_AU.UTF-8/en_AU.UTF-8/en_AU.UTF-8/C/en_AU.UTF-8/en_AU.UTF-8

attached base packages:
 [1] stats4    splines   parallel  stats     graphics  grDevices utils    
 [8] datasets  methods   base     

other attached packages:
 [1] ensembldb_2.12.1        AnnotationFilter_1.12.0 GenomicFeatures_1.40.1 
 [4] AnnotationDbi_1.50.3    Biobase_2.48.0          GenomicRanges_1.40.0   
 [7] GenomeInfoDb_1.24.2     IRanges_2.22.2          S4Vectors_0.26.1       
[10] harmonicmeanp_3.0       FMStable_0.1-2          fgsea_1.14.0           
[13] goseq_1.40.0            geneLenDataBase_1.24.0  BiasedUrn_1.07         
[16] cqn_1.34.0              quantreg_5.86           SparseM_1.81           
[19] preprocessCore_1.50.0   nor1mix_1.3-0           mclust_5.4.7           
[22] edgeR_3.30.3            limma_3.44.3            UpSetR_1.4.0           
[25] RColorBrewer_1.1-2      ggfortify_0.4.12        ggrepel_0.9.1          
[28] ggpubr_0.4.0            pheatmap_1.0.12         scales_1.1.1           
[31] pander_0.6.4            msigdbr_7.4.1           biomaRt_2.44.4         
[34] AnnotationHub_2.20.2    BiocFileCache_1.12.1    dbplyr_2.1.1           
[37] ngsReports_1.4.2        BiocGenerics_0.34.0     readxl_1.3.1           
[40] magrittr_2.0.1          forcats_0.5.1           stringr_1.4.0          
[43] dplyr_1.0.7             purrr_0.3.4             readr_1.4.0            
[46] tidyr_1.1.3             tibble_3.1.2            ggplot2_3.3.6          
[49] tidyverse_1.3.1        

loaded via a namespace (and not attached):
  [1] utf8_1.2.1                    tidyselect_1.1.1             
  [3] RSQLite_2.2.7                 htmlwidgets_1.5.3            
  [5] FactoMineR_2.4                grid_4.0.2                   
  [7] BiocParallel_1.22.0           munsell_0.5.0                
  [9] statmod_1.4.36                DT_0.18                      
 [11] withr_2.4.2                   colorspace_2.0-2             
 [13] highr_0.9                     knitr_1.33                   
 [15] rstudioapi_0.13               leaps_3.1                    
 [17] ggsignif_0.6.2                labeling_0.4.2               
 [19] git2r_0.28.0                  GenomeInfoDbData_1.2.3       
 [21] hwriter_1.3.2                 farver_2.1.0                 
 [23] bit64_4.0.5                   rprojroot_2.0.2              
 [25] vctrs_0.3.8                   generics_0.1.0               
 [27] xfun_0.24                     R6_2.5.0                     
 [29] locfit_1.5-9.4                bitops_1.0-7                 
 [31] cachem_1.0.5                  DelayedArray_0.14.1          
 [33] assertthat_0.2.1              promises_1.2.0.1             
 [35] gtable_0.3.0                  conquer_1.0.2                
 [37] workflowr_1.6.2               rlang_0.4.11                 
 [39] MatrixModels_0.5-0            scatterplot3d_0.3-41         
 [41] rtracklayer_1.48.0            rstatix_0.7.0                
 [43] lazyeval_0.2.2                broom_0.7.8                  
 [45] BiocManager_1.30.16           yaml_2.2.1                   
 [47] reshape2_1.4.4                abind_1.4-5                  
 [49] modelr_0.1.8                  backports_1.2.1              
 [51] httpuv_1.6.1                  tools_4.0.2                  
 [53] ellipsis_0.3.2                jquerylib_0.1.4              
 [55] ggdendro_0.1.22               Rcpp_1.0.7                   
 [57] plyr_1.8.6                    progress_1.2.2               
 [59] zlibbioc_1.34.0               RCurl_1.98-1.3               
 [61] prettyunits_1.1.1             openssl_1.4.4                
 [63] cowplot_1.1.1                 zoo_1.8-9                    
 [65] SummarizedExperiment_1.18.2   haven_2.4.1                  
 [67] cluster_2.1.2                 fs_1.5.0                     
 [69] data.table_1.14.0             openxlsx_4.2.4               
 [71] reprex_2.0.0                  truncnorm_1.0-8              
 [73] ProtGenerics_1.20.0           matrixStats_0.59.0           
 [75] hms_1.1.0                     mime_0.11                    
 [77] evaluate_0.14                 xtable_1.8-4                 
 [79] XML_3.99-0.6                  rio_0.5.27                   
 [81] jpeg_0.1-8.1                  gridExtra_2.3                
 [83] compiler_4.0.2                crayon_1.4.1                 
 [85] htmltools_0.5.1.1             mgcv_1.8-36                  
 [87] later_1.2.0                   lubridate_1.7.10             
 [89] DBI_1.1.1                     MASS_7.3-54                  
 [91] rappdirs_0.3.3                babelgene_21.4               
 [93] ShortRead_1.46.0              Matrix_1.3-4                 
 [95] car_3.0-11                    cli_3.0.0                    
 [97] pkgconfig_2.0.3               flashClust_1.01-2            
 [99] GenomicAlignments_1.24.0      foreign_0.8-81               
[101] plotly_4.9.4.1                xml2_1.3.2                   
[103] bslib_0.2.5.1                 XVector_0.28.0               
[105] rvest_1.0.0                   digest_0.6.27                
[107] Biostrings_2.56.0             fastmatch_1.1-0              
[109] rmarkdown_2.9                 cellranger_1.1.0             
[111] curl_4.3.2                    shiny_1.6.0                  
[113] Rsamtools_2.4.0               nlme_3.1-152                 
[115] lifecycle_1.0.0               jsonlite_1.7.2               
[117] carData_3.0-4                 viridisLite_0.4.0            
[119] askpass_1.1                   fansi_0.5.0                  
[121] pillar_1.6.1                  lattice_0.20-44              
[123] GO.db_3.11.4                  fastmap_1.1.0                
[125] httr_1.4.2                    interactiveDisplayBase_1.26.3
[127] glue_1.4.2                    zip_2.2.0                    
[129] png_0.1-7                     BiocVersion_3.11.1           
[131] bit_4.0.4                     stringi_1.6.2                
[133] sass_0.4.0                    blob_1.2.1                   
[135] latticeExtra_0.6-29           memoise_2.0.0                </code></pre>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
