<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Lachlan Baer" />


<title>DAR Analysis</title>

<script src="site_libs/header-attrs-2.20/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<script src="site_libs/kePrint-0.0.1/kePrint.js"></script>
<link href="site_libs/lightable-0.0.1/lightable.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">2022_MPSIII_3mBrainRNAseq</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="QC.html">QC</a>
</li>
<li>
  <a href="analysis_final.html">Initial analysis</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    DAR analysis
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="dar-analysis_AB.html">MPS-IIIA/B</a>
    </li>
    <li>
      <a href="dar-analysis_AC.html">MPS-IIIA/C</a>
    </li>
    <li>
      <a href="dar-analysis_BC.html">MPS-IIIB/C</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">DAR Analysis</h1>
<h3 class="subtitle">2022_MPSIII_3mBrainRNAseq dataset - AB
comparison</h3>
<h4 class="author">Lachlan Baer</h4>
<h4 class="date">16 March, 2023</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2023-03-16
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 6
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 1
</p>
<p>
<strong>Knit directory:</strong> <code>2022_MPSIII_3mBrainRNAseq/</code>
<span class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.0). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguncommittedchanges">
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> <strong>R Markdown file:</strong> uncommitted
changes </a>
</p>
</div>
<div id="strongRMarkdownfilestronguncommittedchanges"
class="panel-collapse collapse">
<div class="panel-body">
<p>The R Markdown is untracked by Git. To know which version of the R
Markdown file created these results, you’ll want to first commit it to
the Git repo. If you’re still working on the analysis, you can ignore
this warning. When you’re finished, you can run
<code>wflow_publish</code> to commit the R Markdown file and build the
HTML.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20220423code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20220423)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20220423code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20220423)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomkarissab2022MPSIII3mBrainRNAseqtree794a3ed47155f2dad02d8187edf533b4d1eb4dectargetblank794a3eda">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/karissa-b/2022_MPSIII_3mBrainRNAseq/tree/794a3ed47155f2dad02d8187edf533b4d1eb4dec" target="_blank">794a3ed</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomkarissab2022MPSIII3mBrainRNAseqtree794a3ed47155f2dad02d8187edf533b4d1eb4dectargetblank794a3eda"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/karissa-b/2022_MPSIII_3mBrainRNAseq/tree/794a3ed47155f2dad02d8187edf533b4d1eb4dec" target="_blank">794a3ed</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rproj.user/

Untracked files:
    Untracked:  analysis/dar-analysis_AB.Rmd
    Untracked:  data/dar-analysis/AB/

Unstaged changes:
    Modified:   analysis/dar-analysis_BC.Rmd

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
There are no past versions. Publish this analysis with
<code>wflow_publish()</code> to start tracking its development.
</p>
<hr>
</div>
</div>
</div>
<div id="setup" class="section level1">
<h1>Setup</h1>
<p>Load libraries and set global options.</p>
<pre class="r"><code>suppressPackageStartupMessages({
  ## Common
  library(tidyverse)
  library(magrittr)
  library(here)
  library(AnnotationHub)
  library(purrr)
  library(scales)
  library(kableExtra)
  # library(tictoc)  # Causes a bug in rowRanges()
  library(ggrepel)
  library(RColorBrewer)
  library(ggpubr)
  library(pander)
  library(rmarkdown)
  library(viridis)
  ## Project specific
  library(VariantAnnotation)
  library(Gviz)
  library(zoo)
  library(msigdbr)
  library(readxl)
  library(limma)
})</code></pre>
<pre class="r"><code>if (interactive()) setwd(here::here())
theme_set(theme_bw())</code></pre>
<pre class="r"><code># By default Gviz expects UCSC chromosome names
# We&#39;re working with Ensembl so let&#39;s turn this off
options(ucscChromosomeNames = FALSE)</code></pre>
<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>This document contains the Differential Allelic Representation (DAR)
analysis for the MPS-III A/B arm of the experiment.</p>
</div>
<div id="ensdb" class="section level2">
<h2>EnsDb</h2>
<pre class="r"><code>ens_species &lt;- &quot;Danio rerio&quot;
ens_release &lt;- &quot;101&quot;
ens_assembly &lt;- &quot;GRCz11&quot;</code></pre>
<p>Grab genome feature annotations for Danio rerio Ensembl release 101
(genome assembly GRCz11).</p>
<pre class="r"><code>ah &lt;- AnnotationHub() %&gt;%
  subset(species == ens_species) %&gt;%
  subset(rdataclass == &quot;EnsDb&quot;)
ahId &lt;- ah$ah_id[str_detect(ah$title, ens_release)]
ensDb &lt;- ah[[ahId]]</code></pre>
<pre class="r"><code># Filter for primary chromosomes
chrInfo &lt;- getChromInfoFromEnsembl(ens_assembly, release = ens_release) %&gt;%
  dplyr::filter(coord_system == &quot;chromosome&quot;)
primary_chrs &lt;- chrInfo$name</code></pre>
<pre class="r"><code>genes &lt;- genes(ensDb, filter = SeqNameFilter(primary_chrs))
mcols(genes) &lt;- mcols(genes)[
  c(&quot;gene_id&quot;, &quot;gene_name&quot;, &quot;gene_biotype&quot;, &quot;entrezid&quot;)
]</code></pre>
</div>
<div id="metadata" class="section level2">
<h2>Metadata</h2>
<p>Load and clean the sample metadata by renaming columns following
<code>R</code> naming conventions and setting categorical variables as
factors. Also append sample file basenames from a file I created
(<code>data/sample_basenames.csv</code>), as we will need these later.
This metadata contains information for all samples in the experiment, so
let’s filter for only samples in the A/B dataset.</p>
<pre class="r"><code>expArm &lt;- &quot;AB&quot;</code></pre>
<pre class="r"><code>metadata &lt;- read_xlsx(here(&quot;data/meta_final.xlsx&quot;)) %&gt;%
  dplyr::rename(
    rin = RIN,
    sex = Sex,
    rna_batch = rnaBatch,
    mps = MPS,
    uln = ULN,
    home_tank = `Home tank`,
    dob = DOB,
    death = Death
  ) %&gt;%
  mutate(
    sex = as.factor(sex),
    genotype = case_when(
      genotype == &quot;wt&quot; ~ &quot;WT&quot;,
      genotype == &quot;het&quot; ~ &quot;sgsh_het&quot;,
      genotype == &quot;naglu&quot; ~ &quot;MPSIIIB&quot;,
      genotype == &quot;sgsh&quot; ~ &quot;MPSIIIA&quot;,
      genotype == &quot;hgsnat&quot; ~ &quot;MPSIIIC&quot;
    ),
    genotype = as.factor(genotype),
    home_tank = as.factor(home_tank)
  ) %&gt;%
  left_join(read_csv(here(&quot;data/sample_basenames.csv&quot;))) %&gt;%
  dplyr::filter(mps == expArm) %&gt;%
  droplevels()</code></pre>
<pre class="r"><code>metadata %&gt;%
  kable(
    align = &quot;l&quot;,
    caption = &quot;Sample metadata for the MPS-III A/B arm of the experiment&quot;
  ) %&gt;%
  kable_styling(
    bootstrap_options = c(&quot;striped&quot;, &quot;hover&quot;, &quot;condensed&quot;, &quot;responsive&quot;)
  )</code></pre>
<table class="table table-striped table-hover table-condensed table-responsive" style="margin-left: auto; margin-right: auto;">
<caption>
Sample metadata for the MPS-III A/B arm of the experiment
</caption>
<thead>
<tr>
<th style="text-align:left;">
sample
</th>
<th style="text-align:left;">
rin
</th>
<th style="text-align:left;">
sex
</th>
<th style="text-align:left;">
genotype
</th>
<th style="text-align:left;">
rna_batch
</th>
<th style="text-align:left;">
mps
</th>
<th style="text-align:left;">
uln
</th>
<th style="text-align:left;">
home_tank
</th>
<th style="text-align:left;">
dob
</th>
<th style="text-align:left;">
death
</th>
<th style="text-align:left;">
basename
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
AB1
</td>
<td style="text-align:left;">
9.3
</td>
<td style="text-align:left;">
male
</td>
<td style="text-align:left;">
MPSIIIB
</td>
<td style="text-align:left;">
2022-02-10
</td>
<td style="text-align:left;">
AB
</td>
<td style="text-align:left;">
22-00136
</td>
<td style="text-align:left;">
AB_P1_lay1_tank2
</td>
<td style="text-align:left;">
2021-01-15
</td>
<td style="text-align:left;">
2021-04-22
</td>
<td style="text-align:left;">
22-00136_S1
</td>
</tr>
<tr>
<td style="text-align:left;">
AB4
</td>
<td style="text-align:left;">
7.1
</td>
<td style="text-align:left;">
female
</td>
<td style="text-align:left;">
MPSIIIA
</td>
<td style="text-align:left;">
2022-02-10
</td>
<td style="text-align:left;">
AB
</td>
<td style="text-align:left;">
22-00138
</td>
<td style="text-align:left;">
AB_P1_lay1_tank3
</td>
<td style="text-align:left;">
2021-01-15
</td>
<td style="text-align:left;">
2021-04-22
</td>
<td style="text-align:left;">
22-00138_S2
</td>
</tr>
<tr>
<td style="text-align:left;">
AB11
</td>
<td style="text-align:left;">
7.9
</td>
<td style="text-align:left;">
female
</td>
<td style="text-align:left;">
WT
</td>
<td style="text-align:left;">
2022-02-10
</td>
<td style="text-align:left;">
AB
</td>
<td style="text-align:left;">
22-00145
</td>
<td style="text-align:left;">
AB_P1_lay1_tank4
</td>
<td style="text-align:left;">
2021-01-15
</td>
<td style="text-align:left;">
2021-04-22
</td>
<td style="text-align:left;">
22-00145_S3
</td>
</tr>
<tr>
<td style="text-align:left;">
AB13
</td>
<td style="text-align:left;">
7.8
</td>
<td style="text-align:left;">
male
</td>
<td style="text-align:left;">
MPSIIIA
</td>
<td style="text-align:left;">
2022-02-10
</td>
<td style="text-align:left;">
AB
</td>
<td style="text-align:left;">
22-00147
</td>
<td style="text-align:left;">
AB_P1_lay1_tank2
</td>
<td style="text-align:left;">
2021-01-15
</td>
<td style="text-align:left;">
2021-04-22
</td>
<td style="text-align:left;">
22-00147_S4
</td>
</tr>
<tr>
<td style="text-align:left;">
AB14
</td>
<td style="text-align:left;">
9.7
</td>
<td style="text-align:left;">
female
</td>
<td style="text-align:left;">
sgsh_het
</td>
<td style="text-align:left;">
2022-02-10
</td>
<td style="text-align:left;">
AB
</td>
<td style="text-align:left;">
22-00148
</td>
<td style="text-align:left;">
AB_P1_lay1_tank2
</td>
<td style="text-align:left;">
2021-01-15
</td>
<td style="text-align:left;">
2021-04-22
</td>
<td style="text-align:left;">
22-00148_S5
</td>
</tr>
<tr>
<td style="text-align:left;">
AB16
</td>
<td style="text-align:left;">
6.6
</td>
<td style="text-align:left;">
female
</td>
<td style="text-align:left;">
WT
</td>
<td style="text-align:left;">
2022-02-11
</td>
<td style="text-align:left;">
AB
</td>
<td style="text-align:left;">
22-00150
</td>
<td style="text-align:left;">
AB_P1_lay1_tank2
</td>
<td style="text-align:left;">
2021-01-15
</td>
<td style="text-align:left;">
2021-04-22
</td>
<td style="text-align:left;">
22-00150_S6
</td>
</tr>
<tr>
<td style="text-align:left;">
AB17
</td>
<td style="text-align:left;">
8.6
</td>
<td style="text-align:left;">
female
</td>
<td style="text-align:left;">
WT
</td>
<td style="text-align:left;">
2022-02-11
</td>
<td style="text-align:left;">
AB
</td>
<td style="text-align:left;">
22-00151
</td>
<td style="text-align:left;">
AB_P1_lay1_tank2
</td>
<td style="text-align:left;">
2021-01-15
</td>
<td style="text-align:left;">
2021-04-22
</td>
<td style="text-align:left;">
22-00151_S7
</td>
</tr>
<tr>
<td style="text-align:left;">
AB18
</td>
<td style="text-align:left;">
6.6
</td>
<td style="text-align:left;">
male
</td>
<td style="text-align:left;">
sgsh_het
</td>
<td style="text-align:left;">
2022-02-11
</td>
<td style="text-align:left;">
AB
</td>
<td style="text-align:left;">
22-00152
</td>
<td style="text-align:left;">
AB_P1_lay1_tank2
</td>
<td style="text-align:left;">
2021-01-15
</td>
<td style="text-align:left;">
2021-04-22
</td>
<td style="text-align:left;">
22-00152_S8
</td>
</tr>
<tr>
<td style="text-align:left;">
AB20
</td>
<td style="text-align:left;">
8.1
</td>
<td style="text-align:left;">
female
</td>
<td style="text-align:left;">
MPSIIIB
</td>
<td style="text-align:left;">
2022-02-11
</td>
<td style="text-align:left;">
AB
</td>
<td style="text-align:left;">
22-00153
</td>
<td style="text-align:left;">
AB_P1_lay1_tank2
</td>
<td style="text-align:left;">
2021-01-15
</td>
<td style="text-align:left;">
2021-04-22
</td>
<td style="text-align:left;">
22-00153_S9
</td>
</tr>
<tr>
<td style="text-align:left;">
AB21
</td>
<td style="text-align:left;">
9.5
</td>
<td style="text-align:left;">
female
</td>
<td style="text-align:left;">
sgsh_het
</td>
<td style="text-align:left;">
2022-02-11
</td>
<td style="text-align:left;">
AB
</td>
<td style="text-align:left;">
22-00154
</td>
<td style="text-align:left;">
AB_P1_lay1_tank3
</td>
<td style="text-align:left;">
2021-01-15
</td>
<td style="text-align:left;">
2021-04-22
</td>
<td style="text-align:left;">
22-00154_S10
</td>
</tr>
<tr>
<td style="text-align:left;">
AB22
</td>
<td style="text-align:left;">
8.1
</td>
<td style="text-align:left;">
female
</td>
<td style="text-align:left;">
MPSIIIB
</td>
<td style="text-align:left;">
2022-02-11
</td>
<td style="text-align:left;">
AB
</td>
<td style="text-align:left;">
22-00155
</td>
<td style="text-align:left;">
AB_P1_lay1_tank3
</td>
<td style="text-align:left;">
2021-01-15
</td>
<td style="text-align:left;">
2021-04-22
</td>
<td style="text-align:left;">
22-00155_S11
</td>
</tr>
<tr>
<td style="text-align:left;">
AB23
</td>
<td style="text-align:left;">
7.4
</td>
<td style="text-align:left;">
female
</td>
<td style="text-align:left;">
MPSIIIA
</td>
<td style="text-align:left;">
2022-02-11
</td>
<td style="text-align:left;">
AB
</td>
<td style="text-align:left;">
22-00156
</td>
<td style="text-align:left;">
AB_P1_lay1_tank3
</td>
<td style="text-align:left;">
2021-01-15
</td>
<td style="text-align:left;">
2021-04-22
</td>
<td style="text-align:left;">
22-00156_S12
</td>
</tr>
<tr>
<td style="text-align:left;">
AB24
</td>
<td style="text-align:left;">
7.1
</td>
<td style="text-align:left;">
female
</td>
<td style="text-align:left;">
MPSIIIA
</td>
<td style="text-align:left;">
2022-02-11
</td>
<td style="text-align:left;">
AB
</td>
<td style="text-align:left;">
22-00157
</td>
<td style="text-align:left;">
AB_P1_lay1_tank3
</td>
<td style="text-align:left;">
2021-01-15
</td>
<td style="text-align:left;">
2021-04-22
</td>
<td style="text-align:left;">
22-00157_S13
</td>
</tr>
<tr>
<td style="text-align:left;">
AB25
</td>
<td style="text-align:left;">
8.0
</td>
<td style="text-align:left;">
female
</td>
<td style="text-align:left;">
sgsh_het
</td>
<td style="text-align:left;">
2022-02-11
</td>
<td style="text-align:left;">
AB
</td>
<td style="text-align:left;">
22-00158
</td>
<td style="text-align:left;">
AB_P1_lay1_tank3
</td>
<td style="text-align:left;">
2021-01-15
</td>
<td style="text-align:left;">
2021-04-22
</td>
<td style="text-align:left;">
22-00158_S14
</td>
</tr>
<tr>
<td style="text-align:left;">
AB26
</td>
<td style="text-align:left;">
9.1
</td>
<td style="text-align:left;">
female
</td>
<td style="text-align:left;">
MPSIIIB
</td>
<td style="text-align:left;">
2022-02-11
</td>
<td style="text-align:left;">
AB
</td>
<td style="text-align:left;">
22-00159
</td>
<td style="text-align:left;">
AB_P1_lay1_tank3
</td>
<td style="text-align:left;">
2021-01-15
</td>
<td style="text-align:left;">
2021-04-22
</td>
<td style="text-align:left;">
22-00159_S15
</td>
</tr>
<tr>
<td style="text-align:left;">
AB27
</td>
<td style="text-align:left;">
9.6
</td>
<td style="text-align:left;">
female
</td>
<td style="text-align:left;">
WT
</td>
<td style="text-align:left;">
2022-02-11
</td>
<td style="text-align:left;">
AB
</td>
<td style="text-align:left;">
22-00160
</td>
<td style="text-align:left;">
AB_P1_lay1_tank3
</td>
<td style="text-align:left;">
2021-01-15
</td>
<td style="text-align:left;">
2021-04-22
</td>
<td style="text-align:left;">
22-00160_S16
</td>
</tr>
</tbody>
</table>
<p>Set up colours for plotting by genotype. Going with a
colourblind-friendly palette (<code>Set2</code>) for best
accessibility.</p>
<pre class="r"><code>genoCols &lt;- metadata$genotype %&gt;%
  levels() %&gt;%
  length() %&gt;%
  brewer.pal(&quot;Set2&quot;) %&gt;%
  setNames(levels(metadata$genotype))</code></pre>
<p>Now define our samples that belong to each group (genotype). We will
use this later in the analysis when accessing relevant samples from the
<code>VCF</code> file.</p>
<pre class="r"><code>samplesByGroup &lt;- metadata %&gt;%
  split(f = .$genotype) %&gt;%
  sapply(function(x){
    pull(x, basename)
  }, simplify = FALSE)</code></pre>
</div>
</div>
<div id="de-data" class="section level1">
<h1>DE data</h1>
<p>Load in the <code>dgeList</code> and <code>topTable</code>s from
Karissa’s DE analysis.</p>
<pre class="r"><code>topTables &lt;- readRDS(here(&quot;data/R_objects/AB_toptab_cqn.rds&quot;))</code></pre>
<p>Also pull the gene IDs for DE genes.</p>
<pre class="r"><code>deGenes &lt;- sapply(topTables, function(x){
  dplyr::filter(x, DE) %&gt;%
    pull(gene_id)
}, simplify = FALSE)</code></pre>
</div>
<div id="variant-data" class="section level1">
<h1>Variant data</h1>
<p>Variant data from the <code>GVCF</code> file is parsed into a
<code>VCF</code> class object using the <code>VariantAnnotation</code>
package. Note that this file is large and therefore does not exist in
the Git repository.</p>
<pre class="r"><code>vcf_path &lt;- file.path(
  &quot;/hpcfs/users/a1647910/2022_MPSIII_3mBrainRNAseq/code&quot;,
  &quot;analysis-variants_AB/results/09_variants/6_select/all_samples.vcf.gz&quot;
)
svp &lt;- ScanVcfParam(info = &quot;&quot;, geno = c(&quot;GT&quot;, &quot;GQ&quot;))
vcf &lt;- suppressWarnings({
  readVcf(vcf_path, genome = genome(ensDb))
})</code></pre>
<p>A number of accessor functions exist for the <code>VCF</code> object.
We will use the following.</p>
<ul>
<li><p><code>rowRanges()</code>: Contains information from the CHROM,
POS, and ID fields of the VCF file. Returned as a <code>GRanges</code>
object. We will save this object to our <code>rr</code>
variable.</p></li>
<li><p><code>geno()</code>: Contains data described in the FORMAT fields
of the VCF. Returned as a <code>SimpleList</code>. We will extract the
genotype information (GT) and save to the <code>gt</code>
variable.</p></li>
</ul>
<pre class="r"><code>rr &lt;- rowRanges(vcf) %&gt;%
  plyranges::select(-paramRangeID)</code></pre>
<p>Below is a snippet of the <code>rr</code> data:</p>
<pre class="r"><code>head(rr)</code></pre>
<pre><code>GRanges object with 6 ranges and 4 metadata columns:
             seqnames    ranges strand |            REF                ALT
                &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;DNAStringSet&gt; &lt;DNAStringSetList&gt;
  1:6568_G/A        1      6568      * |              G                  A
  1:7028_A/G        1      7028      * |              A                  G
  1:7092_G/T        1      7092      * |              G                  T
  1:7330_T/A        1      7330      * |              T                  A
  1:9791_C/A        1      9791      * |              C                  A
  1:9815_G/A        1      9815      * |              G                  A
                  QUAL      FILTER
             &lt;numeric&gt; &lt;character&gt;
  1:6568_G/A    543.77        PASS
  1:7028_A/G     39.54        PASS
  1:7092_G/T    230.38        PASS
  1:7330_T/A    163.61        PASS
  1:9791_C/A     82.33        PASS
  1:9815_G/A    189.80        PASS
  -------
  seqinfo: 993 sequences from GRCz11 genome</code></pre>
<pre class="r"><code>gt &lt;- geno(vcf)$GT</code></pre>
<p>The genotype data contains phasing information where possible.
Phasing is indicated by the use of <code>|</code> as opposed to
<code>/</code> for SNPs that are not phased. This information is not
required for DAR analysis and complicates downstream processing, so
let’s remove the phasing information and make all fields consistent.</p>
<pre class="r"><code>unphase_gt &lt;- function(x){
  str_replace(x, &quot;\\|&quot;, &quot;\\/&quot;)
}
gt &lt;- gt %&gt;%
  as.data.frame() %&gt;%
  mutate(across(everything(), unphase_gt))</code></pre>
<p>Below is a snippet of the <code>gt</code> data:</p>
<pre class="r"><code>head(gt)</code></pre>
<pre><code>           22-00136_S1 22-00138_S2 22-00145_S3 22-00147_S4 22-00148_S5
1:6568_G/A         ./.         ./.         1/1         ./.         ./.
1:7028_A/G         ./.         ./.         ./.         ./.         0/0
1:7092_G/T         ./.         ./.         ./.         ./.         ./.
1:7330_T/A         ./.         ./.         ./.         1/1         ./.
1:9791_C/A         0/0         ./.         0/0         ./.         0/0
1:9815_G/A         0/1         ./.         0/1         ./.         ./.
           22-00150_S6 22-00151_S7 22-00152_S8 22-00153_S9 22-00154_S10
1:6568_G/A         ./.         ./.         ./.         ./.          ./.
1:7028_A/G         ./.         ./.         ./.         ./.          ./.
1:7092_G/T         ./.         ./.         ./.         1/1          ./.
1:7330_T/A         ./.         ./.         ./.         0/1          ./.
1:9791_C/A         ./.         0/0         ./.         ./.          ./.
1:9815_G/A         ./.         ./.         ./.         ./.          ./.
           22-00155_S11 22-00156_S12 22-00157_S13 22-00158_S14 22-00159_S15
1:6568_G/A          ./.          1/1          ./.          ./.          1/1
1:7028_A/G          ./.          ./.          0/1          ./.          ./.
1:7092_G/T          ./.          ./.          1/1          ./.          ./.
1:7330_T/A          1/1          0/0          0/0          ./.          ./.
1:9791_C/A          1/1          0/0          ./.          ./.          ./.
1:9815_G/A          1/1          0/0          ./.          ./.          ./.
           22-00160_S16
1:6568_G/A          1/1
1:7028_A/G          ./.
1:7092_G/T          1/1
1:7330_T/A          ./.
1:9791_C/A          ./.
1:9815_G/A          ./.</code></pre>
</div>
<div id="differential-allelic-representation" class="section level1">
<h1>Differential Allelic Representation</h1>
<p>Genotypes are reported as numeric indices. <code>0</code> indicates
the reference allele, <code>1</code> is the first alternate allele,
<code>2</code> is the second alternate allele, and so on. The reference
and alternate allele information is contained in our <code>rr</code>
object if ever required. However, for DAR analysis, we can work directly
with the indices as they are consistent across all samples for a single
variant position.</p>
<p>We aim to calculate a DAR metric at each suitable variant locus. This
requires us to firstly summarise the genotype data into counts of the
alleles reported at each variant locus.</p>
<pre class="r"><code>## This code takes a while (~15mins) to execute
## So save as Rds file for quicker processing in future
alleleCounts_path &lt;- file.path(
  here(&quot;data/dar-analysis&quot;),
  expArm,
  &quot;alleleCounts.Rds&quot;
)
if (!file.exists(alleleCounts_path)) {
  if (!dir.exists(dirname(alleleCounts_path))) {
    dir.create(dirname(alleleCounts_path), recursive = TRUE)
  }
  alleleCounts &lt;- lapply(samplesByGroup, function(samples){
    apply(gt[,samples], 1, function(gtInGroup){
      alleles &lt;- gtInGroup %&gt;%
        as.character() %&gt;%
        str_split(&quot;/&quot;) %&gt;%
        unlist()
      tibble(
        n_called = sum(alleles != &quot;.&quot;),
        n_nocall = sum(alleles == &quot;.&quot;),
        n_0 = sum(alleles == &quot;0&quot;),
        n_1 = sum(alleles == &quot;1&quot;),
        n_2 = sum(alleles == &quot;2&quot;),
        n_3 = sum(alleles == &quot;3&quot;)
      )
    }) %&gt;%
      bind_rows()
  })
  saveRDS(alleleCounts, alleleCounts_path)
} else {
  alleleCounts &lt;- readRDS(alleleCounts_path)
}</code></pre>
<p>Not all samples have genotype calls, so we normalise by converting
the allele counts into a proportion of total counts at each variant
locus. We also filter variant loci within each sample group with the
criterion number of called genotypes &gt; number of missing genotypes.
Loci that do not satisfy this criterion have their proportions are
returned as <code>NA</code>’s, so that we can retain the object
structure and join it back to our <code>rowRanges</code> before removing
the filtered loci.</p>
<pre class="r"><code>alleleProps &lt;- lapply(alleleCounts, function(counts){
  counts %&gt;%
    bind_rows() %&gt;%
    mutate(
      across(
        .cols = c(&quot;n_0&quot;, &quot;n_1&quot;, &quot;n_2&quot;, &quot;n_3&quot;),
        .fns = ~ ifelse(n_called &gt; n_nocall, .x / n_called, NA)
      )
    ) %&gt;%
    dplyr::select(-n_called, -n_nocall) %&gt;%
    set_colnames(str_replace(colnames(.), &quot;n_&quot;, &quot;prop_&quot;))
})</code></pre>
<p>Now that we have normalised values of allelic representation at each
variant locus within our sample groups, we can calculate the DAR metric
between groups. Firstly, we manually define our contrasts of
interest.</p>
<pre class="r"><code>contrasts &lt;- makeContrasts(
  A_vs_WT = MPSIIIA - WT,
  B_vs_WT = MPSIIIB - WT,
  het_vs_WT = sgsh_het - WT,
  levels = names(alleleProps)
)</code></pre>
<p>Also choose the number of variant loci to define our elastic window
size for smoothing the DAR metric. This needs to be an odd number, as we
need to include the variant position and an equal number of loci either
side.</p>
<pre class="r"><code>winSize &lt;- 11</code></pre>
<p>With a chosen elastic window size of 11 variants, this will smooth
the DAR metric at each locus with the DAR values of the 5 loci either
side. Note that for loci at the ends of each chromosome, the window size
will be slightly smaller to avoid including DAR values from other
chromosomes.</p>
<p>Now we calculate the Euclidean distance between normalised allelic
representation of the contrasted sample groups. Any variant loci that
were previously filtered due to not containing enough genotype
information are returned as <code>NA</code>, which we can now remove. We
convert the Euclidean distance to the DAR metric by dividing by <span
class="math display">\[\sqrt{2}\]</span>, resulting in an
easy-to-interpret value between 0 and 1, where 0 represents identical
allelic representation and 1 represents complete diversity. Lastly we
smooth the DAR values as described above.</p>
<pre class="r"><code>## This code takes a while (~7mins) to execute
## So save as Rds file for quicker processing in future
DAR_path &lt;- file.path(
  here(&quot;data/dar-analysis&quot;),
  expArm,
  &quot;DAR.Rds&quot;
)
if (!file.exists(DAR_path)) {
  if (!dir.exists(dirname(DAR_path))) {
    dir.create(dirname(DAR_path), recursive = TRUE)
  }
  DAR &lt;- apply(contrasts, 2, function(col){
    ## Grab the sample groups in each contrast
    contrast &lt;- col %&gt;%
      enframe(name = &quot;group&quot;, value = &quot;contrast&quot;) %&gt;%
      dplyr::filter(contrast != 0) %&gt;%
      pull(group)
    stopifnot(length(contrast) == 2)
    ## Return the allele proportions for each sample group in each contrast
    alleleProps[contrast]
  }, simplify = FALSE) %&gt;%
    sapply(function(props){
      ## Calculate the Euclidean distance
      dist &lt;- vapply(seq(length(rr)), function(ind){
        rbind(props[[1]][ind,], props[[2]][ind,]) %&gt;%
          dist() %&gt;%
          as.numeric()
      }, numeric(1))
      ## Bind the calculated Euclidean distance to our rowRanges
      rr %&gt;%
        plyranges::select(-everything()) %&gt;%
        as_tibble() %&gt;%
        mutate(dist = dist) %&gt;%
        ## Remove any loci with NA DAR as this means they were filtered
        ## in at least one of the contrasted sample groups
        dplyr::filter(!is.na(dist)) %&gt;%
        ## Drop levels so we don&#39;t loop over chromosomes/scaffolds with no data
        droplevels() %&gt;%
        ## Split by chromosome to avoid including DAR values from multiple
        ## chromosomes when smoothing
        split(f = .$seqnames) %&gt;%
        lapply(function(chr){
          dplyr::mutate(
            chr,
            ## sqrt(2) is the maximum dist
            dar = dist / sqrt(2),
            dar_smooth = rollapply(dar, width = winSize, mean, partial = TRUE),
            win_start = lag(start, n = (winSize - 1) / 2),
            win_end = lead(start, n = (winSize - 1) / 2),
            ## lag() and lead() will return NA if index is out of range,
            ## so let&#39;s adjust these
            win_start = ifelse(is.na(win_start), start[1], win_start), 
            win_end = ifelse(is.na(win_end), start[nrow(chr)], win_end),
            win_size = win_end - win_start
          )
        }) %&gt;%
        bind_rows() %&gt;%
        GRanges()
    }, simplify = FALSE)
  saveRDS(DAR, DAR_path)
} else {
  DAR &lt;- readRDS(DAR_path)
}</code></pre>
<div id="dar-distribution" class="section level2">
<h2>DAR distribution</h2>
<p>Plotting the cumulative distribution of DAR by chromosome is useful
to determine if the mutant chromosome is affected by high DAR values
more so than other chromosomes.</p>
<p>Let’s first define a plotting function so we can apply it for each
contrast.</p>
<pre class="r"><code>plotCumDAR &lt;- function(DAR, mutChr, title){
  DAR %&gt;%
    split(seqnames(.)) %&gt;%
    lapply(function(chr){
      tibble(
        chromosome = as.vector(seqnames(chr)),
        dar_smooth = chr$dar_smooth
      )
    }) %&gt;%
    purrr::reduce(rbind) %&gt;%
    mutate(
      chromosome = factor(
        chromosome,
        # Draw mutant chr on top
        levels = unique(c(
          chromosome[chromosome != mutChr],
          chromosome[chromosome == mutChr]
        ))
      ),
      mutant_chromosome = chromosome == mutChr
    ) %&gt;%
    ggplot(aes(
      x = dar_smooth,
      group = chromosome,
      colour = mutant_chromosome
    )) +
    stat_ecdf() +
    scale_colour_manual(
      values = c(&quot;red&quot;, &quot;grey50&quot;),
      breaks = c(&quot;TRUE&quot;, &quot;FALSE&quot;)
    ) +
    labs(
      title = title,
      colour = &quot;Mutant chromosome&quot;,
      x = &quot;DAR&quot;,
      y = &quot;F(DAR)&quot;
    )
}</code></pre>
<p>Now let’s see how DAR on the mutant chromosome compares to other
chromosomes.</p>
<pre class="r"><code>plotCumDAR(DAR$A_vs_WT, &quot;22&quot;, &quot;MPS-IIIA vs. WT&quot;)</code></pre>
<p><img src="figure/dar-analysis_AB.Rmd/cumDAR_MPSIIIA-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>plotCumDAR(DAR$B_vs_WT, &quot;24&quot;, &quot;MPS-IIIB vs. WT&quot;)</code></pre>
<p><img src="figure/dar-analysis_AB.Rmd/cumDAR_MPSIIIB-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>plotCumDAR(DAR$het_vs_WT, &quot;22&quot;, &quot;sghs/+ vs. WT&quot;)</code></pre>
<p><img src="figure/dar-analysis_AB.Rmd/cumDAR_sgsh_het-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="elastic-window-ranges" class="section level2">
<h2>Elastic window ranges</h2>
<p>Define a <code>GRanges</code> object that contains the ranges of the
elastic sliding window. We will use this to assign DAR values to genes
for the range(s) they overlap.</p>
<pre class="r"><code>winRanges &lt;- lapply(DAR, function(x){
  x %&gt;%
    as_tibble() %&gt;%
    dplyr::select(
      seqnames, start = win_start, end = win_end, width = win_size, strand,
      dar_smooth
    ) %&gt;%
    split(f = .$seqnames) %&gt;%
    ## Extend the start and ends so that the whole chromosome is covered
    lapply(function(chr){
      seqname &lt;- unique(chr$seqnames)
      chr$start[1] &lt;- 1
      chr$width[1] &lt;- chr$end[1] - chr$start[1]
      last_ind &lt;- nrow(chr)
      chr$end[last_ind] &lt;- chrInfo$length[chrInfo$name == seqname]
      chr$width[last_ind] &lt;- chr$end[last_ind] - chr$start[last_ind]
      chr
    }) %&gt;%
    bind_rows() %&gt;%
    GRanges()
})</code></pre>
<p>Let’s define another function to visualise the distribution of
elastic window sizes.</p>
<pre class="r"><code>plotWinRanges &lt;- function(winRanges, title, bins = 100){
  winRanges %&gt;%
    width() %&gt;%
    enframe(name = NULL, value = &quot;winSize&quot;) %&gt;%
    ggplot(aes(winSize)) +
    geom_histogram(
      bins = bins,
      colour = &quot;black&quot;,
      fill = &quot;grey70&quot;
    ) +
    scale_x_log10() +
    labs(
      title = title,
      x = &quot;Window size (bp)&quot;,
      y = &quot;Count&quot;
    )
}</code></pre>
<pre class="r"><code>plotWinRanges(winRanges$A_vs_WT, title = &quot;MPS-IIIA vs. WT&quot;)</code></pre>
<p><img src="figure/dar-analysis_AB.Rmd/winRanges_MPSIIIA-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>plotWinRanges(winRanges$B_vs_WT, title = &quot;MPS-IIIB vs. WT&quot;)</code></pre>
<p><img src="figure/dar-analysis_AB.Rmd/winRanges_MPSIIIB-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>plotWinRanges(winRanges$het_vs_WT, title = &quot;sgsh/+ vs. WT&quot;)</code></pre>
<p><img src="figure/dar-analysis_AB.Rmd/winRanges_sgsh_het-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="deg-dar-relationship" class="section level2">
<h2>DEG-DAR relationship</h2>
<p>We can visualise whether differentially expressed genes (DEG) may be
prone eQTL artefacts by plotting the DAR values along a chromosomal axis
and overlaying the position of DEGs. Let’s define a plotting function
that we can easily use to select a chromosome and the genomic features
of interest.</p>
<pre class="r"><code>plotChrDAR &lt;- function(
    DAR, chromosome, features = GRanges(), foi = GRanges(),
    darType = &quot;b&quot;, highlight_features = TRUE,
    title = &quot;&quot;
){
  tracks &lt;- list()
  
  if (length(foi)) {
    foi &lt;- foi %&gt;%
      plyranges::filter(seqnames == chromosome) %&gt;%
      plyranges::mutate(symbol = gene_name)
    foi_track &lt;- GeneRegionTrack(
      range = foi,
      transcriptAnnotation = &quot;symbol&quot;,
      col = &quot;white&quot;,
      fill = &quot;white&quot;,
      showTranscriptId = TRUE,
      fontcolor.group = &quot;red&quot;,
      cex.group = 0.6,
      size = 0.5,
      name = NULL
    )
    tracks &lt;- append(tracks, foi_track)
  }
  
  axis_track &lt;- GenomeAxisTrack(
    add53 = TRUE,
    add35 = TRUE,
    name = paste0(&quot;Chr&quot;, chromosome),
    showTitle = TRUE,
    size = 1
  )
  if (length(foi)) {
    axis_track &lt;- HighlightTrack(
      trackList = list(axis_track),
      range = foi
    )
  }
  tracks &lt;- append(tracks, axis_track)
  
  if (length(features)) {
    features &lt;- features %&gt;%
      plyranges::filter(seqnames == chromosome) %&gt;%
      plyranges::mutate(symbol = gene_name)
    features_track &lt;- AnnotationTrack(
      range = features,
      name = &quot;DE&quot;,
      shape = &quot;box&quot;,
      fill = &quot;darkgray&quot;,
      group = features$gene_name,
      groupAnnotation = &quot;group&quot;,
      fontcolor.group = 1,
      cex.group = 0.6,
      size = 0.4
    )
    tracks &lt;- append(tracks, features_track)
  }
  
  dar_track &lt;- DAR[,&quot;dar_smooth&quot;] %&gt;%
    plyranges::filter(seqnames == chromosome) %&gt;%
    DataTrack(
      type = darType,
      name = &quot;DAR&quot;,
      size = 8,
      window = -1,
      windowSize = 1,
      cex = 0.4,
      col = &quot;grey20&quot;,
      col.axis = &quot;black&quot;,
      yTicksAt = seq(0, 1, 0.1),
      ylim = c(0, 1)
    )
  if (highlight_features) {
    if (all(length(features) &amp; length(foi))) {
      ranges &lt;- c(features, foi)
    } else if (length(features)) {
      ranges &lt;- features
    } else if (length(foi)) {
      ranges &lt;- foi
    } else {
      ranges &lt;- GRanges()
    }
    dar_track &lt;- HighlightTrack(
      trackList = list(dar_track),
      range = ranges,
      col = c(rep(&quot;#ffe0e0&quot;, length(features)), rep(&quot;red&quot;, length(foi))),
      fill = c(rep(&quot;#ffe0e0&quot;, length(features)), rep(&quot;red&quot;, length(foi)))
    )
  }
  tracks &lt;- append(tracks, dar_track)
  
  plotTracks(
    trackList = tracks,
    main = title,
    cex.main = 1,
    cex.title = 0.6,
    col.title = &quot;black&quot;,
    background.title = &quot;white&quot;
  )
}</code></pre>
<div id="mps-iiia" class="section level3">
<h3>MPS-IIIA</h3>
<p>First let’s look at the mutant chromosome, as this was identified as
having the highest average DAR from the cumulative distributions we
plotted above.</p>
<pre class="r"><code>plotChrDAR(
  DAR = DAR$A_vs_WT,
  chromosome = &quot;22&quot;,
  features = genes[genes$gene_id %in% deGenes$`MPS-IIIA`],
  foi = genes[genes$gene_name == &quot;sgsh&quot;]
)</code></pre>
<p><img src="figure/dar-analysis_AB.Rmd/DAR_MPSIIIA_chr22-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>And comparing this to a non-mutant chromosome:</p>
<pre class="r"><code>plotChrDAR(
  DAR = DAR$A_vs_WT,
  chromosome = &quot;1&quot;,
  features = genes[genes$gene_id %in% deGenes$`MPS-IIIA`]
)</code></pre>
<p><img src="figure/dar-analysis_AB.Rmd/DAR_MPSIIIA_chr1-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="mps-iiib" class="section level3">
<h3>MPS-IIIB</h3>
<p>The mutant chromosome also showed the highest average DAR in the
MPS-IIIB vs. WT comparison.</p>
<pre class="r"><code>plotChrDAR(
  DAR = DAR$B_vs_WT,
  chromosome = &quot;24&quot;,
  features = genes[genes$gene_id %in% deGenes$`MPS-IIIB`],
  foi = genes[genes$gene_name == &quot;naglu&quot;]
)</code></pre>
<p><img src="figure/dar-analysis_AB.Rmd/DAR_MPSIIIB_chr24-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Interestingly we see a region of low DAR along the mutant chromosome,
however no DEGs exist in these regions. It is likely that many of the
genes classified as DE are eQTLs as they are found within regions of
high DAR.</p>
<p>Again comparing to a non-mutant chromosome:</p>
<pre class="r"><code>plotChrDAR(
  DAR = DAR$B_vs_WT,
  chromosome = &quot;3&quot;,
  features = genes[genes$gene_id %in% deGenes$`MPS-IIIB`]
)</code></pre>
<p><img src="figure/dar-analysis_AB.Rmd/DAR_MPSIIIB_chr2-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="assign-gene-dar" class="section level2">
<h2>Assign gene DAR</h2>
<p>Assigning each gene within the dataset a DAR value is useful for
further analysis such as the weighting of gene rankings for GSEA. Let’s
do this now and save the resulting <code>GRanges</code> object for
further use.</p>
<pre class="r"><code>assignDAR &lt;- function(features, ranges){
  overlaps &lt;- findOverlaps(features, ranges)
  lapply(unique(queryHits(overlaps)), function(query){
    subjects &lt;- subjectHits(overlaps)[queryHits(overlaps) == query]
    dar &lt;- ranges$dar_smooth[subjects] %&gt;%
      mean()
    features[query,] %&gt;%
      as.data.frame() %&gt;%
      dplyr::mutate(dar = dar)
  }) %&gt;%
    bind_rows() %&gt;%
    GRanges()
}</code></pre>
<pre class="r"><code>## This code takes a while (~10mins) to execute
## So save as Rds file for quicker processing in future
geneDAR_path &lt;- file.path(
  here(&quot;data/dar-analysis&quot;),
  expArm,
  &quot;geneDAR.Rds&quot;
)
if (!file.exists(geneDAR_path)) {
  if (!dir.exists(dirname(geneDAR_path))) {
    dir.create(dirname(geneDAR_path), recursive = TRUE)
  }
  geneDAR &lt;- map2(topTables, winRanges, function(tt, wr){
    features &lt;- genes[genes$gene_id %in% tt$gene_id]
    assignDAR(features, wr)
  })
  saveRDS(geneDAR, geneDAR_path)
} else {
  geneDAR &lt;- readRDS(geneDAR_path)
}</code></pre>
</div>
<div id="logfc-comparison-between-mps-iiia-and-sgsh"
class="section level2">
<h2>LogFC comparison between MPS-IIIA and sgsh/+</h2>
<pre class="r"><code>topTables %&gt;% 
  bind_rows() %&gt;% 
  dplyr::filter(coef %in% c(&quot;MPS-IIIA&quot;, &quot;sgsh/+&quot;)) %&gt;% 
  dplyr::filter(chromosome == &quot;22&quot;) %&gt;%
  droplevels() %&gt;%
  mutate(
    chromosome = fct_relevel(
      chromosome,
      c(&quot;22&quot;, levels(chromosome)[levels(chromosome) != &quot;22&quot;]))
  ) %&gt;%
  dplyr::select(gene_id, logFC, coef)  %&gt;% 
  spread(key = &quot;coef&quot;, value = &quot;logFC&quot;) %&gt;% 
  mutate(
    DE = case_when(
      gene_id %in%
        intersect(deGenes$`MPS-IIIA`, deGenes$`sgsh/+`)
      ~ &quot;both&quot;, 
      gene_id %in% deGenes$`sgsh/+`[!deGenes$`sgsh/+` %in% deGenes$`MPS-IIIA`]
      ~ &quot;het only&quot;,
      gene_id %in% deGenes$`MPS-IIIA`[!deGenes$`MPS-IIIA` %in% deGenes$`sgsh/+`]
      ~ &quot;MPS-IIIA only&quot;, 
      TRUE ~ &quot;not DE&quot;
    ) 
  ) %&gt;% 
  left_join(as_tibble(geneDAR$`sgsh/+`)) %&gt;%
  ggplot(aes(x = `MPS-IIIA`, y =`sgsh/+`)) +
  geom_point(
    aes(fill = DE, size = dar),
    colour = &quot;black&quot;,
    shape = 21,
    alpha = 0.5
    ) +
  geom_abline(slope = 1) +
  geom_label_repel(
    aes(label = gene_name), 
    nudge_y = 1,
    data = . %&gt;% dplyr::filter(gene_name == &quot;sgsh&quot;)
  ) +
  scale_x_continuous(limits = c(-5, 5)) +
  scale_y_continuous(limits = c(-2.5, 5)) +
  theme(aspect.ratio = 1) +
  scale_fill_manual(values = brewer.pal(4, &quot;Set2&quot;)) +
  labs(size = &quot;DAR&quot;) +
  ggtitle(&quot;LogFC of genes on chromosome 22 in sgsh het vs hom &quot;)</code></pre>
<p><img src="figure/dar-analysis_AB.Rmd/unnamed-chunk-2-1.png" width="672" style="display: block; margin: auto;" /></p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.2.2 Patched (2022-11-10 r83330)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 20.04.5 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.9.0
LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.9.0

locale:
 [1] LC_CTYPE=en_AU.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_AU.UTF-8        LC_COLLATE=en_AU.UTF-8    
 [5] LC_MONETARY=en_AU.UTF-8    LC_MESSAGES=en_AU.UTF-8   
 [7] LC_PAPER=en_AU.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_AU.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] grid      stats4    stats     graphics  grDevices utils     datasets 
[8] methods   base     

other attached packages:
 [1] ensembldb_2.22.0            AnnotationFilter_1.22.0    
 [3] GenomicFeatures_1.50.4      AnnotationDbi_1.60.0       
 [5] limma_3.54.1                readxl_1.4.2               
 [7] msigdbr_7.5.1               zoo_1.8-11                 
 [9] Gviz_1.42.1                 VariantAnnotation_1.44.1   
[11] Rsamtools_2.14.0            Biostrings_2.66.0          
[13] XVector_0.38.0              SummarizedExperiment_1.28.0
[15] Biobase_2.58.0              GenomicRanges_1.50.2       
[17] GenomeInfoDb_1.34.9         IRanges_2.32.0             
[19] S4Vectors_0.36.1            MatrixGenerics_1.10.0      
[21] matrixStats_0.63.0          viridis_0.6.2              
[23] viridisLite_0.4.1           rmarkdown_2.20             
[25] pander_0.6.5                ggpubr_0.6.0               
[27] RColorBrewer_1.1-3          ggrepel_0.9.3              
[29] kableExtra_1.3.4            scales_1.2.1               
[31] AnnotationHub_3.6.0         BiocFileCache_2.6.1        
[33] dbplyr_2.3.0                BiocGenerics_0.44.0        
[35] here_1.0.1                  magrittr_2.0.3             
[37] lubridate_1.9.2             forcats_1.0.0              
[39] stringr_1.5.0               dplyr_1.1.0                
[41] purrr_1.0.1                 readr_2.1.4                
[43] tidyr_1.3.0                 tibble_3.1.8               
[45] ggplot2_3.4.1               tidyverse_2.0.0            

loaded via a namespace (and not attached):
  [1] utf8_1.2.3                    tidyselect_1.2.0             
  [3] RSQLite_2.3.0                 htmlwidgets_1.6.1            
  [5] BiocParallel_1.32.5           munsell_0.5.0                
  [7] codetools_0.2-19              interp_1.1-3                 
  [9] withr_2.5.0                   colorspace_2.1-0             
 [11] filelock_1.0.2                highr_0.10                   
 [13] knitr_1.42                    rstudioapi_0.14              
 [15] ggsignif_0.6.4                labeling_0.4.2               
 [17] git2r_0.31.0                  GenomeInfoDbData_1.2.9       
 [19] farver_2.1.1                  bit64_4.0.5                  
 [21] rprojroot_2.0.3               vctrs_0.5.2                  
 [23] generics_0.1.3                xfun_0.37                    
 [25] biovizBase_1.46.0             timechange_0.2.0             
 [27] R6_2.5.1                      bitops_1.0-7                 
 [29] cachem_1.0.6                  DelayedArray_0.24.0          
 [31] assertthat_0.2.1              vroom_1.6.1                  
 [33] promises_1.2.0.1              BiocIO_1.8.0                 
 [35] nnet_7.3-18                   gtable_0.3.1                 
 [37] workflowr_1.7.0               rlang_1.0.6                  
 [39] systemfonts_1.0.4             splines_4.2.2                
 [41] rtracklayer_1.58.0            rstatix_0.7.2                
 [43] lazyeval_0.2.2                dichromat_2.0-0.1            
 [45] plyranges_1.18.0              broom_1.0.3                  
 [47] checkmate_2.1.0               BiocManager_1.30.20          
 [49] yaml_2.3.7                    abind_1.4-5                  
 [51] backports_1.4.1               httpuv_1.6.9                 
 [53] Hmisc_4.8-0                   tools_4.2.2                  
 [55] ellipsis_0.3.2                jquerylib_0.1.4              
 [57] Rcpp_1.0.10                   base64enc_0.1-3              
 [59] progress_1.2.2                zlibbioc_1.44.0              
 [61] RCurl_1.98-1.10               prettyunits_1.1.1            
 [63] rpart_4.1.19                  deldir_1.0-6                 
 [65] cluster_2.1.4                 fs_1.6.1                     
 [67] data.table_1.14.8             ProtGenerics_1.30.0          
 [69] hms_1.1.2                     mime_0.12                    
 [71] evaluate_0.20                 xtable_1.8-4                 
 [73] XML_3.99-0.13                 jpeg_0.1-10                  
 [75] gridExtra_2.3                 compiler_4.2.2               
 [77] biomaRt_2.54.0                crayon_1.5.2                 
 [79] htmltools_0.5.4               later_1.3.0                  
 [81] tzdb_0.3.0                    Formula_1.2-4                
 [83] DBI_1.1.3                     rappdirs_0.3.3               
 [85] babelgene_22.9                Matrix_1.5-3                 
 [87] car_3.1-1                     cli_3.6.0                    
 [89] parallel_4.2.2                pkgconfig_2.0.3              
 [91] GenomicAlignments_1.34.0      foreign_0.8-84               
 [93] xml2_1.3.3                    svglite_2.1.1                
 [95] bslib_0.4.2                   webshot_0.5.4                
 [97] rvest_1.0.3                   digest_0.6.31                
 [99] cellranger_1.1.0              htmlTable_2.4.1              
[101] restfulr_0.0.15               curl_5.0.0                   
[103] shiny_1.7.4                   rjson_0.2.21                 
[105] lifecycle_1.0.3               jsonlite_1.8.4               
[107] carData_3.0-5                 BSgenome_1.66.3              
[109] fansi_1.0.4                   pillar_1.8.1                 
[111] lattice_0.20-45               KEGGREST_1.38.0              
[113] fastmap_1.1.0                 httr_1.4.4                   
[115] survival_3.5-3                interactiveDisplayBase_1.36.0
[117] glue_1.6.2                    png_0.1-8                    
[119] BiocVersion_3.16.0            bit_4.0.5                    
[121] stringi_1.7.12                sass_0.4.5                   
[123] blob_1.2.3                    latticeExtra_0.6-30          
[125] memoise_2.0.1                </code></pre>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
