<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Lachlan Baer" />


<title>DAR Analysis</title>

<script src="site_libs/header-attrs-2.22/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/flatly.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<script src="site_libs/kePrint-0.0.1/kePrint.js"></script>
<link href="site_libs/lightable-0.0.1/lightable.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">2022_MPSIII_3mBrainRNAseq</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="QC.html">QC</a>
</li>
<li>
  <a href="analysis_final.html">Initial analysis</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    DAR analysis
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="dar-analysis_AB.html">MPS-IIIA/B</a>
    </li>
    <li>
      <a href="dar-analysis_AC.html">MPS-IIIA/C</a>
    </li>
    <li>
      <a href="dar-analysis_BC.html">MPS-IIIB/C</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Enrichment analysis
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="enrichment_AB.html">MPS-IIIA/B</a>
    </li>
    <li>
      <a href="enrichment_AC.html">MPS-IIIA/C</a>
    </li>
    <li>
      <a href="enrichment_BC.html">MPS-IIIB/C</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" data-bs-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">DAR Analysis</h1>
<h3 class="subtitle">2022_MPSIII_3mBrainRNAseq dataset - AB
comparison</h3>
<h4 class="author">Lachlan Baer</h4>
<h4 class="date">16 August, 2023</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2023-08-16
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 6
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 1
</p>
<p>
<strong>Knit directory:</strong> <code>2022_MPSIII_3mBrainRNAseq/</code>
<span class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.0). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguncommittedchanges">
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> <strong>R Markdown file:</strong> uncommitted
changes </a>
</p>
</div>
<div id="strongRMarkdownfilestronguncommittedchanges"
class="panel-collapse collapse">
<div class="panel-body">
<p>The R Markdown file has unstaged changes. To know which version of
the R Markdown file created these results, you’ll want to first commit
it to the Git repo. If you’re still working on the analysis, you can
ignore this warning. When you’re finished, you can run
<code>wflow_publish</code> to commit the R Markdown file and build the
HTML.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20220423code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20220423)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20220423code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20220423)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomkarissab2022MPSIII3mBrainRNAseqtreef605249d259f82c12f1a833142d993a44d0a64cdtargetblankf605249a">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/karissa-b/2022_MPSIII_3mBrainRNAseq/tree/f605249d259f82c12f1a833142d993a44d0a64cd" target="_blank">f605249</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomkarissab2022MPSIII3mBrainRNAseqtreef605249d259f82c12f1a833142d993a44d0a64cdtargetblankf605249a"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/karissa-b/2022_MPSIII_3mBrainRNAseq/tree/f605249d259f82c12f1a833142d993a44d0a64cd" target="_blank">f605249</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rproj.user/
    Ignored:    analysis/dar-analysis_AB.html
    Ignored:    analysis/dar-analysis_AC.html
    Ignored:    analysis/dar-analysis_BC.html
    Ignored:    analysis/enrichment_AC.html
    Ignored:    analysis/enrichment_BC.html

Untracked files:
    Untracked:  analysis/enrichment_AB.Rmd
    Untracked:  analysis/enrichment_AC.Rmd
    Untracked:  analysis/enrichment_BC.Rmd
    Untracked:  data/dar-analysis/AB/geneDar.Rds
    Untracked:  data/dar-analysis/AC/geneDar.Rds
    Untracked:  data/dar-analysis/BC/geneDar.Rds

Unstaged changes:
    Modified:   analysis/_site.yml
    Modified:   analysis/dar-analysis_AB.Rmd
    Modified:   analysis/dar-analysis_AC.Rmd
    Modified:   analysis/dar-analysis_BC.Rmd
    Deleted:    data/dar-analysis/AB/DAR.Rds
    Deleted:    data/dar-analysis/AB/alleleCounts.Rds
    Deleted:    data/dar-analysis/AB/geneDAR.Rds
    Deleted:    data/dar-analysis/AC/DAR.Rds
    Deleted:    data/dar-analysis/AC/alleleCounts.Rds
    Deleted:    data/dar-analysis/AC/geneDAR.Rds
    Deleted:    data/dar-analysis/BC/DAR.Rds
    Deleted:    data/dar-analysis/BC/alleleCounts.Rds
    Deleted:    data/dar-analysis/BC/geneDAR.Rds

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/dar-analysis_AB.Rmd</code>) and
HTML (<code>docs/dar-analysis_AB.html</code>) files. If you’ve
configured a remote Git repository (see <code>?wflow_git_remote</code>),
click on the hyperlinks in the table below to view the files as they
were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/karissa-b/2022_MPSIII_3mBrainRNAseq/blob/f605249d259f82c12f1a833142d993a44d0a64cd/analysis/dar-analysis_AB.Rmd" target="_blank">f605249</a>
</td>
<td>
baerlachlan
</td>
<td>
2023-03-16
</td>
<td>
Initial DAR analysis on AB
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/karissa-b/2022_MPSIII_3mBrainRNAseq/f605249d259f82c12f1a833142d993a44d0a64cd/docs/dar-analysis_AB.html" target="_blank">f605249</a>
</td>
<td>
baerlachlan
</td>
<td>
2023-03-16
</td>
<td>
Initial DAR analysis on AB
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="setup" class="section level1">
<h1>Setup</h1>
<p>Load libraries and set global options.</p>
<pre class="r"><code>suppressPackageStartupMessages({
  ## Common
  library(tidyverse)
  library(magrittr)
  library(here)
  library(AnnotationHub)
  library(purrr)
  library(scales)
  library(kableExtra)
  # library(tictoc)  # Causes a bug in rowRanges()
  library(ggrepel)
  library(RColorBrewer)
  library(ggpubr)
  library(pander)
  library(rmarkdown)
  library(viridis)
  ## Project specific
  library(VariantAnnotation)
  library(Gviz)
  library(zoo)
  library(msigdbr)
  library(readxl)
  library(limma)
})</code></pre>
<p>An R package for DAR analysis is currently under development and will
be used for the following analysis. The core functionally of the package
is complete (as of 21/06/23). However some subtleties may change and new
features will be added as its developmentment progresses. This means
some of the following code may break, but a simple fix that produces the
same results will always exist. To install the package while it is still
in development, uncomment the code in the chunk below.</p>
<pre class="r"><code># BiocManager::install(&quot;baerlachlan/darr&quot;)
library(darr)</code></pre>
<pre class="r"><code>if (interactive()) setwd(here::here())
theme_set(theme_bw())</code></pre>
<pre class="r"><code># By default Gviz expects UCSC chromosome names
# We&#39;re working with Ensembl so let&#39;s turn this off
options(ucscChromosomeNames = FALSE)</code></pre>
<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>This document contains the Differential Allelic Representation (DAR)
analysis for the MPS-III A/B arm of the experiment.</p>
</div>
<div id="ensdb" class="section level2">
<h2>EnsDb</h2>
<pre class="r"><code>ens_species &lt;- &quot;Danio rerio&quot;
ens_release &lt;- &quot;101&quot;
ens_assembly &lt;- &quot;GRCz11&quot;</code></pre>
<p>Grab genome feature annotations for Danio rerio Ensembl release 101
(genome assembly GRCz11).</p>
<pre class="r"><code>ah &lt;- AnnotationHub() %&gt;%
  subset(species == ens_species) %&gt;%
  subset(rdataclass == &quot;EnsDb&quot;)
ahId &lt;- ah$ah_id[str_detect(ah$title, ens_release)]
ensDb &lt;- ah[[ahId]]</code></pre>
<pre class="r"><code># Filter for primary chromosomes
chrInfo &lt;- getChromInfoFromEnsembl(ens_assembly, release = ens_release) %&gt;%
  dplyr::filter(coord_system == &quot;chromosome&quot;)
primary_chrs &lt;- chrInfo$name</code></pre>
<pre class="r"><code>genes &lt;- genes(ensDb, filter = SeqNameFilter(primary_chrs))
mcols(genes) &lt;- mcols(genes)[
  c(&quot;gene_id&quot;, &quot;gene_name&quot;, &quot;gene_biotype&quot;, &quot;entrezid&quot;)
]</code></pre>
</div>
<div id="metadata" class="section level2">
<h2>Metadata</h2>
<p>Load and clean the sample metadata by renaming columns following
<code>R</code> naming conventions and setting categorical variables as
factors. Also append sample file basenames from a file I created
(<code>data/sample_basenames.csv</code>), as we will need these later.
This metadata contains information for all samples in the experiment, so
let’s filter for only samples in the A/B dataset.</p>
<pre class="r"><code>expArm &lt;- &quot;AB&quot;</code></pre>
<pre class="r"><code>metadata &lt;- read_xlsx(here(&quot;data/meta_final.xlsx&quot;)) %&gt;%
  dplyr::rename(
    rin = RIN,
    sex = Sex,
    rna_batch = rnaBatch,
    mps = MPS,
    uln = ULN,
    home_tank = `Home tank`,
    dob = DOB,
    death = Death
  ) %&gt;%
  mutate(
    sex = as.factor(sex),
    genotype = case_when(
      genotype == &quot;wt&quot; ~ &quot;WT&quot;,
      genotype == &quot;het&quot; ~ &quot;sgsh_het&quot;,
      genotype == &quot;naglu&quot; ~ &quot;MPSIIIB&quot;,
      genotype == &quot;sgsh&quot; ~ &quot;MPSIIIA&quot;,
      genotype == &quot;hgsnat&quot; ~ &quot;MPSIIIC&quot;
    ),
    genotype = as.factor(genotype),
    home_tank = as.factor(home_tank)
  ) %&gt;%
  left_join(read_csv(here(&quot;data/sample_basenames.csv&quot;))) %&gt;%
  dplyr::filter(mps == expArm) %&gt;%
  droplevels()</code></pre>
<pre class="r"><code>metadata %&gt;%
  kable(
    align = &quot;l&quot;,
    caption = &quot;Sample metadata for the MPS-III A/B arm of the experiment&quot;
  ) %&gt;%
  kable_styling(
    bootstrap_options = c(&quot;striped&quot;, &quot;hover&quot;, &quot;condensed&quot;, &quot;responsive&quot;)
  )</code></pre>
<table class="table table-striped table-hover table-condensed table-responsive" style="margin-left: auto; margin-right: auto;">
<caption>
Sample metadata for the MPS-III A/B arm of the experiment
</caption>
<thead>
<tr>
<th style="text-align:left;">
sample
</th>
<th style="text-align:left;">
rin
</th>
<th style="text-align:left;">
sex
</th>
<th style="text-align:left;">
genotype
</th>
<th style="text-align:left;">
rna_batch
</th>
<th style="text-align:left;">
mps
</th>
<th style="text-align:left;">
uln
</th>
<th style="text-align:left;">
home_tank
</th>
<th style="text-align:left;">
dob
</th>
<th style="text-align:left;">
death
</th>
<th style="text-align:left;">
basename
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
AB1
</td>
<td style="text-align:left;">
9.3
</td>
<td style="text-align:left;">
male
</td>
<td style="text-align:left;">
MPSIIIB
</td>
<td style="text-align:left;">
2022-02-10
</td>
<td style="text-align:left;">
AB
</td>
<td style="text-align:left;">
22-00136
</td>
<td style="text-align:left;">
AB_P1_lay1_tank2
</td>
<td style="text-align:left;">
2021-01-15
</td>
<td style="text-align:left;">
2021-04-22
</td>
<td style="text-align:left;">
22-00136_S1
</td>
</tr>
<tr>
<td style="text-align:left;">
AB4
</td>
<td style="text-align:left;">
7.1
</td>
<td style="text-align:left;">
female
</td>
<td style="text-align:left;">
MPSIIIA
</td>
<td style="text-align:left;">
2022-02-10
</td>
<td style="text-align:left;">
AB
</td>
<td style="text-align:left;">
22-00138
</td>
<td style="text-align:left;">
AB_P1_lay1_tank3
</td>
<td style="text-align:left;">
2021-01-15
</td>
<td style="text-align:left;">
2021-04-22
</td>
<td style="text-align:left;">
22-00138_S2
</td>
</tr>
<tr>
<td style="text-align:left;">
AB11
</td>
<td style="text-align:left;">
7.9
</td>
<td style="text-align:left;">
female
</td>
<td style="text-align:left;">
WT
</td>
<td style="text-align:left;">
2022-02-10
</td>
<td style="text-align:left;">
AB
</td>
<td style="text-align:left;">
22-00145
</td>
<td style="text-align:left;">
AB_P1_lay1_tank4
</td>
<td style="text-align:left;">
2021-01-15
</td>
<td style="text-align:left;">
2021-04-22
</td>
<td style="text-align:left;">
22-00145_S3
</td>
</tr>
<tr>
<td style="text-align:left;">
AB13
</td>
<td style="text-align:left;">
7.8
</td>
<td style="text-align:left;">
male
</td>
<td style="text-align:left;">
MPSIIIA
</td>
<td style="text-align:left;">
2022-02-10
</td>
<td style="text-align:left;">
AB
</td>
<td style="text-align:left;">
22-00147
</td>
<td style="text-align:left;">
AB_P1_lay1_tank2
</td>
<td style="text-align:left;">
2021-01-15
</td>
<td style="text-align:left;">
2021-04-22
</td>
<td style="text-align:left;">
22-00147_S4
</td>
</tr>
<tr>
<td style="text-align:left;">
AB14
</td>
<td style="text-align:left;">
9.7
</td>
<td style="text-align:left;">
female
</td>
<td style="text-align:left;">
sgsh_het
</td>
<td style="text-align:left;">
2022-02-10
</td>
<td style="text-align:left;">
AB
</td>
<td style="text-align:left;">
22-00148
</td>
<td style="text-align:left;">
AB_P1_lay1_tank2
</td>
<td style="text-align:left;">
2021-01-15
</td>
<td style="text-align:left;">
2021-04-22
</td>
<td style="text-align:left;">
22-00148_S5
</td>
</tr>
<tr>
<td style="text-align:left;">
AB16
</td>
<td style="text-align:left;">
6.6
</td>
<td style="text-align:left;">
female
</td>
<td style="text-align:left;">
WT
</td>
<td style="text-align:left;">
2022-02-11
</td>
<td style="text-align:left;">
AB
</td>
<td style="text-align:left;">
22-00150
</td>
<td style="text-align:left;">
AB_P1_lay1_tank2
</td>
<td style="text-align:left;">
2021-01-15
</td>
<td style="text-align:left;">
2021-04-22
</td>
<td style="text-align:left;">
22-00150_S6
</td>
</tr>
<tr>
<td style="text-align:left;">
AB17
</td>
<td style="text-align:left;">
8.6
</td>
<td style="text-align:left;">
female
</td>
<td style="text-align:left;">
WT
</td>
<td style="text-align:left;">
2022-02-11
</td>
<td style="text-align:left;">
AB
</td>
<td style="text-align:left;">
22-00151
</td>
<td style="text-align:left;">
AB_P1_lay1_tank2
</td>
<td style="text-align:left;">
2021-01-15
</td>
<td style="text-align:left;">
2021-04-22
</td>
<td style="text-align:left;">
22-00151_S7
</td>
</tr>
<tr>
<td style="text-align:left;">
AB18
</td>
<td style="text-align:left;">
6.6
</td>
<td style="text-align:left;">
male
</td>
<td style="text-align:left;">
sgsh_het
</td>
<td style="text-align:left;">
2022-02-11
</td>
<td style="text-align:left;">
AB
</td>
<td style="text-align:left;">
22-00152
</td>
<td style="text-align:left;">
AB_P1_lay1_tank2
</td>
<td style="text-align:left;">
2021-01-15
</td>
<td style="text-align:left;">
2021-04-22
</td>
<td style="text-align:left;">
22-00152_S8
</td>
</tr>
<tr>
<td style="text-align:left;">
AB20
</td>
<td style="text-align:left;">
8.1
</td>
<td style="text-align:left;">
female
</td>
<td style="text-align:left;">
MPSIIIB
</td>
<td style="text-align:left;">
2022-02-11
</td>
<td style="text-align:left;">
AB
</td>
<td style="text-align:left;">
22-00153
</td>
<td style="text-align:left;">
AB_P1_lay1_tank2
</td>
<td style="text-align:left;">
2021-01-15
</td>
<td style="text-align:left;">
2021-04-22
</td>
<td style="text-align:left;">
22-00153_S9
</td>
</tr>
<tr>
<td style="text-align:left;">
AB21
</td>
<td style="text-align:left;">
9.5
</td>
<td style="text-align:left;">
female
</td>
<td style="text-align:left;">
sgsh_het
</td>
<td style="text-align:left;">
2022-02-11
</td>
<td style="text-align:left;">
AB
</td>
<td style="text-align:left;">
22-00154
</td>
<td style="text-align:left;">
AB_P1_lay1_tank3
</td>
<td style="text-align:left;">
2021-01-15
</td>
<td style="text-align:left;">
2021-04-22
</td>
<td style="text-align:left;">
22-00154_S10
</td>
</tr>
<tr>
<td style="text-align:left;">
AB22
</td>
<td style="text-align:left;">
8.1
</td>
<td style="text-align:left;">
female
</td>
<td style="text-align:left;">
MPSIIIB
</td>
<td style="text-align:left;">
2022-02-11
</td>
<td style="text-align:left;">
AB
</td>
<td style="text-align:left;">
22-00155
</td>
<td style="text-align:left;">
AB_P1_lay1_tank3
</td>
<td style="text-align:left;">
2021-01-15
</td>
<td style="text-align:left;">
2021-04-22
</td>
<td style="text-align:left;">
22-00155_S11
</td>
</tr>
<tr>
<td style="text-align:left;">
AB23
</td>
<td style="text-align:left;">
7.4
</td>
<td style="text-align:left;">
female
</td>
<td style="text-align:left;">
MPSIIIA
</td>
<td style="text-align:left;">
2022-02-11
</td>
<td style="text-align:left;">
AB
</td>
<td style="text-align:left;">
22-00156
</td>
<td style="text-align:left;">
AB_P1_lay1_tank3
</td>
<td style="text-align:left;">
2021-01-15
</td>
<td style="text-align:left;">
2021-04-22
</td>
<td style="text-align:left;">
22-00156_S12
</td>
</tr>
<tr>
<td style="text-align:left;">
AB24
</td>
<td style="text-align:left;">
7.1
</td>
<td style="text-align:left;">
female
</td>
<td style="text-align:left;">
MPSIIIA
</td>
<td style="text-align:left;">
2022-02-11
</td>
<td style="text-align:left;">
AB
</td>
<td style="text-align:left;">
22-00157
</td>
<td style="text-align:left;">
AB_P1_lay1_tank3
</td>
<td style="text-align:left;">
2021-01-15
</td>
<td style="text-align:left;">
2021-04-22
</td>
<td style="text-align:left;">
22-00157_S13
</td>
</tr>
<tr>
<td style="text-align:left;">
AB25
</td>
<td style="text-align:left;">
8.0
</td>
<td style="text-align:left;">
female
</td>
<td style="text-align:left;">
sgsh_het
</td>
<td style="text-align:left;">
2022-02-11
</td>
<td style="text-align:left;">
AB
</td>
<td style="text-align:left;">
22-00158
</td>
<td style="text-align:left;">
AB_P1_lay1_tank3
</td>
<td style="text-align:left;">
2021-01-15
</td>
<td style="text-align:left;">
2021-04-22
</td>
<td style="text-align:left;">
22-00158_S14
</td>
</tr>
<tr>
<td style="text-align:left;">
AB26
</td>
<td style="text-align:left;">
9.1
</td>
<td style="text-align:left;">
female
</td>
<td style="text-align:left;">
MPSIIIB
</td>
<td style="text-align:left;">
2022-02-11
</td>
<td style="text-align:left;">
AB
</td>
<td style="text-align:left;">
22-00159
</td>
<td style="text-align:left;">
AB_P1_lay1_tank3
</td>
<td style="text-align:left;">
2021-01-15
</td>
<td style="text-align:left;">
2021-04-22
</td>
<td style="text-align:left;">
22-00159_S15
</td>
</tr>
<tr>
<td style="text-align:left;">
AB27
</td>
<td style="text-align:left;">
9.6
</td>
<td style="text-align:left;">
female
</td>
<td style="text-align:left;">
WT
</td>
<td style="text-align:left;">
2022-02-11
</td>
<td style="text-align:left;">
AB
</td>
<td style="text-align:left;">
22-00160
</td>
<td style="text-align:left;">
AB_P1_lay1_tank3
</td>
<td style="text-align:left;">
2021-01-15
</td>
<td style="text-align:left;">
2021-04-22
</td>
<td style="text-align:left;">
22-00160_S16
</td>
</tr>
</tbody>
</table>
<p>Set up colours for plotting by genotype. Going with a
colourblind-friendly palette (<code>Set2</code>) for best
accessibility.</p>
<pre class="r"><code>genoCols &lt;- metadata$genotype %&gt;%
  levels() %&gt;%
  length() %&gt;%
  brewer.pal(&quot;Set2&quot;) %&gt;%
  setNames(levels(metadata$genotype))</code></pre>
</div>
</div>
<div id="de-data" class="section level1">
<h1>DE data</h1>
<p>Load in the <code>dgeList</code> and <code>topTable</code>s from
Karissa’s DE analysis.</p>
<pre class="r"><code>dgeList &lt;- readRDS(here(&quot;data/R_objects/dge_AB.rds&quot;))</code></pre>
<pre class="r"><code>topTables &lt;- readRDS(here(&quot;data/R_objects/AB_toptab_cqn.rds&quot;))</code></pre>
<p>Also pull the gene IDs for DE genes.</p>
<pre class="r"><code>deGenes &lt;- sapply(topTables, function(x){
  dplyr::filter(x, DE) %&gt;%
    pull(gene_id)
}, simplify = FALSE)</code></pre>
</div>
<div id="differential-allelic-representation" class="section level1">
<h1>Differential Allelic Representation</h1>
<p>DAR analysis is performed using genotype calls from variant calling
software. Here the <code>GATK</code> best practices workflow for RNAseq
short variant discovery was followed to produce a multi-sample
<code>VCF</code> file containing information for any single nucleotide
locus that varied in at least one sample. Note that specific filters
were applied to only select loci with considerable evidence for
variation (see the Snakemake workflow in this repository for more
information).</p>
<div id="genotypes" class="section level2">
<h2>Genotypes</h2>
<p>Genotype data from the <code>VCF</code> file is parsed into a
<code>GRanges</code> class object using the <code>darr</code> package.
This file is large and therefore does not exist in the Git repository.
Phasing information is also removed when reading the genotypes as it is
not required for DAR analysis and complicates downstream processing.</p>
<pre class="r"><code>vcf_path &lt;- file.path(
  &quot;/hpcfs/users/a1647910/2022_MPSIII_3mBrainRNAseq/code&quot;,
  &quot;analysis-variants_AB/results/09_variants/6_select/all_samples.vcf.gz&quot;
)
genotypes &lt;- readGenotypes(
  file = vcf_path,
  unphase = TRUE,
  genome = genome(ensDb)
)</code></pre>
<p>Genotypes are reported as numeric indices. <code>0</code> indicates
the reference allele, <code>1</code> is the first alternate allele,
<code>2</code> is the second alternate allele, and so on. For DAR
analysis, we work directly with the indices as they are consistent
across all samples for a single variant position.</p>
</div>
<div id="allele-counts" class="section level2">
<h2>Allele counts</h2>
<p>We aim to calculate a DAR metric at each suitable variant locus. This
requires us to firstly summarise the genotype data into counts of the
alleles reported at each variant locus. First we define our sample
grouping structure as a <code>list</code>, where each element contains a
character vector of samples within a single group.</p>
<pre class="r"><code>samplesByGroup &lt;- metadata %&gt;%
  split(f = .$genotype) %&gt;%
  sapply(function(x){
    pull(x, basename)
  }, simplify = FALSE)</code></pre>
<p>Now at each locus we can count the number of alleles that exist
within the groups. This returns a <code>GRangesList</code> with each
element corresponding to a different sample group.</p>
<pre class="r"><code>alleleCounts &lt;- countAlleles(genotypes = genotypes, groups = samplesByGroup)</code></pre>
</div>
<div id="normalisation" class="section level2">
<h2>Normalisation</h2>
<p>Not all samples have genotype calls, so we normalise by converting
the allele counts into a proportion of total counts at each locus. We
also filter variant loci independently within each sample group with the
default criterion: number of samples with called genotypes &gt; number
of samples with missing genotypes.</p>
<pre class="r"><code>alleleProps &lt;- countsToProps(counts = alleleCounts, filter = TRUE)</code></pre>
</div>
<div id="dar" class="section level2">
<h2>DAR</h2>
<p>Now that we have normalised values of allelic representation at each
variant locus within our sample groups, we can calculate the DAR metric
between groups. Firstly, we define our contrasts of interest.</p>
<pre class="r"><code>contrasts &lt;- makeContrasts(
  A_vs_WT = MPSIIIA - WT,
  B_vs_WT = MPSIIIB - WT,
  het_vs_WT = sgsh_het - WT,
  levels = names(alleleProps)
)</code></pre>
<p>DAR is calculated by firstly determining the Euclidean distance
between allelic proportions of the contrasted sample groups. The
Euclidean distance is then converted to the DAR metric by dividing by
<span class="math display">\[\sqrt{2}\]</span>, resulting in an
easy-to-interpret value between 0 and 1, where 0 represents identical
allelic representation and 1 represents complete diversity. This is
handled within the <code>dar()</code> function by passing our allelic
proportions and intended contrasts as arguments.</p>
<p>Two types of DAR values are reported by the <code>dar()</code>
function as metadata columns of the resulting <code>GRanges</code>
objects:</p>
<ul>
<li><code>dar_origin</code>: The raw DAR values calculated at single
nucleotide positions (the origin) between sample groups. This represents
a DAR estimate at a precise locus.</li>
<li><code>dar_region</code>: The mean of raw DAR values in a region
surrounding the origin. The size of the region is controlled using the
<code>winSize</code> argument, which establishes an elastic sliding
window to average the specified number of dar_origin values. This
represents a DAR estimate for the genomic region surrounding the origin
locus.</li>
</ul>
<pre class="r"><code>winSize &lt;- 5</code></pre>
<p>With a chosen elastic window size of 5 loci, this will smooth the DAR
metric at each origin locus with the DAR values of the 2 loci either
side.</p>
<pre class="r"><code>dar &lt;- dar(props = alleleProps, contrasts = contrasts, winSize = winSize)</code></pre>
<p>We now have DAR values for the common loci between the two sample
groups in each contrast. Each element of the resulting
<code>GRangesList</code> object represents a single contrast and are
named as defined in our <code>contrasts</code> object.</p>
</div>
<div id="dar-ecdf" class="section level2">
<h2>DAR ECDF</h2>
<p>Plotting the empirical cumulative distribution function of DAR by
chromosome is useful to determine whether the mutant chromosome contains
higher DAR relative to other chromosomes. Here we highlight the
chromosome containing the mutation to visually identify if this is the
case. The <em>sgsh</em> gene implicated with MPS-IIIA is located on
chromosome 22. The <em>naglu</em> gene implicated with MPS-IIIB is
located on chromosome 24.</p>
<pre class="r"><code>plotDarECDF(dar$A_vs_WT, darVal = &quot;region&quot;, highlight = &quot;22&quot;) +
  scale_colour_manual(values = c(&quot;red&quot;, &quot;grey&quot;)) +
  labs(title = &quot;MPSIIIA vs WT&quot;)</code></pre>
<p><img src="figure/dar-analysis_AB.Rmd/darECDF_A_chr22-1.png" width="768" style="display: block; margin: auto;" /></p>
<p>This dataset provides an interesting scenario where samples were
genotyped at both the <em>sgsh</em> and <em>naglu</em> loci. Samples
within the MPS-IIIA experimental groups were selected so that they were
wildtype at the <em>naglu</em> locus. All samples in the dataset
originated from a single cross of double heterozygous parents, meaning
that all samples in the MPS-IIIA experimental groups should share
identical wildtype homologues of chromosome 24 (the chromosome of the
<em>naglu</em> locus), disregarding recombination. This means that
chromosome 24 should display relatively low DAR in the MPS-IIIA
comparisons, so let’s take a look.</p>
<pre class="r"><code>plotDarECDF(dar$A_vs_WT, darVal = &quot;region&quot;, highlight = &quot;24&quot;) +
  scale_colour_manual(values = c(&quot;red&quot;, &quot;grey&quot;)) +
  labs(title = &quot;MPSIIIA vs WT&quot;)</code></pre>
<p><img src="figure/dar-analysis_AB.Rmd/darECDF_A_chr24-1.png" width="768" style="display: block; margin: auto;" /></p>
<p>Chromosome 24 does indeed have very low DAR relative to the other
chromosomes. This contributes evidence to the argument that
genotype-based selection of samples is a key driver of DAR outcomes in
RNA-seq experiments.</p>
<p>An analogous situation is present in the MPS-IIIB samples, where the
samples should share near-identical wildtype homologues of chromosome
22. Let’s take a look at both.</p>
<pre class="r"><code>plotDarECDF(dar$B_vs_WT, darVal = &quot;region&quot;, highlight = &quot;24&quot;) +
  scale_colour_manual(values = c(&quot;red&quot;, &quot;grey&quot;)) +
  labs(title = &quot;MPSIIIB vs WT&quot;)</code></pre>
<p><img src="figure/dar-analysis_AB.Rmd/darECDF_B_chr24-1.png" width="768" style="display: block; margin: auto;" /></p>
<pre class="r"><code>plotDarECDF(dar$B_vs_WT, darVal = &quot;region&quot;, highlight = &quot;22&quot;) +
  scale_colour_manual(values = c(&quot;red&quot;, &quot;grey&quot;)) +
  labs(title = &quot;MPSIIIB vs WT&quot;)</code></pre>
<p><img src="figure/dar-analysis_AB.Rmd/darECDF_B_chr22-1.png" width="768" style="display: block; margin: auto;" /></p>
<p>Again we see relatively high DAR on the mutant chromosome (24), and
relatively low DAR on the wildtype controlled chromosome (22).</p>
<p>Lastly the MPS-IIIA heterozygotes.</p>
<pre class="r"><code>plotDarECDF(dar$het_vs_WT, darVal = &quot;region&quot;, highlight = &quot;22&quot;) +
  scale_colour_manual(values = c(&quot;red&quot;, &quot;grey&quot;)) +
  labs(title = &quot;sgsh+/- vs WT&quot;)</code></pre>
<p><img src="figure/dar-analysis_AB.Rmd/darECDF_het_chr24-1.png" width="768" style="display: block; margin: auto;" /></p>
<pre class="r"><code>plotDarECDF(dar$het_vs_WT, darVal = &quot;region&quot;, highlight = &quot;24&quot;) +
  scale_colour_manual(values = c(&quot;red&quot;, &quot;grey&quot;)) +
  labs(title = &quot;sgsh+/- vs WT&quot;)</code></pre>
<p><img src="figure/dar-analysis_AB.Rmd/darECDF_het_chr22-1.png" width="768" style="display: block; margin: auto;" /></p>
<p>Unsurprisingly, we see another example of high DAR on the mutant
chromosome, and low DAR on the controlled wildtype chromosome.</p>
<p>From these plots we can predict that for each comparison, an
overrepresentation of DE genes may exist on the respective mutant
chromosomes, due to the potential of eQTL differences. However, the
controlled wildtype chromosomes should exhibit far less of these false
positives due to low DAR.</p>
</div>
<div id="deg-dar-relationship" class="section level2">
<h2>DEG-DAR relationship</h2>
<p>We can produce a localised visualisation to assess whether
differentially expressed genes (DEG) may be prone eQTL artefacts by
plotting the trend in DAR along a chromosomal axis and overlaying the
position of DEGs. This functionality has not yet been implemented in the
<code>darr</code> package, so let’s define a function here instead.</p>
<pre class="r"><code>plotChrDar &lt;- function(
    dar, chromosome, features = GRanges(), foi = GRanges(),
    darType = &quot;b&quot;, highlight_features = TRUE,
    title = &quot;&quot;
){
  tracks &lt;- list()
  
  if (length(foi)) {
    foi &lt;- foi %&gt;%
      plyranges::filter(seqnames == chromosome) %&gt;%
      plyranges::mutate(symbol = gene_name)
    foi_track &lt;- GeneRegionTrack(
      range = foi,
      transcriptAnnotation = &quot;symbol&quot;,
      col = &quot;white&quot;,
      fill = &quot;white&quot;,
      showTranscriptId = TRUE,
      fontcolor.group = &quot;red&quot;,
      cex.group = 0.6,
      size = 0.5,
      name = NULL
    )
    tracks &lt;- append(tracks, foi_track)
  }
  
  axis_track &lt;- GenomeAxisTrack(
    add53 = TRUE,
    add35 = TRUE,
    name = paste0(&quot;Chr&quot;, chromosome),
    showTitle = TRUE,
    size = 1
  )
  if (length(foi)) {
    axis_track &lt;- HighlightTrack(
      trackList = list(axis_track),
      range = foi
    )
  }
  tracks &lt;- append(tracks, axis_track)
  
  if (length(features)) {
    features &lt;- features %&gt;%
      plyranges::filter(seqnames == chromosome) %&gt;%
      plyranges::mutate(symbol = gene_name)
    features_track &lt;- AnnotationTrack(
      range = features,
      name = &quot;DE&quot;,
      shape = &quot;box&quot;,
      fill = &quot;darkgray&quot;,
      group = features$gene_name,
      groupAnnotation = &quot;group&quot;,
      fontcolor.group = 1,
      cex.group = 0.6,
      size = 0.4
    )
    tracks &lt;- append(tracks, features_track)
  }
  
  dar_track &lt;- dar[,&quot;dar_region&quot;] %&gt;%
    plyranges::filter(seqnames == chromosome) %&gt;%
    DataTrack(
      type = darType,
      name = &quot;DAR&quot;,
      size = 8,
      window = -1,
      windowSize = 1,
      cex = 0.4,
      col = &quot;grey20&quot;,
      col.axis = &quot;black&quot;,
      yTicksAt = seq(0, 1, 0.1),
      ylim = c(0, 1)
    )
  if (highlight_features) {
    if (all(length(features) &amp; length(foi))) {
      ranges &lt;- c(features, foi)
    } else if (length(features)) {
      ranges &lt;- features
    } else if (length(foi)) {
      ranges &lt;- foi
    } else {
      ranges &lt;- GRanges()
    }
    dar_track &lt;- HighlightTrack(
      trackList = list(dar_track),
      range = ranges,
      col = c(rep(&quot;#ffe0e0&quot;, length(features)), rep(&quot;red&quot;, length(foi))),
      fill = c(rep(&quot;#ffe0e0&quot;, length(features)), rep(&quot;red&quot;, length(foi)))
    )
  }
  tracks &lt;- append(tracks, dar_track)
  
  plotTracks(
    trackList = tracks,
    main = title,
    cex.main = 1,
    cex.title = 0.6,
    col.title = &quot;black&quot;,
    background.title = &quot;white&quot;
  )
}</code></pre>
<div id="mps-iiia" class="section level3">
<h3>MPS-IIIA</h3>
<p>First let’s look at the mutant chromosome, as this was identified as
having the highest average DAR from the ECDF we plotted above.</p>
<pre class="r"><code>plotChrDar(
  dar = dar$A_vs_WT,
  chromosome = &quot;22&quot;,
  features = genes[genes$gene_id %in% deGenes$`MPS-IIIA`],
  foi = genes[genes$gene_name == &quot;sgsh&quot;]
)</code></pre>
<p><img src="figure/dar-analysis_AB.Rmd/chrDar_A_chr22-1.png" width="768" style="display: block; margin: auto;" /></p>
<p>And comparing this to the controlled wildtype chromosome 24:</p>
<pre class="r"><code>plotChrDar(
  dar = dar$A_vs_WT,
  chromosome = &quot;24&quot;,
  features = genes[genes$gene_id %in% deGenes$`MPS-IIIA`]
)</code></pre>
<p><img src="figure/dar-analysis_AB.Rmd/chrDar_A_chr24-1.png" width="768" style="display: block; margin: auto;" /></p>
</div>
<div id="mps-iiib" class="section level3">
<h3>MPS-IIIB</h3>
<p>The mutant chromosome also showed high DAR in the MPS-IIIB vs. WT
comparison.</p>
<pre class="r"><code>plotChrDar(
  dar = dar$B_vs_WT,
  chromosome = &quot;24&quot;,
  features = genes[genes$gene_id %in% deGenes$`MPS-IIIB`],
  foi = genes[genes$gene_name == &quot;naglu&quot;]
)</code></pre>
<p><img src="figure/dar-analysis_AB.Rmd/chrDar_B_chr24-1.png" width="768" style="display: block; margin: auto;" /></p>
<p>Interestingly we see a region of low DAR along the mutant chromosome,
however no DEGs exist in these regions. It is likely that many of the
genes classified as DE are eQTLs as they are found within regions of
high DAR.</p>
<p>Again comparing to the controlled wildtype chromosome 22:</p>
<pre class="r"><code>plotChrDar(
  dar = dar$B_vs_WT,
  chromosome = &quot;22&quot;,
  features = genes[genes$gene_id %in% deGenes$`MPS-IIIB`]
)</code></pre>
<p><img src="figure/dar-analysis_AB.Rmd/chrDar_B_chr22-1.png" width="768" style="display: block; margin: auto;" /></p>
</div>
<div id="sgsh-" class="section level3">
<h3>sgsh+/-</h3>
<p>The heterozygous mutant <em>sgsh</em> vs WT comparison provides an
interesting situation, where MPS-IIIA disease phenotypes are not
expected to be present. Let’s see how the trend in DAR aligns with DE
genes on the mutant chromosome.</p>
<pre class="r"><code>plotChrDar(
  dar = dar$het_vs_WT,
  chromosome = &quot;22&quot;,
  features = genes[genes$gene_id %in% deGenes$`sgsh/+`],
  foi = genes[genes$gene_name == &quot;sgsh&quot;]
)</code></pre>
<p><img src="figure/dar-analysis_AB.Rmd/chrDar_het_chr22-1.png" width="768" style="display: block; margin: auto;" /></p>
<p>The majority of DE genes are found in a similar region, where DAR
reaches is peak value of ~0.75. These suggests that allelic differences
between sample groups are prevalent around this region and genes may be
DE due to eQTLs.</p>
<p>Let’s look again at the controlled wildtype chromosome 24.</p>
<pre class="r"><code>plotChrDar(
  dar = dar$het_vs_WT,
  chromosome = &quot;24&quot;,
  features = genes[genes$gene_id %in% deGenes$`sgsh/+`]
)</code></pre>
<p><img src="figure/dar-analysis_AB.Rmd/chrDar_het_chr24-1.png" width="768" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="assign-gene-dar" class="section level2">
<h2>Assign gene DAR</h2>
<p>We can assign DAR values to genes within the dataset to provide an
estimate of the potential for a gene’s expression to be impacted by
eQTLs. eQTLs don’t necessarily exist within the gene itself, so it is
preferable to use smoothed DAR values (<code>dar_region</code> metadata
column in the <code>dar</code> object) as an estimate of DAR within a
region. First we need to convert the ranges of our <code>dar</code>
object to represent regions. Here we also extend the edge ranges of each
chromosome to encompass the entire chromosome, such that genes in these
regions also get assigned a DAR value.</p>
<pre class="r"><code>darRegions &lt;- convertRanges(dar, extendEdges = TRUE)</code></pre>
<pre class="r"><code>geneDar &lt;- assignFeatureDar(genes, darRegions, darVal = &quot;region&quot;)</code></pre>
<p>We will save these assigned values for use in further analysis.</p>
<pre class="r"><code>geneDar_path &lt;- file.path(here(&quot;data/dar-analysis&quot;), expArm, &quot;geneDar.Rds&quot;)
saveRDS(geneDar, geneDar_path)</code></pre>
</div>
<div id="logfc-comparison-between-mps-iiia-and-sgsh"
class="section level2">
<h2>LogFC comparison between MPS-IIIA and sgsh/+</h2>
<pre class="r"><code>lfc_comp &lt;- topTables %&gt;%
  bind_rows() %&gt;%
  dplyr::filter(coef %in% c(&quot;MPS-IIIA&quot;, &quot;sgsh/+&quot;)) %&gt;%
  dplyr::filter(chromosome == &quot;22&quot;) %&gt;%
  droplevels() %&gt;%
  dplyr::select(gene_id, logFC, coef) %&gt;%
  spread(key = &quot;coef&quot;, value = &quot;logFC&quot;) %&gt;%
  dplyr::rename(lfc_hom = `MPS-IIIA`, lfc_het = `sgsh/+`) %&gt;%
  mutate(
    DE = case_when(
      gene_id %in%
        intersect(deGenes$`MPS-IIIA`, deGenes$`sgsh/+`)
      ~ &quot;Both&quot;, 
      gene_id %in% deGenes$`sgsh/+`[!deGenes$`sgsh/+` %in% deGenes$`MPS-IIIA`]
      ~ &quot;Het only&quot;,
      gene_id %in% deGenes$`MPS-IIIA`[!deGenes$`MPS-IIIA` %in% deGenes$`sgsh/+`]
      ~ &quot;Hom only&quot;, 
      TRUE ~ &quot;None&quot;
    )
  ) %&gt;%
  left_join(as_tibble(geneDar$het_vs_WT))</code></pre>
<pre class="r"><code>lfc_comp %&gt;%
  ggplot(aes(x = lfc_hom, y = lfc_het)) +
  geom_point(
    aes(fill = DE, colour = DE),
    size = 2.5,
    colour = &quot;black&quot;,
    shape = 21,
    alpha = 0.5
  ) +
  geom_abline(slope = 1) +
  theme(aspect.ratio = 1) +
  scale_fill_manual(values = brewer.pal(3, &quot;Dark2&quot;)) +
  labs(size = &quot;DAR&quot;, x = &quot;MPS-IIIA homozygous&quot;, y = &quot;MPS-IIIA heterozygous&quot;) +
  ggtitle(&quot;LogFC of genes on chromosome 22 in MPS-IIIA hom vs het&quot;)</code></pre>
<p><img src="figure/dar-analysis_AB.Rmd/unnamed-chunk-3-1.png" width="768" style="display: block; margin: auto;" /></p>
<pre class="r"><code>lfc_comp %&gt;%
  dplyr::filter(DE != &quot;None&quot;) %&gt;%
  # dplyr::filter(DE != &quot;None&quot; &amp; lfc_hom &gt; 0) %&gt;%
  ggplot(aes(x = lfc_hom, y = lfc_het)) +
  geom_point(
    aes(fill = DE, colour = DE),
    size = 2.5,
    colour = &quot;black&quot;,
    shape = 21,
    alpha = 0.5
  ) +
  geom_text_repel(
    aes(label = gene_name),
    size = 3,
    data = . %&gt;% dplyr::filter(lfc_hom &gt; 0 | lfc_hom &lt; -5)
  ) +
  geom_abline(slope = 1) +
  theme(aspect.ratio = 1) +
  scale_fill_manual(values = brewer.pal(3, &quot;Dark2&quot;)) +
  labs(size = &quot;DAR&quot;, x = &quot;MPS-IIIA homozygous&quot;, y = &quot;MPS-IIIA heterozygous&quot;) +
  ggtitle(&quot;LogFC of genes on chromosome 22 in MPS-IIIA hom vs het&quot;)</code></pre>
<p><img src="figure/dar-analysis_AB.Rmd/unnamed-chunk-4-1.png" width="768" style="display: block; margin: auto;" /></p>
<p>Notes:</p>
<ul>
<li>All DE genes in hets also DE in homs</li>
<li>All upregulated DE genes have similar logFC between homs and
hets</li>
</ul>
<pre class="r"><code>lfc_goi &lt;- lfc_comp %&gt;%
  # dplyr::filter(DE != &quot;None&quot;)
  # dplyr::filter(lfc_hom &gt; 0)
  dplyr::filter(lfc_hom &gt; 0 &amp; DE != &quot;None&quot;)
  # dplyr::filter(DE == &quot;Hom only&quot; &amp; lfc_hom &gt; 0)
soi &lt;- metadata %&gt;%
  dplyr::filter(genotype == &quot;sgsh_het&quot;) %&gt;%
  # dplyr::filter(genotype == &quot;MPSIIIA&quot;) %&gt;%
  pull(sample)
dgeList_goi &lt;- dgeList[lfc_goi$gene_id, soi]
dgeList_goi$offset &lt;- NULL
cpm_goi &lt;- dgeList_goi %&gt;%
  cpm(log = TRUE)</code></pre>
<pre class="r"><code>apply(cpm_goi, 1, var) %&gt;%
  enframe(name = &quot;gene_id&quot;, value = &quot;cpm_var&quot;) %&gt;%
  left_join(lfc_goi) %&gt;%
  ggplot(aes(x = lfc_hom, y = lfc_het)) +
  geom_point(
    aes(fill = DE, colour = DE, size = cpm_var),
    colour = &quot;black&quot;,
    shape = 21,
    alpha = 0.5
  ) +
  geom_text_repel(
    aes(label = gene_name),
    size = 3,
    data = . %&gt;% dplyr::filter(lfc_hom &gt; 0 | lfc_hom &lt; -5)
  ) +
  geom_abline(slope = 1) +
  theme(aspect.ratio = 1) +
  scale_fill_manual(values = brewer.pal(3, &quot;Dark2&quot;)) +
  labs(size = &quot;logCPM variance in hets&quot;, x = &quot;MPS-IIIA homozygous&quot;, y = &quot;MPS-IIIA heterozygous&quot;) +
  ggtitle(&quot;LogFC of genes on chromosome 22 in MPS-IIIA hom vs het&quot;)</code></pre>
<p><img src="figure/dar-analysis_AB.Rmd/unnamed-chunk-6-1.png" width="768" style="display: block; margin: auto;" /></p>
<pre class="r"><code>pca &lt;- cpm_goi %&gt;%
  t() %&gt;%
  prcomp()
pcaVars &lt;- percent_format(0.1)(summary(pca)$importance[&quot;Proportion of Variance&quot;,])
pca$x %&gt;%
  as.data.frame() %&gt;%
  rownames_to_column(&quot;sample&quot;) %&gt;%
  left_join(metadata) %&gt;%
  as_tibble() %&gt;%
  ggplot(aes(PC1, PC2, colour = genotype, fill = genotype, group = genotype)) +
  geom_point(size = 2) +
  geom_text_repel(aes(label = sample), show.legend = FALSE) +
  # stat_ellipse(geom = &quot;polygon&quot;, alpha = 0.05, show.legend = FALSE) +
  guides(fill = &quot;none&quot;) +
  scale_colour_manual(values = genoCols) +
  labs(
    x = paste0(&quot;PC1 (&quot;, pcaVars[[&quot;PC1&quot;]], &quot;)&quot;),
    y = paste0(&quot;PC2 (&quot;, pcaVars[[&quot;PC2&quot;]], &quot;)&quot;),
    colour = &quot;Genotype&quot;
  )</code></pre>
<p><img src="figure/dar-analysis_AB.Rmd/unnamed-chunk-7-1.png" width="768" style="display: block; margin: auto;" /></p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.3.1 (2023-06-16)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 20.04.6 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.9.0 
LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.9.0

locale:
 [1] LC_CTYPE=en_AU.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_AU.UTF-8        LC_COLLATE=en_AU.UTF-8    
 [5] LC_MONETARY=en_AU.UTF-8    LC_MESSAGES=en_AU.UTF-8   
 [7] LC_PAPER=en_AU.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_AU.UTF-8 LC_IDENTIFICATION=C       

time zone: Australia/Adelaide
tzcode source: system (glibc)

attached base packages:
[1] grid      stats4    stats     graphics  grDevices utils     datasets 
[8] methods   base     

other attached packages:
 [1] edgeR_3.42.4                ensembldb_2.24.0           
 [3] AnnotationFilter_1.24.0     GenomicFeatures_1.52.1     
 [5] AnnotationDbi_1.62.1        darr_0.99.0                
 [7] limma_3.56.2                readxl_1.4.2               
 [9] msigdbr_7.5.1               zoo_1.8-12                 
[11] Gviz_1.44.0                 VariantAnnotation_1.46.0   
[13] Rsamtools_2.16.0            Biostrings_2.68.1          
[15] XVector_0.40.0              SummarizedExperiment_1.30.2
[17] Biobase_2.60.0              GenomicRanges_1.52.0       
[19] GenomeInfoDb_1.36.1         IRanges_2.34.1             
[21] S4Vectors_0.38.1            MatrixGenerics_1.12.2      
[23] matrixStats_1.0.0           viridis_0.6.3              
[25] viridisLite_0.4.2           rmarkdown_2.22             
[27] pander_0.6.5                ggpubr_0.6.0               
[29] RColorBrewer_1.1-3          ggrepel_0.9.3              
[31] kableExtra_1.3.4            scales_1.2.1               
[33] AnnotationHub_3.8.0         BiocFileCache_2.8.0        
[35] dbplyr_2.3.2                BiocGenerics_0.46.0        
[37] here_1.0.1                  magrittr_2.0.3             
[39] lubridate_1.9.2             forcats_1.0.0              
[41] stringr_1.5.0               dplyr_1.1.2                
[43] purrr_1.0.1                 readr_2.1.4                
[45] tidyr_1.3.0                 tibble_3.2.1               
[47] ggplot2_3.4.2               tidyverse_2.0.0            
[49] workflowr_1.7.0            

loaded via a namespace (and not attached):
  [1] later_1.3.1                   BiocIO_1.10.0                
  [3] bitops_1.0-7                  filelock_1.0.2               
  [5] cellranger_1.1.0              XML_3.99-0.14                
  [7] rpart_4.1.19                  lifecycle_1.0.3              
  [9] rstatix_0.7.2                 rprojroot_2.0.3              
 [11] vroom_1.6.3                   processx_3.8.1               
 [13] lattice_0.21-8                backports_1.4.1              
 [15] Hmisc_5.1-0                   sass_0.4.6                   
 [17] jquerylib_0.1.4               yaml_2.3.7                   
 [19] httpuv_1.6.11                 DBI_1.1.3                    
 [21] abind_1.4-5                   zlibbioc_1.46.0              
 [23] rvest_1.0.3                   biovizBase_1.48.0            
 [25] RCurl_1.98-1.12               nnet_7.3-19                  
 [27] rappdirs_0.3.3                git2r_0.32.0                 
 [29] GenomeInfoDbData_1.2.10       svglite_2.1.1                
 [31] codetools_0.2-19              DelayedArray_0.26.3          
 [33] xml2_1.3.4                    tidyselect_1.2.0             
 [35] farver_2.1.1                  base64enc_0.1-3              
 [37] webshot_0.5.4                 GenomicAlignments_1.36.0     
 [39] jsonlite_1.8.5                ellipsis_0.3.2               
 [41] Formula_1.2-5                 systemfonts_1.0.4            
 [43] tools_4.3.1                   progress_1.2.2               
 [45] Rcpp_1.0.10                   glue_1.6.2                   
 [47] gridExtra_2.3                 xfun_0.39                    
 [49] withr_2.5.0                   BiocManager_1.30.21          
 [51] fastmap_1.1.1                 latticeExtra_0.6-30          
 [53] fansi_1.0.4                   callr_3.7.3                  
 [55] digest_0.6.32                 timechange_0.2.0             
 [57] R6_2.5.1                      mime_0.12                    
 [59] colorspace_2.1-0              jpeg_0.1-10                  
 [61] dichromat_2.0-0.1             biomaRt_2.56.1               
 [63] RSQLite_2.3.1                 utf8_1.2.3                   
 [65] generics_0.1.3                data.table_1.14.8            
 [67] rtracklayer_1.60.0            prettyunits_1.1.1            
 [69] httr_1.4.6                    htmlwidgets_1.6.2            
 [71] S4Arrays_1.0.4                whisker_0.4.1                
 [73] pkgconfig_2.0.3               gtable_0.3.3                 
 [75] blob_1.2.4                    htmltools_0.5.5              
 [77] carData_3.0-5                 plyranges_1.20.0             
 [79] ProtGenerics_1.32.0           png_0.1-8                    
 [81] knitr_1.43                    rstudioapi_0.14              
 [83] tzdb_0.4.0                    rjson_0.2.21                 
 [85] checkmate_2.2.0               curl_5.0.1                   
 [87] cachem_1.0.8                  BiocVersion_3.17.1           
 [89] parallel_4.3.1                foreign_0.8-84               
 [91] restfulr_0.0.15               pillar_1.9.0                 
 [93] vctrs_0.6.3                   promises_1.2.0.1             
 [95] car_3.1-2                     xtable_1.8-4                 
 [97] cluster_2.1.4                 htmlTable_2.4.1              
 [99] evaluate_0.21                 locfit_1.5-9.8               
[101] cli_3.6.1                     compiler_4.3.1               
[103] rlang_1.1.1                   crayon_1.5.2                 
[105] ggsignif_0.6.4                labeling_0.4.2               
[107] interp_1.1-4                  ps_1.7.5                     
[109] getPass_0.2-2                 fs_1.6.2                     
[111] stringi_1.7.12                deldir_1.0-9                 
[113] BiocParallel_1.34.2           babelgene_22.9               
[115] munsell_0.5.0                 lazyeval_0.2.2               
[117] Matrix_1.5-4.1                BSgenome_1.68.0              
[119] hms_1.1.3                     bit64_4.0.5                  
[121] KEGGREST_1.40.0               shiny_1.7.4                  
[123] highr_0.10                    interactiveDisplayBase_1.38.0
[125] broom_1.0.5                   memoise_2.0.1                
[127] bslib_0.5.0                   bit_4.0.5                    </code></pre>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
