---
title: "_BCAC"
author: "Karissa Barthelson"
date: "2022-12-12"
output: workflowr::wflow_html
editor_options:
  chunk_output_type: console
---
## Introduction
```{r setup, include=FALSE}
knitr::opts_chunk$set(
  autodep = TRUE,
  echo = TRUE,
  warning = FALSE,
  message = FALSE,
  fig.align = "center", 
  out.width ="75%", 
  out.height = "75%"
)
```

```{r loadLibs}
library(tidyverse)
library(magrittr)
library(readxl)
library(ngsReports)
library(AnnotationHub)
library(biomaRt)
library(msigdbr)
library(pander)
library(scales)
library(pheatmap)
library(ggpubr)
library(ggrepel)
library(ggfortify)
library(RColorBrewer)
library(UpSetR)
library(edgeR)
library(cqn)
library(goseq)
library(fgsea)
library(harmonicmeanp)
theme_set(theme_bw())
```

```{r anno}
ah <- AnnotationHub() %>%
	subset(species == "Danio rerio") %>%
	subset(rdataclass == "EnsDb")

ensDb <- ah[["AH83189"]] # for release 101, latest version in Ah
grTrans <- transcripts(ensDb)
trLengths <- exonsBy(ensDb, "tx") %>%
	width() %>%
	vapply(sum, integer(1))
mcols(grTrans)$length <- trLengths[names(grTrans)]
gcGene <- grTrans %>%
  mcols() %>%
  as.data.frame() %>%
  dplyr::select(gene_id, tx_id, gc_content, length) %>%
  as_tibble() %>%
  group_by(gene_id) %>%
  summarise(
    gc_content = sum(gc_content*length) / sum(length),
    length = ceiling(median(length))
  )
grGenes <- genes(ensDb)
mcols(grGenes) %<>%
  as.data.frame() %>%
  left_join(gcGene) %>%
  as.data.frame() %>%
  DataFrame()
```

```{r meta}
meta <- 
  read_excel("data/meta_final.xlsx") %>% 
  mutate(genotype = case_when(
    genotype == "wt" ~ "WT", 
    genotype == "het" ~ "sgsh/+",
    genotype == "naglu" ~ "MPS-IIIB", 
    genotype == "sgsh" ~ "MPS-IIIA",
    genotype == "hgsnat" ~ "MPS-IIIC"
  ) %>% 
    factor(levels = c("WT", "sgsh/+", "MPS-IIIA", "MPS-IIIB", "MPS-IIIC")),
  Tank = as.factor(`Home tank`)
  ) #%>% 
  # mutate(Tank = as.factor(Tank),
  #        batchKilled = as.factor(batchKilled), 
  #        Sex = as.factor(Sex)) %>% 
  # dplyr::filter(!is.na(ULN))

featureCounts <- 
  read_delim(file = "data/mergedBCAC/fc/counts.out", delim = "\t", skip = 1) %>%
  set_names(basename(names(.))) %>% 
  set_names(names(.) %>% str_remove(pattern = "_S[0-9]+_merged.Aligned.sortedByCoord.dedup.out.bam")) %>% 
  as_tibble() %>% 
  dplyr::select(-c(Chr, Start, End, Length, Strand)) %>% 
  gather(key = "ULN", value = "counts", starts_with("22")) %>% 
  left_join(meta) %>% 
  dplyr::select(Geneid, counts, sample) %>% 
  spread(key = "sample", value = "counts") %>% 
  column_to_rownames("Geneid") %>% 
  .[,-(dim(.)[2])]# omit the SAGC neg control lane, the final col of the df. 
```

## filter lowly expressed genes

```{r}
min.lib.size = featureCounts %>% 
    as.matrix() %>% 
    DGEList() %>%
    calcNormFactors() %>% 
  .$samples %>% 
  .$lib.size %>% 
  min()
```

Genes which are lowly expressed are uninformative for differential gene expression analysis. A general rule of thumb for considering a gene to be lowly expressed if it contains a count per million (CPM) of more than 10/(min_lib_size/1000000) in at least one of the experimental groups. This will allow us . The min lib size in this analysis is `r comma(min.lib.size)`. So the minimum CPM in this analysis will be `r 10/(min.lib.size/1000000)`.
 

```{r filt}
a <- featureCounts %>% 
    cpm(log = TRUE) %>%
    as.data.frame() %>% 
  pivot_longer(
    cols = everything(),
    names_to = "sample",
    values_to = "logCPM"
  ) %>%
  split(f = .$sample) %>%
  lapply(function(x){
    d <- density(x$logCPM)
    tibble(
      sample = unique(x$sample),
      x = d$x,
      y = d$y
    )
  }) %>%
  bind_rows() %>%
  left_join(meta) %>% 
  ggplot(aes(x, y, colour = genotype, group = sample)) +
  geom_line() +
  scale_y_continuous(limits = c(0,0.35)) +
  labs(
    x = "logCPM",
    y = "Density",
    colour = "Genotype"
  )+
  ggtitle("Before filtering") +
  theme_bw()

b <- featureCounts %>% 
  .[rowSums(cpm(.) >= 2.435138) >= 6,] %>% 
    cpm(log = TRUE) %>%
    as.data.frame() %>% 
  pivot_longer(
    cols = everything(),
    names_to = "sample",
    values_to = "logCPM"
  ) %>%
  split(f = .$sample) %>%
  lapply(function(x){
    d <- density(x$logCPM)
    tibble(
      sample = unique(x$sample),
      x = d$x,
      y = d$y
    )
  }) %>%
  bind_rows() %>%
  left_join(meta) %>% 
  ggplot(aes(x, y, colour = genotype, group = sample)) +
  geom_line() +
  scale_y_continuous(limits = c(0,0.35)) +
  labs(
    x = "logCPM",
    y = "Density",
    colour = "Genotype"
  )+
  ggtitle("After filtering")

ggarrange(a, b, common.legend = TRUE)
```


```{r dge}
x <- featureCounts %>% 
  as.matrix() %>% 
  .[rowSums(cpm(.) >= 3) >= 6,] %>% 
  DGEList(
    samples = tibble(sample = colnames(.)) %>%
      left_join(meta),
    genes = grGenes[rownames(.)] %>%
      as.data.frame() %>%
      dplyr::select(
        chromosome = seqnames, start, end, 
        gene_id, gene_name, gene_biotype, description, entrezid
      ) %>% 
      left_join(gcGene) %>% 
      as_tibble()
  ) %>%
  calcNormFactors()
```

## PCA
### Both AC and BC

2 distinct groups are observed across PC2. And this is explained by the 2 families of fish which comprise this dataset. Therefore, I should analyse them seperately

```{r}
x$counts %>% 
  cpm(log=TRUE) %>% 
  t() %>% 
  prcomp() %>% 
  autoplot(data = tibble(sample = rownames(.$x)) %>%
      left_join(x$samples),
    colour = "MPS", 
    size = 6
  ) +
  theme(aspect.ratio = 1, 
        legend.position = "bottom") +
  labs(colour = "MPS comparison family") +
  scale_color_brewer(palette = "Set2")
```

### AC only

The AC group came from 1 family, raised across 5 tanks side by side in the same system. The MPS-IIIA samples appear to be clustering away from the other 2. 

```{r}
BCsamps <- meta %>% 
  dplyr::filter(MPS == "BC") %>% 
  .$sample

ACsamps <- meta %>% 
  dplyr::filter(MPS == "AC") %>% 
  .$sample

ggarrange(
x$counts %>% 
  .[,ACsamps] %>% 
  cpm(log=TRUE) %>% 
  t() %>% 
  prcomp() %>% 
  autoplot(data = tibble(sample = rownames(.$x)) %>%
      left_join(x$samples),
    colour = "genotype", 
    size = 6
  ) +
  theme(aspect.ratio = 1) +
  ggtitle("by genotype") +
  labs(colour = ""),

x$counts %>% 
  .[,ACsamps] %>% 
  cpm(log=TRUE) %>% 
  t() %>% 
  prcomp() %>% 
  autoplot(data = tibble(sample = rownames(.$x)) %>%
      left_join(x$samples),
    colour = "Tank", 
    size = 6
  ) +
  theme(aspect.ratio = 1) +
  labs(colour = "") +
  ggtitle("by tank")
)
```

### BC only

The BC group of fish came from 2 sets of parents, across multiple lays/ This is evident in the plot below. The effect of "Home Tank" cannot be ignored in a DE analysis. 
```{r}
ggarrange(
x$counts %>% 
  .[,BCsamps] %>% 
  cpm(log=TRUE) %>% 
  t() %>% 
  prcomp() %>% 
  autoplot(data = tibble(sample = rownames(.$x)) %>%
      left_join(x$samples),
    colour = "genotype", 
    size = 6
  ) +
  theme(aspect.ratio = 1) +
  ggtitle("by genotype") +
  labs(colour = ""),

x$counts %>% 
  .[,BCsamps] %>% 
  cpm(log=TRUE) %>% 
  t() %>% 
  prcomp() %>% 
  autoplot(data = tibble(sample = rownames(.$x)) %>%
      left_join(x$samples),
    colour = "Tank", 
    size = 6
  ) +
  theme(aspect.ratio = 1) +
  labs(colour = "") +
  ggtitle("by tank")
)
```
From hereout, I will process the 2 comparisons separately. 

# AC
## initial DE analysis

```{r}
x.ac <- x[,ACsamps]

design.ac <- model.matrix(~genotype + Tank, data = x.ac$samples %>% 
                            droplevels()) %>% # get rid of the unused factor levels
  set_colnames(gsub(colnames(.), pattern = "genotype", replacement = ""))

fit_1_ac <- x.ac %>% 
  estimateDisp(design.ac) %>% 
  glmFit(design.ac)

toptable_1.ac <- design.ac %>% colnames() %>% .[2:3] %>% 
  sapply(function(x){
    glmLRT(fit_1_ac, coef = x) %>%
      topTags(n = Inf) %>%
      .[["table"]] %>%
      as_tibble() %>%
      arrange(PValue) %>%
       mutate(
        coef = x,
        DE = FDR < 0.05
      ) %>% 
      dplyr::select(
        gene_name, logFC, logCPM, PValue, FDR, DE, everything()  
      )
      
  }, simplify = FALSE)
```

### vis
### Number of DE genes
```{r}
toptable_1.ac %>% 
  lapply(function(y) {
    y %>% 
      dplyr::filter(DE == T) %>% 
      nrow()
  }) %>% 
  bind_rows() %>% 
  as_tibble() %>% 
  gather(key = "coef", value = "numDE") %>% 
  ggplot(aes(x = numDE, y = coef)) +
  geom_col(aes(fill = coef)) +
  geom_text(aes(label = numDE)) +
  ggtitle("Number of DE genes in each comparison",
          subtitle = "The MPS-IIIA zebrafish samples have the most DE genes")
```

### Volcano plots
```{r}
toptable_1.ac %>%
  bind_rows() %>% 
  ggplot(aes(y = -log10(PValue), x = logFC, colour = DE)) +
  geom_point(
    alpha = 0.5, size = 1.25
  ) +
  facet_wrap(~coef) +
  # geom_label_repel(
  #   aes(label = gene_name),
  #   data = .  %>% dplyr::filter(FDR < 0.05),
  #   show.legend = FALSE
  #   ) +
  # coord_cartesian(xlim = c(-2.5,2.5), 
  #                 ylim = c(0, 20) )+
  theme(legend.position = "none") +
  scale_color_manual(values = c("grey50", "red")) 
```

### MD plots
```{r}
toptable_1.ac %>%
  bind_rows() %>% 
  ggplot(aes(x = logCPM, y = logFC)) +
  geom_point(
    aes(colour = DE),
    alpha = 0.5
  ) +
  facet_wrap(~coef, ncol = 1) +
  theme(legend.position = "none") +
  scale_color_manual(values = c("grey50", "red"))
```

### %GC AND LENGTH CHECK
```{r}
ggarrange(
toptable_1.ac %>% 
  bind_rows() %>% 
  mutate(rankstat = sign(logFC)*-log10(PValue)) %>% 
  ggplot(aes(x = length, y = rankstat)) +
  geom_point(
    aes(colour = DE),
    alpha = 0.5
  ) +
  geom_smooth(se = FALSE, method = "gam") +
  facet_grid(rows = vars(coef)) +
  theme_bw() +
  theme(legend.position = "none") +
  scale_color_manual(values = c("grey50", "red")) +
  scale_x_log10()+
  labs(x = "Average transcript length per gene",
       colour = "Differentially expressed?",
       y = "sign(logFC)*-log10(PValue)") +
  coord_cartesian(ylim = c(-10, 10)),

toptable_1.ac %>% 
  bind_rows() %>% 
  mutate(rankstat = sign(logFC)*-log10(PValue)) %>% 
  ggplot(aes(x = gc_content, y = rankstat)) +
  geom_point(
    aes(colour = DE),
    alpha = 0.5
  ) +
  geom_smooth(se = FALSE, method = "gam") +
  facet_grid(rows = vars(coef)) +
  theme_bw() +
  theme(legend.position = "none") +
  scale_color_manual(values = c("grey50", "red")) +
  coord_cartesian(ylim = c(-10,10)) + 
  labs(x = "Weighted average GC content (%) per gene", 
       colour = "Differentially expressed?", 
       y = "sign(logFC)*-log10(PValue)"),
common.legend = TRUE, 
labels = "AUTO"
) 
```



## CQN

A little bit of bias is obsevred for %GC content and gene length, so will apply the `cqn` correction to adjust for this. 

```{r}
cqn.ac <-   
x.ac %>% 
  with(
    cqn(
      counts = counts,
      x = genes$gc_content,
      lengths = genes$length,
      sizeFactors = samples$lib.size
    )
  )
logCPM.ac <- cqn.ac$y + cqn.ac$offset
```

## plot out the model fits
```{r plotCQN_AC, fig.cap ="*Model fits for GC content and gene length under the CQN model. Variability is clearly visible at either end*"}
par(mfrow = c(1, 2))
cqnplot(cqn.ac, n = 1, xlab = "GC Content")
cqnplot(cqn.ac, n = 2, xlab = "Length")
par(mfrow = c(1, 1))
```

## PCA
PCA analysis was repeated. Similar clustering of samples was observed across PC1. However, samples appeared closer together. 


```{r}
ggarrange(
  cpm(x.ac, log = T) %>%
    t() %>%
    prcomp() %>% 
    autoplot(data = tibble(sample = rownames(.$x)) %>%
               left_join(x$samples),
             colour = "genotype", 
             size = 4
    ) +
    scale_color_viridis_d(option = "turbo", end = 0.8) +
    ggtitle("Before CQN"),
  
  logCPM.ac %>% 
    t() %>% 
    prcomp() %>%
    autoplot(data = tibble(sample = rownames(.$x)) %>%
               left_join(x$samples),
             colour = "genotype", 
             size = 4
    ) +
    scale_color_viridis_d(option = "turbo", end = 0.8) +
    ggtitle("After CQN"),
  common.legend = T, 
  labels = "AUTO"
) 
#ggsave("plotsforpub/PCA_cqn.png", width = 18, height = 8, units = "cm", scale = 1.4)
```

# DE with cqn

```{r}
# Add the offset to the dge object 
x.ac$offset <- cqn.ac$glm.offset

# fit the glm
fit_cqn.ac <- x.ac %>% 
  estimateDisp(design.ac) %>% 
  glmFit(design.ac)

toptables_cqn.ac <- design.ac %>% colnames() %>% .[2:3] %>% 
  sapply(function(x){
    glmLRT(fit_cqn.ac, coef = x) %>%
      topTags(n = Inf) %>%
      .[["table"]] %>%
      as_tibble() %>%
      arrange(PValue) %>%
       mutate(
        coef = x,
        DE = FDR < 0.05
      ) %>% 
      dplyr::select(
        gene_name, logFC, logCPM, PValue, FDR, DE, everything()  
      )
      
  }, simplify = FALSE)
```

## Vis with CQN
### Num DE genes
```{r}
toptables_cqn.ac %>% 
  lapply(function(x){
    x %>%
      mutate(Direction = case_when(
        sign(logFC) == "1" ~ "up",
        sign(logFC) == "-1" ~ "down"
      ))
  }) %>%
  bind_rows() %>%
  dplyr::filter(DE == TRUE) %>%
  ggplot(aes(y = `coef`)) +
  geom_bar(stat = "count",
           width = 0.5,
           aes(fill = Direction),
           position = "dodge") +
  theme(legend.position = "bottom") +
  ggtitle("Number of DE genes in each comparison") +
  labs(x = "Number of DE genes",
       y = "Contrast",
       colour = "Direction of change") 
```

### Volcano plots
```{r}
toptables_cqn.ac %>%
  bind_rows() %>% 
  ggplot(aes(y = -log10(PValue), x = logFC, colour = DE)) +
  geom_point(
    alpha = 0.5, size = 1.25
  ) +
  facet_wrap(~coef) +
  # geom_label_repel(
  #   aes(label = gene_name),
  #   data = .  %>% dplyr::filter(FDR < 0.05),
  #   show.legend = FALSE
  #   ) +
  # coord_cartesian(xlim = c(-2.5,2.5), 
  #                 ylim = c(0, 20) )+
  theme(legend.position = "none") +
  scale_color_manual(values = c("grey50", "red")) 
```

### MD plots
```{r}
toptables_cqn.ac %>% 
  bind_rows() %>% 
  ggplot(aes(x = logCPM, y = logFC)) +
  geom_point(
    aes(colour = DE),
    alpha = 0.5
  ) +
  facet_wrap(~coef, ncol = 1) +
  theme(legend.position = "none") +
  scale_color_manual(values = c("grey50", "red"))
```

### Recheck the %GC and length bias

```{r}
ggarrange(
toptables_cqn.ac %>% 
  bind_rows() %>% 
  mutate(rankstat = sign(logFC)*-log10(PValue)) %>% 
  ggplot(aes(x = length, y = rankstat)) +
  geom_point(
    aes(colour = DE),
    alpha = 0.5
  ) +
  geom_smooth(se = FALSE, method = "gam") +
  facet_grid(rows = vars(coef)) +
  theme_bw() +
  theme(legend.position = "none") +
  scale_color_manual(values = c("grey50", "red")) +
  scale_x_log10()+
  labs(x = "Average transcript length per gene",
       colour = "Differentially expressed?",
       y = "sign(logFC)*-log10(PValue)") +
  coord_cartesian(ylim = c(-10, 10)),

toptables_cqn.ac %>% 
  bind_rows() %>% 
  mutate(rankstat = sign(logFC)*-log10(PValue)) %>% 
  ggplot(aes(x = gc_content, y = rankstat)) +
  geom_point(
    aes(colour = DE),
    alpha = 0.5
  ) +
  geom_smooth(se = FALSE, method = "gam") +
  facet_grid(rows = vars(coef)) +
  theme_bw() +
  theme(legend.position = "none") +
  scale_color_manual(values = c("grey50", "red")) +
  coord_cartesian(ylim = c(-10,10)) + 
  labs(x = "Weighted average GC content (%) per gene", 
       colour = "Differentially expressed?", 
       y = "sign(logFC)*-log10(PValue)"),
common.legend = TRUE, 
labels = "AUTO"
) 
```

## Gene set enrichment analysis

### Gene sets

```{r}
KEGG <- msigdbr("Danio rerio", category = "C2", subcategory = "CP:KEGG") %>% 
  left_join(x$genes, by = c("gene_symbol" = "gene_name")) %>% 
  dplyr::filter(gene_id %in% rownames(x)) %>% 
  distinct(gs_name, gene_id, .keep_all = TRUE) %>% 
  split(f = .$gs_name) %>%
  lapply(extract2, "gene_id")

sizes <- KEGG %>% 
  lapply(length) %>% 
  unlist %>% 
  as.data.frame() %>% 
  set_colnames( "n_genes") %>% 
  rownames_to_column("gs")


# retain gene sets with at least 10 genes in it
KEGG <- KEGG[sizes %>% dplyr::filter(n_genes > 10) %>% .$gs]

# GO Terms
goSummaries <- url("https://uofabioinformaticshub.github.io/summaries2GO/data/goSummaries.RDS") %>%
  readRDS() %>%
  mutate(
    Term = Term(id),
    gs_name = Term %>% str_to_upper() %>% str_replace_all("[ -]", "_"),
    gs_name = paste0("GO_", gs_name)
    )

minPath <- 3

GO <- 
  msigdbr("Danio rerio", category = "C5") %>% 
  dplyr::filter(grepl(gs_name, pattern = "^GO")) %>% 
  left_join(x$genes, by = c("gene_symbol" = "gene_name")) %>% 
  dplyr::filter(gene_id %in% rownames(x)) %>% 
  mutate(gs_name = str_replace(gs_name, pattern = "GOBP_", replacement = "GO_")) %>% 
  mutate(gs_name = str_replace(gs_name, pattern = "GOMF_", replacement = "GO_")) %>%     
  mutate(gs_name = str_replace(gs_name, pattern = "GOCC_", replacement = "GO_")) %>%     
  left_join(goSummaries) %>% 
  dplyr::filter(shortest_path >= minPath) %>%
  distinct(gs_name, gene_id, .keep_all = TRUE) %>% 
  split(f = .$gs_name) %>%
  lapply(extract2, "gene_id")

GOdf <-  msigdbr("Danio rerio", category = "C5") %>% 
  dplyr::filter(grepl(gs_name, pattern = "^GO")) %>% 
  left_join(x$genes, by = c("gene_symbol" = "gene_name")) %>% 
  dplyr::filter(gene_id %in% rownames(x)) %>% 
  mutate(gs_name = str_replace(gs_name, pattern = "GOBP_", replacement = "GO_")) %>% 
  mutate(gs_name = str_replace(gs_name, pattern = "GOMF_", replacement = "GO_")) %>%     
  mutate(gs_name = str_replace(gs_name, pattern = "GOCC_", replacement = "GO_")) %>%     
  left_join(goSummaries) %>% 
  dplyr::filter(shortest_path >= minPath) %>%
  distinct(gs_name, gene_id, .keep_all = TRUE) %>% 
  dplyr::select(gs_name, gene_id)

sizes_GO <- GO %>%
  lapply(length) %>%
  unlist %>%
  as.data.frame() %>%
  set_colnames( "n_genes") %>%
  rownames_to_column("gs")


# retain gene sets with at least 5 genes in it
GO <- GO[sizes_GO %>% dplyr::filter(n_genes > 5) %>% .$gs]

ireGenes <- 
  read_xlsx("data/gene_sets/ireGenes.xlsx", sheet = "Zebrafish IREs") %>% 
  gather %>% 
  set_colnames(c("IRE", "gene_id")) %>% 
  mutate(IRE = case_when(
    IRE == "All 3' IREs" ~ "ire3_all", 
    IRE == "All 5' IREs" ~ "ire5_all", 
    IRE == "HQ 3' IREs" ~ "ire3_HQ", 
    IRE == "HQ 5' IREs" ~ "ire5_HQ"
  )) %>% 
  dplyr::filter(gene_id %in% rownames(x)) %>% 
  split(f = .$IRE) %>% 
  lapply(magrittr::extract2,"gene_id") 

chr <- x$genes %>% 
  dplyr::select(gene_id, chromosome) %>% 
  split(f = .$chromosome) %>%
  lapply(extract2, "gene_id") %>% 
  .[c("MT", seq(1:25))] # omit the ALT chromosmes and scaffolds

#biomart to get the conversion between mouse and zebrafish IDs
mart <- useMart("ensembl", "drerio_gene_ensembl")
getFromBiomart <- c("ensembl_gene_id", "mmusculus_homolog_associated_gene_name")
zf_id_2_mm_genename <- getBM(attributes = getFromBiomart,
                             mart = mart) %>% 
  as_tibble() %>% 
  set_colnames(c("gene_id", "mouse_gene_name"))

cell_type_markers <- 
  read_xls("data/gene_sets/cell_type_genes41586_2007_BFnature05453_MOESM11_ESM.xls") %>%
  as_tibble() %>%
  dplyr::select(c("Neuron-enriched", "Astrocyte-enriched", "Oligodendrocyte-enriched")) %>%
  gather(key = "cell_type", value = "mouse_gene_name") %>% 
  left_join(zf_id_2_mm_genename) %>% 
  na.omit %>% 
  split(f = .$cell_type) %>% 
  lapply(extract2, "gene_id")

cell_type_markers$`Microglia-enriched` <- 
  read_excel("data/gene_sets/microglia_oosterofetal.xlsx", sheet = 1) %>%
  rbind(read_excel("data/gene_sets/microglia_oosterofetal.xlsx", sheet = 2)) %>%
  rbind(read_excel("data/gene_sets/microglia_oosterofetal.xlsx", sheet = 3)) %>%
  .$Zebrafish.Ensembl.Gene.ID %>% 
  unique

cell_type_markers %<>% 
  lapply(function(y) {
    y[y %in% rownames(x)]
  })
```

## GOSEQ 

```{r}
pwf <- toptables_cqn.ac$`MPS-IIIA` %>%
      with(
        nullp(
          DEgenes = structure(DE, names = gene_id), 
          bias.data = length
        )
      )

goseq(pwf, gene2cat = GO) %>% 
  as_tibble() %>% 
  dplyr::filter(numDEInCat > 0) %>% 
  mutate(FDR = p.adjust(over_represented_pvalue, method = "fdr")) %>% 
  dplyr::select(-under_represented_pvalue) %>% 
  dplyr::filter(FDR < 0.05) 

pwf <- toptables_cqn.ac$`MPS-IIIC` %>%
      with(
        nullp(
          DEgenes = structure(DE, names = gene_id), 
          bias.data = length
        )
      )

goseq(pwf, gene2cat = GO) %>% 
  as_tibble() %>% 
  dplyr::filter(numDEInCat > 0) %>% 
  mutate(FDR = p.adjust(over_represented_pvalue, method = "fdr")) %>% 
  dplyr::select(-under_represented_pvalue) %>% 
  dplyr::filter(FDR < 0.05) %>% 
  head(10)

  
```



## KEGG 
```{r}
fryKEGG <- 
  design.ac %>% colnames() %>% .[2:3] %>% 
  sapply(function(y) {
    logCPM.ac %>% 
      fry(
        index = KEGG,
        design = design.ac, 
        contrast = y, 
        sort = "mixed"
      ) %>% 
      rownames_to_column("pathway") %>% 
      as_tibble() %>% 
      mutate(coef = y)
  }, simplify = FALSE)

fryKEGG$`MPS-IIIA` %>% 
  dplyr::filter(FDR.Mixed < 0.05) %>%  
  mutate(pathway = str_remove(pathway, pattern = "KEGG_")) %>% 
  ggplot(aes(x = -log10(PValue.Mixed), y = reorder(pathway, -FDR.Mixed))) +
  geom_col(aes(fill = FDR.Mixed < 0.05)) +
  scale_fill_manual(values = c("grey50", 'red')) +
  theme_pubr() +
  labs(fill = "FDR < 0.05", 
       x = "-log10(PValue)", 
       y = "")

fryKEGG$`MPS-IIIC` %>% 
  dplyr::filter(FDR.Mixed < 0.05) %>%  
  mutate(pathway = str_remove(pathway, pattern = "KEGG_")) %>% 
  ggplot(aes(x = -log10(PValue.Mixed), y = reorder(pathway, -FDR.Mixed))) +
  geom_col(aes(fill = FDR.Mixed < 0.05)) +
  scale_fill_manual(values = c("grey50", 'red')) +
  theme_pubr() +
  labs(fill = "FDR < 0.05", 
       x = "-log10(PValue)", 
       y = "")  

sigpaths <- fryKEGG %>% 
  bind_rows() %>% 
  dplyr::filter(FDR.Mixed < 0.05) %>% 
    .$pathway %>% 
    unique()
  
fryKEGG %>% 
  bind_rows() %>% 
  dplyr::filter(pathway %in% sigpaths) %>% 
  mutate(pathway = str_remove(pathway, pattern = "KEGG_")) %>% 
  ggplot(aes(x = -log10(PValue.Mixed), y = reorder(pathway, -FDR.Mixed))) +
  geom_col(aes(fill = coef, alpha = FDR.Mixed < 0.05), position = "dodge") +
  scale_fill_viridis_d(end = 0.75) +
  scale_alpha_manual(values = c(0.5,1)) +
  theme_pubr() +
  geom_vline(xintercept = 2.3) +
  labs(fill = "FDR < 0.05", 
       x = "-log10(PValue)", 
       y = "")
  #ggsave("output/plots/AC_KEGG.png", width = 20, height = 20, units = "cm", dpi = 400, scale = 2)
```


### GO
```{r}
fry_GO <-  design.ac %>% colnames() %>% .[2:3] %>% 
  sapply(function(y) {
    logCPM.ac %>% 
      fry(
        index = GO,
        design = design.ac, 
        contrast = y, 
        sort = "mixed"
      ) %>% 
      rownames_to_column("pathway") %>% 
      as_tibble() %>% 
      mutate(coef = y, 
             sig = FDR.Mixed < 0.05)
  }, simplify = FALSE)
```


```{r}
toptables_cqn.ac %>% 
  bind_rows() %>% 
  dplyr::filter(gene_id %in% KEGG$KEGG_LYSOSOME) %>% 
  dplyr::select(gene_name, logFC, coef) %>% 
  dplyr::distinct(gene_name, coef, .keep_all = T) %>% 
  spread(key = "coef", value = "logFC") %>% 
  column_to_rownames("gene_name") %>% 
  t() %>% 
  pheatmap(color = colorRampPalette(rev(brewer.pal(n = 5,
                                               name = "RdBu")))(100), 
           main = "KEGG_LYSOSOME: FDR = 6.44e-5", 
           breaks = c(seq(min(.), 0, length.out=ceiling(100/2) + 1), 
              seq(max(.)/100, max(.), length.out=50)), 
           cellheight = 30, cellwidth = 12,angle_col = 45,
           treeheight_row = 0, treeheight_col = 0
           )

```
### IRE

```{r}
fryIRE.ac <- design.ac %>% colnames() %>% .[2:3] %>% 
  sapply(function(y) {
    logCPM.ac %>% 
      fry(
        index = ireGenes,
        design = design.ac, 
        contrast = y, 
        sort = "directional"
      ) %>% 
      rownames_to_column("pathway") %>% 
      as_tibble() %>% 
      mutate(coef = y)
  }, simplify = FALSE)

fryIRE.ac %>% 
  bind_rows() %>% 
  dplyr::select(pathway, coef, pvalue = PValue.Mixed, FDR = FDR.Mixed) %>% 
  ggplot(aes(y = pathway, x = -log10(pvalue))) + 
  geom_col(aes(fill = coef, alpha = FDR < 0.05), 
           position = "dodge") +
  scale_fill_viridis_d(end = 0.75) +
  geom_vline(xintercept = 1.2) +
  scale_alpha_manual(values = c(0.5,1)) +
  annotate(geom = "label", x = 1.1, y = "ire5_HQ", label = "FDR = 0.05") 
  #ggsave("output/plots/AC_ire.png", width = 20, height = 10, units = "cm", dpi = 400, scale = 2)


toptables_cqn.ac %>% 
  bind_rows() %>% 
    dplyr::filter(gene_id %in% ireGenes$ire3_all) %>% 
  dplyr::select(gene_name, logFC, coef) %>% 
  dplyr::distinct(gene_name, coef, .keep_all = T) %>% 
  spread(key = "coef", value = "logFC") %>% 
  column_to_rownames("gene_name") %>% 
  t() %>% 
  pheatmap(color = colorRampPalette(rev(brewer.pal(n = 5,
                                               name = "RdBu")))(100), 
           main = "3' IRE genes: FDR = 0.066", 
           breaks = c(seq(min(.), 0, length.out=ceiling(100/2) + 1), 
              seq(max(.)/100, max(.), length.out=50)), 
           # cellheight = 30, 
           # cellwidth = 1,
           angle_col = 45,
           treeheight_row = 0, 
           treeheight_col = 0, 
           show_colnames = F
           )
```

### Cell type prop check

Since this experiment was performed on whole brains, we need to have a look at whether changes to cell type proportions could explain cause some artefactual changes to gene expression. We have done this previously by assessing the expression of marker genes of specific cell types. So this is what I will do here. I have taken the lists of genes frmo day 5 only from the zebrafish single cell atlas (Farnsworth et al. 2020) which contain at least 10 genes. This leaves  `r length(cell_type_markers)` gene sets (cell types) to test.

```{r}
celltype.ac <- design.ac %>% colnames() %>% .[2:3] %>% 
  sapply(function(y) {
    logCPM.ac %>% 
      fry(
        index = cell_type_markers,
        design = design.ac, 
        contrast = y, 
        sort = "directional"
      ) %>% 
      rownames_to_column("pathway") %>% 
      as_tibble() %>% 
      mutate(coef = y)
  }, simplify = FALSE)

celltype.ac %>% 
  bind_rows() %>% 
  ggplot(aes(y = pathway, x = -log10(PValue))) + 
  geom_col(aes(fill = coef, alpha = FDR < 0.05), 
           position = "dodge") +
  scale_fill_viridis_d(end = 0.75) +
  geom_vline(xintercept = 1.8) +
  annotate(geom = "label", x = 2, y = 4.5, label = "FDR = 0.05") 
```

Statistical significance of the oligodendrocyte gene set is observed in both models. 

```{r}
toptables_cqn.ac %>% 
  bind_rows() %>% 
  dplyr::filter(gene_id %in% cell_type_markers$`Oligodendrocyte-enriched`) %>% 
  dplyr::select(gene_name, logFC, coef) %>% 
  dplyr::filter(coef == "MPS-IIIA") %>% 
  dplyr::distinct(gene_name, coef, .keep_all = T) %>% 
  spread(key = "coef", value = "logFC") %>% 
  column_to_rownames("gene_name") %>% 
  arrange(desc(`MPS-IIIA`)) %>% 
  t() %>% 
  pheatmap(color = colorRampPalette(rev(brewer.pal(n = 5,
                                               name = "RdBu")))(100), 
           main = "Oligodendrocytes in MPS-IIIA", 
           breaks = c(seq(min(.), 0, length.out=ceiling(100/2) + 1), 
              seq(max(.)/100, max(.), length.out=50)), 
           cellheight = 30, cellwidth = 12,
           angle_col = 45,
           cluster_rows = F,
           cluster_cols = F,
           treeheight_row = 0, treeheight_col = 0
           )

toptables_cqn.ac %>% 
  bind_rows() %>% 
  dplyr::filter(gene_id %in% cell_type_markers$`Oligodendrocyte-enriched`) %>% 
  dplyr::select(gene_name, logFC, coef) %>% 
  dplyr::filter(coef == "MPS-IIIC") %>% 
  dplyr::distinct(gene_name, coef, .keep_all = T) %>% 
  spread(key = "coef", value = "logFC") %>% 
  column_to_rownames("gene_name") %>% 
  arrange(desc(`MPS-IIIC`)) %>% 
  t() %>% 
  pheatmap(color = colorRampPalette(rev(brewer.pal(n = 5,
                                               name = "RdBu")))(100), 
           main = "Oligodendrocytes in MPS-IIIC", 
           breaks = c(seq(min(.), 0, length.out=ceiling(100/2) + 1), 
              seq(max(.)/100, max(.), length.out=50)), 
           cellheight = 30, cellwidth = 12,
           angle_col = 45,
           cluster_rows = F,
           cluster_cols = F,
           treeheight_row = 0, treeheight_col = 0
           )
```


### Chrosomosomes gene sets

```{r}
fry_chr.ac <- design.ac %>% colnames() %>% .[2:3] %>% 
  sapply(function(y) {
    logCPM.ac %>% 
      fry(
        index = c(chr),
        design = design.ac, 
        contrast = y, 
        sort = "mixed"
      ) %>% 
      rownames_to_column("pathway") %>% 
      as_tibble() %>% 
      mutate(coef = y)
  }, simplify = FALSE)


fry_chr.ac %>% 
  bind_rows() %>% 
  mutate(pathway = factor(pathway, levels = c("MT", seq(1:25)))) %>% 
  ggplot(aes(y = pathway, x = -log10(PValue.Mixed), fill = coef)) +
  geom_col(position = "dodge")
  
```

An enrichment of DE genes is observed on the same chromosome as the mutation. This is obsevred in the manhat plots below. 

```{r}
manhat_input <- 
	toptables_cqn.ac %>% 
  bind_rows() %>% 
	left_join(x$genes %>% 
			  	as_tibble() %>% 
			  	dplyr::select(start, end, chromosome, gene_id) %>% 
			  	group_by(gene_id) %>% 
			  	mutate(mid = mean(c(start, end)))
	) %>% 
  dplyr::filter(chromosome %in% seq(1:25)) %>% 
  group_by(gene_id) %>% 
  mutate(mid = mean(c(start, end))) %>% 
  ungroup %>% 
  mutate(chromosome = factor(chromosome, levels = seq(1:25)))  %>% 
  group_by(chromosome) %>% 
  summarise(chrLen = max(mid)) %>% 
  mutate(chrSt = cumsum(chrLen)-chrLen) %>% 
	dplyr::select(-chrLen) %>% 
  left_join(toptables_cqn.ac %>% 
              bind_rows() %>% 
              left_join(x$genes %>% 
                          as_tibble() %>% 
                          dplyr::select(start, end, chromosome, gene_id) %>% 
                          group_by(gene_id) %>% 
                          mutate(mid = mean(c(start, end)))
              )
  ) %>% 
  group_by(gene_id) %>% 
  mutate(mid = mean(c(start, end))) %>% 
  ungroup %>% 
  dplyr::arrange(mid, chromosome) %>% 
  mutate(midCum = chrSt + mid)

axis <- manhat_input %>%
  group_by(chromosome) %>%
  dplyr::arrange(chromosome) %>%
  summarize(center = (max(midCum) + min(midCum)) / 2) %>%
  mutate(
    seqnames = as.numeric(chromosome),
    colour = rep(c("grey40", "black"), length.out = 25)
  )

ggplot(manhat_input, aes(x = midCum, y = -log10(PValue))) +
  geom_point(aes(color=chromosome), alpha = 0.5, size = 1) +
  geom_point(alpha = 0.5, size = 1, colour = "red", 
             data = . %>% 
               dplyr::filter(coef == "MPS-IIIC" & chromosome == 1)) +
  geom_point(alpha = 0.5, size = 1, colour = "red", 
             data = . %>% 
               dplyr::filter(coef ==  "MPS-IIIA" & chromosome == 22)) +
  scale_color_manual(values = axis$colour) +
  scale_x_continuous(label = axis$chromosome, breaks = axis$center) +
  scale_y_continuous(limits =c(0, 10)) + # zoom in
  facet_wrap(~coef, ncol = 1) +
  labs(x = "Chromosome", y = expression(paste(-log[10], "(p)"))) +
  theme( 
    legend.position="none",
    panel.border = element_blank(),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.x = element_blank()
  ) 
```

# BC
```{r}
x.bc <- x[,BCsamps]

design.bc <- model.matrix(~genotype + Tank, data = x.bc$samples %>% 
                            droplevels()) %>% # get rid of the unused factor levels
  set_colnames(gsub(colnames(.), pattern = "genotype", replacement = ""))

fit_1_bc <- x.bc %>% 
  estimateDisp(design.bc) %>% 
  glmFit(design.bc)

toptable_1.bc <- design.bc %>% colnames() %>% .[2:3] %>% 
  sapply(function(x){
    glmLRT(fit_1_bc, coef = x) %>%
      topTags(n = Inf) %>%
      .[["table"]] %>%
      as_tibble() %>%
      arrange(PValue) %>%
       mutate(
        coef = x,
        DE = FDR < 0.05
      ) %>% 
      dplyr::select(
        gene_name, logFC, logCPM, PValue, FDR, DE, everything()  
      )
      
  }, simplify = FALSE)
```

### vis
### Number of DE genes
```{r}
toptable_1.bc %>% 
  lapply(function(y) {
    y %>% 
      dplyr::filter(DE == T) %>% 
      nrow()
  }) %>% 
  bind_rows() %>% 
  as_tibble() %>% 
  gather(key = "coef", value = "numDE") %>% 
  ggplot(aes(x = numDE, y = coef)) +
  geom_col(aes(fill = coef)) +
  geom_text(aes(label = numDE)) +
  ggtitle("Number of DE genes in each comparison",
          subtitle = "The MPS-IIIA zebrafish samples have the most DE genes")
```

### Volcano plots

The volcano plot below is zoomed in for vis purposes. 
```{r}
toptable_1.bc %>%
  bind_rows() %>% 
  ggplot(aes(y = -log10(PValue), x = logFC, colour = DE)) +
  geom_point(
    alpha = 0.5, size = 1.25
  ) +
  facet_wrap(~coef) +
  # geom_label_repel(
  #   aes(label = gene_name),
  #   data = .  %>% dplyr::filter(FDR < 0.05),
  #   show.legend = FALSE
  #   ) +
  coord_cartesian(xlim = c(-2.5,2.5),
                  ylim = c(0, 20) )+
  theme(legend.position = "none") +
  scale_color_manual(values = c("grey50", "red")) 
```

### MD plots
```{r}
toptable_1.bc %>%
  bind_rows() %>% 
  ggplot(aes(x = logCPM, y = logFC)) +
  geom_point(
    aes(colour = DE),
    alpha = 0.5
  ) +
  facet_wrap(~coef, ncol = 1) +
  theme(legend.position = "none") +
  scale_color_manual(values = c("grey50", "red"))
```

### %GC AND LENGTH CHECK
```{r}
ggarrange(
toptable_1.bc %>% 
  bind_rows() %>% 
  mutate(rankstat = sign(logFC)*-log10(PValue)) %>% 
  ggplot(aes(x = length, y = rankstat)) +
  geom_point(
    aes(colour = DE),
    alpha = 0.5
  ) +
  geom_smooth(se = FALSE, method = "gam") +
  facet_grid(rows = vars(coef)) +
  theme_bw() +
  theme(legend.position = "none") +
  scale_color_manual(values = c("grey50", "red")) +
  scale_x_log10()+
  labs(x = "Average transcript length per gene",
       colour = "Differentially expressed?",
       y = "sign(logFC)*-log10(PValue)") +
  coord_cartesian(ylim = c(-10, 10)),

toptable_1.bc %>% 
  bind_rows() %>% 
  mutate(rankstat = sign(logFC)*-log10(PValue)) %>% 
  ggplot(aes(x = gc_content, y = rankstat)) +
  geom_point(
    aes(colour = DE),
    alpha = 0.5
  ) +
  geom_smooth(se = FALSE, method = "gam") +
  facet_grid(rows = vars(coef)) +
  theme_bw() +
  theme(legend.position = "none") +
  scale_color_manual(values = c("grey50", "red")) +
  coord_cartesian(ylim = c(-10,10)) + 
  labs(x = "Weighted average GC content (%) per gene", 
       colour = "Differentially expressed?", 
       y = "sign(logFC)*-log10(PValue)"),
common.legend = TRUE, 
labels = "AUTO"
) 
```



## CQN

A  bit of bias is obsevred for %GC content and gene length, so will apply the `cqn` correction to adjust for this. 

```{r}
cqn.bc <-   
x.bc %>% 
  with(
    cqn(
      counts = counts,
      x = genes$gc_content,
      lengths = genes$length,
      sizeFactors = samples$lib.size
    )
  )
logCPM.bc <- cqn.bc$y + cqn.bc$offset
```

## plot out the model fits
```{r plotCQN, fig.cap ="*Model fits for GC content and gene length under the CQN model. Variability is clearly visible at either end*"}
par(mfrow = c(1, 2))
cqnplot(cqn.bc, n = 1, xlab = "GC Content")
cqnplot(cqn.bc, n = 2, xlab = "Length")
par(mfrow = c(1, 1))
```

## PCA
PCA analysis was repeated. Similar clustering of samples was observed across PC1. However, samples appeared closer together. 


```{r}
ggarrange(
  cpm(x.bc, log = T) %>%
    t() %>%
    prcomp() %>% 
    autoplot(data = tibble(sample = rownames(.$x)) %>%
               left_join(x$samples),
             colour = "genotype", 
             size = 4
    ) +
    scale_color_viridis_d(option = "turbo", end = 0.8) +
    ggtitle("Before CQN"),
  
  logCPM.bc %>% 
    t() %>% 
    prcomp() %>%
    autoplot(data = tibble(sample = rownames(.$x)) %>%
               left_join(x$samples),
             colour = "genotype", 
             size = 4
    ) +
    scale_color_viridis_d(option = "turbo", end = 0.8) +
    ggtitle("After CQN"),
  common.legend = T, 
  labels = "AUTO"
) 
```

# DE with cqn

```{r}
# Add the offset to the dge object 
x.bc$offset <- cqn.bc$glm.offset

# fit the glm
fit_cqn.bc <- x.bc %>% 
  estimateDisp(design.bc) %>% 
  glmFit(design.bc)

toptables_cqn.bc <- design.bc %>% colnames() %>% .[2:3] %>% 
  sapply(function(x){
    glmLRT(fit_cqn.bc, coef = x) %>%
      topTags(n = Inf) %>%
      .[["table"]] %>%
      as_tibble() %>%
      arrange(PValue) %>%
       mutate(
        coef = x,
        DE = FDR < 0.05
      ) %>% 
      dplyr::select(
        gene_name, logFC, logCPM, PValue, FDR, DE, everything()  
      )
      
  }, simplify = FALSE)
```

## Vis with CQN
### Num DE genes
```{r}
toptables_cqn.bc %>% 
  lapply(function(x){
    x %>%
      mutate(Direction = case_when(
        sign(logFC) == "1" ~ "up",
        sign(logFC) == "-1" ~ "down"
      ))
  }) %>%
  bind_rows() %>%
  dplyr::filter(DE == TRUE) %>%
  ggplot(aes(y = `coef`)) +
  geom_bar(stat = "count",
           width = 0.5,
           aes(fill = Direction),
           position = "dodge") +
  theme(legend.position = "bottom") +
  ggtitle("Number of DE genes in each comparison") +
  labs(x = "Number of DE genes",
       y = "Contrast",
       colour = "Direction of change") 
```

### Volcano plots

These are zoomed in again for vis purposes
```{r}
toptables_cqn.bc %>%
  bind_rows() %>% 
  ggplot(aes(y = -log10(PValue), x = logFC, colour = DE)) +
  geom_point(
    alpha = 0.5, size = 1.25
  ) +
  facet_wrap(~coef) +
  # geom_label_repel(
  #   aes(label = gene_name),
  #   data = .  %>% dplyr::filter(FDR < 0.05),
  #   show.legend = FALSE
  #   ) +
  coord_cartesian(xlim = c(-2.5,2.5),
                  ylim = c(0, 20) )+
  theme(legend.position = "none") +
  scale_color_manual(values = c("grey50", "red")) 
```

### MD plots
```{r}
toptables_cqn.bc %>% 
  bind_rows() %>% 
  ggplot(aes(x = logCPM, y = logFC)) +
  geom_point(
    aes(colour = DE),
    alpha = 0.5
  ) +
  facet_wrap(~coef, ncol = 1) +
  theme(legend.position = "none") +
  scale_color_manual(values = c("grey50", "red"))
```

### Recheck the %GC and length bias

```{r}
ggarrange(
toptables_cqn.bc %>% 
  bind_rows() %>% 
  mutate(rankstat = sign(logFC)*-log10(PValue)) %>% 
  ggplot(aes(x = length, y = rankstat)) +
  geom_point(
    aes(colour = DE),
    alpha = 0.5
  ) +
  geom_smooth(se = FALSE, method = "gam") +
  facet_grid(rows = vars(coef)) +
  theme_bw() +
  theme(legend.position = "none") +
  scale_color_manual(values = c("grey50", "red")) +
  scale_x_log10()+
  labs(x = "Average transcript length per gene",
       colour = "Differentially expressed?",
       y = "sign(logFC)*-log10(PValue)") +
  coord_cartesian(ylim = c(-10, 10)),

toptables_cqn.bc %>% 
  bind_rows() %>% 
  mutate(rankstat = sign(logFC)*-log10(PValue)) %>% 
  ggplot(aes(x = gc_content, y = rankstat)) +
  geom_point(
    aes(colour = DE),
    alpha = 0.5
  ) +
  geom_smooth(se = FALSE, method = "gam") +
  facet_grid(rows = vars(coef)) +
  theme_bw() +
  theme(legend.position = "none") +
  scale_color_manual(values = c("grey50", "red")) +
  coord_cartesian(ylim = c(-10,10)) + 
  labs(x = "Weighted average GC content (%) per gene", 
       colour = "Differentially expressed?", 
       y = "sign(logFC)*-log10(PValue)"),
common.legend = TRUE, 
labels = "AUTO"
) 
```

## Gene set enrichment analysis

### Gene sets

```{r}
KEGG <- msigdbr("Danio rerio", category = "C2", subcategory = "CP:KEGG") %>% 
  left_join(x$genes, by = c("gene_symbol" = "gene_name")) %>% 
  dplyr::filter(gene_id %in% rownames(x)) %>% 
  distinct(gs_name, gene_id, .keep_all = TRUE) %>% 
  split(f = .$gs_name) %>%
  lapply(extract2, "gene_id")

sizes <- KEGG %>% 
  lapply(length) %>% 
  unlist %>% 
  as.data.frame() %>% 
  set_colnames( "n_genes") %>% 
  rownames_to_column("gs")


# retain gene sets with at least 10 genes in it
KEGG <- KEGG[sizes %>% dplyr::filter(n_genes > 10) %>% .$gs]

# GO Terms
goSummaries <- url("https://uofabioinformaticshub.github.io/summaries2GO/data/goSummaries.RDS") %>%
  readRDS() %>%
  mutate(
    Term = Term(id),
    gs_name = Term %>% str_to_upper() %>% str_replace_all("[ -]", "_"),
    gs_name = paste0("GO_", gs_name)
    )

minPath <- 3

GO <- 
  msigdbr("Danio rerio", category = "C5") %>% 
  dplyr::filter(grepl(gs_name, pattern = "^GO")) %>% 
  left_join(x$genes, by = c("gene_symbol" = "gene_name")) %>% 
  dplyr::filter(gene_id %in% rownames(x)) %>% 
  mutate(gs_name = str_replace(gs_name, pattern = "GOBP_", replacement = "GO_")) %>% 
  mutate(gs_name = str_replace(gs_name, pattern = "GOMF_", replacement = "GO_")) %>%     
  mutate(gs_name = str_replace(gs_name, pattern = "GOCC_", replacement = "GO_")) %>%     
  left_join(goSummaries) %>% 
  dplyr::filter(shortest_path >= minPath) %>%
  distinct(gs_name, gene_id, .keep_all = TRUE) %>% 
  split(f = .$gs_name) %>%
  lapply(extract2, "gene_id")

GOdf <-  msigdbr("Danio rerio", category = "C5") %>% 
  dplyr::filter(grepl(gs_name, pattern = "^GO")) %>% 
  left_join(x$genes, by = c("gene_symbol" = "gene_name")) %>% 
  dplyr::filter(gene_id %in% rownames(x)) %>% 
  mutate(gs_name = str_replace(gs_name, pattern = "GOBP_", replacement = "GO_")) %>% 
  mutate(gs_name = str_replace(gs_name, pattern = "GOMF_", replacement = "GO_")) %>%     
  mutate(gs_name = str_replace(gs_name, pattern = "GOCC_", replacement = "GO_")) %>%     
  left_join(goSummaries) %>% 
  dplyr::filter(shortest_path >= minPath) %>%
  distinct(gs_name, gene_id, .keep_all = TRUE) %>% 
  dplyr::select(gs_name, gene_id)

sizes_GO <- GO %>%
  lapply(length) %>%
  unlist %>%
  as.data.frame() %>%
  set_colnames( "n_genes") %>%
  rownames_to_column("gs")


# retain gene sets with at least 5 genes in it
GO <- GO[sizes_GO %>% dplyr::filter(n_genes > 5) %>% .$gs]

ireGenes <- 
  read_xlsx("data/gene_sets/ireGenes.xlsx", sheet = "Zebrafish IREs") %>% 
  gather %>% 
  set_colnames(c("IRE", "gene_id")) %>% 
  mutate(IRE = case_when(
    IRE == "All 3' IREs" ~ "ire3_all", 
    IRE == "All 5' IREs" ~ "ire5_all", 
    IRE == "HQ 3' IREs" ~ "ire3_HQ", 
    IRE == "HQ 5' IREs" ~ "ire5_HQ"
  )) %>% 
  dplyr::filter(gene_id %in% rownames(x)) %>% 
  split(f = .$IRE) %>% 
  lapply(magrittr::extract2,"gene_id") 

chr <- x$genes %>% 
  dplyr::select(gene_id, chromosome) %>% 
  split(f = .$chromosome) %>%
  lapply(extract2, "gene_id") %>% 
  .[c("MT", seq(1:25))] # omit the ALT chromosmes and scaffolds

#biomart to get the conversion between mouse and zebrafish IDs
mart <- useMart("ensembl", "drerio_gene_ensembl")
getFromBiomart <- c("ensembl_gene_id", "mmusculus_homolog_associated_gene_name")
zf_id_2_mm_genename <- getBM(attributes = getFromBiomart,
                             mart = mart) %>% 
  as_tibble() %>% 
  set_colnames(c("gene_id", "mouse_gene_name"))

cell_type_markers <- 
  read_xls("data/gene_sets/cell_type_genes41586_2007_BFnature05453_MOESM11_ESM.xls") %>%
  as_tibble() %>%
  dplyr::select(c("Neuron-enriched", "Astrocyte-enriched", "Oligodendrocyte-enriched")) %>%
  gather(key = "cell_type", value = "mouse_gene_name") %>% 
  left_join(zf_id_2_mm_genename) %>% 
  na.omit %>% 
  split(f = .$cell_type) %>% 
  lapply(extract2, "gene_id")

cell_type_markers$`Microglia-enriched` <- 
  read_excel("data/gene_sets/microglia_oosterofetal.xlsx", sheet = 1) %>%
  rbind(read_excel("data/gene_sets/microglia_oosterofetal.xlsx", sheet = 2)) %>%
  rbind(read_excel("data/gene_sets/microglia_oosterofetal.xlsx", sheet = 3)) %>%
  .$Zebrafish.Ensembl.Gene.ID %>% 
  unique

cell_type_markers %<>% 
  lapply(function(y) {
    y[y %in% rownames(x)]
  })
```

## GOSEQ 

```{r}
pwf <- toptables_cqn.bc$`MPS-IIIB` %>%
      with(
        nullp(
          DEgenes = structure(DE, names = gene_id), 
          bias.data = length
        )
      )

goseq(pwf, gene2cat = GO) %>% 
  as_tibble() %>% 
  dplyr::filter(numDEInCat > 0) %>% 
  mutate(FDR = p.adjust(over_represented_pvalue, method = "fdr")) %>% 
  dplyr::select(-under_represented_pvalue) %>% 
  dplyr::filter(FDR < 0.05) 

pwf <- toptables_cqn.bc$`MPS-IIIC` %>%
      with(
        nullp(
          DEgenes = structure(DE, names = gene_id), 
          bias.data = length
        )
      )

goseq(pwf, gene2cat = GO) %>% 
  as_tibble() %>% 
  dplyr::filter(numDEInCat > 0) %>% 
  mutate(FDR = p.adjust(over_represented_pvalue, method = "fdr")) %>% 
  dplyr::select(-under_represented_pvalue) %>% 
  dplyr::filter(FDR < 0.05) %>% 
  head(10)

  
```



## KEGG 
```{r}
fryKEGG <- 
  design.bc %>% colnames() %>% .[2:3] %>% 
  sapply(function(y) {
    logCPM.bc %>% 
      fry(
        index = KEGG,
        design = design.bc, 
        contrast = y, 
        sort = "mixed"
      ) %>% 
      rownames_to_column("pathway") %>% 
      as_tibble() %>% 
      mutate(coef = y)
  }, simplify = FALSE)

fryKEGG$`MPS-IIIB` %>% 
  dplyr::filter(FDR.Mixed < 0.05) %>%  
  mutate(pathway = str_remove(pathway, pattern = "KEGG_")) %>% 
  ggplot(aes(x = -log10(PValue.Mixed), y = reorder(pathway, -FDR.Mixed))) +
  geom_col(aes(fill = FDR.Mixed < 0.05)) +
  scale_fill_manual(values = c("grey50", 'red')) +
  theme_pubr() +
  labs(fill = "FDR < 0.05", 
       x = "-log10(PValue)", 
       y = "")

fryKEGG$`MPS-IIIC` %>% 
  dplyr::filter(FDR.Mixed < 0.05) %>%  
  mutate(pathway = str_remove(pathway, pattern = "KEGG_")) %>% 
  ggplot(aes(x = -log10(PValue.Mixed), y = reorder(pathway, -FDR.Mixed))) +
  geom_col(aes(fill = FDR.Mixed < 0.05)) +
  scale_fill_manual(values = c("grey50", 'red')) +
  theme_pubr() +
  labs(fill = "FDR < 0.05", 
       x = "-log10(PValue)", 
       y = "")  

sigpaths <- fryKEGG %>% 
  bind_rows() %>% 
  dplyr::filter(FDR.Mixed < 0.05) %>% 
    .$pathway %>% 
    unique()
  
fryKEGG %>% 
  bind_rows() %>% 
  dplyr::filter(pathway %in% sigpaths) %>% 
  mutate(pathway = str_remove(pathway, pattern = "KEGG_")) %>% 
  ggplot(aes(x = -log10(PValue.Mixed), y = reorder(pathway, -FDR.Mixed))) +
  geom_col(aes(fill = coef, alpha = FDR.Mixed < 0.05), position = "dodge") +
  scale_fill_viridis_d(end = 0.75) +
  scale_alpha_manual(values = c(0.5,1)) +
  theme_pubr() +
  geom_vline(xintercept = 2.3) +
  labs(fill = "FDR < 0.05", 
       x = "-log10(PValue)", 
       y = "")
  #ggsave("output/plots/AC_KEGG.png", width = 20, height = 20, units = "cm", dpi = 400, scale = 2)
```


### GO
```{r}
fry_GO <-  design.ac %>% colnames() %>% .[2:3] %>% 
  sapply(function(y) {
    logCPM.ac %>% 
      fry(
        index = GO,
        design = design.ac, 
        contrast = y, 
        sort = "mixed"
      ) %>% 
      rownames_to_column("pathway") %>% 
      as_tibble() %>% 
      mutate(coef = y, 
             sig = FDR.Mixed < 0.05)
  }, simplify = FALSE)
```


```{r}
toptables_cqn.ac %>% 
  bind_rows() %>% 
  dplyr::filter(gene_id %in% KEGG$KEGG_LYSOSOME) %>% 
  dplyr::select(gene_name, logFC, coef) %>% 
  dplyr::distinct(gene_name, coef, .keep_all = T) %>% 
  spread(key = "coef", value = "logFC") %>% 
  column_to_rownames("gene_name") %>% 
  t() %>% 
  pheatmap(color = colorRampPalette(rev(brewer.pal(n = 5,
                                               name = "RdBu")))(100), 
           main = "KEGG_LYSOSOME: FDR = 6.44e-5", 
           breaks = c(seq(min(.), 0, length.out=ceiling(100/2) + 1), 
              seq(max(.)/100, max(.), length.out=50)), 
           cellheight = 30, cellwidth = 12,angle_col = 45,
           treeheight_row = 0, treeheight_col = 0
           )

```
### IRE

```{r}
fryIRE.bc <- design.bc %>% colnames() %>% .[2:3] %>% 
  sapply(function(y) {
    logCPM.bc %>% 
      fry(
        index = ireGenes,
        design = design.bc, 
        contrast = y, 
        sort = "directional"
      ) %>% 
      rownames_to_column("pathway") %>% 
      as_tibble() %>% 
      mutate(coef = y)
  }, simplify = FALSE)

fryIRE.bc %>% 
  bind_rows() %>% 
  dplyr::select(pathway, coef, pvalue = PValue.Mixed, FDR = FDR.Mixed) %>% 
  ggplot(aes(y = pathway, x = -log10(pvalue))) + 
  geom_col(aes(fill = coef, alpha = FDR < 0.05), 
           position = "dodge") +
  scale_fill_viridis_d(end = 0.75) +
  scale_alpha_manual(values = c(0.5, 1)) +
  geom_vline(xintercept = 1.5) +
  annotate(geom = "label", x = 1.8, y = "ire5_HQ", label = "FDR = 0.05") 
  #ggsave("output/plots/AC_ire.png", width = 20, height = 10, units = "cm", dpi = 400, scale = 2)


toptables_cqn.bc %>% 
  bind_rows() %>% 
    dplyr::filter(gene_id %in% ireGenes$ire3_all) %>% 
  dplyr::select(gene_name, logFC, coef) %>% 
  dplyr::distinct(gene_name, coef, .keep_all = T) %>% 
  spread(key = "coef", value = "logFC") %>% 
  column_to_rownames("gene_name") %>% 
  t() %>% 
  pheatmap(color = colorRampPalette(rev(brewer.pal(n = 5,
                                               name = "RdBu")))(100), 
           main = "3' IRE genes: FDR = 0.066", 
           breaks = c(seq(min(.), 0, length.out=ceiling(100/2) + 1), 
              seq(max(.)/100, max(.), length.out=50)), 
           # cellheight = 30, 
           # cellwidth = 1,
           angle_col = 45,
           treeheight_row = 0, 
           treeheight_col = 0, 
           show_colnames = F
           )
```

### Cell type prop check

Since this experiment was performed on whole brains, we need to have a look at whether changes to cell type proportions could explain cause some artefactual changes to gene expression. We have done this previously by assessing the expression of marker genes of specific cell types. So this is what I will do here. I have taken the lists of genes frmo day 5 only from the zebrafish single cell atlas (Farnsworth et al. 2020) which contain at least 10 genes. This leaves  `r length(cell_type_markers)` gene sets (cell types) to test.

```{r}
celltype.bc <- design.bc %>% colnames() %>% .[2:3] %>% 
  sapply(function(y) {
    logCPM.bc %>% 
      fry(
        index = cell_type_markers,
        design = design.bc, 
        contrast = y, 
        sort = "directional"
      ) %>% 
      rownames_to_column("pathway") %>% 
      as_tibble() %>% 
      mutate(coef = y)
  }, simplify = FALSE)

celltype.bc %>% 
  bind_rows() %>% 
  ggplot(aes(y = pathway, x = -log10(PValue))) + 
  geom_col(aes(fill = coef), 
           position = "dodge") +
  scale_fill_viridis_d(end = 0.75) +
  geom_vline(xintercept = 1.7) +
  annotate(geom = "label", x = 1.7, y = 4.5, label = "FDR = 0.05") 
```


### Chrosomosomes gene sets

```{r}
fry_chr.bc <- design.bc %>% colnames() %>% .[2:3] %>% 
  sapply(function(y) {
    logCPM.bc %>% 
      fry(
        index = c(chr),
        design = design.bc, 
        contrast = y, 
        sort = "mixed"
      ) %>% 
      rownames_to_column("pathway") %>% 
      as_tibble() %>% 
      mutate(coef = y)
  }, simplify = FALSE)


fry_chr.bc %>% 
  bind_rows() %>% 
  mutate(pathway = factor(pathway, levels = c("MT", seq(1:25)))) %>% 
  ggplot(aes(y = pathway, x = -log10(PValue.Mixed), fill = coef)) +
  geom_col(position = "dodge")
  
```

An enrichment of DE genes is observed on the same chromosome as the mutation. This is obsevred in the manhat plots below. 

```{r}
manhat_input.bc <- 
	toptables_cqn.bc %>% 
  bind_rows() %>% 
	left_join(x$genes %>% 
			  	as_tibble() %>% 
			  	dplyr::select(start, end, chromosome, gene_id) %>% 
			  	group_by(gene_id) %>% 
			  	mutate(mid = mean(c(start, end)))
	) %>% 
  dplyr::filter(chromosome %in% seq(1:25)) %>% 
  group_by(gene_id) %>% 
  mutate(mid = mean(c(start, end))) %>% 
  ungroup %>% 
  mutate(chromosome = factor(chromosome, levels = seq(1:25)))  %>% 
  group_by(chromosome) %>% 
  summarise(chrLen = max(mid)) %>% 
  mutate(chrSt = cumsum(chrLen)-chrLen) %>% 
	dplyr::select(-chrLen) %>% 
  left_join(toptables_cqn.bc %>% 
              bind_rows() %>% 
              left_join(x$genes %>% 
                          as_tibble() %>% 
                          dplyr::select(start, end, chromosome, gene_id) %>% 
                          group_by(gene_id) %>% 
                          mutate(mid = mean(c(start, end)))
              )
  ) %>% 
  group_by(gene_id) %>% 
  mutate(mid = mean(c(start, end))) %>% 
  ungroup %>% 
  dplyr::arrange(mid, chromosome) %>% 
  mutate(midCum = chrSt + mid)

axis.bc <- manhat_input.bc %>%
  group_by(chromosome) %>%
  dplyr::arrange(chromosome) %>%
  summarize(center = (max(midCum) + min(midCum)) / 2) %>%
  mutate(
    seqnames = as.numeric(chromosome),
    colour = rep(c("grey40", "black"), length.out = 25)
  )

ggplot(manhat_input.bc, aes(x = midCum, y = -log10(PValue))) +
  geom_point(aes(color=chromosome), alpha = 0.5, size = 1) +
  geom_point(alpha = 0.5, size = 1, colour = "red", 
             data = . %>% 
               dplyr::filter(coef == "MPS-IIIC" & chromosome == 1)) +
  geom_point(alpha = 0.5, size = 1, colour = "red", 
             data = . %>% 
               dplyr::filter(coef ==  "MPS-IIIB" & chromosome == 24)) +
  scale_color_manual(values = axis.bc$colour) +
  scale_x_continuous(label = axis.bc$chromosome, breaks = axis.bc$center) +
  scale_y_continuous(limits =c(0, 10)) + # zoom in
  facet_wrap(~coef, ncol = 1) +
  labs(x = "Chromosome", y = expression(paste(-log[10], "(p)"))) +
  theme( 
    legend.position="none",
    panel.border = element_blank(),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.x = element_blank()
  ) 
```


## Data export
```{r}
x.ac %>% saveRDS("data/R_objects/dge_ac.rds")
toptables_cqn.ac %>% saveRDS("data/R_objects/toptab_ac_cqn.rds")
fryKEGG %>% saveRDS("data/R_objects/AC_fryKEGG.rds")
fryIRE.ac %>% saveRDS("data/R_objects/AC_fryIRE.rds")
celltype.ac %>% saveRDS("data/R_objects/AC_frycellTypes.rds")



x.bc %>% saveRDS("data/R_objects/dge_bc.rds")
toptables_cqn.bc %>% saveRDS("data/R_objects/toptab_bc_cqn.rds")
fryKEGG %>% saveRDS("data/R_objects/BC_fryKEGG.rds")
fryIRE.bc %>% saveRDS("data/R_objects/BC_fryIRE.rds")
celltype.bc %>% saveRDS("data/R_objects/BC_frycellTypes.rds")

```

