---
title: "analysis_final"
author: "Karissa Barthelson"
date: "2022-12-22"
output: workflowr::wflow_html
editor_options:
  chunk_output_type: console
---

## Introduction
```{r setup, include=FALSE}
knitr::opts_chunk$set(
  autodep = TRUE,
  echo = TRUE,
  warning = FALSE,
  message = FALSE,
  fig.align = "center", 
  out.width ="75%", 
  out.height = "75%"
)
```

```{r loadLibs}
# data wrangling
library(tidyverse)
library(magrittr)
library(readxl)

# annotations
library(AnnotationHub)
library(biomaRt)
library(msigdbr)

# table formatting
library(pander)
library(kableExtra)

# data vis
library(scales)
library(pheatmap)
library(ggpubr)
library(ggrepel)
library(ggfortify)
library(RColorBrewer)
library(UpSetR)

# bioc RNAseq packages
library(edgeR)
library(cqn)
library(goseq)
library(fgsea)
library(harmonicmeanp)
library(ssizeRNA)

theme_set(theme_bw())
```

```{r anno}
ah <- AnnotationHub() %>%
	subset(species == "Danio rerio") %>%
	subset(rdataclass == "EnsDb")

ensDb <- ah[["AH83189"]] # for release 101, latest version in Ah
grTrans <- transcripts(ensDb)
trLengths <- exonsBy(ensDb, "tx") %>%
	width() %>%
	vapply(sum, integer(1))
mcols(grTrans)$length <- trLengths[names(grTrans)]
gcGene <- grTrans %>%
  mcols() %>%
  as.data.frame() %>%
  dplyr::select(gene_id, tx_id, gc_content, length) %>%
  as_tibble() %>%
  group_by(gene_id) %>%
  summarise(
    gc_content = sum(gc_content*length) / sum(length),
    length = ceiling(median(length))
  )
grGenes <- genes(ensDb)
mcols(grGenes) %<>%
  as.data.frame() %>%
  left_join(gcGene) %>%
  as.data.frame() %>%
  DataFrame()
```

```{r}
# all the metadata from the fish in analysis
meta <- 
  read_excel("data/meta_final.xlsx") %>% 
  mutate(genotype = case_when(
    genotype == "wt" ~ "WT", 
    genotype == "het" ~ "sgsh/+",
    genotype == "naglu" ~ "MPS-IIIB", 
    genotype == "sgsh" ~ "MPS-IIIA",
    genotype == "hgsnat" ~ "MPS-IIIC"
  ) %>% 
    factor(levels = c("WT", "sgsh/+", "MPS-IIIA", "MPS-IIIB", "MPS-IIIC")),
  Tank = as.factor(`Home tank`)
  )
```

Each RNA-seq experiment will be analysed seperately 

# A and B analysis.

In this family, there were > 100 sibing fish all raised together side in the same system

```{r}
featureCounts <- 
  read_delim(file = "data/AB/featurecounts/counts.out", delim = "\t", skip = 1) %>%
  set_names(basename(names(.))) %>% 
  set_names(names(.) %>% str_remove(pattern = "_S[0-9]+_merged.Aligned.sortedByCoord.dedup.out.bam")) %>% 
  as_tibble() %>% 
  dplyr::select(-c(Chr, Start, End, Length, Strand)) %>% 
  gather(key = "ULN", value = "counts", starts_with("22")) %>% 
  left_join(meta) %>% 
  dplyr::select(Geneid, counts, sample) %>% 
  spread(key = "sample", value = "counts") %>% 
  column_to_rownames("Geneid") %>% 
  .[,-(dim(.)[2])]
```

## filter lowly expressed genes

following the 10/min lib size rule
```{r filt}
a <- featureCounts %>% 
    cpm(log = TRUE) %>%
    as.data.frame() %>% 
  pivot_longer(
    cols = everything(),
    names_to = "sample",
    values_to = "logCPM"
  ) %>%
  split(f = .$sample) %>%
  lapply(function(x){
    d <- density(x$logCPM)
    tibble(
      sample = unique(x$sample),
      x = d$x,
      y = d$y
    )
  }) %>%
  bind_rows() %>%
  left_join(meta) %>% 
  ggplot(aes(x, y, colour = genotype, group = sample)) +
  geom_line() +
  labs(
    x = "logCPM",
    y = "Density",
    colour = "Genotype"
  )+
  ggtitle("Before filtering") +
  theme_bw()

b <- featureCounts %>% 
  .[rowSums(cpm(.) >= 0.5) >= 4,] %>% 
    cpm(log = TRUE) %>%
    as.data.frame() %>% 
  pivot_longer(
    cols = everything(),
    names_to = "sample",
    values_to = "logCPM"
  ) %>%
  split(f = .$sample) %>%
  lapply(function(x){
    d <- density(x$logCPM)
    tibble(
      sample = unique(x$sample),
      x = d$x,
      y = d$y
    )
  }) %>%
  bind_rows() %>%
  left_join(meta) %>% 
  ggplot(aes(x, y, colour = genotype, group = sample)) +
  geom_line() +
  labs(
    x = "logCPM",
    y = "Density",
    colour = "Genotype"
  )+
  ggtitle("After filtering")

ggarrange(a, b, common.legend = TRUE)
```

```{r dge}
x.ab <- featureCounts %>% 
  as.matrix() %>% 
  .[rowSums(cpm(.) >= 0.5) >= 4,] %>%
  DGEList(
    samples = tibble(sample = colnames(.)) %>%
      left_join(meta),
    genes = grGenes[rownames(.)] %>%
      as.data.frame() %>%
      dplyr::select(
        chromosome = seqnames, start, end, 
        gene_id, gene_name, gene_biotype, description, entrezid
      ) %>% 
      left_join(gcGene) %>% 
      as_tibble()
  ) %>%
  calcNormFactors()
```

## PCA
Principal component analysis was performed to assess the overall similarity between samples and to assess any batch effects. 
```{r}
x.ab$counts %>% 
  cpm(log=TRUE) %>% 
  t() %>% 
  prcomp() %>% 
  autoplot(data = tibble(sample = rownames(.$x)) %>%
      left_join(x.ab$samples),
    colour = "genotype", 
    size = 6
  ) +
  theme(aspect.ratio = 1) +
  labs(colour = "Genotype", 
       title = "By genotype") +
  scale_color_brewer(palette = "Set2")


x.ab$counts %>% 
  cpm(log=TRUE) %>% 
  t() %>% 
  prcomp() %>% 
  autoplot(data = tibble(sample = rownames(.$x)) %>%
      left_join(x.ab$samples),
    colour = "Sex", 
    size = 6
  ) +
  theme(aspect.ratio = 1) +
  labs(colour = "sex", 
       title = "By sex")

x.ab$counts %>% 
  cpm(log=TRUE) %>% 
  t() %>% 
  prcomp() %>% 
  autoplot(data = tibble(sample = rownames(.$x)) %>%
        left_join(x.ab$samples),
      colour = "Home.tank", 
      shape = "genotype",
    size = 6
  ) +
  theme(aspect.ratio = 1) +
  labs(colour = "Home.tank",
       title = "By home tank"
       )
```

No clear batch effects are observed. However. one outlier sgsh heterozygous fish is possible. Will keep this in mind when looking at the differentially expressed genes and gene sets later to see if it is driving the changes to gene expression. 

## %GC and Length check

In RNA-seq data, there is a known phenomenon of that genes can tend to be differentially expressed (DE) based on its % GC content and length. Here, I will perform an initial DE analysis using generlaised linear models and likelihood ratio tests in `edgeR.`A design matrix was constructed which specifies the effect of genotype accounting for any effects of home.tank. 

```{r}
# construct a design matrix and tidy up the columnnames
design.ab <- model.matrix(~genotype + Home.tank, data = x.ab$samples %>% 
                            droplevels()) %>% # need to remove the unused factor level using droplevels
  set_colnames(gsub(colnames(.), pattern = "genotype", replacement = ""))

# fit the glm
fit_1.ab <- x.ab %>% 
  estimateDisp(design.ab) %>% 
  glmFit(design.ab)

# call the DE genes
toptable_1.ab <- design.ab %>% colnames() %>% .[2:4] %>% 
  sapply(function(x){
    glmLRT(fit_1.ab, coef = x) %>%
      topTags(n = Inf) %>%
      .[["table"]] %>%
      as_tibble() %>%
      arrange(PValue) %>%
       mutate(
        coef = x,
        DE = FDR < 0.05
      ) %>% 
      dplyr::select(
        gene_name, logFC, logCPM, PValue, FDR, DE, everything()  
      )
      
  }, simplify = FALSE)
```

### vis
#### Number of DE genes
```{r}
toptable_1.ab %>% 
  lapply(function(y) {
    y %>% 
      dplyr::filter(DE == T) %>% 
      nrow()
  }) %>% 
  bind_rows() %>% 
  as_tibble() %>% 
  gather(key = "coef", value = "numDE") %>% 
  ggplot(aes(x = numDE, y = coef)) +
  geom_col(aes(fill = coef)) +
  geom_text(aes(label = numDE)) +
  ggtitle("Number of DE genes in each comparison",
          subtitle = "The MPS-IIIA zebrafish samples have the most DE genes")
```

#### Volcano plot
```{r}
toptable_1.ab %>%
  bind_rows() %>% 
  ggplot(aes(y = -log10(PValue), x = logFC, colour = DE)) +
  geom_point(
    alpha = 0.5, size = 1.25
  ) +
  facet_wrap(~coef) +
  # geom_label_repel(
  #   aes(label = gene_name),
  #   data = .  %>% dplyr::filter(FDR < 0.05),
  #   show.legend = FALSE
  #   ) +
  coord_cartesian(xlim = c(-2.5,2.5), 
                  ylim = c(0, 20) )+
  theme(legend.position = "none") +
  scale_color_manual(values = c("grey50", "red")) 
  #ggsave("output/plots/volvano.png", width = 18, height = 10, units = "cm", dpi = 300, scale = 1.5)
```

#### MD plot
```{r}
toptable_1.ab %>%
  bind_rows() %>% 
  ggplot(aes(x = logCPM, y = logFC)) +
  geom_point(
    aes(colour = DE),
    alpha = 0.5
  ) +
  facet_wrap(~coef, ncol = 1) +
  theme(legend.position = "none") +
  scale_color_manual(values = c("grey50", "red")) +
  geom_smooth(se = F)

```

### %GC and Length

These plots are zoomed in between -10 and 10 on the y-axisfor vis purposes. the blue lines of best fit are a bit wonky, meaning there are some biases here for %GC and Length. Therefore, conditional quantile normailsation is required. 
```{r}
ggarrange(
toptable_1.ab %>% 
  bind_rows() %>% 
  mutate(rankstat = sign(logFC)*-log10(PValue)) %>% 
  ggplot(aes(x = length, y = rankstat)) +
  geom_point(
    aes(colour = DE),
    alpha = 0.5
  ) +
  geom_smooth(se = FALSE, method = "gam") +
  facet_grid(rows = vars(coef)) +
  theme_bw() +
  theme(legend.position = "none") +
  scale_color_manual(values = c("grey50", "red")) +
  scale_x_log10()+
  labs(x = "Average transcript length per gene",
       colour = "Differentially expressed?",
       y = "sign(logFC)*-log10(PValue)") +
  coord_cartesian(ylim = c(-10, 10)),

toptable_1.ab %>% 
  bind_rows() %>% 
  mutate(rankstat = sign(logFC)*-log10(PValue)) %>% 
  ggplot(aes(x = gc_content, y = rankstat)) +
  geom_point(
    aes(colour = DE),
    alpha = 0.5
  ) +
  geom_smooth(se = FALSE, method = "gam") +
  facet_grid(rows = vars(coef)) +
  theme_bw() +
  theme(legend.position = "none") +
  scale_color_manual(values = c("grey50", "red")) +
  coord_cartesian(ylim = c(-10,10)) + 
  labs(x = "Weighted average GC content (%) per gene", 
       colour = "Differentially expressed?", 
       y = "sign(logFC)*-log10(PValue)"),
common.legend = TRUE, 
labels = "AUTO"
) 
```

## CQN

A little bit of bias is obsevred for %GC content and gene length, so will apply the `cqn` correction to adjust for this. 

```{r}
cqn.ab <-   
x.ab %>% 
  with(
    cqn(
      counts = counts,
      x = genes$gc_content,
      lengths = genes$length,
      sizeFactors = samples$lib.size
    )
  )
logCPM.ab <- cqn.ab$y + cqn.ab$offset
```

### plot out the model fits
```{r, fig.cap ="*Model fits for GC content and gene length under the CQN model. Variability is clearly visible at either end*"}
par(mfrow = c(1, 2))
cqnplot(cqn.ab, n = 1, xlab = "GC Content")
cqnplot(cqn.ab, n = 2, xlab = "Length")
par(mfrow = c(1, 1))
```

### PCA after cqn
PCA analysis was repeated. Similar clustering of samples was observed, with MPS-IIIA samples appearing to form a more distinct cluster from the rest of the samples. The sgsh/+ samples appear much more variable. 

```{r}
ggarrange(
  cpm(x.ab, log = T) %>%
    t() %>%
    prcomp() %>% 
    autoplot(data = tibble(sample = rownames(.$x)) %>%
               left_join(x.ab$samples),
             colour = "genotype", 
             shape = "Sex", 
             size = 4
    ) +
    scale_color_viridis_d(option = "turbo", end = 0.8) +
    theme(aspect.ratio = 1) +
    ggtitle("Before CQN"),
  
  logCPM.ab %>% 
    t() %>% 
    prcomp() %>%
    autoplot(data = tibble(sample = rownames(.$x)) %>%
               left_join(x.ab$samples),
             colour = "genotype", 
             shape = "Sex", 
             size = 4
    ) +
    scale_color_viridis_d(option = "turbo", end = 0.8) +
    theme(aspect.ratio = 1) +
    ggtitle("After CQN"),
  common.legend = T, 
  labels = "AUTO"
)
```

## DE with cqn

The DE analysis was then repeated, this time including the offset generated by `cqn` in the model. 
```{r}
# Add the offset to the dge object 
x.ab$offset <- cqn.ab$glm.offset

# fit the glm
fit_cqn.ab <- x.ab %>% 
  estimateDisp(design.ab) %>% 
  glmFit(design.ab)

toptables_cqn.ab <- design.ab %>% colnames() %>% .[2:4] %>% 
  sapply(function(x){
    glmLRT(fit_cqn.ab, coef = x) %>%
      topTags(n = Inf) %>%
      .[["table"]] %>%
      as_tibble() %>%
      arrange(PValue) %>%
       mutate(
        coef = x,
        DE = FDR < 0.05
      ) %>% 
      dplyr::select(
        gene_name, logFC, logCPM, PValue, FDR, DE, everything()  
      )
      
  }, simplify = FALSE)
```

## Visualisation
### Number of DE genes
```{r}
toptables_cqn.ab %>% 
  lapply(function(x){
    x %>%
      mutate(Direction = case_when(
        sign(logFC) == "1" ~ "up",
        sign(logFC) == "-1" ~ "down"
      ))
  }) %>%
  bind_rows() %>%
  dplyr::filter(DE == TRUE) %>%
  ggplot(aes(y = `coef`)) +
  geom_bar(stat = "count",
           width = 0.5,
           aes(fill = Direction),
           position = "dodge") +
  theme(legend.position = "bottom") +
  ggtitle("Number of DE genes in each comparison") +
  labs(x = "Number of DE genes",
       y = "Contrast",
       colour = "Direction of change", 
       title = "Number of DE genes in each comparison")
```
### Volcano plot
```{r}
toptables_cqn.ab %>%
  bind_rows() %>% 
  ggplot(aes(y = -log10(PValue), x = logFC, colour = DE)) +
  geom_point(
    alpha = 0.5, size = 1.25
  ) +
  facet_wrap(~coef) +
  # geom_label_repel(
  #   aes(label = gene_name),
  #   data = .  %>% dplyr::filter(FDR < 0.05),
  #   show.legend = FALSE
  #   ) +
  coord_cartesian(xlim = c(-2.5,2.5), 
                  ylim = c(0, 20) )+
  theme(legend.position = "none") +
  scale_color_manual(values = c("grey50", "red")) 
  #ggsave("output/plots/volvano.png", width = 18, height = 10, units = "cm", dpi = 300, scale = 1.5)
```

### MD plot
```{r}
toptables_cqn.ab %>%
  bind_rows() %>% 
  ggplot(aes(x = logCPM, y = logFC)) +
  geom_point(
    aes(colour = DE),
    alpha = 0.5
  ) +
  facet_wrap(~coef, ncol = 1) +
  theme(legend.position = "none") +
  scale_color_manual(values = c("grey50", "red")) +
  geom_smooth(se = F)
```

## Recheck the %GC and length bias

The %GC and length bias data was checked again. Some gene length bias still remains. But this seems to be driven only at the ends, where there is less genes. So this should be ok.
```{r}
ggarrange(
toptables_cqn.ab %>% 
  bind_rows() %>% 
  mutate(rankstat = sign(logFC)*-log10(PValue)) %>% 
  ggplot(aes(x = length, y = rankstat)) +
  geom_point(
    aes(colour = DE),
    alpha = 0.5
  ) +
  geom_smooth(se = FALSE, method = "gam") +
  facet_grid(rows = vars(coef)) +
  theme_bw() +
  theme(legend.position = "none") +
  scale_color_manual(values = c("grey50", "red")) +
  scale_x_log10()+
  labs(x = "Average transcript length per gene",
       colour = "Differentially expressed?",
       y = "sign(logFC)*-log10(PValue)") +
  coord_cartesian(ylim = c(-10, 10)),

toptables_cqn.ab %>% 
  bind_rows() %>% 
  mutate(rankstat = sign(logFC)*-log10(PValue)) %>% 
  ggplot(aes(x = gc_content, y = rankstat)) +
  geom_point(
    aes(colour = DE),
    alpha = 0.5
  ) +
  geom_smooth(se = FALSE, method = "gam") +
  facet_grid(rows = vars(coef)) +
  theme_bw() +
  theme(legend.position = "none") +
  scale_color_manual(values = c("grey50", "red")) +
  coord_cartesian(ylim = c(-10,10)) + 
  labs(x = "Weighted average GC content (%) per gene", 
       colour = "Differentially expressed?", 
       y = "sign(logFC)*-log10(PValue)"),
common.legend = TRUE, 
labels = "AUTO"
) 
```

## Gene set enrichment analysis

I now am going to test the KEGG, gene ontology (GO), iron-responsive element (IRE), cell type markers and chromosomal location gene sets to determine whether any of them display changes to gene expression as a group. The gene sets were obtained from MSIGDB using the msigdbr package. 

The KEGG gene sets were filtered to keep only the KEGG gene sets which contained at least 10 detectable genes in the dataset. 

The GO term list was filtered to keep only the GO terms which contained at least 5 detectable genes in the dataset. 
### Gene sets

```{r}
KEGG <- msigdbr("Danio rerio", category = "C2", subcategory = "CP:KEGG") %>% 
  left_join(x.ab$genes, by = c("gene_symbol" = "gene_name")) %>% 
  dplyr::filter(gene_id %in% rownames(x.ab)) %>% 
  distinct(gs_name, gene_id, .keep_all = TRUE) %>% 
  split(f = .$gs_name) %>%
  lapply(extract2, "gene_id")

sizes <- KEGG %>% 
  lapply(length) %>% 
  unlist %>% 
  as.data.frame() %>% 
  set_colnames( "n_genes") %>% 
  rownames_to_column("gs")


# retain gene sets with at least 10 genes in it
KEGG <- KEGG[sizes %>% dplyr::filter(n_genes > 10) %>% .$gs]

# GO Terms
goSummaries <- url("https://uofabioinformaticshub.github.io/summaries2GO/data/goSummaries.RDS") %>%
  readRDS() %>%
  mutate(
    Term = Term(id),
    gs_name = Term %>% str_to_upper() %>% str_replace_all("[ -]", "_"),
    gs_name = paste0("GO_", gs_name)
    )

minPath <- 3

GO <- 
  msigdbr("Danio rerio", category = "C5") %>% 
  dplyr::filter(grepl(gs_name, pattern = "^GO")) %>% 
  left_join(x.ab$genes, by = c("gene_symbol" = "gene_name")) %>% 
  dplyr::filter(gene_id %in% rownames(x.ab)) %>% 
  mutate(gs_name = str_replace(gs_name, pattern = "GOBP_", replacement = "GO_")) %>% 
  mutate(gs_name = str_replace(gs_name, pattern = "GOMF_", replacement = "GO_")) %>%     
  mutate(gs_name = str_replace(gs_name, pattern = "GOCC_", replacement = "GO_")) %>%     
  left_join(goSummaries) %>% 
  dplyr::filter(shortest_path >= minPath) %>%
  distinct(gs_name, gene_id, .keep_all = TRUE) %>% 
  split(f = .$gs_name) %>%
  lapply(extract2, "gene_id")

# GOdf <-  msigdbr("Danio rerio", category = "C5") %>% 
#   dplyr::filter(grepl(gs_name, pattern = "^GO")) %>% 
#   left_join(x.ab$genes, by = c("gene_symbol" = "gene_name")) %>% 
#   dplyr::filter(gene_id %in% rownames(x)) %>% 
#   mutate(gs_name = str_replace(gs_name, pattern = "GOBP_", replacement = "GO_")) %>% 
#   mutate(gs_name = str_replace(gs_name, pattern = "GOMF_", replacement = "GO_")) %>%     
#   mutate(gs_name = str_replace(gs_name, pattern = "GOCC_", replacement = "GO_")) %>%     
#   left_join(goSummaries) %>% 
#   dplyr::filter(shortest_path >= minPath) %>%
#   distinct(gs_name, gene_id, .keep_all = TRUE) %>% 
#   dplyr::select(gs_name, gene_id)

sizes_GO <- GO %>%
  lapply(length) %>%
  unlist %>%
  as.data.frame() %>%
  set_colnames( "n_genes") %>%
  rownames_to_column("gs")


# retain gene sets with at least 5 genes in it
GO <- GO[sizes_GO %>% dplyr::filter(n_genes > 5) %>% .$gs]

ireGenes <- 
  read_xlsx("data/gene_sets/ireGenes.xlsx", sheet = "Zebrafish IREs") %>% 
  gather %>% 
  set_colnames(c("IRE", "gene_id")) %>% 
  mutate(IRE = case_when(
    IRE == "All 3' IREs" ~ "ire3_all", 
    IRE == "All 5' IREs" ~ "ire5_all", 
    IRE == "HQ 3' IREs" ~ "ire3_HQ", 
    IRE == "HQ 5' IREs" ~ "ire5_HQ"
  )) %>% 
  dplyr::filter(gene_id %in% rownames(x.ab)) %>% 
  split(f = .$IRE) %>% 
  lapply(magrittr::extract2,"gene_id") 

chr <- x.ab$genes %>% 
  dplyr::select(gene_id, chromosome) %>% 
  split(f = .$chromosome) %>%
  lapply(extract2, "gene_id") %>% 
  .[c("MT", seq(1:25))] # omit the ALT chromosmes and scaffolds


cell_type_markers <- 
  read_xlsx("data/gene_sets/Suppdata2_jiangetal_2021_fcelldev.xlsx", sheet = "Brain") %>% 
  dplyr::rename("gene_name" = gene) %>% 
  left_join(x.ab$genes) %>% 
  dplyr::filter(gene_id %in% rownames(x.ab)) %>% 
  split(f = .$`cell type`) %>% 
  lapply(extract2, "gene_id")
  

cell_type_markers %<>% 
  lapply(function(y) {
    y[y %in% rownames(x.ab)]
  })

allC2.ab <- msigdbr("Danio rerio", category = "C2") %>% 
  left_join(x.ab$genes, by = c("gene_symbol" = "gene_name")) %>% 
  dplyr::filter(gene_id %in% rownames(x.ab)) %>% 
  distinct(gs_name, gene_id, .keep_all = TRUE) %>% 
  split(f = .$gs_name) %>%
  lapply(extract2, "gene_id")
```

## GOSEQ 

I first will look at enrichment of GO terms within the DE genes using goseq. I like goseq as it lets you include a covariate. Since there was still a bit of bias left after cqn for gene length, i will use this as the covariate.  
```{r}

pwf <- toptables_cqn.ab$`MPS-IIIB` %>%
      with(
        nullp(
          DEgenes = structure(DE, names = gene_id), 
          bias.data = length
        )
      )

goseq(pwf, gene2cat = GO) %>% 
  as_tibble() %>% 
  dplyr::filter(numDEInCat > 0) %>% 
  mutate(FDR = p.adjust(over_represented_pvalue, method = "fdr")) %>% 
  dplyr::select(-under_represented_pvalue) %>% 
  #write_csv("output/results/goseq_cqn/famAB_Bcoef.csv")
  dplyr::slice(1:10) %>% 
  kable(caption = "Top 10 over-represented GO terms in MPS-IIIB DE genes. 
        None reach statistical significance") %>% 
  kable_styling(full_width = FALSE)

pwf <- toptables_cqn.ab$`MPS-IIIA` %>%
      with(
        nullp(
          DEgenes = structure(DE, names = gene_id), 
          bias.data = length
        )
      )

goseq(pwf, gene2cat = GO) %>% 
  as_tibble() %>% 
  dplyr::filter(numDEInCat > 0) %>% 
  mutate(FDR = p.adjust(over_represented_pvalue, method = "fdr")) %>% 
  dplyr::select(-under_represented_pvalue) %>% 
  # write_csv("output/results/goseq_cqn/famAB_Acoef.csv")
  dplyr::slice(1:10)%>% 
  kable(caption = "Top 10 over-represented GO terms in MPS-IIIA DE genes. 
        Some reach statistical significance with FDR < 0.05") %>% 
  kable_styling(full_width = FALSE)

pwf <- toptables_cqn.ab$`sgsh/+` %>%
      with(
        nullp(
          DEgenes = structure(DE, names = gene_id), 
          bias.data = length
        )
      )

goseq(pwf, gene2cat = GO) %>% 
  as_tibble() %>% 
  dplyr::filter(numDEInCat > 0) %>% 
  mutate(FDR = p.adjust(over_represented_pvalue, method = "fdr")) %>% 
  dplyr::select(-under_represented_pvalue) %>% 
  #write_csv("output/results/goseq_cqn/famAB_sgshHetcoef.csv")
  dplyr::slice(1:10)
  
  
  kable(caption = "Top 10 over-represented GO terms in sgsh het DE genes. 
        Some reach statistical significance") %>% 
  kable_styling(full_width = FALSE)
  
```



### KEGG 
```{r}
fryKEGG.ab <- 
  design.ab %>% colnames() %>% .[2:4] %>% 
  sapply(function(y) {
    logCPM.ab %>% 
      fry(
        index = KEGG,
        design = design.ab, 
        contrast = y, 
        sort = "mixed"
      ) %>% 
      rownames_to_column("pathway") %>% 
      as_tibble() %>% 
      mutate(coef = y)
  }, simplify = FALSE)

fryKEGG.ab$`MPS-IIIB` %>% 
  dplyr::slice(1:10) %>% 
  mutate(pathway = str_remove(pathway, pattern = "KEGG_")) %>% 
  ggplot(aes(x = -log10(PValue.Mixed), y = reorder(pathway, -FDR.Mixed))) +
  geom_col(aes(fill = FDR.Mixed < 0.05)) +
  scale_fill_manual(values = c("grey50", 'red')) +
  theme_pubr() +
  labs(fill = "FDR < 0.05", 
       x = "-log10(PValue)", 
       y = "")

fryKEGG.ab$`MPS-IIIA` %>% 
  dplyr::slice(1:10) %>% 
  mutate(pathway = str_remove(pathway, pattern = "KEGG_")) %>% 
  ggplot(aes(x = -log10(PValue.Mixed), y = reorder(pathway, -FDR.Mixed))) +
  geom_col(aes(fill = FDR.Mixed < 0.05)) +
  scale_fill_manual(values = c("grey50", 'red')) +
  theme_pubr() +
  labs(fill = "FDR < 0.05", 
       x = "-log10(PValue)", 
       y = "")  

fryKEGG.ab$`sgsh/+` %>% 
  dplyr::slice(1:10) %>% 
  mutate(pathway = str_remove(pathway, pattern = "KEGG_")) %>% 
  ggplot(aes(x = -log10(PValue.Mixed), y = reorder(pathway, -PValue.Mixed))) +
  geom_col(aes(fill = FDR.Mixed < 0.05)) +
  scale_fill_manual(values = c("grey50", 'red')) +
  theme_pubr() +
  labs(fill = "FDR < 0.05", 
       x = "-log10(PValue)", 
       y = "")

sigpaths <- fryKEGG.ab %>% 
  bind_rows() %>% 
  dplyr::filter(FDR.Mixed < 0.05) %>% 
    .$pathway %>% 
    unique()
  
fryKEGG.ab %>% 
  bind_rows() %>% 
  dplyr::filter(pathway %in% sigpaths) %>% 
  mutate(pathway = str_remove(pathway, pattern = "KEGG_")) %>% 
  ggplot(aes(x = -log10(PValue.Mixed), y = reorder(pathway, -FDR.Mixed))) +
  geom_col(aes(fill = coef, alpha = FDR.Mixed < 0.05), position = "dodge") +
  scale_fill_viridis_d(end = 0.75) +
  scale_alpha_manual(values = c(0.5,1)) +
  theme_pubr() +
  geom_vline(xintercept = 2.3) +
  labs(fill = "FDR < 0.05", 
       x = "-log10(PValue)", 
       y = "")
  #ggsave("output/plots/AB_KEGG.png", width = 20, height = 10, units = "cm", dpi = 400, scale = 2)

```

```{r}
fryC2.ab <- 
  design.ab %>% colnames() %>% .[2:4] %>% 
  sapply(function(y) {
    logCPM.ab %>% 
      fry(
        index = allC2.ab,
        design = design.ab, 
        contrast = y, 
        sort = "mixed"
      ) %>% 
      rownames_to_column("pathway") %>% 
      as_tibble() %>% 
      mutate(coef = y)
  }, simplify = FALSE)


sigpaths <- fryC2.ab %>% 
  bind_rows() %>% 
  dplyr::filter(FDR.Mixed < 0.05) %>% 
    .$pathway %>% 
    unique()
  
fryC2.ab %>% 
  bind_rows() %>% 
  dplyr::filter(pathway %in% sigpaths) %>% 
  mutate(pathway = str_remove(pathway, pattern = "KEGG_")) %>% 
  ggplot(aes(x = -log10(PValue.Mixed), y = reorder(pathway, -FDR.Mixed))) +
  geom_col(aes(fill = coef, alpha = FDR.Mixed < 0.05), position = "dodge") +
  scale_fill_viridis_d(end = 0.75) +
  scale_alpha_manual(values = c(0.5,1)) +
  theme_pubr() +
  geom_vline(xintercept = 2.3) +
  labs(fill = "FDR < 0.05", 
       x = "-log10(PValue)", 
       y = "")
```


### GO
```{r}
fry_GO.ab <-  design.ab %>% colnames() %>% .[2:4] %>% 
  sapply(function(y) {
    logCPM.ab %>% 
      fry(
        index = GO,
        design = design.ab, 
        contrast = y, 
        sort = "mixed"
      ) %>% 
      rownames_to_column("pathway") %>% 
      as_tibble() %>% 
      mutate(coef = y, 
             sig = FDR.Mixed < 0.05)
  }, simplify = FALSE)
```


```{r}
toptables_cqn %>% 
  bind_rows() %>% 
  dplyr::filter(gene_id %in% GO$GO_REGULATION_OF_MICROGLIAL_CELL_ACTIVATION) %>% 
  dplyr::select(gene_name, logFC, coef) %>% 
  dplyr::distinct(gene_name, coef, .keep_all = T) %>% 
  spread(key = "coef", value = "logFC") %>% 
  column_to_rownames("gene_name") %>% 
  t() %>% 
  pheatmap(color = colorRampPalette(rev(brewer.pal(n = 5,
                                               name = "RdBu")))(100), 
           main = "GO_REGULATION_OF_MICROGLIAL_CELL_ACTIVATION: FDR = 6.44e-5", 
           breaks = c(seq(min(.), 0, length.out=ceiling(100/2) + 1), 
              seq(max(.)/100, max(.), length.out=50)), 
           cellheight = 30, cellwidth = 12,angle_col = 45,
           treeheight_row = 0, treeheight_col = 0
           )


toptables_cqn %>% 
  bind_rows() %>% 
  dplyr::filter(gene_id %in% KEGG$KEGG_OXIDATIVE_PHOSPHORYLATION) %>% 
  dplyr::select(gene_name, logFC, coef) %>% 
  dplyr::distinct(gene_name, coef, .keep_all = T) %>% 
  spread(key = "coef", value = "logFC") %>% 
  column_to_rownames("gene_name") %>% 
  t() %>% 
  pheatmap(color = colorRampPalette(rev(brewer.pal(n = 5,
                                               name = "RdBu")))(100), 
           main = "KEGG_LYSOSOME: FDR = 6.44e-5", 
           breaks = c(seq(min(.), 0, length.out=ceiling(100/2) + 1), 
              seq(max(.)/100, max(.), length.out=50)), 
           cellheight = 30, cellwidth = 12,angle_col = 45,
           treeheight_row = 0, treeheight_col = 0
           )
```

### IRE

No obsvered evidence for changes to gene expression in the IRE genes
```{r}
fryIRE.ab <- design.ab %>% colnames() %>% .[2:4] %>% 
  sapply(function(y) {
    logCPM.ab %>% 
      fry(
        index = ireGenes,
        design = design.ab, 
        contrast = y, 
        sort = "directional"
      ) %>% 
      rownames_to_column("pathway") %>% 
      as_tibble() %>% 
      mutate(coef = y)
  }, simplify = FALSE)

fryIRE.ab %>% 
  bind_rows() %>% 
  dplyr::select(pathway, coef, pvalue = PValue.Mixed, FDR = FDR.Mixed) %>% 
  ggplot(aes(y = pathway, x = -log10(pvalue))) + 
  geom_col(aes(fill = coef, alpha = ), 
           position = "dodge") +
  scale_fill_viridis_d(end = 0.75) +
  geom_vline(xintercept = 1.88) +
  annotate(geom = "label", x = 1.8, y = "ire5_HQ", label = "FDR = 0.05")

```

### Cell type prop check

Since this experiment was performed on whole larvae, we need to have a look at whether changes to cell type proportions could explain cause some artefactual changes to gene expression. We have done this previously by assessing the expression of marker genes of specific cell types. So this is what I will do here. I have taken the lists of genes frmo day 5 only from the zebrafish single cell atlas (Farnsworth et al. 2020) which contain at least 10 genes. This leaves  `r length(cell_type_markers)` gene sets (cell types) to test.

```{r}
celltypes.ab <- design.ab %>% colnames() %>% .[2:4] %>% 
  sapply(function(y) {
    logCPM.ab %>% 
      fry(
        index = cell_type_markers,
        design = design.ab, 
        contrast = y, 
        sort = "directional"
      ) %>% 
      rownames_to_column("pathway") %>% 
      as_tibble() %>% 
      dplyr::select(1:5) %>% 
      mutate(coef = y)
  }, simplify = FALSE)

celltypes.ab %>% 
  bind_rows() %>% 
  ggplot(aes(y = pathway, x = -log10(PValue))) + 
  geom_col(aes(fill = coef, alpha = FDR < 0.05), 
           position = "dodge") +
  scale_fill_viridis_d(end = 0.75) +
  scale_alpha_manual(values = c(0.5,1)) +
  geom_vline(xintercept = -log10(0.012)) +
  annotate(geom = "label", x = -log10(0.01), y = 23, label = "FDR = 0.05") 
```


### Chrosomosomes gene sets

```{r}
fry_chr.ab <- design.ab %>% colnames() %>% .[2:4] %>% 
  sapply(function(y) {
    logCPM.ab %>% 
      fry(
        index = c(chr),
        design = design.ab, 
        contrast = y, 
        sort = "mixed"
      ) %>% 
      rownames_to_column("pathway") %>% 
      as_tibble() %>% 
      mutate(coef = y)
  }, simplify = FALSE)


fry_chr.ab %>% 
  bind_rows() %>% 
  mutate(pathway = factor(pathway, levels = c("MT", seq(1:25)))) %>% 
  ggplot(aes(y = pathway, x = -log10(PValue.Mixed), fill = coef)) +
  geom_col(position = "dodge")
  
```

An enrichment of DE genes is observed on the same chromosome as the mutation. This is obsevred in the manhat plots below. 

```{r}
manhat_input.ab <- 
	toptables_cqn.ab %>% 
  bind_rows() %>% 
	left_join(x.ab$genes %>% 
			  	as_tibble() %>% 
			  	dplyr::select(start, end, chromosome, gene_id) %>% 
			  	group_by(gene_id) %>% 
			  	mutate(mid = mean(c(start, end)))
	) %>% 
  dplyr::filter(chromosome %in% seq(1:25)) %>% 
  group_by(gene_id) %>% 
  mutate(mid = mean(c(start, end))) %>% 
  ungroup %>% 
  mutate(chromosome = factor(chromosome, levels = seq(1:25)))  %>% 
  group_by(chromosome) %>% 
  summarise(chrLen = max(mid)) %>% 
  mutate(chrSt = cumsum(chrLen)-chrLen) %>% 
	dplyr::select(-chrLen) %>% 
  left_join(toptables_cqn.ab %>% 
              bind_rows() %>% 
              left_join(x.ab$genes %>% 
                          as_tibble() %>% 
                          dplyr::select(start, end, chromosome, gene_id) %>% 
                          group_by(gene_id) %>% 
                          mutate(mid = mean(c(start, end)))
              )
  ) %>% 
  group_by(gene_id) %>% 
  mutate(mid = mean(c(start, end))) %>% 
  ungroup %>% 
  dplyr::arrange(mid, chromosome) %>% 
  mutate(midCum = chrSt + mid)

axis.ab <- manhat_input.ab %>%
  group_by(chromosome) %>%
  dplyr::arrange(chromosome) %>%
  summarize(center = (max(midCum) + min(midCum)) / 2) %>%
  mutate(
    seqnames = as.numeric(chromosome),
    colour = rep(c("grey40", "black"), length.out = 25)
  )

ggplot(manhat_input.ab, aes(x = midCum, y = -log10(PValue))) +
  geom_point(aes(color=chromosome), alpha = 0.5, size = 1) +
  geom_point(alpha = 0.5, size = 1, colour = "red", 
             data = . %>% 
               dplyr::filter(coef == "MPS-IIIB" & chromosome == 24)) +
  geom_point(alpha = 0.5, size = 1, colour = "red", 
             data = . %>% 
               dplyr::filter(coef ==  "MPS-IIIA" & chromosome == 22)) +
  geom_point(alpha = 0.5, size = 1, colour = "red", 
             data = . %>% 
               dplyr::filter(coef ==  "sgsh/+" & chromosome == 22)) +
  scale_color_manual(values = axis.ab$colour) +
  scale_x_continuous(label = axis.ab$chromosome, breaks = axis.ab$center) +
  scale_y_continuous(limits =c(0, 10)) + # zoom in
  facet_wrap(~coef, ncol = 1) +
  labs(x = "Chromosome", y = expression(paste(-log[10], "(p)"))) +
  theme( 
    legend.position="none",
    panel.border = element_blank(),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.x = element_blank()
  ) 
```


I also compared the log2 fold changes of genes on chr22 in the heterozygous and homozygous sgsh samples.

```{r}
DEgenesIIIA <- toptables_cqn.ab$`MPS-IIIA` %>% 
  dplyr::filter(DE == T) %>% 
  .$gene_id

DEgenesIIIB <- toptables_cqn.ab$`MPS-IIIB` %>% 
  dplyr::filter(DE == T) %>% 
  .$gene_id

DEgeneshet <- toptables_cqn.ab$`sgsh/+` %>% 
  dplyr::filter(DE == T) %>% 
  .$gene_id

toptables_cqn.ab %>% 
  bind_rows() %>% 
  dplyr::filter(coef %in% c("MPS-IIIA", "sgsh/+")) %>% 
  dplyr::filter(chromosome == "22") %>% 
  dplyr::select(gene_id, logFC, coef)  %>% 
  spread(key = "coef", value = "logFC") %>% 
  mutate(DE = case_when(
    gene_id %in% intersect(DEgenesIIIA, DEgeneshet) ~ "both", 
    gene_id %in% DEgeneshet[!DEgeneshet %in% DEgenesIIIA] ~ "het only",
    gene_id %in% DEgenesIIIA[!DEgenesIIIA %in% DEgeneshet] ~ "MPS-IIIA only", 
    TRUE ~ "not DE"
  ) 
  ) %>% 
  left_join(x.ab$genes) %>% 
  group_by(gene_id) %>% 
  mutate(mid = mean(c(start, end))) %>% 
  ggplot(aes(x = `MPS-IIIA`, y =`sgsh/+`)) +
  geom_point(aes(colour = DE), 
             alpha = 0.5, 
             size = 4) +
  geom_abline(slope = 1) +
  geom_label_repel(aes(label = gene_name), 
                   nudge_y = 1,
                   data = . %>% 
                     dplyr::filter(gene_name == "sgsh")) +
  scale_x_continuous(limits = c(-5, 5)) +
  scale_y_continuous(limits = c(-2.5, 5)) +
  theme(aspect.ratio = 1) +
  scale_color_manual(values = c("red", "blue", "grey50")) +
  ggtitle("LogFC of genes on chromosome 22 in sgsh het vs hom ")

```

# is there a functional enrichment of genes only DE in the MPS-IIIA
```{r}
kegga_DE <- toptables_cqn.ab %>% 
  bind_rows() %>% 
  dplyr::filter(coef %in% c("MPS-IIIA", "sgsh/+")) %>% 
  dplyr::filter(chromosome == "22") %>% # only chr22 genes
  dplyr::filter(gene_id %in% DEgenesIIIA) %>%#  DE in the homs
  dplyr::filter(!(gene_id %in% DEgeneshet) ) %>% # and not DE in hets 
  .$gene_id %>% 
  unique
  
#KEGG using a background of all genes
kegga(de = kegga_DE, 
      
      geneid = x.ab$genes %>% 
        dplyr::filter(chromosome == 22) %>% 
        .$gene_id,
      
      universe = x.ab$genes %>% 
        dplyr::filter(chromosome == 22) %>% 
        .$gene_id,
      
      gene.pathway = KEGG %>%
        unlist %>%
        as.data.frame() %>%
        rownames_to_column("pathway") %>%
        set_colnames(c("pathway", "gene_id")) %>%
        mutate(pathway = pathway %>% str_remove_all("[0-9]+$")) %>%
        dplyr::select(gene_id, pathway)  
) %>% 
   topKEGG() %>% 
    as.data.frame() %>% 
    rownames_to_column("pathway") %>% 
    mutate(FDR = p.adjust(P.DE, "fdr")) %>% 
    dplyr::select(pathway, raw.p = P.DE, FDR, `No. DE` = DE, `No. in geneset` = N) %>%
    head(10) 
```


# A and C , B and C analyses
```{r}
featureCounts2 <- 
  read_delim(file = "data/mergedBCAC/fc/counts.out", delim = "\t", skip = 1) %>%
  set_names(basename(names(.))) %>% 
  set_names(names(.) %>% str_remove(pattern = "_S[0-9]+_merged.Aligned.sortedByCoord.dedup.out.bam")) %>% 
  as_tibble() %>% 
  dplyr::select(-c(Chr, Start, End, Length, Strand)) %>% 
  gather(key = "ULN", value = "counts", starts_with("22")) %>% 
  left_join(meta) %>% 
  dplyr::select(Geneid, counts, sample ) %>% 
  spread(key = "sample", value = "counts") %>% 
  column_to_rownames("Geneid") %>% 
  .[,-(dim(.)[2])]# omit the SAGC neg control lane, the final col of the df. 
```


## filter lowly expressed genes

```{r}
min.lib.size = featureCounts2 %>% 
    as.matrix() %>% 
    DGEList() %>%
    calcNormFactors() %>% 
  .$samples %>% 
  .$lib.size %>% 
  min()
```

Genes which are lowly expressed are uninformative for differential gene expression analysis. A general rule of thumb for considering a gene to be lowly expressed if it contains a count per million (CPM) of more than 10/(min_lib_size/1000000) in at least one of the experimental groups. This will allow us . The min lib size in this analysis is `r comma(min.lib.size)`. So the minimum CPM in this analysis will be `r 10/(min.lib.size/1000000)`.
 

```{r filt}
a <- featureCounts2 %>% 
    cpm(log = TRUE) %>%
    as.data.frame() %>% 
  pivot_longer(
    cols = everything(),
    names_to = "sample",
    values_to = "logCPM"
  ) %>%
  split(f = .$sample) %>%
  lapply(function(x){
    d <- density(x$logCPM)
    tibble(
      sample = unique(x$sample),
      x = d$x,
      y = d$y
    )
  }) %>%
  bind_rows() %>%
  left_join(meta) %>% 
  ggplot(aes(x, y, colour = genotype, group = sample)) +
  geom_line() +
  scale_y_continuous(limits = c(0,0.35)) +
  labs(
    x = "logCPM",
    y = "Density",
    colour = "Genotype"
  )+
  ggtitle("Before filtering") +
  theme_bw()

b <- featureCounts2 %>% 
  .[rowSums(cpm(.) >= 2.435138) >= 6,] %>% 
    cpm(log = TRUE) %>%
    as.data.frame() %>% 
  pivot_longer(
    cols = everything(),
    names_to = "sample",
    values_to = "logCPM"
  ) %>%
  split(f = .$sample) %>%
  lapply(function(x){
    d <- density(x$logCPM)
    tibble(
      sample = unique(x$sample),
      x = d$x,
      y = d$y
    )
  }) %>%
  bind_rows() %>%
  left_join(meta) %>% 
  ggplot(aes(x, y, colour = genotype, group = sample)) +
  geom_line() +
  scale_y_continuous(limits = c(0,0.35)) +
  labs(
    x = "logCPM",
    y = "Density",
    colour = "Genotype"
  )+
  ggtitle("After filtering")

ggarrange(a, b, common.legend = TRUE)
```


```{r dge}
x2 <- featureCounts2 %>% 
  as.matrix() %>% 
  .[rowSums(cpm(.) >= 2.435138) >= 6,] %>% 
  DGEList(
    samples = tibble(sample = colnames(.)) %>%
      left_join(meta),
    genes = grGenes[rownames(.)] %>%
      as.data.frame() %>%
      dplyr::select(
        chromosome = seqnames, start, end, 
        gene_id, gene_name, gene_biotype, description, entrezid
      ) %>% 
      left_join(gcGene) %>% 
      as_tibble()
  ) %>%
  calcNormFactors()
```

## PCA
### Both AC and BC

2 distinct groups are observed across PC2. And this is explained by the 2 families of fish which comprise this dataset. Therefore, I should analyse them seperately

```{r}
x2$counts %>% 
  cpm(log=TRUE) %>% 
  t() %>% 
  prcomp() %>% 
  autoplot(data = tibble(sample = rownames(.$x)) %>%
      left_join(x2$samples),
    colour = "MPS", 
    size = 6
  ) +
  theme(aspect.ratio = 1, 
        legend.position = "bottom") +
  labs(colour = "MPS comparison family") +
  scale_color_brewer(palette = "Set2")
```

### AC only

The AC group came from 1 family, raised across 5 tanks side by side in the same system. The MPS-IIIA samples appear to be clustering away from the other 2. 

```{r}
BCsamps <- meta %>% 
  dplyr::filter(MPS == "BC") %>% 
  .$sample

ACsamps <- meta %>% 
  dplyr::filter(MPS == "AC") %>% 
  .$sample

ggarrange(
x2$counts %>% 
  .[,ACsamps] %>% 
  cpm(log=TRUE) %>% 
  t() %>% 
  prcomp() %>% 
  autoplot(data = tibble(sample = rownames(.$x)) %>%
      left_join(x2$samples),
    colour = "genotype", 
    size = 6
  ) +
  theme(aspect.ratio = 1) +
  ggtitle("by genotype") +
  labs(colour = ""),

x2$counts %>% 
  .[,ACsamps] %>% 
  cpm(log=TRUE) %>% 
  t() %>% 
  prcomp() %>% 
  autoplot(data = tibble(sample = rownames(.$x)) %>%
      left_join(x2$samples),
    colour = "Tank", 
    size = 6
  ) +
  theme(aspect.ratio = 1) +
  labs(colour = "") +
  ggtitle("by tank")
)
```

### BC only

The BC group of fish came from 2 sets of parents, across multiple lays/ This is evident in the plot below. The effect of "Home Tank" cannot be ignored in a DE analysis. 
```{r}
ggarrange(
x2$counts %>% 
  .[,BCsamps] %>% 
  cpm(log=TRUE) %>% 
  t() %>% 
  prcomp() %>% 
  autoplot(data = tibble(sample = rownames(.$x)) %>%
      left_join(x2$samples),
    colour = "genotype", 
    size = 6
  ) +
  theme(aspect.ratio = 1) +
  ggtitle("by genotype") +
  labs(colour = ""),

x2$counts %>% 
  .[,BCsamps] %>% 
  cpm(log=TRUE) %>% 
  t() %>% 
  prcomp() %>% 
  autoplot(data = tibble(sample = rownames(.$x)) %>%
      left_join(x2$samples),
    colour = "Tank", 
    size = 6
  ) +
  theme(aspect.ratio = 1) +
  labs(colour = "") +
  ggtitle("by tank")
)
```
From hereout, I will process the 2 comparisons separately. 

# A and C
## initial DE analysis

```{r}
x.ac <- x2[,ACsamps]

design.ac <- model.matrix(~genotype + Tank, data = x.ac$samples %>% 
                            droplevels()) %>% # get rid of the unused factor levels
  set_colnames(gsub(colnames(.), pattern = "genotype", replacement = ""))

fit_1_ac <- x.ac %>% 
  estimateDisp(design.ac) %>% 
  glmFit(design.ac)

toptable_1.ac <- design.ac %>% colnames() %>% .[2:3] %>% 
  sapply(function(x){
    glmLRT(fit_1_ac, coef = x) %>%
      topTags(n = Inf) %>%
      .[["table"]] %>%
      as_tibble() %>%
      arrange(PValue) %>%
       mutate(
        coef = x,
        DE = FDR < 0.05
      ) %>% 
      dplyr::select(
        gene_name, logFC, logCPM, PValue, FDR, DE, everything()  
      )
      
  }, simplify = FALSE)
```

### vis
### Number of DE genes
```{r}
toptable_1.ac %>% 
  lapply(function(y) {
    y %>% 
      dplyr::filter(DE == T) %>% 
      nrow()
  }) %>% 
  bind_rows() %>% 
  as_tibble() %>% 
  gather(key = "coef", value = "numDE") %>% 
  ggplot(aes(x = numDE, y = coef)) +
  geom_col(aes(fill = coef)) +
  geom_text(aes(label = numDE)) +
  ggtitle("Number of DE genes in each comparison",
          subtitle = "The MPS-IIIA zebrafish samples have the most DE genes")
```

### Volcano plots
```{r}
toptable_1.ac %>%
  bind_rows() %>% 
  ggplot(aes(y = -log10(PValue), x = logFC, colour = DE)) +
  geom_point(
    alpha = 0.5, size = 1.25
  ) +
  facet_wrap(~coef) +
  # geom_label_repel(
  #   aes(label = gene_name),
  #   data = .  %>% dplyr::filter(FDR < 0.05),
  #   show.legend = FALSE
  #   ) +
  # coord_cartesian(xlim = c(-2.5,2.5), 
  #                 ylim = c(0, 20) )+
  theme(legend.position = "none") +
  scale_color_manual(values = c("grey50", "red")) 
```

### MD plots
```{r}
toptable_1.ac %>%
  bind_rows() %>% 
  ggplot(aes(x = logCPM, y = logFC)) +
  geom_point(
    aes(colour = DE),
    alpha = 0.5
  ) +
  facet_wrap(~coef, ncol = 1) +
  theme(legend.position = "none") +
  scale_color_manual(values = c("grey50", "red"))
```

### %GC AND LENGTH CHECK
```{r}
ggarrange(
toptable_1.ac %>% 
  bind_rows() %>% 
  mutate(rankstat = sign(logFC)*-log10(PValue)) %>% 
  ggplot(aes(x = length, y = rankstat)) +
  geom_point(
    aes(colour = DE),
    alpha = 0.5
  ) +
  geom_smooth(se = FALSE, method = "gam") +
  facet_grid(rows = vars(coef)) +
  theme_bw() +
  theme(legend.position = "none") +
  scale_color_manual(values = c("grey50", "red")) +
  scale_x_log10()+
  labs(x = "Average transcript length per gene",
       colour = "Differentially expressed?",
       y = "sign(logFC)*-log10(PValue)") +
  coord_cartesian(ylim = c(-10, 10)),

toptable_1.ac %>% 
  bind_rows() %>% 
  mutate(rankstat = sign(logFC)*-log10(PValue)) %>% 
  ggplot(aes(x = gc_content, y = rankstat)) +
  geom_point(
    aes(colour = DE),
    alpha = 0.5
  ) +
  geom_smooth(se = FALSE, method = "gam") +
  facet_grid(rows = vars(coef)) +
  theme_bw() +
  theme(legend.position = "none") +
  scale_color_manual(values = c("grey50", "red")) +
  coord_cartesian(ylim = c(-10,10)) + 
  labs(x = "Weighted average GC content (%) per gene", 
       colour = "Differentially expressed?", 
       y = "sign(logFC)*-log10(PValue)"),
common.legend = TRUE, 
labels = "AUTO"
) 
```



## CQN

A little bit of bias is obsevred for %GC content and gene length, so will apply the `cqn` correction to adjust for this. 

```{r}
cqn.ac <-   
x.ac %>% 
  with(
    cqn(
      counts = counts,
      x = genes$gc_content,
      lengths = genes$length,
      sizeFactors = samples$lib.size
    )
  )
logCPM.ac <- cqn.ac$y + cqn.ac$offset
```

## plot out the model fits
```{r plotCQN_AC, fig.cap ="*Model fits for GC content and gene length under the CQN model. Variability is clearly visible at either end*"}
par(mfrow = c(1, 2))
cqnplot(cqn.ac, n = 1, xlab = "GC Content")
cqnplot(cqn.ac, n = 2, xlab = "Length")
par(mfrow = c(1, 1))
```

## PCA
PCA analysis was repeated. Similar clustering of samples was observed across PC1. However, samples appeared closer together. 


```{r}
ggarrange(
  cpm(x.ac, log = T) %>%
    t() %>%
    prcomp() %>% 
    autoplot(data = tibble(sample = rownames(.$x)) %>%
               left_join(x$samples),
             colour = "genotype", 
             size = 4
    ) +
    scale_color_viridis_d(option = "turbo", end = 0.8) +
    ggtitle("Before CQN"),
  
  logCPM.ac %>% 
    t() %>% 
    prcomp() %>%
    autoplot(data = tibble(sample = rownames(.$x)) %>%
               left_join(x$samples),
             colour = "genotype", 
             size = 4
    ) +
    scale_color_viridis_d(option = "turbo", end = 0.8) +
    ggtitle("After CQN"),
  common.legend = T, 
  labels = "AUTO"
) 
#ggsave("plotsforpub/PCA_cqn.png", width = 18, height = 8, units = "cm", scale = 1.4)
```

# DE with cqn

```{r}
# Add the offset to the dge object 
x.ac$offset <- cqn.ac$glm.offset

# fit the glm
fit_cqn.ac <- x.ac %>% 
  estimateDisp(design.ac) %>% 
  glmFit(design.ac)

# call the toptable
toptables_cqn.ac <- design.ac %>% colnames() %>% .[2:3] %>% 
  sapply(function(x){
    glmLRT(fit_cqn.ac, coef = x) %>%
      topTags(n = Inf) %>%
      .[["table"]] %>%
      as_tibble() %>%
      arrange(PValue) %>%
       mutate(
        coef = x,
        DE = FDR < 0.05
      ) %>% 
      dplyr::select(
        gene_name, logFC, logCPM, PValue, FDR, DE, everything()  
      )
      
  }, simplify = FALSE)
```

## Vis with CQN
### Num DE genes
```{r}
toptables_cqn.ac %>% 
  lapply(function(x){
    x %>%
      mutate(Direction = case_when(
        sign(logFC) == "1" ~ "up",
        sign(logFC) == "-1" ~ "down"
      ))
  }) %>%
  bind_rows() %>%
  dplyr::filter(DE == TRUE) %>%
  ggplot(aes(y = `coef`)) +
  geom_bar(stat = "count",
           width = 0.5,
           aes(fill = Direction),
           position = "dodge") +
  theme(legend.position = "bottom") +
  ggtitle("Number of DE genes in each comparison") +
  labs(x = "Number of DE genes",
       y = "Contrast",
       colour = "Direction of change") 
```

### Volcano plots
```{r}
toptables_cqn.ac %>%
  bind_rows() %>% 
  ggplot(aes(y = -log10(PValue), x = logFC, colour = DE)) +
  geom_point(
    alpha = 0.5, size = 1.25
  ) +
  facet_wrap(~coef) +
  # geom_label_repel(
  #   aes(label = gene_name),
  #   data = .  %>% dplyr::filter(FDR < 0.05),
  #   show.legend = FALSE
  #   ) +
  # coord_cartesian(xlim = c(-2.5,2.5), 
  #                 ylim = c(0, 20) )+
  theme(legend.position = "none") +
  scale_color_manual(values = c("grey50", "red")) 
```

### MD plots
```{r}
toptables_cqn.ac %>% 
  bind_rows() %>% 
  ggplot(aes(x = logCPM, y = logFC)) +
  geom_point(
    aes(colour = DE),
    alpha = 0.5
  ) +
  facet_wrap(~coef, ncol = 1) +
  theme(legend.position = "none") +
  scale_color_manual(values = c("grey50", "red"))
```

### Recheck the %GC and length bias

```{r}
ggarrange(
toptables_cqn.ac %>% 
  bind_rows() %>% 
  mutate(rankstat = sign(logFC)*-log10(PValue)) %>% 
  ggplot(aes(x = length, y = rankstat)) +
  geom_point(
    aes(colour = DE),
    alpha = 0.5
  ) +
  geom_smooth(se = FALSE, method = "gam") +
  facet_grid(rows = vars(coef)) +
  theme_bw() +
  theme(legend.position = "none") +
  scale_color_manual(values = c("grey50", "red")) +
  scale_x_log10()+
  labs(x = "Average transcript length per gene",
       colour = "Differentially expressed?",
       y = "sign(logFC)*-log10(PValue)") +
  coord_cartesian(ylim = c(-10, 10)),

toptables_cqn.ac %>% 
  bind_rows() %>% 
  mutate(rankstat = sign(logFC)*-log10(PValue)) %>% 
  ggplot(aes(x = gc_content, y = rankstat)) +
  geom_point(
    aes(colour = DE),
    alpha = 0.5
  ) +
  geom_smooth(se = FALSE, method = "gam") +
  facet_grid(rows = vars(coef)) +
  theme_bw() +
  theme(legend.position = "none") +
  scale_color_manual(values = c("grey50", "red")) +
  coord_cartesian(ylim = c(-10,10)) + 
  labs(x = "Weighted average GC content (%) per gene", 
       colour = "Differentially expressed?", 
       y = "sign(logFC)*-log10(PValue)"),
common.legend = TRUE, 
labels = "AUTO"
) 
```

## Gene set enrichment analysis

### Gene sets

```{r}
KEGG <- msigdbr("Danio rerio", category = "C2", subcategory = "CP:KEGG") %>% 
  left_join(x2$genes, by = c("gene_symbol" = "gene_name")) %>% 
  dplyr::filter(gene_id %in% rownames(x2)) %>% 
  distinct(gs_name, gene_id, .keep_all = TRUE) %>% 
  split(f = .$gs_name) %>%
  lapply(extract2, "gene_id")

sizes <- KEGG %>% 
  lapply(length) %>% 
  unlist %>% 
  as.data.frame() %>% 
  set_colnames( "n_genes") %>% 
  rownames_to_column("gs")


# retain gene sets with at least 10 genes in it
KEGG <- KEGG[sizes %>% dplyr::filter(n_genes > 10) %>% .$gs]

# GO Terms
goSummaries <- url("https://uofabioinformaticshub.github.io/summaries2GO/data/goSummaries.RDS") %>%
  readRDS() %>%
  mutate(
    Term = Term(id),
    gs_name = Term %>% str_to_upper() %>% str_replace_all("[ -]", "_"),
    gs_name = paste0("GO_", gs_name)
    )

minPath <- 3

GO <- 
  msigdbr("Danio rerio", category = "C5") %>% 
  dplyr::filter(grepl(gs_name, pattern = "^GO")) %>% 
  left_join(x2$genes, by = c("gene_symbol" = "gene_name")) %>% 
  dplyr::filter(gene_id %in% rownames(x2)) %>% 
  mutate(gs_name = str_replace(gs_name, pattern = "GOBP_", replacement = "GO_")) %>% 
  mutate(gs_name = str_replace(gs_name, pattern = "GOMF_", replacement = "GO_")) %>%     
  mutate(gs_name = str_replace(gs_name, pattern = "GOCC_", replacement = "GO_")) %>%     
  left_join(goSummaries) %>% 
  dplyr::filter(shortest_path >= minPath) %>%
  distinct(gs_name, gene_id, .keep_all = TRUE) %>% 
  split(f = .$gs_name) %>%
  lapply(extract2, "gene_id")

# GOdf <-  msigdbr("Danio rerio", category = "C5") %>% 
#   dplyr::filter(grepl(gs_name, pattern = "^GO")) %>% 
#   left_join(x2$genes, by = c("gene_symbol" = "gene_name")) %>% 
#   dplyr::filter(gene_id %in% rownames(x2)) %>% 
#   mutate(gs_name = str_replace(gs_name, pattern = "GOBP_", replacement = "GO_")) %>% 
#   mutate(gs_name = str_replace(gs_name, pattern = "GOMF_", replacement = "GO_")) %>%     
#   mutate(gs_name = str_replace(gs_name, pattern = "GOCC_", replacement = "GO_")) %>%     
#   left_join(goSummaries) %>% 
#   dplyr::filter(shortest_path >= minPath) %>%
#   distinct(gs_name, gene_id, .keep_all = TRUE) %>% 
#   dplyr::select(gs_name, gene_id)

  sizes_GO <- GO %>%
    lapply(length) %>%
    unlist %>%
    as.data.frame() %>%
    set_colnames( "n_genes") %>%
    rownames_to_column("gs")


# retain gene sets with at least 5 genes in it
GO <- GO[sizes_GO %>% dplyr::filter(n_genes > 5) %>% .$gs]

ireGenes <- 
  read_xlsx("data/gene_sets/ireGenes.xlsx", sheet = "Zebrafish IREs") %>% 
  gather %>% 
  set_colnames(c("IRE", "gene_id")) %>% 
  mutate(IRE = case_when(
    IRE == "All 3' IREs" ~ "ire3_all", 
    IRE == "All 5' IREs" ~ "ire5_all", 
    IRE == "HQ 3' IREs" ~ "ire3_HQ", 
    IRE == "HQ 5' IREs" ~ "ire5_HQ"
  )) %>% 
  dplyr::filter(gene_id %in% rownames(x2)) %>% 
  split(f = .$IRE) %>% 
  lapply(magrittr::extract2,"gene_id") 

chr <- x2$genes %>% 
  dplyr::select(gene_id, chromosome) %>% 
  split(f = .$chromosome) %>%
  lapply(extract2, "gene_id") %>% 
  .[c("MT", seq(1:25))] # omit the ALT chromosmes and scaffolds

cell_type_markers <-  read_xlsx("data/gene_sets/Suppdata2_jiangetal_2021_fcelldev.xlsx", sheet = "Brain") %>% 
  dplyr::rename("gene_name" = gene) %>% 
  left_join(x.ac$genes) %>% 
  dplyr::filter(gene_id %in% rownames(x.ac)) %>% 
  split(f = .$`cell type`) %>% 
  lapply(extract2, "gene_id")

cell_type_markers %<>% 
  lapply(function(y) {
    y[y %in% rownames(x2)]
  })

allC2.bcac <- msigdbr("Danio rerio", category = "C2") %>% 
  left_join(x2$genes, by = c("gene_symbol" = "gene_name")) %>% 
  dplyr::filter(gene_id %in% rownames(x2)) %>% 
  distinct(gs_name, gene_id, .keep_all = TRUE) %>% 
  split(f = .$gs_name) %>%
  lapply(extract2, "gene_id")

sizes_allc2 <- allC2.bcac %>%
    lapply(length) %>%
    unlist %>%
    as.data.frame() %>%
    set_colnames( "n_genes") %>%
    rownames_to_column("gs")


# retain gene sets with at least 5 genes in it
allC2.bcac <- allC2.bcac[sizes_allc2 %>% dplyr::filter(n_genes > 5) %>% .$gs]
```

## GOSEQ 

```{r}
pwf <- toptables_cqn.ac$`MPS-IIIA` %>%
      with(
        nullp(
          DEgenes = structure(DE, names = gene_id), 
          bias.data = length
        )
      )

goseqac.a <- goseq(pwf, gene2cat = c(GO)) %>% 
  as_tibble() %>% 
  dplyr::filter(numDEInCat > 0) %>% 
  mutate(FDR = p.adjust(over_represented_pvalue, method = "fdr")) %>% 
  dplyr::select(-under_represented_pvalue) 
  #write_csv("output/results/goseq_cqn/famAC_Acoef.csv")
 # dplyr::filter(FDR < 0.05) 

pwf <- toptables_cqn.ac$`MPS-IIIC` %>%
      with(
        nullp(
          DEgenes = structure(DE, names = gene_id), 
          bias.data = length
        )
      )

goseqac.c <- goseq(pwf, gene2cat = GO) %>% 
  as_tibble() %>% 
  dplyr::filter(numDEInCat > 0) %>% 
  mutate(FDR = p.adjust(over_represented_pvalue, method = "fdr")) %>% 
  dplyr::select(-under_represented_pvalue) 
  # #write_csv("output/results/goseq_cqn/famAC_Ccoef.csv")
  # dplyr::filter(FDR < 0.05) %>% 
  # head(10)

goseqac.a %>% 
  mutate(coef = "MPS-IIIA") %>% 
  bind_rows(goseqac.c %>% 
              mutate(coef = "MPS-IIIC")) %>% 
  dplyr::filter(FDR < 0.05) %>% 
  ggplot(aes(x = coef, y = category)) + 
  geom_tile(aes(fill = -log10(over_represented_pvalue)))
```



## KEGG 
```{r}
fryKEGG.ac <- 
  design.ac %>% colnames() %>% .[2:3] %>% 
  sapply(function(y) {
    logCPM.ac %>% 
      fry(
        index = KEGG,
        design = design.ac, 
        contrast = y, 
        sort = "mixed"
      ) %>% 
      rownames_to_column("pathway") %>% 
      as_tibble() %>% 
      mutate(coef = y)
  }, simplify = FALSE)

fryKEGG.ac$`MPS-IIIA` %>% 
  dplyr::filter(FDR.Mixed < 0.05) %>%  
  mutate(pathway = str_remove(pathway, pattern = "KEGG_")) %>% 
  ggplot(aes(x = -log10(PValue.Mixed), y = reorder(pathway, -FDR.Mixed))) +
  geom_col(aes(fill = FDR.Mixed < 0.05)) +
  scale_fill_manual(values = c("grey50", 'red')) +
  theme_pubr() +
  labs(fill = "FDR < 0.05", 
       x = "-log10(PValue)", 
       y = "")

fryKEGG.ac$`MPS-IIIC` %>% 
  dplyr::filter(FDR.Mixed < 0.05) %>%  
  mutate(pathway = str_remove(pathway, pattern = "KEGG_")) %>% 
  ggplot(aes(x = -log10(PValue.Mixed), y = reorder(pathway, -FDR.Mixed))) +
  geom_col(aes(fill = FDR.Mixed < 0.05)) +
  scale_fill_manual(values = c("grey50", 'red')) +
  theme_pubr() +
  labs(fill = "FDR < 0.05", 
       x = "-log10(PValue)", 
       y = "")  

sigpaths.ac <- fryKEGG.ac %>% 
  bind_rows() %>% 
  dplyr::filter(FDR.Mixed < 0.05) %>% 
    .$pathway %>% 
    unique()
  
fryKEGG.ac %>% 
  bind_rows() %>% 
  dplyr::filter(pathway %in% sigpaths.ac) %>% 
  mutate(pathway = str_remove(pathway, pattern = "KEGG_")) %>% 
  ggplot(aes(x = -log10(PValue.Mixed), y = reorder(pathway, -FDR.Mixed))) +
  geom_col(aes(fill = coef, alpha = FDR.Mixed < 0.05), position = "dodge") +
  scale_fill_viridis_d(end = 0.75) +
  scale_alpha_manual(values = c(0.5,1)) +
  theme_pubr() +
  geom_vline(xintercept = 2.3) +
  labs(fill = "FDR < 0.05", 
       x = "-log10(PValue)", 
       y = "")
  #ggsave("output/plots/AC_KEGG.png", width = 20, height = 20, units = "cm", dpi = 400, scale = 2)
```


<!-- ### GO -->
<!-- ```{r} -->
<!-- fry_GO <-  design.ac %>% colnames() %>% .[2:3] %>%  -->
<!--   sapply(function(y) { -->
<!--     logCPM.ac %>%  -->
<!--       fry( -->
<!--         index = GO, -->
<!--         design = design.ac,  -->
<!--         contrast = y,  -->
<!--         sort = "mixed" -->
<!--       ) %>%  -->
<!--       rownames_to_column("pathway") %>%  -->
<!--       as_tibble() %>%  -->
<!--       mutate(coef = y,  -->
<!--              sig = FDR.Mixed < 0.05) -->
<!--   }, simplify = FALSE) -->
<!-- ``` -->


```{r}
toptables_cqn.ac %>% 
  bind_rows() %>% 
  dplyr::filter(gene_id %in% KEGG$KEGG_AMINO_SUGAR_AND_NUCLEOTIDE_SUGAR_METABOLISM) %>% 
  dplyr::select(gene_name, logFC, coef) %>% 
  dplyr::distinct(gene_name, coef, .keep_all = T) %>% 
  spread(key = "coef", value = "logFC") %>% 
  column_to_rownames("gene_name") %>% 
  t() %>% 
  pheatmap(color = colorRampPalette(rev(brewer.pal(n = 5,
                                               name = "RdBu")))(100), 
           main = "KEGG_AMINO_SUGAR_AND_NUCLEOTIDE_SUGAR_METABOLISM", 
           breaks = c(seq(min(.), 0, length.out=ceiling(100/2) + 1), 
              seq(max(.)/100, max(.), length.out=50)), 
           cellheight = 30, 
           cellwidth = 10,
           angle_col = 45,
           treeheight_row = 0, treeheight_col = 0
           )

```
### IRE

```{r}
fryIRE.ac <- design.ac %>% colnames() %>% .[2:3] %>% 
  sapply(function(y) {
    logCPM.ac %>% 
      fry(
        index = ireGenes,
        design = design.ac, 
        contrast = y, 
        sort = "directional"
      ) %>% 
      rownames_to_column("pathway") %>% 
      as_tibble() %>% 
      mutate(coef = y)
  }, simplify = FALSE)

fryIRE.ac %>% 
  bind_rows() %>% 
  dplyr::select(pathway, coef, pvalue = PValue.Mixed, FDR = FDR.Mixed) %>% 
  ggplot(aes(y = pathway, x = -log10(pvalue))) + 
  geom_col(aes(fill = coef, alpha = FDR < 0.05), 
           position = "dodge") +
  scale_fill_viridis_d(end = 0.75) +
  geom_vline(xintercept = 1.2) +
  scale_alpha_manual(values = c(0.5,1)) +
  annotate(geom = "label", x = 1.1, y = "ire5_HQ", label = "FDR = 0.05") 
  #ggsave("output/plots/AC_ire.png", width = 20, height = 10, units = "cm", dpi = 400, scale = 2)


toptables_cqn.ac %>% 
  bind_rows() %>% 
  dplyr::filter(gene_id %in% ireGenes$ire5_HQ) %>% 
  dplyr::select(gene_name, logFC, coef) %>% 
  dplyr::distinct(gene_name, coef, .keep_all = T) %>% 
  spread(key = "coef", value = "logFC") %>% 
  column_to_rownames("gene_name") %>% 
  t() %>% 
  pheatmap(color = colorRampPalette(rev(brewer.pal(n = 5,
                                               name = "RdBu")))(100), 
           main = "3' IRE genes: FDR = 0.066", 
           breaks = c(seq(min(.), 0, length.out=ceiling(100/2) + 1), 
              seq(max(.)/100, max(.), length.out=50)), 
           # cellheight = 30, 
           # cellwidth = 1,
           angle_col = 45,
           treeheight_row = 0, 
           treeheight_col = 0, 
           show_colnames = T
           )
```

### Cell type prop check

Since this experiment was performed on whole brains, we need to have a look at whether changes to cell type proportions could explain cause some artefactual changes to gene expression. We have done this previously by assessing the expression of marker genes of specific cell types. So this is what I will do here. I have taken the lists of genes frmo day 5 only from the zebrafish single cell atlas (Farnsworth et al. 2020) which contain at least 10 genes. This leaves  `r length(cell_type_markers)` gene sets (cell types) to test.

```{r}
celltype.ac <- design.ac %>% colnames() %>% .[2:3] %>% 
  sapply(function(y) {
    logCPM.ac %>% 
      fry(
        index = cell_type_markers,
        design = design.ac, 
        contrast = y, 
        sort = "directional"
      ) %>% 
      rownames_to_column("pathway") %>% 
      as_tibble() %>% 
      dplyr::select(1:5) %>% 
      mutate(coef = y)
  }, simplify = FALSE)

celltype.ac %>% 
  bind_rows() %>% 
  ggplot(aes(y = pathway, x = -log10(PValue))) + 
  geom_col(aes(fill = coef, alpha = FDR < 0.05), 
           position = "dodge") +
  scale_fill_viridis_d(end = 0.75) +
  scale_alpha_manual(values = c(0.5, 1)) +
  geom_vline(xintercept = 2.1) +
  annotate(geom = "label", x = 2, y = 4.5, label = "FDR = 0.05") 
```



```{r}
toptables_cqn.ac %>% 
  bind_rows() %>% 
  dplyr::filter(gene_id %in% cell_type_markers$Oligodendrocyte) %>% 
  dplyr::select(gene_name, logFC, coef) %>% 
  dplyr::filter(coef == "MPS-IIIA") %>% 
  dplyr::distinct(gene_name, coef, .keep_all = T) %>% 
  spread(key = "coef", value = "logFC") %>% 
  column_to_rownames("gene_name") %>% 
  arrange(desc(`MPS-IIIA`)) %>% 
  t() %>% 
  pheatmap(color = colorRampPalette(rev(brewer.pal(n = 5,
                                               name = "RdBu")))(100), 
           main = "Oligodendrocytes in MPS-IIIA", 
           breaks = c(seq(min(.), 0, length.out=ceiling(100/2) + 1), 
              seq(max(.)/100, max(.), length.out=50)), 
           cellheight = 30, cellwidth = 12,
           angle_col = 45,
           cluster_rows = F,
           cluster_cols = F,
           treeheight_row = 0, treeheight_col = 0
           )

toptables_cqn.ac %>% 
  bind_rows() %>% 
  dplyr::filter(gene_id %in% cell_type_markers$Oligodendrocyte) %>% 
  dplyr::select(gene_name, logFC, coef) %>% 
  dplyr::filter(coef == "MPS-IIIC") %>% 
  dplyr::distinct(gene_name, coef, .keep_all = T) %>% 
  spread(key = "coef", value = "logFC") %>% 
  column_to_rownames("gene_name") %>% 
  arrange(desc(`MPS-IIIC`)) %>% 
  t() %>% 
  pheatmap(color = colorRampPalette(rev(brewer.pal(n = 5,
                                               name = "RdBu")))(100), 
           main = "Oligodendrocytes in MPS-IIIC", 
           breaks = c(seq(min(.), 0, length.out=ceiling(100/2) + 1), 
              seq(max(.)/100, max(.), length.out=50)), 
           cellheight = 30, cellwidth = 12,
           angle_col = 45,
           cluster_rows = F,
           cluster_cols = F,
           treeheight_row = 0, treeheight_col = 0
           )
```


### Chrosomosomes gene sets

```{r}
fry_chr.ac <- design.ac %>% colnames() %>% .[2:3] %>% 
  sapply(function(y) {
    logCPM.ac %>% 
      fry(
        index = c(chr),
        design = design.ac, 
        contrast = y, 
        sort = "mixed"
      ) %>% 
      rownames_to_column("pathway") %>% 
      as_tibble() %>% 
      mutate(coef = y)
  }, simplify = FALSE)


fry_chr.ac %>% 
  bind_rows() %>% 
  mutate(pathway = factor(pathway, levels = c("MT", seq(1:25)))) %>% 
  ggplot(aes(y = pathway, x = -log10(PValue.Mixed), fill = coef)) +
  geom_col(position = "dodge")
  
```

An enrichment of DE genes is observed on the same chromosome as the mutation. This is obsevred in the manhat plots below. 

```{r}
manhat_input.ac <- 
	toptables_cqn.ac %>% 
  bind_rows() %>% 
	left_join(x.ac$genes %>% 
			  	as_tibble() %>% 
			  	dplyr::select(start, end, chromosome, gene_id) %>% 
			  	group_by(gene_id) %>% 
			  	mutate(mid = mean(c(start, end)))
	) %>% 
  dplyr::filter(chromosome %in% seq(1:25)) %>% 
  group_by(gene_id) %>% 
  mutate(mid = mean(c(start, end))) %>% 
  ungroup %>% 
  mutate(chromosome = factor(chromosome, levels = seq(1:25)))  %>% 
  group_by(chromosome) %>% 
  summarise(chrLen = max(mid)) %>% 
  mutate(chrSt = cumsum(chrLen)-chrLen) %>% 
	dplyr::select(-chrLen) %>% 
  left_join(toptables_cqn.ac %>% 
              bind_rows() %>% 
              left_join(x.ac$genes %>% 
                          as_tibble() %>% 
                          dplyr::select(start, end, chromosome, gene_id) %>% 
                          group_by(gene_id) %>% 
                          mutate(mid = mean(c(start, end)))
              )
  ) %>% 
  group_by(gene_id) %>% 
  mutate(mid = mean(c(start, end))) %>% 
  ungroup %>% 
  dplyr::arrange(mid, chromosome) %>% 
  mutate(midCum = chrSt + mid)

axis.ac <- manhat_input %>%
  group_by(chromosome) %>%
  dplyr::arrange(chromosome) %>%
  summarize(center = (max(midCum) + min(midCum)) / 2) %>%
  mutate(
    seqnames = as.numeric(chromosome),
    colour = rep(c("grey40", "black"), length.out = 25)
  )

ggplot(manhat_input, aes(x = midCum, y = -log10(PValue))) +
  geom_point(aes(color=chromosome), alpha = 0.5, size = 1) +
  geom_point(alpha = 0.5, size = 1, colour = "red", 
             data = . %>% 
               dplyr::filter(coef == "MPS-IIIC" & chromosome == 1)) +
  geom_point(alpha = 0.5, size = 1, colour = "red", 
             data = . %>% 
               dplyr::filter(coef ==  "MPS-IIIA" & chromosome == 22)) +
  scale_color_manual(values = axis$colour) +
  scale_x_continuous(label = axis$chromosome, breaks = axis$center) +
  scale_y_continuous(limits =c(0, 10)) + # zoom in
  facet_wrap(~coef, ncol = 1) +
  labs(x = "Chromosome", y = expression(paste(-log[10], "(p)"))) +
  theme( 
    legend.position="none",
    panel.border = element_blank(),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.x = element_blank()
  ) 
```

# BC
```{r}
x.bc <- x2[,BCsamps]

design.bc <- model.matrix(~genotype + Tank, data = x.bc$samples %>% 
                            droplevels()) %>% # get rid of the unused factor levels
  set_colnames(gsub(colnames(.), pattern = "genotype", replacement = ""))

fit_1_bc <- x.bc %>% 
  estimateDisp(design.bc) %>% 
  glmFit(design.bc)

toptable_1.bc <- design.bc %>% colnames() %>% .[2:3] %>% 
  sapply(function(x){
    glmLRT(fit_1_bc, coef = x) %>%
      topTags(n = Inf) %>%
      .[["table"]] %>%
      as_tibble() %>%
      arrange(PValue) %>%
       mutate(
        coef = x,
        DE = FDR < 0.05
      ) %>% 
      dplyr::select(
        gene_name, logFC, logCPM, PValue, FDR, DE, everything()  
      )
      
  }, simplify = FALSE)
```

### vis
### Number of DE genes
```{r}
toptable_1.bc %>% 
  lapply(function(y) {
    y %>% 
      dplyr::filter(DE == T) %>% 
      nrow()
  }) %>% 
  bind_rows() %>% 
  as_tibble() %>% 
  gather(key = "coef", value = "numDE") %>% 
  ggplot(aes(x = numDE, y = coef)) +
  geom_col(aes(fill = coef)) +
  geom_text(aes(label = numDE)) +
  ggtitle("Number of DE genes in each comparison",
          subtitle = "The MPS-IIIA zebrafish samples have the most DE genes")
```

### Volcano plots

The volcano plot below is zoomed in for vis purposes. 
```{r}
toptable_1.bc %>%
  bind_rows() %>% 
  ggplot(aes(y = -log10(PValue), x = logFC, colour = DE)) +
  geom_point(
    alpha = 0.5, size = 1.25
  ) +
  facet_wrap(~coef) +
  # geom_label_repel(
  #   aes(label = gene_name),
  #   data = .  %>% dplyr::filter(FDR < 0.05),
  #   show.legend = FALSE
  #   ) +
  coord_cartesian(xlim = c(-2.5,2.5),
                  ylim = c(0, 20) )+
  theme(legend.position = "none") +
  scale_color_manual(values = c("grey50", "red")) 
```

### MD plots
```{r}
toptable_1.bc %>%
  bind_rows() %>% 
  ggplot(aes(x = logCPM, y = logFC)) +
  geom_point(
    aes(colour = DE),
    alpha = 0.5
  ) +
  facet_wrap(~coef, ncol = 1) +
  theme(legend.position = "none") +
  scale_color_manual(values = c("grey50", "red"))
```

### %GC AND LENGTH CHECK
```{r}
ggarrange(
toptable_1.bc %>% 
  bind_rows() %>% 
  mutate(rankstat = sign(logFC)*-log10(PValue)) %>% 
  ggplot(aes(x = length, y = rankstat)) +
  geom_point(
    aes(colour = DE),
    alpha = 0.5
  ) +
  geom_smooth(se = FALSE, method = "gam") +
  facet_grid(rows = vars(coef)) +
  theme_bw() +
  theme(legend.position = "none") +
  scale_color_manual(values = c("grey50", "red")) +
  scale_x_log10()+
  labs(x = "Average transcript length per gene",
       colour = "Differentially expressed?",
       y = "sign(logFC)*-log10(PValue)") +
  coord_cartesian(ylim = c(-10, 10)),

toptable_1.bc %>% 
  bind_rows() %>% 
  mutate(rankstat = sign(logFC)*-log10(PValue)) %>% 
  ggplot(aes(x = gc_content, y = rankstat)) +
  geom_point(
    aes(colour = DE),
    alpha = 0.5
  ) +
  geom_smooth(se = FALSE, method = "gam") +
  facet_grid(rows = vars(coef)) +
  theme_bw() +
  theme(legend.position = "none") +
  scale_color_manual(values = c("grey50", "red")) +
  coord_cartesian(ylim = c(-10,10)) + 
  labs(x = "Weighted average GC content (%) per gene", 
       colour = "Differentially expressed?", 
       y = "sign(logFC)*-log10(PValue)"),
common.legend = TRUE, 
labels = "AUTO"
) 
```



## CQN

A  bit of bias is obsevred for %GC content and gene length, so will apply the `cqn` correction to adjust for this. 

```{r}
cqn.bc <-   
x.bc %>% 
  with(
    cqn(
      counts = counts,
      x = genes$gc_content,
      lengths = genes$length,
      sizeFactors = samples$lib.size
    )
  )
logCPM.bc <- cqn.bc$y + cqn.bc$offset
```

## plot out the model fits
```{r plotCQN, fig.cap ="*Model fits for GC content and gene length under the CQN model. Variability is clearly visible at either end*"}
par(mfrow = c(1, 2))
cqnplot(cqn.bc, n = 1, xlab = "GC Content")
cqnplot(cqn.bc, n = 2, xlab = "Length")
par(mfrow = c(1, 1))
```

## PCA
PCA analysis was repeated. Similar clustering of samples was observed across PC1. However, samples appeared closer together. 


```{r}
ggarrange(
  cpm(x.bc, log = T) %>%
    t() %>%
    prcomp() %>% 
    autoplot(data = tibble(sample = rownames(.$x)) %>%
               left_join(x$samples),
             colour = "genotype", 
             size = 4
    ) +
    scale_color_viridis_d(option = "turbo", end = 0.8) +
    ggtitle("Before CQN"),
  
  logCPM.bc %>% 
    t() %>% 
    prcomp() %>%
    autoplot(data = tibble(sample = rownames(.$x)) %>%
               left_join(x$samples),
             colour = "genotype", 
             size = 4
    ) +
    scale_color_viridis_d(option = "turbo", end = 0.8) +
    ggtitle("After CQN"),
  common.legend = T, 
  labels = "AUTO"
) 
```

# DE with cqn

```{r}
# Add the offset to the dge object 
x.bc$offset <- cqn.bc$glm.offset

# fit the glm
fit_cqn.bc <- x.bc %>% 
  estimateDisp(design.bc) %>% 
  glmFit(design.bc)

toptables_cqn.bc <- design.bc %>% colnames() %>% .[2:3] %>% 
  sapply(function(x){
    glmLRT(fit_cqn.bc, coef = x) %>%
      topTags(n = Inf) %>%
      .[["table"]] %>%
      as_tibble() %>%
      arrange(PValue) %>%
       mutate(
        coef = x,
        DE = FDR < 0.05
      ) %>% 
      dplyr::select(
        gene_name, logFC, logCPM, PValue, FDR, DE, everything()  
      )
      
  }, simplify = FALSE)
```

## Vis with CQN
### Num DE genes
```{r}
toptables_cqn.bc %>% 
  lapply(function(x){
    x %>%
      mutate(Direction = case_when(
        sign(logFC) == "1" ~ "up",
        sign(logFC) == "-1" ~ "down"
      ))
  }) %>%
  bind_rows() %>%
  dplyr::filter(DE == TRUE) %>%
  ggplot(aes(y = `coef`)) +
  geom_bar(stat = "count",
           width = 0.5,
           aes(fill = Direction),
           position = "dodge") +
  theme(legend.position = "bottom") +
  ggtitle("Number of DE genes in each comparison") +
  labs(x = "Number of DE genes",
       y = "Contrast",
       colour = "Direction of change") 
```

### Volcano plots

These are zoomed in again for vis purposes
```{r}
toptables_cqn.bc %>%
  bind_rows() %>% 
  ggplot(aes(y = -log10(PValue), x = logFC, colour = DE)) +
  geom_point(
    alpha = 0.5, size = 1.25
  ) +
  facet_wrap(~coef) +
  # geom_label_repel(
  #   aes(label = gene_name),
  #   data = .  %>% dplyr::filter(FDR < 0.05),
  #   show.legend = FALSE
  #   ) +
  coord_cartesian(xlim = c(-2.5,2.5),
                  ylim = c(0, 20) )+
  theme(legend.position = "none") +
  scale_color_manual(values = c("grey50", "red")) 
```

### MD plots
```{r}
toptables_cqn.bc %>% 
  bind_rows() %>% 
  ggplot(aes(x = logCPM, y = logFC)) +
  geom_point(
    aes(colour = DE),
    alpha = 0.5
  ) +
  facet_wrap(~coef, ncol = 1) +
  theme(legend.position = "none") +
  scale_color_manual(values = c("grey50", "red"))
```

### Recheck the %GC and length bias

```{r}
ggarrange(
toptables_cqn.bc %>% 
  bind_rows() %>% 
  mutate(rankstat = sign(logFC)*-log10(PValue)) %>% 
  ggplot(aes(x = length, y = rankstat)) +
  geom_point(
    aes(colour = DE),
    alpha = 0.5
  ) +
  geom_smooth(se = FALSE, method = "gam") +
  facet_grid(rows = vars(coef)) +
  theme_bw() +
  theme(legend.position = "none") +
  scale_color_manual(values = c("grey50", "red")) +
  scale_x_log10()+
  labs(x = "Average transcript length per gene",
       colour = "Differentially expressed?",
       y = "sign(logFC)*-log10(PValue)") +
  coord_cartesian(ylim = c(-10, 10)),

toptables_cqn.bc %>% 
  bind_rows() %>% 
  mutate(rankstat = sign(logFC)*-log10(PValue)) %>% 
  ggplot(aes(x = gc_content, y = rankstat)) +
  geom_point(
    aes(colour = DE),
    alpha = 0.5
  ) +
  geom_smooth(se = FALSE, method = "gam") +
  facet_grid(rows = vars(coef)) +
  theme_bw() +
  theme(legend.position = "none") +
  scale_color_manual(values = c("grey50", "red")) +
  coord_cartesian(ylim = c(-10,10)) + 
  labs(x = "Weighted average GC content (%) per gene", 
       colour = "Differentially expressed?", 
       y = "sign(logFC)*-log10(PValue)"),
common.legend = TRUE, 
labels = "AUTO"
) 
```

## Gene set enrichment analysis


## GOSEQ 

```{r}
pwf <- toptables_cqn.bc$`MPS-IIIB` %>%
      with(
        nullp(
          DEgenes = structure(DE, names = gene_id), 
          bias.data = length
        )
      )

goseq(pwf, gene2cat = GO) %>% 
  as_tibble() %>% 
  dplyr::filter(numDEInCat > 0) %>% 
  mutate(FDR = p.adjust(over_represented_pvalue, method = "fdr")) %>% 
  dplyr::select(-under_represented_pvalue) %>% 
  #write_csv("output/results/goseq_cqn/famBC_Bcoef.csv")
  dplyr::filter(FDR < 0.05) 

pwf <- toptables_cqn.bc$`MPS-IIIC` %>%
      with(
        nullp(
          DEgenes = structure(DE, names = gene_id), 
          bias.data = length
        )
      )

goseq(pwf, gene2cat = GO) %>% 
  as_tibble() %>% 
  dplyr::filter(numDEInCat > 0) %>% 
  mutate(FDR = p.adjust(over_represented_pvalue, method = "fdr")) %>% 
  dplyr::select(-under_represented_pvalue) %>% 
  #write_csv("output/results/goseq_cqn/famBC_Ccoef.csv")
  dplyr::filter(FDR < 0.05) %>% 
  head(10)

  
```



## KEGG 
```{r}
fryKEGG.bc <- 
  design.bc %>% colnames() %>% .[2:3] %>% 
  sapply(function(y) {
    logCPM.bc %>% 
      fry(
        index = KEGG,
        design = design.bc, 
        contrast = y, 
        sort = "mixed"
      ) %>% 
      rownames_to_column("pathway") %>% 
      as_tibble() %>% 
      mutate(coef = y)
  }, simplify = FALSE)

fryKEGG.bc$`MPS-IIIB` %>% 
  dplyr::filter(FDR.Mixed < 0.05) %>%  
  mutate(pathway = str_remove(pathway, pattern = "KEGG_")) %>% 
  ggplot(aes(x = -log10(PValue.Mixed), y = reorder(pathway, -FDR.Mixed))) +
  geom_col(aes(fill = FDR.Mixed < 0.05)) +
  scale_fill_manual(values = c("grey50", 'red')) +
  theme_pubr() +
  labs(fill = "FDR < 0.05", 
       x = "-log10(PValue)", 
       y = "")

fryKEGG.bc$`MPS-IIIC` %>% 
  dplyr::filter(FDR.Mixed < 0.05) %>%  
  mutate(pathway = str_remove(pathway, pattern = "KEGG_")) %>% 
  ggplot(aes(x = -log10(PValue.Mixed), y = reorder(pathway, -FDR.Mixed))) +
  geom_col(aes(fill = FDR.Mixed < 0.05)) +
  scale_fill_manual(values = c("grey50", 'red')) +
  theme_pubr() +
  labs(fill = "FDR < 0.05", 
       x = "-log10(PValue)", 
       y = "")  


```


### GO
```{r}
fry_GO.bc <-  design.bc %>% colnames() %>% .[2:3] %>% 
  sapply(function(y) {
    logCPM.bc %>% 
      fry(
        index = GO,
        design = design.bc, 
        contrast = y, 
        sort = "mixed"
      ) %>% 
      rownames_to_column("pathway") %>% 
      as_tibble() %>% 
      mutate(coef = y, 
             sig = FDR.Mixed < 0.05)
  }, simplify = FALSE)
```


### IRE

```{r}
fryIRE.bc <- design.bc %>% colnames() %>% .[2:3] %>% 
  sapply(function(y) {
    logCPM.bc %>% 
      fry(
        index = ireGenes,
        design = design.bc, 
        contrast = y, 
        sort = "directional"
      ) %>% 
      rownames_to_column("pathway") %>% 
      as_tibble() %>% 
      mutate(coef = y)
  }, simplify = FALSE)

fryIRE.bc %>% 
  bind_rows() %>% 
  dplyr::select(pathway, coef, pvalue = PValue.Mixed, FDR = FDR.Mixed) %>% 
  ggplot(aes(y = pathway, x = -log10(pvalue))) + 
  geom_col(aes(fill = coef, alpha = FDR < 0.05), 
           position = "dodge") +
  scale_fill_viridis_d(end = 0.75) +
  scale_alpha_manual(values = c(0.5, 1)) +
  geom_vline(xintercept = 1.5) +
  annotate(geom = "label", x = 1.8, y = "ire5_HQ", label = "FDR = 0.05") 
  #ggsave("output/plots/AC_ire.png", width = 20, height = 10, units = "cm", dpi = 400, scale = 2)


toptables_cqn.bc %>% 
  bind_rows() %>% 
    dplyr::filter(gene_id %in% ireGenes$ire3_all) %>% 
  dplyr::select(gene_name, logFC, coef) %>% 
  dplyr::distinct(gene_name, coef, .keep_all = T) %>% 
  spread(key = "coef", value = "logFC") %>% 
  column_to_rownames("gene_name") %>% 
  t() %>% 
  pheatmap(color = colorRampPalette(rev(brewer.pal(n = 5,
                                               name = "RdBu")))(100), 
           main = "3' IRE genes: FDR = 0.066", 
           breaks = c(seq(min(.), 0, length.out=ceiling(100/2) + 1), 
              seq(max(.)/100, max(.), length.out=50)), 
           # cellheight = 30, 
           # cellwidth = 1,
           angle_col = 45,
           treeheight_row = 0, 
           treeheight_col = 0, 
           show_colnames = F
           )
```

### Cell type prop check

Since this experiment was performed on whole brains, we need to have a look at whether changes to cell type proportions could explain cause some artefactual changes to gene expression. We have done this previously by assessing the expression of marker genes of specific cell types. So this is what I will do here. I have taken the lists of genes frmo day 5 only from the zebrafish single cell atlas (Farnsworth et al. 2020) which contain at least 10 genes. This leaves  `r length(cell_type_markers)` gene sets (cell types) to test.

```{r}
celltype.bc <- design.bc %>% colnames() %>% .[2:3] %>% 
  sapply(function(y) {
    logCPM.bc %>% 
      fry(
        index = cell_type_markers,
        design = design.bc, 
        contrast = y, 
        sort = "directional"
      ) %>% 
      rownames_to_column("pathway") %>% 
      as_tibble() %>% 
      dplyr::select(1:5) %>% 
      mutate(coef = y)
  }, simplify = FALSE)

celltype.bc %>% 
  bind_rows() %>% 
  ggplot(aes(y = pathway, x = -log10(PValue))) + 
  geom_col(aes(fill = coef, alpha = FDR < 0.05), 
           position = "dodge") +
  scale_fill_viridis_d(end = 0.75) +
  scale_alpha_manual(values = c(0.5, 1)) +
  geom_vline(xintercept = 2.1) +
  annotate(geom = "label", x = 2, y = 4.5, label = "FDR = 0.05") 
```


### Chrosomosomes gene sets

```{r}
fry_chr.bc <- design.bc %>% colnames() %>% .[2:3] %>% 
  sapply(function(y) {
    logCPM.bc %>% 
      fry(
        index = c(chr),
        design = design.bc, 
        contrast = y, 
        sort = "mixed"
      ) %>% 
      rownames_to_column("pathway") %>% 
      as_tibble() %>% 
      mutate(coef = y)
  }, simplify = FALSE)


fry_chr.bc %>% 
  bind_rows() %>% 
  mutate(pathway = factor(pathway, levels = c("MT", seq(1:25)))) %>% 
  ggplot(aes(y = pathway, x = -log10(PValue.Mixed), fill = coef)) +
  geom_col(position = "dodge")
  
```

An enrichment of DE genes is observed on the same chromosome as the mutation. This is obsevred in the manhat plots below. 

```{r}
manhat_input.bc <- 
	toptables_cqn.bc %>% 
  bind_rows() %>% 
	left_join(x2$genes %>% 
			  	as_tibble() %>% 
			  	dplyr::select(start, end, chromosome, gene_id) %>% 
			  	group_by(gene_id) %>% 
			  	mutate(mid = mean(c(start, end)))
	) %>% 
  dplyr::filter(chromosome %in% seq(1:25)) %>% 
  group_by(gene_id) %>% 
  mutate(mid = mean(c(start, end))) %>% 
  ungroup %>% 
  mutate(chromosome = factor(chromosome, levels = seq(1:25)))  %>% 
  group_by(chromosome) %>% 
  summarise(chrLen = max(mid)) %>% 
  mutate(chrSt = cumsum(chrLen)-chrLen) %>% 
	dplyr::select(-chrLen) %>% 
  left_join(toptables_cqn.bc %>% 
              bind_rows() %>% 
              left_join(x2$genes %>% 
                          as_tibble() %>% 
                          dplyr::select(start, end, chromosome, gene_id) %>% 
                          group_by(gene_id) %>% 
                          mutate(mid = mean(c(start, end)))
              )
  ) %>% 
  group_by(gene_id) %>% 
  mutate(mid = mean(c(start, end))) %>% 
  ungroup %>% 
  dplyr::arrange(mid, chromosome) %>% 
  mutate(midCum = chrSt + mid)

axis.bc <- manhat_input.bc %>%
  group_by(chromosome) %>%
  dplyr::arrange(chromosome) %>%
  summarize(center = (max(midCum) + min(midCum)) / 2) %>%
  mutate(
    seqnames = as.numeric(chromosome),
    colour = rep(c("grey40", "black"), length.out = 25)
  )

ggplot(manhat_input.bc, aes(x = midCum, y = -log10(PValue))) +
  geom_point(aes(color=chromosome), alpha = 0.5, size = 1) +
  geom_point(alpha = 0.5, size = 1, colour = "red", 
             data = . %>% 
               dplyr::filter(coef == "MPS-IIIC" & chromosome == 1)) +
  geom_point(alpha = 0.5, size = 1, colour = "red", 
             data = . %>% 
               dplyr::filter(coef ==  "MPS-IIIB" & chromosome == 24)) +
  scale_color_manual(values = axis.bc$colour) +
  scale_x_continuous(label = axis.bc$chromosome, breaks = axis.bc$center) +
  scale_y_continuous(limits =c(0, 10)) + # zoom in
  facet_wrap(~coef, ncol = 1) +
  labs(x = "Chromosome", y = expression(paste(-log[10], "(p)"))) +
  theme( 
    legend.position="none",
    panel.border = element_blank(),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.x = element_blank()
  ) 
```

# Extra MPS-IIIB from MPS-IIIB v Q96 exp

```{r}
meta.extrab <- 
  read_xlsx("data/nagluvQ96data/2022_nagluA603fs_v_psen1Q96_adultBrainMetadata.xlsx") %>% 
  mutate(genotype = case_when(
    `usable genotype?`  == "wt" ~ "WT", 
    `usable genotype?` == "MPS-III" ~ "MPS-IIIB", 
    `usable genotype?` == "EOfAD" ~ "EOfAD-like" 
  ) %>% 
    factor(levels = c("WT","MPS-IIIB", "EOfAD-like")),
  Tank = as.factor(tank), 
  Sex = as.factor(sex),
  sample = paste0("B_", fish) 
  )
  
featureCounts.B <- 
  read_delim(file = "data/nagluvQ96data/05_featureCounts/counts.out", delim = "\t", skip = 1) %>%
  set_names(basename(names(.))) %>% 
  set_names(names(.) %>% str_remove(pattern = "_S[0-9]+_merged.Aligned.sortedByCoord.dedup.out.bam")) %>% 
  as_tibble() %>% 
  dplyr::select(-c(Chr, Start, End, Length, Strand)) %>% 
  gather(key = "ULN", value = "counts", starts_with("22")) %>% 
  left_join(meta.extrab) %>% 
  dplyr::select(Geneid, counts, sample) %>% 
  spread(key = "sample", value = "counts") %>% 
  column_to_rownames("Geneid") %>% 
  .[,-(dim(.)[2])]
```
## filter lowly expressed genes
```{r filt}
a <- featureCounts.B %>% 
    cpm(log = TRUE) %>%
    as.data.frame() %>% 
  pivot_longer(
    cols = everything(),
    names_to = "sample",
    values_to = "logCPM"
  ) %>%
  split(f = .$sample) %>%
  lapply(function(x){
    d <- density(x$logCPM)
    tibble(
      sample = unique(x$sample),
      x = d$x,
      y = d$y
    )
  }) %>%
  bind_rows() %>%
  left_join(meta.extrab) %>% 
  ggplot(aes(x, y, colour = genotype, group = sample)) +
  geom_line() +
  labs(
    x = "logCPM",
    y = "Density",
    colour = "Genotype"
  )+
  ggtitle("Before filtering") +
  theme_bw()

b <- featureCounts.B %>% 
  .[rowSums(cpm(.) >= 0.5) >= 8,] %>% 
    cpm(log = TRUE) %>%
    as.data.frame() %>% 
  pivot_longer(
    cols = everything(),
    names_to = "sample",
    values_to = "logCPM"
  ) %>%
  split(f = .$sample) %>%
  lapply(function(x){
    d <- density(x$logCPM)
    tibble(
      sample = unique(x$sample),
      x = d$x,
      y = d$y
    )
  }) %>%
  bind_rows() %>%
  left_join(meta.extrab) %>% 
  ggplot(aes(x, y, colour = genotype, group = sample)) +
  geom_line() +
  labs(
    x = "logCPM",
    y = "Density",
    colour = "Genotype"
  )+
  ggtitle("After filtering")

ggarrange(a, b, common.legend = TRUE)
```

```{r dge}
x.b2 <- featureCounts.B %>% 
  as.matrix() %>% 
  .[rowSums(cpm(.) >= 0.5) >= 4,] %>%
  DGEList(
    samples = tibble(sample = colnames(.)) %>%
      left_join(meta.extrab),
    genes = grGenes[rownames(.)] %>%
      as.data.frame() %>%
      dplyr::select(
        chromosome = seqnames, start, end, 
        gene_id, gene_name, gene_biotype, description, entrezid
      ) %>% 
      left_join(gcGene) %>% 
      as_tibble()
  ) %>%
  calcNormFactors()
```
I used PCA to see whether tank or sex is influening the brain transcriptome. Nothing really obvious. 

```{r}
ggarrange(
 cpm(x.b2, log = T) %>%
    t() %>%
    prcomp() %>% 
    autoplot(data = tibble(sample = rownames(.$x)) %>%
               left_join(x.b2$samples),
             colour = "genotype", 
             size = 4
    )+
    scale_color_viridis_d(option = "turbo", end = 0.8) , 
    
    cpm(x.b2, log = T) %>%
    t() %>%
    prcomp() %>% 
    autoplot(data = tibble(sample = rownames(.$x)) %>%
               left_join(x.b2$samples),
             colour = "Tank", 
             size = 4
    )+
    scale_color_viridis_d(option = "turbo", end = 0.8) , 
 
     cpm(x.b2, log = T) %>%
    t() %>%
    prcomp() %>% 
    autoplot(data = tibble(sample = rownames(.$x)) %>%
               left_join(x.b2$samples),
             colour = "sex", 
             size = 4
    )+
    scale_color_viridis_d(option = "turbo", end = 0.8) 
)
```


```{r}
design.b2 <- model.matrix(~genotype, data = x.b2$samples %>% 
                            droplevels()) %>% # get rid of the unused factor levels
  set_colnames(gsub(colnames(.), pattern = "genotype", replacement = ""))
```




## CQN

A  bit of bias is obsevred for %GC content and gene length, so will apply the `cqn` correction to adjust for this. 

```{r}
cqn.b2 <-   
x.b2 %>% 
  with(
    cqn(
      counts = counts,
      x = genes$gc_content,
      lengths = genes$length,
      sizeFactors = samples$lib.size
    )
  )
logCPM.b2 <- cqn.b2$y + cqn.b2$offset
```

## plot out the model fits
```{r plotCQN, fig.cap ="*Model fits for GC content and gene length under the CQN model. Variability is clearly visible at either end*"}
par(mfrow = c(1, 2))
cqnplot(cqn.b2, n = 1, xlab = "GC Content")
cqnplot(cqn.b2, n = 2, xlab = "Length")
par(mfrow = c(1, 1))
```


# DE with cqn

```{r}
# Add the offset to the dge object 
x.b2$offset <- cqn.b2$glm.offset

# fit the glm
fit_cqn.b2 <- x.b2 %>% 
  estimateDisp(design.b2) %>% 
  glmFit(design.b2)

toptables_cqn.b2 <- design.b2 %>% colnames() %>% .[2] %>% 
  sapply(function(x){
    glmLRT(fit_cqn.b2, coef = x) %>%
      topTags(n = Inf) %>%
      .[["table"]] %>%
      as_tibble() %>%
      arrange(PValue) %>%
       mutate(
        coef = x,
        DE = FDR < 0.05
      ) %>% 
      dplyr::select(
        gene_name, logFC, logCPM, PValue, FDR, DE, everything()  
      )
      
  }, simplify = FALSE)
```

## Vis with CQN
### Num DE genes
```{r}
toptables_cqn.b2 %>% 
  lapply(function(x){
    x %>%
      mutate(Direction = case_when(
        sign(logFC) == "1" ~ "up",
        sign(logFC) == "-1" ~ "down"
      ))
  }) %>%
  bind_rows() %>%
  dplyr::filter(DE == TRUE) %>%
  ggplot(aes(y = `coef`)) +
  geom_bar(stat = "count",
           width = 0.5,
           aes(fill = Direction),
           position = "dodge") +
  theme(legend.position = "bottom") +
  ggtitle("Number of DE genes in each comparison") +
  labs(x = "Number of DE genes",
       y = "Contrast",
       colour = "Direction of change") 
```

### Volcano plots

These are zoomed in again for vis purposes
```{r}
toptables_cqn.b2 %>%
  bind_rows() %>% 
  ggplot(aes(y = -log10(PValue), x = logFC, colour = DE)) +
  geom_point(
    alpha = 0.5, size = 1.25
  ) +
  facet_wrap(~coef) +
  coord_cartesian(xlim = c(-2.5,2.5),
                  ylim = c(0, 20) )+
  theme(legend.position = "none") +
  scale_color_manual(values = c("grey50", "red")) 
```

### MD plots
```{r}
toptables_cqn.b2 %>% 
  bind_rows() %>% 
  ggplot(aes(x = logCPM, y = logFC)) +
  geom_point(
    aes(colour = DE),
    alpha = 0.5
  ) +
  facet_wrap(~coef, ncol = 1) +
  theme(legend.position = "none") +
  scale_color_manual(values = c("grey50", "red"))
```

## Gene set enrichment analysis
#### GENE SETS
```{r}
 
KEGG <- msigdbr("Danio rerio", category = "C2", subcategory = "CP:KEGG") %>% 
  left_join(x.b2$genes, by = c("gene_symbol" = "gene_name")) %>% 
  dplyr::filter(gene_id %in% rownames(x.b2)) %>% 
  distinct(gs_name, gene_id, .keep_all = TRUE) %>% 
  split(f = .$gs_name) %>%
  lapply(extract2, "gene_id")

sizes <- KEGG %>% 
  lapply(length) %>% 
  unlist %>% 
  as.data.frame() %>% 
  set_colnames( "n_genes") %>% 
  rownames_to_column("gs")


# retain gene sets with at least 10 genes in it
KEGG <- KEGG[sizes %>% dplyr::filter(n_genes > 10) %>% .$gs]




GO <- 
  msigdbr("Danio rerio", category = "C5") %>% 
  dplyr::filter(grepl(gs_name, pattern = "^GO")) %>% 
  left_join(x.ab$genes, by = c("gene_symbol" = "gene_name")) %>% 
  dplyr::filter(gene_id %in% rownames(x.b2)) %>% 
  mutate(gs_name = str_replace(gs_name, pattern = "GOBP_", replacement = "GO_")) %>% 
  mutate(gs_name = str_replace(gs_name, pattern = "GOMF_", replacement = "GO_")) %>%     
  mutate(gs_name = str_replace(gs_name, pattern = "GOCC_", replacement = "GO_")) %>%     
  left_join(goSummaries) %>% 
  dplyr::filter(shortest_path >= minPath) %>%
  distinct(gs_name, gene_id, .keep_all = TRUE) %>% 
  split(f = .$gs_name) %>%
  lapply(extract2, "gene_id")


sizes_GO <- GO %>%
  lapply(length) %>%
  unlist %>%
  as.data.frame() %>%
  set_colnames( "n_genes") %>%
  rownames_to_column("gs")


# retain gene sets with at least 5 genes in it
GO <- GO[sizes_GO %>% dplyr::filter(n_genes > 5) %>% .$gs]

ireGenes <- 
  read_xlsx("data/gene_sets/ireGenes.xlsx", sheet = "Zebrafish IREs") %>% 
  gather %>% 
  set_colnames(c("IRE", "gene_id")) %>% 
  mutate(IRE = case_when(
    IRE == "All 3' IREs" ~ "ire3_all", 
    IRE == "All 5' IREs" ~ "ire5_all", 
    IRE == "HQ 3' IREs" ~ "ire3_HQ", 
    IRE == "HQ 5' IREs" ~ "ire5_HQ"
  )) %>% 
  dplyr::filter(gene_id %in% rownames(x.b2)) %>% 
  split(f = .$IRE) %>% 
  lapply(magrittr::extract2,"gene_id") 

chr <- x.b2$genes %>% 
  dplyr::select(gene_id, chromosome) %>% 
  split(f = .$chromosome) %>%
  lapply(extract2, "gene_id") %>% 
  .[c("MT", seq(1:25))] # omit the ALT chromosmes and scaffolds


cell_type_markers <- 
  read_xlsx("data/gene_sets/Suppdata2_jiangetal_2021_fcelldev.xlsx", sheet = "Brain") %>% 
  dplyr::rename("gene_name" = gene) %>% 
  left_join(x.b2$genes) %>% 
  dplyr::filter(gene_id %in% rownames(x.b2)) %>% 
  split(f = .$`cell type`) %>% 
  lapply(extract2, "gene_id")
  

cell_type_markers %<>% 
  lapply(function(y) {
    y[y %in% rownames(x.b2)]
  })
```


## GOSEQ 

```{r}
pwf <- toptables_cqn.b2$`MPS-IIIB` %>%
      with(
        nullp(
          DEgenes = structure(DE, names = gene_id), 
          bias.data = length
        )
      )

goseq(pwf, gene2cat = GO) %>% 
  as_tibble() %>% 
  dplyr::filter(numDEInCat > 0) %>% 
  mutate(FDR = p.adjust(over_represented_pvalue, method = "fdr")) %>% 
  dplyr::select(-under_represented_pvalue) %>% 
  # write_csv("output/results/goseq_cqn/famB2_Bcoef.csv")
  dplyr::filter(FDR < 0.05) 
```



## KEGG 
```{r}
fryKEGG.b2 <- 
  design.b2 %>% colnames() %>% .[2] %>% 
  sapply(function(y) {
    logCPM.b2 %>% 
      fry(
        index = KEGG,
        design = design.b2, 
        contrast = y, 
        sort = "mixed"
      ) %>% 
      rownames_to_column("pathway") %>% 
      as_tibble() %>% 
      mutate(coef = y)
  }, simplify = FALSE)

fryKEGG.b2$`MPS-IIIB` %>% 
  dplyr::filter(FDR.Mixed < 0.05) %>%  
  mutate(pathway = str_remove(pathway, pattern = "KEGG_")) %>% 
  ggplot(aes(x = -log10(PValue.Mixed), y = reorder(pathway, -FDR.Mixed))) +
  geom_col(aes(fill = FDR.Mixed < 0.05)) +
  scale_fill_manual(values = c("grey50", 'red')) +
  theme_pubr() +
  labs(fill = "FDR < 0.05", 
       x = "-log10(PValue)", 
       y = "")
```


### IRE

```{r}
fryIRE.b2 <- design.b2 %>% colnames() %>% .[2] %>% 
  sapply(function(y) {
    logCPM.b2 %>% 
      fry(
        index = ireGenes,
        design = design.b2, 
        contrast = y, 
        sort = "directional"
      ) %>% 
      rownames_to_column("pathway") %>% 
      as_tibble() %>% 
      mutate(coef = y)
  }, simplify = FALSE)

fryIRE.b2 %>% 
  bind_rows() %>% 
  dplyr::select(pathway, coef, pvalue = PValue.Mixed, FDR = FDR.Mixed) %>% 
  ggplot(aes(y = pathway, x = -log10(pvalue))) + 
  geom_col(aes(fill = coef, alpha = FDR < 0.05), 
           position = "dodge") +
  scale_fill_viridis_d(end = 0.75) +
  scale_alpha_manual(values = c(0.5, 1)) +
  geom_vline(xintercept = 1.5) +
  annotate(geom = "label", x = 1.8, y = "ire5_HQ", label = "FDR = 0.05") 

```

### Cell type prop check

Since this experiment was performed on whole brains, we need to have a look at whether changes to cell type proportions could explain cause some artefactual changes to gene expression. We have done this previously by assessing the expression of marker genes of specific cell types. So this is what I will do here. I have taken the lists of genes frmo day 5 only from the zebrafish single cell atlas (Farnsworth et al. 2020) which contain at least 10 genes. This leaves  `r length(cell_type_markers)` gene sets (cell types) to test.

```{r}
celltype.b2 <- design.b2 %>% colnames() %>% .[2] %>% 
  sapply(function(y) {
    logCPM.b2 %>% 
      fry(
        index = cell_type_markers,
        design = design.b2, 
        contrast = y, 
        sort = "directional"
      ) %>% 
      rownames_to_column("pathway") %>% 
      as_tibble() %>% 
      dplyr::select(1:5) %>% 
      mutate(coef = y)
  }, simplify = FALSE)

celltype.b2 %>% 
  bind_rows() %>% 
  ggplot(aes(y = pathway, x = -log10(PValue))) + 
  geom_col(aes(fill = coef, alpha = FDR < 0.05), 
           position = "dodge") +
  scale_fill_viridis_d(end = 0.75) +
  scale_alpha_manual(values = c(0.5, 1)) +
  geom_vline(xintercept = 2.1) +
  annotate(geom = "label", x = 2, y = 4.5, label = "FDR = 0.05") 
```


### Chrosomosomes gene sets

```{r}
fry_chr.b2 <- design.b2 %>% colnames() %>% .[2] %>% 
  sapply(function(y) {
    logCPM.b2 %>% 
      fry(
        index = c(chr),
        design = design.b2, 
        contrast = y, 
        sort = "mixed"
      ) %>% 
      rownames_to_column("pathway") %>% 
      as_tibble() %>% 
      mutate(coef = y)
  }, simplify = FALSE)


fry_chr.b2 %>% 
  bind_rows() %>% 
  mutate(pathway = factor(pathway, levels = c("MT", seq(1:25)))) %>% 
  ggplot(aes(y = pathway, x = -log10(PValue.Mixed), fill = coef)) +
  geom_col(position = "dodge")
  
```

An enrichment of DE genes is observed on the same chromosome as the mutation. This is obsevred in the manhat plots below. 

```{r}
manhat_input.bc <- 
	toptables_cqn.bc %>% 
  bind_rows() %>% 
	left_join(x2$genes %>% 
			  	as_tibble() %>% 
			  	dplyr::select(start, end, chromosome, gene_id) %>% 
			  	group_by(gene_id) %>% 
			  	mutate(mid = mean(c(start, end)))
	) %>% 
  dplyr::filter(chromosome %in% seq(1:25)) %>% 
  group_by(gene_id) %>% 
  mutate(mid = mean(c(start, end))) %>% 
  ungroup %>% 
  mutate(chromosome = factor(chromosome, levels = seq(1:25)))  %>% 
  group_by(chromosome) %>% 
  summarise(chrLen = max(mid)) %>% 
  mutate(chrSt = cumsum(chrLen)-chrLen) %>% 
	dplyr::select(-chrLen) %>% 
  left_join(toptables_cqn.bc %>% 
              bind_rows() %>% 
              left_join(x2$genes %>% 
                          as_tibble() %>% 
                          dplyr::select(start, end, chromosome, gene_id) %>% 
                          group_by(gene_id) %>% 
                          mutate(mid = mean(c(start, end)))
              )
  ) %>% 
  group_by(gene_id) %>% 
  mutate(mid = mean(c(start, end))) %>% 
  ungroup %>% 
  dplyr::arrange(mid, chromosome) %>% 
  mutate(midCum = chrSt + mid)

axis.bc <- manhat_input.bc %>%
  group_by(chromosome) %>%
  dplyr::arrange(chromosome) %>%
  summarize(center = (max(midCum) + min(midCum)) / 2) %>%
  mutate(
    seqnames = as.numeric(chromosome),
    colour = rep(c("grey40", "black"), length.out = 25)
  )

ggplot(manhat_input.bc, aes(x = midCum, y = -log10(PValue))) +
  geom_point(aes(color=chromosome), alpha = 0.5, size = 1) +
  geom_point(alpha = 0.5, size = 1, colour = "red", 
             data = . %>% 
               dplyr::filter(coef == "MPS-IIIC" & chromosome == 1)) +
  geom_point(alpha = 0.5, size = 1, colour = "red", 
             data = . %>% 
               dplyr::filter(coef ==  "MPS-IIIB" & chromosome == 24)) +
  scale_color_manual(values = axis.bc$colour) +
  scale_x_continuous(label = axis.bc$chromosome, breaks = axis.bc$center) +
  scale_y_continuous(limits =c(0, 10)) + # zoom in
  facet_wrap(~coef, ncol = 1) +
  labs(x = "Chromosome", y = expression(paste(-log[10], "(p)"))) +
  theme( 
    legend.position="none",
    panel.border = element_blank(),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.x = element_blank()
  ) 
```


# post-hoc power analysis

```{r}
ab.A.samps <- x.ab$samples %>% 
  dplyr::filter(genotype == "WT") %>% 
  .$sample

mu.ab <- x.ab$counts %>%
  as.data.frame() %>%
  .[ab.A.samps] %>% 
  rowMeans()

disp.ab <-  x.ab %>% estimateDisp(design.ab) %>% . $tagwise.dispersion

fc.ab <- function(x){exp(rnorm(x, log(1), 0.5*log(2)))}

abPowerPlot <- ssizeRNA_vary(nGenes = dim(x.ab$counts)[1], # Num detectable genes
              pi0 = 0.9, # Proportion of non-DE genes
              m = 40, # pseudo sample size
              mu = mu.ab, # Average read count for each gene in control group. Calculated from this current dataset)
              disp = disp.ab, # Dispersion parameter for each gene
              fc = fc.ab, # Fold change per gene
              fdr = 0.05, # FDR level to control
              power = 0.7, # power level
              maxN = 20,
              side = "two-sided",
              replace = TRUE
                )

ac.A.samps <- x.ac$samples %>% 
  dplyr::filter(genotype == "WT") %>% 
  .$sample

mu.ac <- x.ac$counts %>%
  as.data.frame() %>%
  .[ac.A.samps] %>% 
  rowMeans()

disp.ac <-  x.ac %>% estimateDisp() %>% . $tagwise.dispersion

fc.ac <- function(x){exp(rnorm(x, log(1), 0.5*log(2)))}

dim(x.ac$counts)

# set a seed for a reproducible result
set.seed(2223)
ac.powerplot <- ssizeRNA_vary(nGenes = dim(x.ac$counts)[1], # Num detectable genes
              pi0 = 0.9, # Proportion of non-DE genes
              m = 30, # pseudo sample size
              mu = mu.ac, # Average read count for each gene in control group. Calculated from this current dataset)
              disp = disp.ac, # Dispersion parameter for each gene
              fc = fc.ac, # Fold change per gene
              fdr = 0.05, # FDR level to control
              power = 0.7, # power level
              maxN = 20, 
              side = "two-sided",
              replace = T
                )

# BC sans
bc.A.samps <- x.bc$samples %>% 
  dplyr::filter(genotype == "WT") %>% 
  .$sample

mu.bc <- x.bc$counts %>%
  as.data.frame() %>%
  .[bc.A.samps] %>% 
  rowMeans()

disp.bc <-  x.bc %>% estimateDisp() %>% . $tagwise.dispersion

fc.bc <- function(x){exp(rnorm(x, log(1), 0.5*log(2)))}

dim(x.bc$counts)

bcPowerPlot <- ssizeRNA_vary(nGenes = dim(x.bc$counts)[1], # Num detectable genes
              pi0 = 0.9, # Proportion of non-DE genes
              m = 20, # pseudo sample size
              mu = mu.bc, # Average read count for each gene in control group. Calculated from this current dataset)
              disp = disp.bc, # Dispersion parameter for each gene
              fc = fc.bc, # Fold change per gene
              fdr = 0.05, # FDR level to control
              power = 0.7, # power level
              maxN = 20,
              side = "two-sided",
              replace = T
                )



abPowerPlot$power %>% 
  as_tibble() %>% 
  mutate(exp = "A and B") %>% 
  bind_rows(ac.powerplot$power %>% 
              as_tibble() %>% 
              mutate(exp = "A and C")) %>% 
  bind_rows(bcPowerPlot$power %>% 
              as_tibble() %>% 
              mutate(exp = "B and C")
  ) %>% 
  set_colnames(c("Sample size (n)", "Power", "Experiment")) %>% 
  ggplot(aes(y = Power, x = `Sample size (n)`, colour = Experiment)) +
  geom_point() +
  geom_line() +
  scale_x_continuous(limits = c(0,15), breaks = seq(0,15, by = 1)) +
  theme(panel.grid = element_blank()) +
  ggsave("output/plots/powerplots.png", width = 10, height = 2.5, units = "cm", dpi = 600, scale  =2)


ggarrange(abpp, acpp, bcpp,
          ncol = 1, 
          labels = "AUTO") +
  ggsave("output/plots/powercalc.png", width = 10, height = 5, units = "cm", dpi = 500, scale = 5)
```


# Data export
```{r}
x.ac %>% saveRDS("data/R_objects/dge_ac.rds")
toptables_cqn.ac %>% saveRDS("data/R_objects/toptab_ac_cqn.rds")
fryKEGG.ac %>% saveRDS("data/R_objects/AC_fryKEGG.rds")
fryIRE.ac %>% saveRDS("data/R_objects/AC_fryIRE.rds")
celltype.ac %>% saveRDS("data/R_objects/AC_frycellTypes.rds")

x.ab %>% saveRDS("data/R_objects/dge_AB.rds")
fryKEGG.ab %>% saveRDS("data/R_objects/AB_fryKEGG.rds")
fryIRE.ab %>% saveRDS("data/R_objects/AB_fryIRE.rds")
celltypes.ab %>% saveRDS("data/R_objects/AB_frycellTypes.rds")
toptables_cqn.ab %>% saveRDS("data/R_objects/AB_toptab_cqn.rds")

x.bc %>% saveRDS("data/R_objects/dge_bc.rds")
toptables_cqn.bc %>% saveRDS("data/R_objects/toptab_bc_cqn.rds")
fryKEGG.bc %>% saveRDS("data/R_objects/BC_fryKEGG.rds")
fryIRE.bc %>% saveRDS("data/R_objects/BC_fryIRE.rds")
celltype.bc %>% saveRDS("data/R_objects/BC_frycellTypes.rds")

x.b2 %>% saveRDS("data/R_objects/dge_B2.rds")
fryKEGG.b2 %>% saveRDS("data/R_objects/B2_fryKEGG.rds")
fryIRE.b2 %>% saveRDS("data/R_objects/B2_fryIRE.rds")
celltype.b2 %>% saveRDS("data/R_objects/B2_frycellTypes.rds")
toptables_cqn.b2 %>% saveRDS("data/R_objects/B2_toptab_cqn.rds")
```

