---
title: "Enrichment Analysis"
subtitle: "2022_MPSIII_3mBrainRNAseq dataset - AB comparison"
author: "Lachlan Baer"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  workflowr::wflow_html:
    theme: flatly
    toc: true
    toc_float: true
    toc_depth: 2
    fig_width: 8
    fig_height: 6
    fig_align: "center"
    code_folding: hide
editor_options:
  chunk_output_type: console
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  message = FALSE,
  warning = FALSE,
  error = FALSE,
  fig.align = "center"
)
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
```

# Setup

```{r packages}
suppressPackageStartupMessages({
  ## Common
  library(tidyverse)
  library(magrittr)
  library(future.apply)
  library(here)
  library(AnnotationHub)
  library(purrr)
  library(scales)
  library(kableExtra)
  library(tictoc)
  library(ggrepel)
  library(RColorBrewer)
  library(ggpubr)
  library(pander)
  library(rmarkdown)
  ## Project specific
  library(edgeR)
  library(limma)
  library(pheatmap)
  library(cqn)
  library(DT)
  library(htmltools)
  library(msigdbr)
  library(fgsea)
  library(readxl)
})
```

```{r options}
if (interactive()) setwd(here::here())
theme_set(theme_bw())
```

```{r funcs}
## Remove AsIs class
removeAsIs <- function(x){
  if ("AsIs" %in% class(x)) {
    class(x) <- class(x)[-match("AsIs", class(x))]
  }
  x
}

## Simple DT::datatable
## - for small tables needing simple presentation
lbSimpleDT <- function(tbl, cap, rownames = FALSE){
  datatable(
    tbl,
    caption = htmltools::tags$caption(
      cap,
      style = "text-align: left;"
    ),
    options = list(
      dom = "t",  # Available options: tflipr
      columnDefs = list(list(className = 'dt-left', targets = "_all")),  # Align all columns left
      pageLength = -1,  # Show all results
      scrollCollapse = TRUE,  # Removes whitespace when table is small
      scrollY = 350,  # Sets max height of table before scrollbar is added
      scrollX = TRUE,  # Enable horizontal scrolling when too wide
      sScrollX = "100%"  # Stops last column title being cut off
    ),
    rownames = rownames,
    height = "100%",  # Seems to work in tandem with scrollCollapse
    width = "100%",  # Fills width when viewed in RStudio viewer
    class = "stripe hover row-border order-column compact nowrap"
  )
}

## Complex DT::datatable
## - for large tables needing pagination and search
lbComplexDT <- function(tbl, cap, rownames = FALSE){
  datatable(
    tbl,
    caption = htmltools::tags$caption(
      cap,
      style = "text-align: left;"
    ),
    options = list(
      dom = "tflipr",  # Available options: tflipr
      columnDefs = list(list(className = 'dt-left', targets = "_all")),  # Align all columns left
      pageLength = 10,  # Show all results
      scrollCollapse = TRUE,  # Removes whitespace when table is small
      scrollY = "100%",  # Sets max height of table before scrollbar is added
      scrollX = TRUE,  # Enable horizontal scrolling when too wide
      sScrollX = "100%"  # Stops last column title being cut off
    ),
    rownames = rownames,
    height = "100%",  # Seems to work in tandem with scrollCollapse
    width = "100%",  # Fills width when viewed in RStudio viewer
    class = "stripe hover row-border order-column compact nowrap"
  )
}
```

## Introduction

This document contains the enrichment analyses before and after DAR weighting for the MPS-III A/B arm of the experiment.

## EnsDb

```{r ensParams}
ens_species <- "Danio rerio"
ens_release <- "101"
ens_assembly <- "GRCz11"
```

Grab genome feature annotations for `r ens_species` Ensembl release `r ens_release` (genome assembly `r ens_assembly`).

```{r ensDb}
ah <- AnnotationHub() %>%
  subset(species == ens_species) %>%
  subset(rdataclass == "EnsDb")
ahId <- ah$ah_id[str_detect(ah$title, ens_release)]
ensDb <- ah[[ahId]]
```

```{r chrInfo}
# Filter for primary chromosomes
chrInfo <- getChromInfoFromEnsembl(ens_assembly, release = ens_release) %>%
  dplyr::filter(coord_system == "chromosome")
primary_chrs <- chrInfo$name
```

```{r genes}
genes <- genes(ensDb, filter = SeqNameFilter(primary_chrs))
mcols(genes) <- mcols(genes)[
  c("gene_id", "gene_name", "gene_biotype", "entrezid")
]
```

```{r entrezGenes}
entrezGenes <- genes %>%
  as.data.frame() %>%
  dplyr::filter(!is.na(entrezid)) %>%
  mutate(entrezid = removeAsIs(entrezid)) %>%
  unnest(entrezid) %>%
  dplyr::rename(entrez_gene = entrezid)
```

```{r geneChrs}
geneChrs <- as_tibble(genes) %>%
  dplyr::select(gene_id, chromosome = seqnames)
```

## Metadata

Load and clean the sample metadata by renaming columns following `R` naming conventions and setting categorical variables as factors.+
Also append sample file basenames from a file I created (`data/sample_basenames.csv`), as we will need these later.
This metadata contains information for all samples in the experiment, so let's filter for only samples in the A/B dataset.

```{r experiment_arm}
expArm <- "AB"
```

```{r metadata}
metadata <- read_xlsx(here("data/meta_final.xlsx")) %>%
  dplyr::rename(
    rin = RIN,
    sex = Sex,
    rna_batch = rnaBatch,
    mps = MPS,
    uln = ULN,
    home_tank = `Home tank`,
    dob = DOB,
    death = Death
  ) %>%
  mutate(
    sex = as.factor(sex),
    genotype = case_when(
      genotype == "wt" ~ "WT",
      genotype == "het" ~ "sgsh_het",
      genotype == "naglu" ~ "MPSIIIB",
      genotype == "sgsh" ~ "MPSIIIA",
      genotype == "hgsnat" ~ "MPSIIIC"
    ),
    genotype = as.factor(genotype),
    home_tank = as.factor(home_tank)
  ) %>%
  left_join(read_csv(here("data/sample_basenames.csv"))) %>%
  dplyr::filter(mps == expArm) %>%
  droplevels()
```

```{r metadataVis}
metadata %>%
  kable(
    align = "l",
    caption = "Sample metadata for the MPS-III A/B arm of the experiment"
  ) %>%
  kable_styling(
    bootstrap_options = c("striped", "hover", "condensed", "responsive")
  )
```

Set up colours for plotting by genotype.
Going with a colourblind-friendly palette (`Set2`) for best accessibility.

```{r genoCols}
genoCols <- metadata$genotype %>%
  levels() %>%
  length() %>%
  brewer.pal("Set2") %>%
  setNames(levels(metadata$genotype))
```

## Differential expression data

```{r topTables}
topTables <- readRDS(here("data/R_objects/AB_toptab_cqn.rds")) %>%
  set_names(c("het_vs_WT", "A_vs_WT", "B_vs_WT"))
```

```{r deGenes}
deGenes <- lapply(topTables, function(tt){
  dplyr::filter(tt, DE) %>%
    pull(gene_id)
})
```

## Gene DAR

```{r}
geneDar <- readRDS(here("data/dar-analysis", expArm, "geneDar.Rds"))
```

# Gene sets

Multiple gene set databases were accessed with the `msigdbr` package.
For this analysis the Hallmark, KEGG, Wikipathways and chromosomal based gene sets were chosen.

```{r hm}
hm <- msigdbr(ens_species, category = "H") %>%
  dplyr::filter(!is.na(ensembl_gene)) %>%
  distinct(gs_name, ensembl_gene, .keep_all = TRUE) %>%
  mutate(gs_name = str_remove(gs_name, "^HALLMARK_"))
hmByGene <- hm %>%
  split(f = .$ensembl_gene) %>%
  lapply(extract2, "gs_name")
hmByID <- hm %>%
  split(f = .$gs_name) %>%
  lapply(extract2, "ensembl_gene")
```

- The Hallmark gene sets consists of `r length(hmByID)` gene sets ranging from sizes of `r min(sapply(hmByID, length))` to `r max(sapply(hmByID, length))` genes.

```{r kg}
kg <- msigdbr(ens_species, category = "C2", subcategory = "CP:KEGG") %>%
  dplyr::filter(!is.na(ensembl_gene)) %>%
  distinct(gs_name, ensembl_gene, .keep_all = TRUE) %>%
  mutate(gs_name = str_remove(gs_name, "^KEGG_"))
kgByGene <- kg %>%
  split(f = .$ensembl_gene) %>%
  lapply(extract2, "gs_name")
kgByID <- kg %>%
  split(f = .$gs_name) %>%
  lapply(extract2, "ensembl_gene")
```

- The KEGG gene sets consists of `r length(kgByID)` gene sets ranging from sizes of `r min(sapply(kgByID, length))` to `r max(sapply(kgByID, length))` genes.

```{r wk}
wk <- msigdbr(ens_species, category = "C2", subcategory = "CP:WIKIPATHWAYS") %>%
  dplyr::filter(!is.na(ensembl_gene)) %>%
  distinct(gs_name, ensembl_gene, .keep_all = TRUE) %>%
  mutate(gs_name = str_remove(gs_name, "^WP_"))
wkByGene <- wk %>%
  split(f = .$ensembl_gene) %>%
  lapply(extract2, "gs_name")
wkByID <- wk %>%
  split(f = .$gs_name) %>%
  lapply(extract2, "ensembl_gene")
```

- The Wikipathways gene sets consists of `r length(wkByID)` gene sets ranging from sizes of `r min(sapply(wkByID, length))` to `r max(sapply(wkByID, length))` genes.

# Fisher's exact test for clustering

## Chromosome clustering

```{r}
byChr_fisher <- sapply(topTables, function(tt){
  # Check for DE genes
  if (sum(tt$DE)) {
    sapply(chrInfo$name[!(chrInfo$name %in% c("MT", "X", "Y"))], function(chr){
      table(
        tt$DE,
        tt$chromosome == chr
      ) %>%
        fisher.test()
    }, simplify = FALSE) %>%
      .[as.character(sort(as.numeric(names(.)), ))]
  }
}, simplify = FALSE)
```

```{r}
byChr_pvals <- sapply(byChr_fisher, function(geno){
  lapply(geno, function(fisherRes){
    fisherRes$p.value
  }) %>%
    unlist() %>%
    enframe(name = "chr", value = "p") %>%
    mutate(
      bonf = p.adjust(p, method = "bonf"),
      sig = bonf < 0.05,
      p = signif(p, digits = 3),
      bonf = signif(bonf, digits = 3),
    ) %>%
    dplyr::filter(sig)
}, simplify = FALSE)
```

```{r}
tagList({
  lapply(names(byChr_pvals), function(x){
    lbSimpleDT(
      byChr_pvals[[x]],
      cap = paste(
        "Chromosome(s) showing significant enrichment for DE genes in the",
        x, "comparison (Bonferonni p < 0.05)."
      )
    ) %>%
      formatSignif(
        columns = c("p", "bonf"), digits = 2
      )
  })
})
```

# Enrichment testing

## GSEA {.tabset .tabset-pills}

Set up directional rankings with/without DAR weighting.

Note that mitochondrial genes have been excluded as DAR analysis was not performed on the mitochondrial DNA.

```{r ranks}
ranks_d <- lapply(topTables, function(x){
  x %>%
    dplyr::filter(chromosome != "MT") %>%  ## Remove mitochondrial genes to keep consistency with weighted gene list
    mutate(stat = -log10(PValue) * sign(logFC)) %>%
    dplyr::arrange(desc(stat)) %>%
    with(structure(stat, names = gene_id))
})
ranks_d_dar <- lapply(names(topTables), function(x){
  mcols(geneDar[[x]])[,c("gene_id", "dar")] %>%
    as_tibble() %>%
    right_join(topTables[[x]]) %>%
    dplyr::filter(chromosome != "MT") %>%  ## Remove mitochondrial genes as no DAR assigned
    mutate(stat = (-log10(PValue) * sign(logFC)) * (1 - dar)) %>%
    dplyr::arrange(desc(stat)) %>%
    with(structure(stat, names = gene_id))
}) %>%
  set_names(names(topTables))
```

We can assess the impact of weighting gene rankings with DAR by plotting the ranks before and after weighting:

```{r compareRanks, fig.height=9}
lapply(names(ranks_d), function(x){
  before <- enframe(ranks_d[[x]], name = "gene_id", value = "before") %>%
    mutate(
      rank_before = seq(1, length(gene_id), by = 1),
      contrast = x
    )
  after <- enframe(ranks_d_dar[[x]], name = "gene_id", value = "after") %>%
    mutate(
      rank_after = seq(1, length(gene_id), by = 1),
      contrast = x
    )
  left_join(before, after)
}) %>%
  bind_rows() %>%
  ggplot(aes(rank_before, rank_after)) +
  geom_point() +
  geom_abline(slope = 1, intercept = 0, colour = "blue") +
  labs(x = "Before weighting", y = "After weighting") +
  facet_wrap(~ contrast, ncol = 1)
```

### Hallmark {.tabset}

```{r gsea_hm}
set.seed(230815)
gsea_hm <- lapply(ranks_d, function(x){
  fgseaMultilevel(hmByID, x) %>%
    as_tibble() %>%
    dplyr::rename(FDR = padj, Pathway = pathway) %>%
    dplyr::arrange(pval) %>%
    mutate(
      edgeSize = vapply(leadingEdge, length, numeric(1)),
      Direction = case_when(
        sign(ES) == 1 ~ "Up",
        sign(ES) == -1 ~ "Down",
        sign(ES) == 0 ~ "None"
      ),
      rank = row_number(),
      bonf_p = p.adjust(pval, "bonferroni"),
      sig_FDR = FDR < 0.05,
      sig_bonf = bonf_p < 0.05
    )
})
```

```{r gsea_hm_dar}
set.seed(230815)
gsea_hm_dar <- lapply(ranks_d_dar, function(x){
  fgseaMultilevel(hmByID, x) %>%
    as_tibble() %>%
    dplyr::rename(FDR = padj, Pathway = pathway) %>%
    dplyr::arrange(pval) %>%
    mutate(
      edgeSize = vapply(leadingEdge, length, numeric(1)),
      Direction = case_when(
        sign(ES) == 1 ~ "Up",
        sign(ES) == -1 ~ "Down",
        sign(ES) == 0 ~ "None"
      ),
      rank = row_number(),
      bonf_p = p.adjust(pval, "bonferroni"),
      sig_FDR = FDR < 0.05,
      sig_bonf = bonf_p < 0.05
    )
})
```

#### MPS-IIIA

##### Standard enrichment

```{r gsea_hm_A}
gsea_hm$A_vs_WT %>%
  mutate(Pathway = str_trunc(Pathway, 20)) %>%
  dplyr::select(
    Pathway, setSize = size, edgeSize, Direction,
    raw_p = pval, FDR, sig_FDR, bonf_p, sig_bonf
  ) %>%
  lbComplexDT(gsea_hm, cap = "") %>%
  formatSignif(
    columns = c("raw_p", "FDR", "bonf_p"), digits = 2
  ) %>%
  formatStyle(
    columns = c("sig_FDR", "sig_bonf"),
    fontWeight = styleEqual(levels = TRUE, values = "bold")
  )
```

##### Enrichment with DAR weighting

```{r gsea_hm_dar_A}
gsea_hm_dar$A_vs_WT %>%
  mutate(Pathway = str_trunc(Pathway, 20)) %>%
  dplyr::select(
    Pathway, setSize = size, edgeSize, Direction,
    raw_p = pval, FDR, sig_FDR, bonf_p, sig_bonf
  ) %>%
  lbComplexDT(gsea_hm, cap = "") %>%
  formatSignif(
    columns = c("raw_p", "FDR", "bonf_p"), digits = 2
  ) %>%
  formatStyle(
    columns = c("sig_FDR", "sig_bonf"),
    fontWeight = styleEqual(levels = TRUE, values = "bold")
  )
```

#### MPS-IIIB

##### Standard enrichment

```{r gsea_hm_B}
gsea_hm$B_vs_WT %>%
  mutate(Pathway = str_trunc(Pathway, 20)) %>%
  dplyr::select(
    Pathway, setSize = size, edgeSize, Direction,
    raw_p = pval, FDR, sig_FDR, bonf_p, sig_bonf
  ) %>%
  lbComplexDT(gsea_hm, cap = "") %>%
  formatSignif(
    columns = c("raw_p", "FDR", "bonf_p"), digits = 2
  ) %>%
  formatStyle(
    columns = c("sig_FDR", "sig_bonf"),
    fontWeight = styleEqual(levels = TRUE, values = "bold")
  )
```

##### Enrichment with DAR weighting

```{r gsea_hm_dar_B}
gsea_hm_dar$B_vs_WT %>%
  mutate(Pathway = str_trunc(Pathway, 20)) %>%
  dplyr::select(
    Pathway, setSize = size, edgeSize, Direction,
    raw_p = pval, FDR, sig_FDR, bonf_p, sig_bonf
  ) %>%
  lbComplexDT(gsea_hm, cap = "") %>%
  formatSignif(
    columns = c("raw_p", "FDR", "bonf_p"), digits = 2
  ) %>%
  formatStyle(
    columns = c("sig_FDR", "sig_bonf"),
    fontWeight = styleEqual(levels = TRUE, values = "bold")
  )
```

#### sgsh/+

##### Standard enrichment

```{r gsea_hm_het}
gsea_hm$het_vs_WT %>%
  mutate(Pathway = str_trunc(Pathway, 20)) %>%
  dplyr::select(
    Pathway, setSize = size, edgeSize, Direction,
    raw_p = pval, FDR, sig_FDR, bonf_p, sig_bonf
  ) %>%
  lbComplexDT(gsea_hm, cap = "") %>%
  formatSignif(
    columns = c("raw_p", "FDR", "bonf_p"), digits = 2
  ) %>%
  formatStyle(
    columns = c("sig_FDR", "sig_bonf"),
    fontWeight = styleEqual(levels = TRUE, values = "bold")
  )
```

##### Enrichment with DAR weighting

```{r gsea_hm_dar_het}
gsea_hm_dar$het_vs_WT %>%
  mutate(Pathway = str_trunc(Pathway, 20)) %>%
  dplyr::select(
    Pathway, setSize = size, edgeSize, Direction,
    raw_p = pval, FDR, sig_FDR, bonf_p, sig_bonf
  ) %>%
  lbComplexDT(gsea_hm, cap = "") %>%
  formatSignif(
    columns = c("raw_p", "FDR", "bonf_p"), digits = 2
  ) %>%
  formatStyle(
    columns = c("sig_FDR", "sig_bonf"),
    fontWeight = styleEqual(levels = TRUE, values = "bold")
  )
```

### KEGG {.tabset}

```{r gsea_kg}
set.seed(230815)
gsea_kg <- lapply(ranks_d, function(x){
  fgseaMultilevel(kgByID, x) %>%
    as_tibble() %>%
    dplyr::rename(FDR = padj, Pathway = pathway) %>%
    dplyr::arrange(pval) %>%
    mutate(
      edgeSize = vapply(leadingEdge, length, numeric(1)),
      Direction = case_when(
        sign(ES) == 1 ~ "Up",
        sign(ES) == -1 ~ "Down",
        sign(ES) == 0 ~ "None"
      ),
      rank = row_number(),
      bonf_p = p.adjust(pval, "bonferroni"),
      sig_FDR = FDR < 0.05,
      sig_bonf = bonf_p < 0.05
    )
})
```

```{r gsea_kg_dar}
set.seed(230815)
gsea_kg_dar <- lapply(ranks_d_dar, function(x){
  fgseaMultilevel(kgByID, x) %>%
    as_tibble() %>%
    dplyr::rename(FDR = padj, Pathway = pathway) %>%
    dplyr::arrange(pval) %>%
    mutate(
      edgeSize = vapply(leadingEdge, length, numeric(1)),
      Direction = case_when(
        sign(ES) == 1 ~ "Up",
        sign(ES) == -1 ~ "Down",
        sign(ES) == 0 ~ "None"
      ),
      rank = row_number(),
      bonf_p = p.adjust(pval, "bonferroni"),
      sig_FDR = FDR < 0.05,
      sig_bonf = bonf_p < 0.05
    )
})
```

#### MPS-IIIA

##### Standard enrichment

```{r gsea_kg_A}
gsea_kg$A_vs_WT %>%
  mutate(Pathway = str_trunc(Pathway, 20)) %>%
  dplyr::select(
    Pathway, setSize = size, edgeSize, Direction,
    raw_p = pval, FDR, sig_FDR, bonf_p, sig_bonf
  ) %>%
  lbComplexDT(gsea_hm, cap = "") %>%
  formatSignif(
    columns = c("raw_p", "FDR", "bonf_p"), digits = 2
  ) %>%
  formatStyle(
    columns = c("sig_FDR", "sig_bonf"),
    fontWeight = styleEqual(levels = TRUE, values = "bold")
  )
```

##### Enrichment with DAR weighting

```{r gsea_kg_dar_A}
gsea_kg_dar$A_vs_WT %>%
  mutate(Pathway = str_trunc(Pathway, 20)) %>%
  dplyr::select(
    Pathway, setSize = size, edgeSize, Direction,
    raw_p = pval, FDR, sig_FDR, bonf_p, sig_bonf
  ) %>%
  lbComplexDT(gsea_hm, cap = "") %>%
  formatSignif(
    columns = c("raw_p", "FDR", "bonf_p"), digits = 2
  ) %>%
  formatStyle(
    columns = c("sig_FDR", "sig_bonf"),
    fontWeight = styleEqual(levels = TRUE, values = "bold")
  )
```

#### MPS-IIIB

##### Standard enrichment

```{r gsea_kg_B}
gsea_kg$B_vs_WT %>%
  mutate(Pathway = str_trunc(Pathway, 20)) %>%
  dplyr::select(
    Pathway, setSize = size, edgeSize, Direction,
    raw_p = pval, FDR, sig_FDR, bonf_p, sig_bonf
  ) %>%
  lbComplexDT(gsea_hm, cap = "") %>%
  formatSignif(
    columns = c("raw_p", "FDR", "bonf_p"), digits = 2
  ) %>%
  formatStyle(
    columns = c("sig_FDR", "sig_bonf"),
    fontWeight = styleEqual(levels = TRUE, values = "bold")
  )
```

##### Enrichment with DAR weighting

```{r gsea_kg_dar_B}
gsea_kg_dar$B_vs_WT %>%
  mutate(Pathway = str_trunc(Pathway, 20)) %>%
  dplyr::select(
    Pathway, setSize = size, edgeSize, Direction,
    raw_p = pval, FDR, sig_FDR, bonf_p, sig_bonf
  ) %>%
  lbComplexDT(gsea_hm, cap = "") %>%
  formatSignif(
    columns = c("raw_p", "FDR", "bonf_p"), digits = 2
  ) %>%
  formatStyle(
    columns = c("sig_FDR", "sig_bonf"),
    fontWeight = styleEqual(levels = TRUE, values = "bold")
  )
```

#### sgsh/+

##### Standard enrichment

```{r gsea_kg_het}
gsea_kg$het_vs_WT %>%
  mutate(Pathway = str_trunc(Pathway, 20)) %>%
  dplyr::select(
    Pathway, setSize = size, edgeSize, Direction,
    raw_p = pval, FDR, sig_FDR, bonf_p, sig_bonf
  ) %>%
  lbComplexDT(gsea_hm, cap = "") %>%
  formatSignif(
    columns = c("raw_p", "FDR", "bonf_p"), digits = 2
  ) %>%
  formatStyle(
    columns = c("sig_FDR", "sig_bonf"),
    fontWeight = styleEqual(levels = TRUE, values = "bold")
  )
```

##### Enrichment with DAR weighting

```{r gsea_kg_dar_het}
gsea_kg_dar$het_vs_WT %>%
  mutate(Pathway = str_trunc(Pathway, 20)) %>%
  dplyr::select(
    Pathway, setSize = size, edgeSize, Direction,
    raw_p = pval, FDR, sig_FDR, bonf_p, sig_bonf
  ) %>%
  lbComplexDT(gsea_hm, cap = "") %>%
  formatSignif(
    columns = c("raw_p", "FDR", "bonf_p"), digits = 2
  ) %>%
  formatStyle(
    columns = c("sig_FDR", "sig_bonf"),
    fontWeight = styleEqual(levels = TRUE, values = "bold")
  )
```

### Wikipathways {.tabset}

```{r gsea_wk}
set.seed(230815)
gsea_wk <- lapply(ranks_d, function(x){
  fgseaMultilevel(wkByID, x) %>%
    as_tibble() %>%
    dplyr::rename(FDR = padj, Pathway = pathway) %>%
    dplyr::arrange(pval) %>%
    mutate(
      edgeSize = vapply(leadingEdge, length, numeric(1)),
      Direction = case_when(
        sign(ES) == 1 ~ "Up",
        sign(ES) == -1 ~ "Down",
        sign(ES) == 0 ~ "None"
      ),
      rank = row_number(),
      bonf_p = p.adjust(pval, "bonferroni"),
      sig_FDR = FDR < 0.05,
      sig_bonf = bonf_p < 0.05
    )
})
```

```{r gsea_wk_dar}
set.seed(230815)
gsea_wk_dar <- lapply(ranks_d_dar, function(x){
  fgseaMultilevel(wkByID, x) %>%
    as_tibble() %>%
    dplyr::rename(FDR = padj, Pathway = pathway) %>%
    dplyr::arrange(pval) %>%
    mutate(
      edgeSize = vapply(leadingEdge, length, numeric(1)),
      Direction = case_when(
        sign(ES) == 1 ~ "Up",
        sign(ES) == -1 ~ "Down",
        sign(ES) == 0 ~ "None"
      ),
      rank = row_number(),
      bonf_p = p.adjust(pval, "bonferroni"),
      sig_FDR = FDR < 0.05,
      sig_bonf = bonf_p < 0.05
    )
})
```

#### MPS-IIIA

##### Standard enrichment

```{r gsea_wk_A}
gsea_wk$A_vs_WT %>%
  mutate(Pathway = str_trunc(Pathway, 20)) %>%
  dplyr::select(
    Pathway, setSize = size, edgeSize, Direction,
    raw_p = pval, FDR, sig_FDR, bonf_p, sig_bonf
  ) %>%
  lbComplexDT(gsea_hm, cap = "") %>%
  formatSignif(
    columns = c("raw_p", "FDR", "bonf_p"), digits = 2
  ) %>%
  formatStyle(
    columns = c("sig_FDR", "sig_bonf"),
    fontWeight = styleEqual(levels = TRUE, values = "bold")
  )
```

##### Enrichment with DAR weighting

```{r gsea_wk_dar_A}
gsea_wk_dar$A_vs_WT %>%
  mutate(Pathway = str_trunc(Pathway, 20)) %>%
  dplyr::select(
    Pathway, setSize = size, edgeSize, Direction,
    raw_p = pval, FDR, sig_FDR, bonf_p, sig_bonf
  ) %>%
  lbComplexDT(gsea_hm, cap = "") %>%
  formatSignif(
    columns = c("raw_p", "FDR", "bonf_p"), digits = 2
  ) %>%
  formatStyle(
    columns = c("sig_FDR", "sig_bonf"),
    fontWeight = styleEqual(levels = TRUE, values = "bold")
  )
```

#### MPS-IIIB

##### Standard enrichment

```{r gsea_wk_B}
gsea_wk$B_vs_WT %>%
  mutate(Pathway = str_trunc(Pathway, 20)) %>%
  dplyr::select(
    Pathway, setSize = size, edgeSize, Direction,
    raw_p = pval, FDR, sig_FDR, bonf_p, sig_bonf
  ) %>%
  lbComplexDT(gsea_hm, cap = "") %>%
  formatSignif(
    columns = c("raw_p", "FDR", "bonf_p"), digits = 2
  ) %>%
  formatStyle(
    columns = c("sig_FDR", "sig_bonf"),
    fontWeight = styleEqual(levels = TRUE, values = "bold")
  )
```

##### Enrichment with DAR weighting

```{r gsea_wk_dar_B}
gsea_wk_dar$B_vs_WT %>%
  mutate(Pathway = str_trunc(Pathway, 20)) %>%
  dplyr::select(
    Pathway, setSize = size, edgeSize, Direction,
    raw_p = pval, FDR, sig_FDR, bonf_p, sig_bonf
  ) %>%
  lbComplexDT(gsea_hm, cap = "") %>%
  formatSignif(
    columns = c("raw_p", "FDR", "bonf_p"), digits = 2
  ) %>%
  formatStyle(
    columns = c("sig_FDR", "sig_bonf"),
    fontWeight = styleEqual(levels = TRUE, values = "bold")
  )
```

#### sgsh/+

##### Standard enrichment

```{r gsea_wk_het}
gsea_wk$het_vs_WT %>%
  mutate(Pathway = str_trunc(Pathway, 20)) %>%
  dplyr::select(
    Pathway, setSize = size, edgeSize, Direction,
    raw_p = pval, FDR, sig_FDR, bonf_p, sig_bonf
  ) %>%
  lbComplexDT(gsea_hm, cap = "") %>%
  formatSignif(
    columns = c("raw_p", "FDR", "bonf_p"), digits = 2
  ) %>%
  formatStyle(
    columns = c("sig_FDR", "sig_bonf"),
    fontWeight = styleEqual(levels = TRUE, values = "bold")
  )
```

##### Enrichment with DAR weighting

```{r gsea_wk_dar_het}
gsea_wk_dar$het_vs_WT %>%
  mutate(Pathway = str_trunc(Pathway, 20)) %>%
  dplyr::select(
    Pathway, setSize = size, edgeSize, Direction,
    raw_p = pval, FDR, sig_FDR, bonf_p, sig_bonf
  ) %>%
  lbComplexDT(gsea_hm, cap = "") %>%
  formatSignif(
    columns = c("raw_p", "FDR", "bonf_p"), digits = 2
  ) %>%
  formatStyle(
    columns = c("sig_FDR", "sig_bonf"),
    fontWeight = styleEqual(levels = TRUE, values = "bold")
  )
```
