---
title: "explorationCQN"
author: "Karissa Barthelson"
date: "2022-04-24"
output: workflowr::wflow_html
editor_options:
  chunk_output_type: console
---

## Introduction
```{r setup, include=FALSE}
knitr::opts_chunk$set(
  autodep = TRUE,
  echo = TRUE,
  warning = FALSE,
  message = FALSE,
  fig.align = "center", 
  out.width ="75%", 
  out.height = "75%"
)
```

```{r loadLibs}
library(tidyverse)
library(magrittr)
library(readxl)
library(ngsReports)
library(AnnotationHub)
library(biomaRt)
library(msigdbr)
library(pander)
library(scales)
library(pheatmap)
library(ggpubr)
library(ggrepel)
library(ggfortify)
library(RColorBrewer)
library(UpSetR)
library(edgeR)
library(cqn)
library(goseq)
library(fgsea)
library(vegan) # for PCoA
library(ape)   # for PCoA
library(harmonicmeanp)
theme_set(theme_bw())
```

```{r anno}
ah <- AnnotationHub() %>%
	subset(species == "Danio rerio") %>%
	subset(rdataclass == "EnsDb")

ensDb <- ah[["AH83189"]] # for release 101, latest version in Ah
grTrans <- transcripts(ensDb)
trLengths <- exonsBy(ensDb, "tx") %>%
	width() %>%
	vapply(sum, integer(1))
mcols(grTrans)$length <- trLengths[names(grTrans)]
gcGene <- grTrans %>%
  mcols() %>%
  as.data.frame() %>%
  dplyr::select(gene_id, tx_id, gc_content, length) %>%
  as_tibble() %>%
  group_by(gene_id) %>%
  summarise(
    gc_content = sum(gc_content*length) / sum(length),
    length = ceiling(median(length))
  )
grGenes <- genes(ensDb)
mcols(grGenes) %<>%
  as.data.frame() %>%
  left_join(gcGene) %>%
  as.data.frame() %>%
  DataFrame()
```

```{r meta}
meta <- 
  read_excel("data/RNASeqMetaData.xlsx", sheet = "SAGC AB1 batch") %>% 
  mutate(genotype = case_when(
    genotype == "wt" ~ "WT", 
    genotype == "het" ~ "sgsh/+",
    genotype == "naglu hom" ~ "MPS-IIIB", 
    genotype == "sgsh hom" ~ "MPS-IIIA"
  ) %>% 
    factor(levels = c("WT", "sgsh/+", "MPS-IIIA", "MPS-IIIB"))
  ) %>% 
  mutate(sample = SampleName) %>% 
  left_join(read_excel("data/RNASeqMetaData.xlsx", sheet = "AB 1.1") %>% 
              dplyr::select(batchKilled, Tank, Sex, sample), 
            by = "sample"
  ) %>% 
  mutate(Tank = as.factor(Tank),
         batchKilled = as.factor(batchKilled), 
         Sex = as.factor(Sex)) %>% 
  dplyr::filter(!is.na(ULN))

featureCounts <- 
  read_delim(file = "data/AB/featurecounts/counts.out", delim = "\t", skip = 1) %>%
  set_names(basename(names(.))) %>% 
  set_names(names(.) %>% str_remove(pattern = "_S[0-9]+_merged.Aligned.sortedByCoord.dedup.out.bam")) %>% 
  as_tibble() %>% 
  dplyr::select(-c(Chr, Start, End, Length, Strand)) %>% 
  gather(key = "ULN", value = "counts", starts_with("22")) %>% 
  left_join(meta) %>% 
  dplyr::select(Geneid, counts, sample) %>% 
  spread(key = "sample", value = "counts") %>% 
  column_to_rownames("Geneid") %>% 
  .[,-(dim(.)[2])]
```

## filter lowly expressed genes
```{r filt}
a <- featureCounts %>% 
    cpm(log = TRUE) %>%
    as.data.frame() %>% 
  pivot_longer(
    cols = everything(),
    names_to = "sample",
    values_to = "logCPM"
  ) %>%
  split(f = .$sample) %>%
  lapply(function(x){
    d <- density(x$logCPM)
    tibble(
      sample = unique(x$sample),
      x = d$x,
      y = d$y
    )
  }) %>%
  bind_rows() %>%
  left_join(meta) %>% 
  ggplot(aes(x, y, colour = genotype, group = sample)) +
  geom_line() +
  labs(
    x = "logCPM",
    y = "Density",
    colour = "Genotype"
  )+
  ggtitle("Before filtering") +
  theme_bw()

b <- featureCounts %>% 
  .[rowSums(cpm(.) >= 0.5) >= 4,] %>% 
    cpm(log = TRUE) %>%
    as.data.frame() %>% 
  pivot_longer(
    cols = everything(),
    names_to = "sample",
    values_to = "logCPM"
  ) %>%
  split(f = .$sample) %>%
  lapply(function(x){
    d <- density(x$logCPM)
    tibble(
      sample = unique(x$sample),
      x = d$x,
      y = d$y
    )
  }) %>%
  bind_rows() %>%
  left_join(meta) %>% 
  ggplot(aes(x, y, colour = genotype, group = sample)) +
  geom_line() +
  labs(
    x = "logCPM",
    y = "Density",
    colour = "Genotype"
  )+
  ggtitle("After filtering")

ggarrange(a, b, common.legend = TRUE)
```

```{r dge}
x <- featureCounts %>% 
  as.matrix() %>% 
  .[rowSums(cpm(.) >= 0.5) >= 4,] %>%
  DGEList(
    samples = tibble(sample = colnames(.)) %>%
      left_join(meta),
    genes = grGenes[rownames(.)] %>%
      as.data.frame() %>%
      dplyr::select(
        chromosome = seqnames, start, end, 
        gene_id, gene_name, gene_biotype, description, entrezid
      ) %>% 
      left_join(gcGene) %>% 
      as_tibble()
  ) %>%
  calcNormFactors()
```

## PCA

```{r}
x$counts %>% 
  cpm(log=TRUE) %>% 
  t() %>% 
  prcomp() %>% 
  autoplot(data = tibble(sample = rownames(.$x)) %>%
      left_join(x$samples),
    colour = "genotype", 
    size = 6
  ) +
  theme_bw() + 
  theme(aspect.ratio = 1) +
  labs(colour = "Genotype") +
  scale_color_brewer(palette = "Set2")


x$counts %>% 
  cpm(log=TRUE) %>% 
  t() %>% 
  prcomp() %>% 
  autoplot(data = tibble(sample = rownames(.$x)) %>%
      left_join(x$samples),
    colour = "Sex", 
    size = 6
  ) +
  theme_bw() + 
  theme(aspect.ratio = 1) +
  labs(colour = "sex")
```


## %GC and Length check

```{r}
design <- model.matrix(~genotype, data = x$samples) %>% 
  set_colnames(gsub(colnames(.), pattern = "genotype", replacement = ""))

fit_1 <- x %>% 
  estimateDisp(design) %>% 
  glmFit(design)

toptable_1 <- design %>% colnames() %>% .[2:4] %>% 
  sapply(function(x){
    glmLRT(fit_1, coef = x) %>%
      topTags(n = Inf) %>%
      .[["table"]] %>%
      as_tibble() %>%
      arrange(PValue) %>%
       mutate(
        coef = x,
        DE = FDR < 0.05
      ) %>% 
      dplyr::select(
        gene_name, logFC, logCPM, PValue, FDR, DE, everything()  
      )
      
  }, simplify = FALSE)

toptable_1 %>%
  bind_rows() %>% 
  ggplot(aes(y = -log10(PValue), x = logFC, colour = DE)) +
  geom_point(
    alpha = 0.5, size = 1.25
  ) +
  facet_wrap(~coef) +
  # geom_label_repel(
  #   aes(label = gene_name),
  #   data = .  %>% dplyr::filter(FDR < 0.05),
  #   show.legend = FALSE
  #   ) +
  coord_cartesian(xlim = c(-2.5,2.5), 
                  ylim = c(0, 20) )+
  theme(legend.position = "none") +
  scale_color_manual(values = c("grey50", "red")) 
  #ggsave("output/plots/volvano.png", width = 18, height = 10, units = "cm", dpi = 300, scale = 1.5)

toptable_1 %>%
  bind_rows() %>% 
  ggplot(aes(x = logCPM, y = logFC,  colour = DE)) +
  geom_point(
    alpha = 0.5
  ) +
  facet_wrap(~coef) +
  theme(legend.position = "none") +
  scale_color_manual(values = c("grey50", "red"))

```

### Vis

```{r}
ggarrange(
toptable_1 %>% 
  bind_rows() %>% 
  mutate(rankstat = sign(logFC)*-log10(PValue)) %>% 
  ggplot(aes(x = length, y = rankstat)) +
  geom_point(
    aes(colour = DE),
    alpha = 0.5
  ) +
  geom_smooth(se = FALSE, method = "gam") +
  facet_grid(rows = vars(coef)) +
  theme_bw() +
  theme(legend.position = "none") +
  scale_color_manual(values = c("grey50", "red")) +
  scale_x_log10()+
  labs(x = "Average transcript length per gene",
       colour = "Differentially expressed?",
       y = "sign(logFC)*-log10(PValue)") +
  coord_cartesian(ylim = c(-10, 10)),

toptable_1 %>% 
  bind_rows() %>% 
  mutate(rankstat = sign(logFC)*-log10(PValue)) %>% 
  ggplot(aes(x = gc_content, y = rankstat)) +
  geom_point(
    aes(colour = DE),
    alpha = 0.5
  ) +
  geom_smooth(se = FALSE, method = "gam") +
  facet_grid(rows = vars(coef)) +
  theme_bw() +
  theme(legend.position = "none") +
  scale_color_manual(values = c("grey50", "red")) +
  coord_cartesian(ylim = c(-10,10)) + 
  labs(x = "Weighted average GC content (%) per gene", 
       colour = "Differentially expressed?", 
       y = "sign(logFC)*-log10(PValue)"),
common.legend = TRUE, 
labels = "AUTO"
) 
```

# CQN

A little bit of bias is obsevred for %GC content and gene length, so will apply the `cqn` correction to adjust for this. 

```{r}
cqn <-   
x %>% 
  with(
    cqn(
      counts = counts,
      x = genes$gc_content,
      lengths = genes$length,
      sizeFactors = samples$lib.size
    )
  )
logCPM <- cqn$y + cqn$offset
```

## plot out the model fits
```{r plotCQN, fig.cap ="*Model fits for GC content and gene length under the CQN model. Variability is clearly visible at either end*"}
par(mfrow = c(1, 2))
cqnplot(cqn, n = 1, xlab = "GC Content")
cqnplot(cqn, n = 2, xlab = "Length")
par(mfrow = c(1, 1))
```

## PCA
PCA analysis was repeated. Similar clustering of samples was observed across PC1. However, samples appeared closer together. 


```{r}
ggarrange(
  cpm(x, log = T) %>%
    t() %>%
    prcomp() %>% 
    autoplot(data = tibble(sample = rownames(.$x)) %>%
               left_join(x$samples),
             colour = "genotype", 
             shape = "Sex", 
             size = 4
    ) +
    scale_color_viridis_d(option = "turbo", end = 0.8) +
    ggtitle("Before CQN"),
  
  logCPM %>% 
    t() %>% 
    prcomp() %>%
    autoplot(data = tibble(sample = rownames(.$x)) %>%
               left_join(x$samples),
             colour = "genotype", 
             shape = "Sex", 
             size = 4
    ) +
    scale_color_viridis_d(option = "turbo", end = 0.8) +
    ggtitle("After CQN"),
  common.legend = T, 
  labels = "AUTO"
) 
#ggsave("plotsforpub/PCA_cqn.png", width = 18, height = 8, units = "cm", scale = 1.4)
```

# DE with cqn

```{r}
# Add the offset to the dge object 
x$offset <- cqn$glm.offset

# fit the glm
fit_cqn <- x %>% 
  estimateDisp(design) %>% 
  glmFit(design)

toptables_cqn <- design %>% colnames() %>% .[2:4] %>% 
  sapply(function(x){
    glmLRT(fit_1, coef = x) %>%
      topTags(n = Inf) %>%
      .[["table"]] %>%
      as_tibble() %>%
      arrange(PValue) %>%
       mutate(
        coef = x,
        DE = FDR < 0.05
      ) %>% 
      dplyr::select(
        gene_name, logFC, logCPM, PValue, FDR, DE, everything()  
      )
      
  }, simplify = FALSE)

toptables_cqn %>% 
  lapply(function(x){
    x %>%
      mutate(Direction = case_when(
        sign(logFC) == "1" ~ "up",
        sign(logFC) == "-1" ~ "down"
      ))
  }) %>%
  bind_rows() %>%
  dplyr::filter(DE == TRUE) %>%
  ggplot(aes(y = `coef`)) +
  geom_bar(stat = "count",
           width = 0.5,
           aes(fill = Direction),
           position = "dodge") +
  theme(legend.position = "bottom") +
  ggtitle("Number of DE genes in each comparison") +
  labs(x = "Number of DE genes",
       y = "Contrast",
       colour = "Direction of change") 
```

## Recheck the GC and length bias

```{r}
ggarrange(
toptables_cqn %>% 
  bind_rows() %>% 
  mutate(rankstat = sign(logFC)*-log10(PValue)) %>% 
  ggplot(aes(x = length, y = rankstat)) +
  geom_point(
    aes(colour = DE),
    alpha = 0.5
  ) +
  geom_smooth(se = FALSE, method = "gam") +
  facet_grid(rows = vars(coef)) +
  theme_bw() +
  theme(legend.position = "none") +
  scale_color_manual(values = c("grey50", "red")) +
  scale_x_log10()+
  labs(x = "Average transcript length per gene",
       colour = "Differentially expressed?",
       y = "sign(logFC)*-log10(PValue)") +
  coord_cartesian(ylim = c(-10, 10)),

toptables_cqn %>% 
  bind_rows() %>% 
  mutate(rankstat = sign(logFC)*-log10(PValue)) %>% 
  ggplot(aes(x = gc_content, y = rankstat)) +
  geom_point(
    aes(colour = DE),
    alpha = 0.5
  ) +
  geom_smooth(se = FALSE, method = "gam") +
  facet_grid(rows = vars(coef)) +
  theme_bw() +
  theme(legend.position = "none") +
  scale_color_manual(values = c("grey50", "red")) +
  coord_cartesian(ylim = c(-10,10)) + 
  labs(x = "Weighted average GC content (%) per gene", 
       colour = "Differentially expressed?", 
       y = "sign(logFC)*-log10(PValue)"),
common.legend = TRUE, 
labels = "AUTO"
) 
```

## Gene set enrichment analysis

### Gene sets

```{r}
KEGG <- msigdbr("Danio rerio", category = "C2", subcategory = "CP:KEGG") %>% 
  left_join(x$genes, by = c("gene_symbol" = "gene_name")) %>% 
  dplyr::filter(gene_id %in% rownames(x)) %>% 
  distinct(gs_name, gene_id, .keep_all = TRUE) %>% 
  split(f = .$gs_name) %>%
  lapply(extract2, "gene_id")

sizes <- KEGG %>% 
  lapply(length) %>% 
  unlist %>% 
  as.data.frame() %>% 
  set_colnames( "n_genes") %>% 
  rownames_to_column("gs")


# retain gene sets with at least 10 genes in it
KEGG <- KEGG[sizes %>% dplyr::filter(n_genes > 10) %>% .$gs]

# GO Terms
goSummaries <- url("https://uofabioinformaticshub.github.io/summaries2GO/data/goSummaries.RDS") %>%
  readRDS() %>%
  mutate(
    Term = Term(id),
    gs_name = Term %>% str_to_upper() %>% str_replace_all("[ -]", "_"),
    gs_name = paste0("GO_", gs_name)
    )

minPath <- 3

GO <- 
  msigdbr("Danio rerio", category = "C5") %>% 
  dplyr::filter(grepl(gs_name, pattern = "^GO")) %>% 
  left_join(x$genes, by = c("gene_symbol" = "gene_name")) %>% 
  dplyr::filter(gene_id %in% rownames(x)) %>% 
  mutate(gs_name = str_replace(gs_name, pattern = "GOBP_", replacement = "GO_")) %>% 
  mutate(gs_name = str_replace(gs_name, pattern = "GOMF_", replacement = "GO_")) %>%     
  mutate(gs_name = str_replace(gs_name, pattern = "GOCC_", replacement = "GO_")) %>%     
  left_join(goSummaries) %>% 
  dplyr::filter(shortest_path >= minPath) %>%
  distinct(gs_name, gene_id, .keep_all = TRUE) %>% 
  split(f = .$gs_name) %>%
  lapply(extract2, "gene_id")

GOdf <-  msigdbr("Danio rerio", category = "C5") %>% 
  dplyr::filter(grepl(gs_name, pattern = "^GO")) %>% 
  left_join(x$genes, by = c("gene_symbol" = "gene_name")) %>% 
  dplyr::filter(gene_id %in% rownames(x)) %>% 
  mutate(gs_name = str_replace(gs_name, pattern = "GOBP_", replacement = "GO_")) %>% 
  mutate(gs_name = str_replace(gs_name, pattern = "GOMF_", replacement = "GO_")) %>%     
  mutate(gs_name = str_replace(gs_name, pattern = "GOCC_", replacement = "GO_")) %>%     
  left_join(goSummaries) %>% 
  dplyr::filter(shortest_path >= minPath) %>%
  distinct(gs_name, gene_id, .keep_all = TRUE) %>% 
  dplyr::select(gs_name, gene_id)

sizes_GO <- GO %>%
  lapply(length) %>%
  unlist %>%
  as.data.frame() %>%
  set_colnames( "n_genes") %>%
  rownames_to_column("gs")


# retain gene sets with at least 5 genes in it
GO <- GO[sizes_GO %>% dplyr::filter(n_genes > 5) %>% .$gs]

ireGenes <- 
  read_xlsx("data/gene_sets/ireGenes.xlsx", sheet = "Zebrafish IREs") %>% 
  gather %>% 
  set_colnames(c("IRE", "gene_id")) %>% 
  mutate(IRE = case_when(
    IRE == "All 3' IREs" ~ "ire3_all", 
    IRE == "All 5' IREs" ~ "ire5_all", 
    IRE == "HQ 3' IREs" ~ "ire3_HQ", 
    IRE == "HQ 5' IREs" ~ "ire5_HQ"
  )) %>% 
  dplyr::filter(gene_id %in% rownames(x)) %>% 
  split(f = .$IRE) %>% 
  lapply(magrittr::extract2,"gene_id") 

chr <- x$genes %>% 
  dplyr::select(gene_id, chromosome) %>% 
  split(f = .$chromosome) %>%
  lapply(extract2, "gene_id") %>% 
  .[c("MT", seq(1:25))] # omit the ALT chromosmes and scaffolds

#biomart to get the conversion between mouse and zebrafish IDs
mart <- useMart("ensembl", "drerio_gene_ensembl")
getFromBiomart <- c("ensembl_gene_id", "mmusculus_homolog_associated_gene_name")
zf_id_2_mm_genename <- getBM(attributes = getFromBiomart,
                             mart = mart) %>% 
  as_tibble() %>% 
  set_colnames(c("gene_id", "mouse_gene_name"))

cell_type_markers <- 
  read_xls("data/gene_sets/cell_type_genes41586_2007_BFnature05453_MOESM11_ESM.xls") %>%
  as_tibble() %>%
  dplyr::select(c("Neuron-enriched", "Astrocyte-enriched", "Oligodendrocyte-enriched")) %>%
  gather(key = "cell_type", value = "mouse_gene_name") %>% 
  left_join(zf_id_2_mm_genename) %>% 
  na.omit %>% 
  split(f = .$cell_type) %>% 
  lapply(extract2, "gene_id")

cell_type_markers$`Microglia-enriched` <- 
  read_excel("data/gene_sets/microglia_oosterofetal.xlsx", sheet = 1) %>%
  rbind(read_excel("data/gene_sets/microglia_oosterofetal.xlsx", sheet = 2)) %>%
  rbind(read_excel("data/gene_sets/microglia_oosterofetal.xlsx", sheet = 3)) %>%
  .$Zebrafish.Ensembl.Gene.ID %>% 
  unique

cell_type_markers %<>% 
  lapply(function(y) {
    y[y %in% rownames(x)]
  })
```

## GOSEQ 

```{r}
pwf <- toptables_cqn$`MPS-IIIB` %>%
      with(
        nullp(
          DEgenes = structure(DE, names = gene_id), 
          bias.data = length
        )
      )

goseq(pwf, gene2cat = GO) %>% 
  as_tibble() %>% 
  dplyr::filter(numDEInCat > 0) %>% 
  mutate(FDR = p.adjust(over_represented_pvalue, method = "fdr")) %>% 
  dplyr::select(-under_represented_pvalue) %>% 
  dplyr::slice(1:10)

pwf <- toptables_cqn$`MPS-IIIA` %>%
      with(
        nullp(
          DEgenes = structure(DE, names = gene_id), 
          bias.data = length
        )
      )

goseq(pwf, gene2cat = GO) %>% 
  as_tibble() %>% 
  dplyr::filter(numDEInCat > 0) %>% 
  mutate(FDR = p.adjust(over_represented_pvalue, method = "fdr")) %>% 
  dplyr::select(-under_represented_pvalue) %>% 
  dplyr::slice(1:10)

pwf <- toptables_cqn$`sgsh/+` %>%
      with(
        nullp(
          DEgenes = structure(DE, names = gene_id), 
          bias.data = length
        )
      )

goseq(pwf, gene2cat = GO) %>% 
  as_tibble() %>% 
  dplyr::filter(numDEInCat > 0) %>% 
  mutate(FDR = p.adjust(over_represented_pvalue, method = "fdr")) %>% 
  dplyr::select(-under_represented_pvalue) %>% 
  dplyr::slice(1:10)
  
```



### KEGG 
```{r}
fryKEGG <- 
  design %>% colnames() %>% .[2:4] %>% 
  sapply(function(y) {
    logCPM %>% 
      fry(
        index = KEGG,
        design = design, 
        contrast = y, 
        sort = "mixed"
      ) %>% 
      rownames_to_column("pathway") %>% 
      as_tibble() %>% 
      mutate(coef = y)
  }, simplify = FALSE)

fryKEGG$`MPS-IIIB` %>% 
  dplyr::slice(1:10) %>% 
  mutate(pathway = str_remove(pathway, pattern = "KEGG_")) %>% 
  ggplot(aes(x = -log10(PValue.Mixed), y = reorder(pathway, -FDR.Mixed))) +
  geom_col(aes(fill = FDR.Mixed < 0.05)) +
  scale_fill_manual(values = c("grey50", 'red')) +
  theme_pubr() +
  labs(fill = "FDR < 0.05", 
       x = "-log10(PValue)", 
       y = "")

fryKEGG$`MPS-IIIA` %>% 
  dplyr::slice(1:10) %>% 
  mutate(pathway = str_remove(pathway, pattern = "KEGG_")) %>% 
  ggplot(aes(x = -log10(PValue.Mixed), y = reorder(pathway, -FDR.Mixed))) +
  geom_col(aes(fill = FDR.Mixed < 0.05)) +
  scale_fill_manual(values = c("grey50", 'red')) +
  theme_pubr() +
  labs(fill = "FDR < 0.05", 
       x = "-log10(PValue)", 
       y = "")  

fryKEGG$`sgsh/+` %>% 
  dplyr::slice(1:10) %>% 
  mutate(pathway = str_remove(pathway, pattern = "KEGG_")) %>% 
  ggplot(aes(x = -log10(PValue.Mixed), y = reorder(pathway, -FDR.Mixed))) +
  geom_col(aes(fill = FDR.Mixed < 0.05)) +
  scale_fill_manual(values = c("grey50", 'red')) +
  theme_pubr() +
  labs(fill = "FDR < 0.05", 
       x = "-log10(PValue)", 
       y = "")

sigpaths <- fryKEGG %>% 
  bind_rows() %>% 
  dplyr::filter(FDR.Mixed < 0.05) %>% 
    .$pathway %>% 
    unique()
  
fryKEGG %>% 
  bind_rows() %>% 
  dplyr::filter(pathway %in% sigpaths) %>% 
  mutate(pathway = str_remove(pathway, pattern = "KEGG_")) %>% 
  ggplot(aes(x = -log10(PValue.Mixed), y = reorder(pathway, -FDR.Mixed))) +
  geom_col(aes(fill = coef, alpha = FDR.Mixed < 0.05), position = "dodge") +
  scale_fill_viridis_d(end = 0.75) +
  scale_alpha_manual(values = c(0.5,1)) +
  theme_pubr() +
  geom_vline(xintercept = 2.3) +
  labs(fill = "FDR < 0.05", 
       x = "-log10(PValue)", 
       y = "") +
  ggsave("output/plots/AB_KEGG.png", width = 20, height = 10, units = "cm", dpi = 400, scale = 2)

```


### GO
```{r}
fry_GO <-  design %>% colnames() %>% .[2:4] %>% 
  sapply(function(y) {
    logCPM %>% 
      fry(
        index = GO,
        design = design, 
        contrast = y, 
        sort = "mixed"
      ) %>% 
      rownames_to_column("pathway") %>% 
      as_tibble() %>% 
      mutate(coef = y, 
             sig = FDR.Mixed < 0.05)
  }, simplify = FALSE)
```


```{r}
toptables_cqn %>% 
  bind_rows() %>% 
  dplyr::filter(gene_id %in% GO$GO_REGULATION_OF_MICROGLIAL_CELL_ACTIVATION) %>% 
  dplyr::select(gene_name, logFC, coef) %>% 
  dplyr::distinct(gene_name, coef, .keep_all = T) %>% 
  spread(key = "coef", value = "logFC") %>% 
  column_to_rownames("gene_name") %>% 
  t() %>% 
  pheatmap(color = colorRampPalette(rev(brewer.pal(n = 5,
                                               name = "RdBu")))(100), 
           main = "GO_REGULATION_OF_MICROGLIAL_CELL_ACTIVATION: FDR = 6.44e-5", 
           breaks = c(seq(min(.), 0, length.out=ceiling(100/2) + 1), 
              seq(max(.)/100, max(.), length.out=50)), 
           cellheight = 30, cellwidth = 12,angle_col = 45,
           treeheight_row = 0, treeheight_col = 0
           )


toptables_cqn %>% 
  bind_rows() %>% 
  dplyr::filter(gene_id %in% KEGG$KEGG_OXIDATIVE_PHOSPHORYLATION) %>% 
  dplyr::select(gene_name, logFC, coef) %>% 
  dplyr::distinct(gene_name, coef, .keep_all = T) %>% 
  spread(key = "coef", value = "logFC") %>% 
  column_to_rownames("gene_name") %>% 
  t() %>% 
  pheatmap(color = colorRampPalette(rev(brewer.pal(n = 5,
                                               name = "RdBu")))(100), 
           main = "KEGG_LYSOSOME: FDR = 6.44e-5", 
           breaks = c(seq(min(.), 0, length.out=ceiling(100/2) + 1), 
              seq(max(.)/100, max(.), length.out=50)), 
           cellheight = 30, cellwidth = 12,angle_col = 45,
           treeheight_row = 0, treeheight_col = 0
           )
```

### IRE

```{r}
fryIRE <- design %>% colnames() %>% .[2:4] %>% 
  sapply(function(y) {
    logCPM %>% 
      fry(
        index = ireGenes,
        design = design, 
        contrast = y, 
        sort = "directional"
      ) %>% 
      rownames_to_column("pathway") %>% 
      as_tibble() %>% 
      mutate(coef = y)
  }, simplify = FALSE)

fryIRE %>% 
  bind_rows() %>% 
  dplyr::select(pathway, coef, pvalue = PValue.Mixed, FDR = FDR.Mixed) %>% 
  ggplot(aes(y = pathway, x = -log10(pvalue))) + 
  geom_col(aes(fill = coef, alpha = ), 
           position = "dodge") +
  scale_fill_viridis_d(end = 0.75) +
  geom_vline(xintercept = 1.88) +
  annotate(geom = "label", x = 1.8, y = "ire5_HQ", label = "FDR = 0.05") +
  ggsave("output/plots/AB_ire.png", width = 20, height = 10, units = "cm", dpi = 400, scale = 2)


toptables_cqn %>% 
  bind_rows() %>% 
    dplyr::filter(gene_id %in% ireGenes$ire3_all) %>% 
  dplyr::select(gene_name, logFC, coef) %>% 
  dplyr::distinct(gene_name, coef, .keep_all = T) %>% 
  spread(key = "coef", value = "logFC") %>% 
  column_to_rownames("gene_name") %>% 
  t() %>% 
  pheatmap(color = colorRampPalette(rev(brewer.pal(n = 5,
                                               name = "RdBu")))(100), 
           main = "3' IRE genes: FDR = 0.066", 
           breaks = c(seq(min(.), 0, length.out=ceiling(100/2) + 1), 
              seq(max(.)/100, max(.), length.out=50)), 
           cellheight = 30, 
           cellwidth = 1,
           angle_col = 45,
           treeheight_row = 0, 
           treeheight_col = 0, 
           show_colnames = F
           )
```

### Cell type prop check

Since this experiment was performed on whole larvae, we need to have a look at whether changes to cell type proportions could explain cause some artefactual changes to gene expression. We have done this previously by assessing the expression of marker genes of specific cell types. So this is what I will do here. I have taken the lists of genes frmo day 5 only from the zebrafish single cell atlas (Farnsworth et al. 2020) which contain at least 10 genes. This leaves  `r length(cell_type_markers)` gene sets (cell types) to test.

```{r}
celltype <- design %>% colnames() %>% .[2:4] %>% 
  sapply(function(y) {
    logCPM %>% 
      fry(
        index = cell_type_markers,
        design = design, 
        contrast = y, 
        sort = "directional"
      ) %>% 
      rownames_to_column("pathway") %>% 
      as_tibble() %>% 
      mutate(coef = y)
  }, simplify = FALSE)

celltype %>% 
  bind_rows() %>% 
  ggplot(aes(y = pathway, x = -log10(PValue))) + 
  geom_col(aes(fill = coef), 
           position = "dodge") +
  scale_fill_viridis_d(end = 0.75) +
  geom_vline(xintercept = 1.3) +
  annotate(geom = "label", x = 1.2, y = 4.4, label = "FDR = 0.05") 
```


### Chrosomosomes gene sets

```{r}
fry_chr <- design %>% colnames() %>% .[2:4] %>% 
  sapply(function(y) {
    logCPM %>% 
      fry(
        index = c(chr),
        design = design, 
        contrast = y, 
        sort = "mixed"
      ) %>% 
      rownames_to_column("pathway") %>% 
      as_tibble() %>% 
      mutate(coef = y)
  }, simplify = FALSE)


fry_chr %>% 
  bind_rows() %>% 
  mutate(pathway = factor(pathway, levels = c("MT", seq(1:25)))) %>% 
  ggplot(aes(y = pathway, x = -log10(PValue.Mixed), fill = coef)) +
  geom_col(position = "dodge")
  
```

An enrichment of DE genes is observed on the same chromosome as the mutation. This is obsevred in the manhat plots below. 

```{r}
manhat_input <- 
	toptables_cqn %>% 
  bind_rows() %>% 
	left_join(x$genes %>% 
			  	as_tibble() %>% 
			  	dplyr::select(start, end, chromosome, gene_id) %>% 
			  	group_by(gene_id) %>% 
			  	mutate(mid = mean(c(start, end)))
	) %>% 
  dplyr::filter(chromosome %in% seq(1:25)) %>% 
  group_by(gene_id) %>% 
  mutate(mid = mean(c(start, end))) %>% 
  ungroup %>% 
  mutate(chromosome = factor(chromosome, levels = seq(1:25)))  %>% 
  group_by(chromosome) %>% 
  summarise(chrLen = max(mid)) %>% 
  mutate(chrSt = cumsum(chrLen)-chrLen) %>% 
	dplyr::select(-chrLen) %>% 
  left_join(toptable_1 %>% 
              bind_rows() %>% 
              left_join(x$genes %>% 
                          as_tibble() %>% 
                          dplyr::select(start, end, chromosome, gene_id) %>% 
                          group_by(gene_id) %>% 
                          mutate(mid = mean(c(start, end)))
              )
  ) %>% 
  group_by(gene_id) %>% 
  mutate(mid = mean(c(start, end))) %>% 
  ungroup %>% 
  dplyr::arrange(mid, chromosome) %>% 
  mutate(midCum = chrSt + mid)

axis <- manhat_input %>%
  group_by(chromosome) %>%
  dplyr::arrange(chromosome) %>%
  summarize(center = (max(midCum) + min(midCum)) / 2) %>%
  mutate(
    seqnames = as.numeric(chromosome),
    colour = rep(c("grey40", "black"), length.out = 25)
  )

ggplot(manhat_input, aes(x = midCum, y = -log10(PValue))) +
  geom_point(aes(color=chromosome), alpha = 0.5, size = 1) +
  geom_point(alpha = 0.5, size = 1, colour = "red", 
             data = . %>% 
               dplyr::filter(coef == "MPS-IIIB" & chromosome == 24)) +
  geom_point(alpha = 0.5, size = 1, colour = "red", 
             data = . %>% 
               dplyr::filter(coef ==  "MPS-IIIA" & chromosome == 22)) +
  geom_point(alpha = 0.5, size = 1, colour = "red", 
             data = . %>% 
               dplyr::filter(coef ==  "sgsh/+" & chromosome == 22)) +
  scale_color_manual(values = axis$colour) +
  scale_x_continuous(label = axis$chromosome, breaks = axis$center) +
  scale_y_continuous(limits =c(0, 10)) + # zoom in
  facet_wrap(~coef, ncol = 1) +
  labs(x = "Chromosome", y = expression(paste(-log[10], "(p)"))) +
  theme( 
    legend.position="none",
    panel.border = element_blank(),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.x = element_blank()
  ) 
```


I also compared the log2 fold changes of genes on chr22 in the heterozygous and homozygous sgsh samples.

```{r}
DEgenesIIIA <- toptables_cqn$`MPS-IIIA` %>% 
  dplyr::filter(DE == T) %>% 
  .$gene_id

DEgenesIIIB <- toptables_cqn$`MPS-IIIB` %>% 
  dplyr::filter(DE == T) %>% 
  .$gene_id

DEgeneshet <- toptables_cqn$`sgsh/+` %>% 
  dplyr::filter(DE == T) %>% 
  .$gene_id

toptables_cqn %>% 
  bind_rows() %>% 
  dplyr::filter(coef %in% c("MPS-IIIA", "sgsh/+")) %>% 
  dplyr::filter(chromosome == "22") %>% 
  dplyr::select(gene_id, logFC, coef)  %>% 
  spread(key = "coef", value = "logFC") %>% 
  mutate(DE = case_when(
    gene_id %in% intersect(DEgenesIIIA, DEgeneshet) ~ "both", 
    gene_id %in% DEgeneshet[!DEgeneshet %in% DEgenesIIIA] ~ "het only",
    gene_id %in% DEgenesIIIA[!DEgenesIIIA %in% DEgeneshet] ~ "MPS-IIIA only", 
    TRUE ~ "not DE"
  ) 
  ) %>% 
  left_join(x$genes) %>% 
  group_by(gene_id) %>% 
  mutate(mid = mean(c(start, end))) %>% 
  ggplot(aes(x = `MPS-IIIA`, y =`sgsh/+`)) +
  geom_point(aes(colour = DE), 
             alpha = 0.5, 
             size = 4) +
  # geom_abline(slope = 1) +
  geom_label_repel(aes(label = gene_name), nudge_y = 1,
                   data = . %>% 
                     dplyr::filter(gene_name == "sgsh")) +
  scale_x_continuous(limits = c(-5, 5)) +
  scale_y_continuous(limits = c(-2.5, 5)) +
  theme(aspect.ratio = 1) +
  # scale_color_viridis_c(option = "turbo") +
  ggtitle("LogFC of genes on chromosome 22 in sgsh het vs hom ")

```

# is there a functional enrichment of genes only DE in the MPS-IIIA
```{r}
kegga_DE <- toptables_cqn %>% 
  bind_rows() %>% 
  dplyr::filter(coef %in% c("MPS-IIIA", "sgsh/+")) %>% 
  dplyr::filter(chromosome == "22") %>% # only chr22 genes
  dplyr::filter(gene_id %in% DEgenesIIIA) %>%#  DE in the homs
  dplyr::filter(!(gene_id %in% DEgeneshet) ) %>% # and not DE in hets 
  .$gene_id %>% 
  unique
  
#KEGG using a background of all genes
kegga(de = kegga_DE, 
      
      geneid = x$genes %>% 
        dplyr::filter(chromosome == 22) %>% 
        .$gene_id,
      
      universe = x$genes %>% 
        dplyr::filter(chromosome == 22) %>% 
        .$gene_id,
      
      gene.pathway = KEGG %>%
        unlist %>%
        as.data.frame() %>%
        rownames_to_column("pathway") %>%
        set_colnames(c("pathway", "gene_id")) %>%
        mutate(pathway = pathway %>% str_remove_all("[0-9]+$")) %>%
        dplyr::select(gene_id, pathway)  
) %>% 
   topKEGG() %>% 
    as.data.frame() %>% 
    rownames_to_column("pathway") %>% 
    mutate(FDR = p.adjust(P.DE, "fdr")) %>% 
    dplyr::select(pathway, raw.p = P.DE, FDR, `No. DE` = DE, `No. in geneset` = N) %>%
    head(10) 
```


## Exploration with HMP

```{r}
fry <- 
  design %>% colnames() %>% .[2:4] %>% 
  sapply(function(y) {
    logCPM %>% 
      fry(
        index = c(KEGG, ireGenes),
        design = design, 
        contrast = y, 
        sort = "mixed"
      ) %>% 
      rownames_to_column("pathway") %>% 
      as_tibble() %>% 
      mutate(coef = y)
  }, simplify = FALSE)

camera <- 
  design %>% colnames() %>% .[2:4] %>% 
  sapply(function(y) {
    logCPM %>% 
      camera(
        index = c(KEGG, ireGenes),
        design = design, 
        contrast = y
      ) %>% 
      rownames_to_column("pathway") %>% 
      as_tibble() %>% 
      mutate(coef = y)
  }, simplify = FALSE)

ranks <- 
   sapply(toptables_cqn, function(y) {
     y %>% 
       mutate(rankstat = sign(logFC) * log10(1/PValue)) %>% 
       arrange(rankstat) %>% 
       dplyr::select(c("gene_id", "rankstat")) %>% #only want the Pvalue with sign
       with(structure(rankstat, names = gene_id)) %>% 
       rev() # reverse so the start of the list is upregulated genes
   }, simplify = FALSE)

# Run fgsea

# set a seed for a reproducible result
set.seed(33)
fgsea <- ranks %>%
  sapply(function(x){
    fgseaMultilevel(stats = x, pathways = c(KEGG, ireGenes)) %>%
      as_tibble() %>%
      dplyr::rename(FDR = padj) %>%
      mutate(padj = p.adjust(pval, "bonferroni")) %>%
      dplyr::select(pathway, pval, FDR, padj, everything()) %>%
      arrange(pval) %>%
      mutate(sig = padj < 0.05)
  }, simplify = F)

fgsea$`MPS-IIIB` %<>% 
  mutate(coef = "MPS-IIIB")
fgsea$`MPS-IIIA` %<>% 
  mutate(coef = "MPS-IIIA")
fgsea$`sgsh/+` %<>% 
  mutate(coef = "sgsh/+")


HMP <- fry %>% 
  bind_rows() %>% 
  dplyr::select(pathway, PValue.Mixed, coef) %>% 
  dplyr::rename(fry_p = PValue.Mixed) %>% 
  left_join(camera %>% 
              bind_rows() %>% 
              dplyr::select(pathway, PValue, coef), 
            by = c("pathway", "coef")) %>% 
  dplyr::rename(camera_p = PValue) %>% 
  left_join(fgsea %>% 
              bind_rows() %>% 
              dplyr::select(pathway, pval, coef), 
            by = c("pathway", "coef")) %>% 
  dplyr::rename(fgsea_p = pval) %>% 
  bind_rows() %>% 
  nest(p = one_of(c("fry_p", "camera_p", "fgsea_p"))) %>% 
 mutate(harmonic_p = vapply(p, function(x){
        x <- unlist(x)
        x <- x[!is.na(x)]
        p.hmp(x, L = 4)
      }, numeric(1))
 ) %>% 
  unnest() %>% 
  mutate(harmonic_p_FDR = p.adjust(harmonic_p, "fdr"),
         sig = harmonic_p_FDR < 0.05) %>% 
  arrange(harmonic_p) 


sigpaths <- HMP %>% 
  dplyr::filter(sig == TRUE) %>% 
  .$pathway

HMP %>% 
  dplyr::filter(pathway %in% sigpaths) %>% 
  ggplot(aes(y = pathway, x = -log10(harmonic_p))) +
  geom_col(aes(fill = coef,alpha = sig), position = "dodge") +
  geom_vline(xintercept = -log10(3.0000e-03)) + 
  scale_alpha_manual(values = c(0.5, 1))
```

# Exploration of traditional ire containing genes involved in iron metabolism. 

ferritin, transferrin, dmt1, steap, 

```{r}
ferritinGenes <- x$genes %>% 
  dplyr::filter(grepl(description, pattern = "ferritin")) %>% 
  .$gene_id

# x$genes %>% dplyr::filter(grepl(description, pattern = "transferrin"))
transferrinreceptors <- c("ENSDARG00000016771", "ENSDARG00000077372", "ENSDARG00000101322", "ENSDARG00000103727")

toptables_cqn %>% 
  bind_rows() %>% 
  dplyr::filter(gene_id %in% c(ferritinGenes, transferrinreceptors)) %>% 
  dplyr::select(gene_name, logFC, coef) %>% 
  dplyr::distinct(gene_name, coef, .keep_all = T) %>% 
  spread(key = "coef", value = "logFC") %>% 
  column_to_rownames("gene_name") %>% 
  t() %>% 
  pheatmap(color = colorRampPalette(rev(brewer.pal(n = 5,
                                               name = "RdBu")))(100), 
           main = "ferririn and Transferrin genes", 
           breaks = c(seq(min(.), 0, length.out=ceiling(100/2) + 1), 
              seq(max(.)/100, max(.), length.out=50)), 
           cellheight = 30, cellwidth = 12,angle_col = 45,
           treeheight_row = 0, treeheight_col = 0
           )

toptables_cqn %>% 
  bind_rows() %>% 
  dplyr::filter(gene_id %in% GO$GO_REGULATION_OF_MICROGLIAL_CELL_ACTIVATION) %>% 
  dplyr::select(gene_name, logFC, coef) %>% 
  dplyr::distinct(gene_name, coef, .keep_all = T) %>% 
  spread(key = "coef", value = "logFC") %>% 
  column_to_rownames("gene_name") %>% 
  t() %>% 
  pheatmap(color = colorRampPalette(rev(brewer.pal(n = 5,
                                               name = "RdBu")))(100), 
           main = "GO_REGULATION_OF_MICROGLIAL_CELL_ACTIVATION", 
           breaks = c(seq(min(.), 0, length.out=ceiling(100/2) + 1), 
              seq(max(.)/100, max(.), length.out=50)), 
           cellheight = 30, cellwidth = 12,angle_col = 45,
           treeheight_row = 0, treeheight_col = 0
           )

cpm(x, log = T) %>% 
  .[GO$GO_PH_REDUCTION,] %>% 
  as.data.frame() %>% 
  rownames_to_column("gene_id") %>% 
  gather(key = "sample", value = "logCPM", colnames(x)) %>% 
  left_join(x$samples) %>% 
  ggplot(aes(x = genotype, y = logCPM)) +
  geom_jitter() +
  geom_boxplot() +
  facet_wrap(~gene_id, scales = "free")
```


## Data export
```{r}

x %>% saveRDS("data/R_objects/dge_AB.rds")
fryKEGG %>% saveRDS("data/R_objects/AB_fryKEGG.rds")
fryIRE %>% saveRDS("data/R_objects/AB_fryIRE.rds")
celltype %>% saveRDS("data/R_objects/AB_frycellTypes.rds")
toptables_cqn %>% saveRDS("data/R_objects/AB_toptab_cqn.rds")

```


